// sushi/_cjs/kvm/chain/chains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KvmChainId = exports.kvmChainIds = exports.kvmChains = void 0;
exports.isKvmChainId = isKvmChainId;
exports.getKvmChainById = getKvmChainById;
exports.isKvmMainnetChainId = isKvmMainnetChainId;
exports.isKvmChainKey = isKvmChainKey;
exports.getKvmChainByKey = getKvmChainByKey;
const define_chain_js_1 = require("./define-chain.js");
const kadena = (0, define_chain_js_1.defineKvmChain)({
    chainId: -3,
    key: 'kadena',
    shortName: 'kda',
    name: 'Kadena',
    blockExplorers: {
        default: {
            name: 'Kadena Explorer',
            url: 'https://explorer.chainweb.com/mainnet',
        },
    },
    netType: 'mainnet',
});
exports.kvmChains = [kadena];
exports.kvmChainIds = exports.kvmChains.map((chain) => chain.chainId);
exports.KvmChainId = new Proxy({}, {
    get: (_, key) => {
        const _key = key.toLowerCase().replace(/_/g, '-');
        const chain = exports.kvmChains.find((chain) => chain.key === _key);
        if (!chain)
            return undefined;
        return chain.chainId;
    },
});
function isKvmChainId(chainId) {
    return exports.kvmChains.some((chain) => chain.chainId === chainId);
}
function getKvmChainById(chainId) {
    return exports.kvmChains.find((chain) => chain.chainId === chainId);
}
function isKvmMainnetChainId(chainId) {
    return exports.kvmChains.some((chain) => chain.chainId === chainId && chain.netType === 'mainnet');
}
function isKvmChainKey(key) {
    return exports.kvmChains.some((chain) => chain.key === key);
}
function getKvmChainByKey(key) {
    return exports.kvmChains.find((chain) => chain.key === key);
}
