// sushi/_cjs/evm/pool/sushiswap-v3/entities/NonfungiblePositionManager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NonfungiblePositionManager = void 0;
const tiny_invariant_1 = __importDefault(require("tiny-invariant"));
const viem_1 = require("viem");
const amount_js_1 = require("../../../../generic/currency/amount.js");
const index_js_1 = require("../../../abi/nonfungiblePositionManagerAbi/index.js");
const index_js_2 = require("../utils/index.js");
const Multicall_js_1 = require("./Multicall.js");
const Payments_js_1 = require("./Payments.js");
const Position_js_1 = require("./Position.js");
const SelfPermit_js_1 = require("./SelfPermit.js");
function isMint(options) {
    return Object.keys(options).some((k) => k === 'recipient');
}
class NonfungiblePositionManager {
    constructor() { }
    static encodeCreate(pool) {
        return (0, viem_1.encodeFunctionData)({
            abi: index_js_1.nonfungiblePositionManagerAbi_createAndInitializePoolIfNecessary,
            functionName: 'createAndInitializePoolIfNecessary',
            args: [
                pool.token0.address,
                pool.token1.address,
                pool.fee,
                pool.sqrtRatioX96,
            ],
        });
    }
    static createCallParameters(pool) {
        return {
            calldata: this.encodeCreate(pool),
            value: (0, viem_1.toHex)(0),
        };
    }
    static addCallParameters(position, options) {
        (0, tiny_invariant_1.default)(position.liquidity > 0n, 'ZERO_LIQUIDITY');
        const calldatas = [];
        const { amount0: amount0Desired, amount1: amount1Desired } = position.mintAmounts;
        const minimumAmounts = position.mintAmountsWithSlippage(options.slippageTolerance);
        const amount0Min = minimumAmounts.amount0;
        const amount1Min = minimumAmounts.amount1;
        const deadline = BigInt(options.deadline);
        if (isMint(options) && options.createPool) {
            calldatas.push(this.encodeCreate(position.pool));
        }
        if (options.token0Permit) {
            calldatas.push(SelfPermit_js_1.SelfPermit.encodePermit(position.pool.token0, options.token0Permit));
        }
        if (options.token1Permit) {
            calldatas.push(SelfPermit_js_1.SelfPermit.encodePermit(position.pool.token1, options.token1Permit));
        }
        if (isMint(options)) {
            const recipient = (0, index_js_2.validateAndParseAddress)(options.recipient);
            calldatas.push((0, viem_1.encodeFunctionData)({
                abi: index_js_1.nonfungiblePositionManagerAbi_mint,
                functionName: 'mint',
                args: [
                    {
                        token0: position.pool.token0.address,
                        token1: position.pool.token1.address,
                        fee: position.pool.fee,
                        tickLower: position.tickLower,
                        tickUpper: position.tickUpper,
                        amount0Desired: amount0Desired,
                        amount1Desired: amount1Desired,
                        amount0Min,
                        amount1Min,
                        recipient,
                        deadline,
                    },
                ],
            }));
        }
        else {
            calldatas.push((0, viem_1.encodeFunctionData)({
                abi: index_js_1.nonfungiblePositionManagerAbi_increaseLiquidity,
                functionName: 'increaseLiquidity',
                args: [
                    {
                        tokenId: BigInt(options.tokenId),
                        amount0Desired: amount0Desired,
                        amount1Desired: amount1Desired,
                        amount0Min,
                        amount1Min,
                        deadline,
                    },
                ],
            }));
        }
        let value = (0, viem_1.toHex)(0);
        if (options.useNative) {
            const wrapped = options.useNative.wrap();
            (0, tiny_invariant_1.default)(position.pool.token0.isSame(wrapped) ||
                position.pool.token1.isSame(wrapped), 'NO_WETH');
            const wrappedValue = position.pool.token0.isSame(wrapped)
                ? amount0Desired
                : amount1Desired;
            if (wrappedValue > 0n) {
                calldatas.push(Payments_js_1.Payments.encodeRefundETH());
            }
            value = (0, viem_1.toHex)(wrappedValue);
        }
        return {
            calldata: Multicall_js_1.Multicall.encodeMulticall(calldatas),
            value,
        };
    }
    static encodeCollect(collectOptions) {
        const calldatas = [];
        const nativeBalancesByRecipient = new Map();
        const sweepBalancesByRecipient = new Map();
        for (const options of collectOptions) {
            const tokenId = BigInt(options.tokenId);
            const involvesETH = options.expectedCurrencyOwed0.currency.type === 'native' ||
                options.expectedCurrencyOwed1.currency.type === 'native';
            const recipient = (0, index_js_2.validateAndParseAddress)(options.recipient);
            calldatas.push((0, viem_1.encodeFunctionData)({
                abi: index_js_1.nonfungiblePositionManagerAbi_collect,
                functionName: 'collect',
                args: [
                    {
                        tokenId,
                        recipient: involvesETH ? viem_1.zeroAddress : recipient,
                        amount0Max: viem_1.maxUint128,
                        amount1Max: viem_1.maxUint128,
                    },
                ],
            }));
            if (involvesETH) {
                const ethAmount = options.expectedCurrencyOwed0.currency.type === 'native'
                    ? options.expectedCurrencyOwed0.amount
                    : options.expectedCurrencyOwed1.amount;
                const token = options.expectedCurrencyOwed0.currency.type === 'native'
                    ? options.expectedCurrencyOwed1.currency
                    : options.expectedCurrencyOwed0.currency;
                const tokenAmount = options.expectedCurrencyOwed0.currency.type === 'native'
                    ? options.expectedCurrencyOwed1.amount
                    : options.expectedCurrencyOwed0.amount;
                const currentNativeBalance = nativeBalancesByRecipient.get(recipient) ?? 0n;
                nativeBalancesByRecipient.set(recipient, currentNativeBalance + ethAmount);
                const balanceKey = `${token.address}-${recipient}`;
                const accumulatedBalance = sweepBalancesByRecipient.get(balanceKey);
                if (accumulatedBalance) {
                    accumulatedBalance.amount += tokenAmount;
                }
                else {
                    sweepBalancesByRecipient.set(balanceKey, {
                        token,
                        amount: tokenAmount,
                        recipient,
                    });
                }
            }
        }
        for (const [recipient, nativeAmount,] of nativeBalancesByRecipient.entries()) {
            calldatas.push(Payments_js_1.Payments.encodeUnwrapWETH9(nativeAmount, recipient));
        }
        for (const { token, amount, recipient, } of sweepBalancesByRecipient.values()) {
            calldatas.push(Payments_js_1.Payments.encodeSweepToken(token, amount, recipient));
        }
        return calldatas;
    }
    static collectCallParameters(options) {
        const calldatas = NonfungiblePositionManager.encodeCollect(Array.isArray(options) ? options : [options]);
        return {
            calldata: Multicall_js_1.Multicall.encodeMulticall(calldatas),
            value: (0, viem_1.toHex)(0),
        };
    }
    static removeCallParameters(position, options) {
        const calldatas = [];
        const deadline = BigInt(options.deadline);
        const tokenId = BigInt(options.tokenId);
        const partialPosition = new Position_js_1.Position({
            pool: position.pool,
            liquidity: options.liquidityPercentage.mul(position.liquidity).quotient,
            tickLower: position.tickLower,
            tickUpper: position.tickUpper,
        });
        (0, tiny_invariant_1.default)(partialPosition.liquidity > 0n, 'ZERO_LIQUIDITY');
        const { amount0: amount0Min, amount1: amount1Min } = partialPosition.burnAmountsWithSlippage(options.slippageTolerance);
        if (options.permit) {
            calldatas.push((0, viem_1.encodeFunctionData)({
                abi: index_js_1.nonfungiblePositionManagerAbi_permit,
                functionName: 'permit',
                args: [
                    (0, index_js_2.validateAndParseAddress)(options.permit.spender),
                    tokenId,
                    deadline,
                    options.permit.v,
                    options.permit.r,
                    options.permit.s,
                ],
            }));
        }
        calldatas.push((0, viem_1.encodeFunctionData)({
            abi: index_js_1.nonfungiblePositionManagerAbi_decreaseLiquidity,
            functionName: 'decreaseLiquidity',
            args: [
                {
                    tokenId,
                    liquidity: partialPosition.liquidity,
                    amount0Min,
                    amount1Min,
                    deadline,
                },
            ],
        }));
        const { expectedCurrencyOwed0, expectedCurrencyOwed1, ...rest } = options.collectOptions;
        calldatas.push(...NonfungiblePositionManager.encodeCollect([
            {
                tokenId: (0, viem_1.toHex)(BigInt(options.tokenId)),
                expectedCurrencyOwed0: expectedCurrencyOwed0.add(new amount_js_1.Amount(expectedCurrencyOwed0.currency, amount0Min)),
                expectedCurrencyOwed1: expectedCurrencyOwed1.add(new amount_js_1.Amount(expectedCurrencyOwed1.currency, amount1Min)),
                ...rest,
            },
        ]));
        if (options.liquidityPercentage.eq(1n)) {
            if (options.burnToken) {
                calldatas.push((0, viem_1.encodeFunctionData)({
                    abi: index_js_1.nonfungiblePositionManagerAbi_burn,
                    functionName: 'burn',
                    args: [tokenId],
                }));
            }
        }
        else {
            (0, tiny_invariant_1.default)(options.burnToken !== true, 'CANNOT_BURN');
        }
        return {
            calldata: Multicall_js_1.Multicall.encodeMulticall(calldatas),
            value: (0, viem_1.toHex)(0),
        };
    }
    static safeTransferFromParameters(options) {
        const recipient = (0, index_js_2.validateAndParseAddress)(options.recipient);
        const sender = (0, index_js_2.validateAndParseAddress)(options.sender);
        let calldata;
        if (options.data) {
            calldata = (0, viem_1.encodeFunctionData)({
                abi: index_js_1.nonfungiblePositionManagerAbi_safeTransferFrom,
                functionName: 'safeTransferFrom',
                args: [sender, recipient, BigInt(options.tokenId), options.data],
            });
        }
        else {
            calldata = (0, viem_1.encodeFunctionData)({
                abi: index_js_1.nonfungiblePositionManagerAbi_safeTransferFrom,
                functionName: 'safeTransferFrom',
                args: [sender, recipient, BigInt(options.tokenId)],
            });
        }
        return {
            calldata: calldata,
            value: (0, viem_1.toHex)(0),
        };
    }
}
exports.NonfungiblePositionManager = NonfungiblePositionManager;
