// sushi/_cjs/evm/pool/sushiswap-v3/entities/Payments.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Payments = void 0;
const viem_1 = require("viem");
const index_js_1 = require("../../../abi/peripheryPaymentsWithFeeAbi/index.js");
const index_js_2 = require("../utils/index.js");
class Payments {
    constructor() { }
    static encodeFeeBips(fee) {
        return fee.mul(10_000).quotient;
    }
    static encodeUnwrapWETH9(amountMinimum, recipient, feeOptions) {
        recipient = (0, index_js_2.validateAndParseAddress)(recipient);
        if (feeOptions) {
            const feeBips = this.encodeFeeBips(feeOptions.fee);
            const feeRecipient = (0, index_js_2.validateAndParseAddress)(feeOptions.recipient);
            return (0, viem_1.encodeFunctionData)({
                abi: index_js_1.peripheryPaymentsWithFeeAbi_unwrapWETH9WithFee,
                functionName: 'unwrapWETH9WithFee',
                args: [amountMinimum, recipient, feeBips, feeRecipient],
            });
        }
        else {
            return (0, viem_1.encodeFunctionData)({
                abi: index_js_1.peripheryPaymentsWithFeeAbi_unwrapWETH9,
                functionName: 'unwrapWETH9',
                args: [amountMinimum, recipient],
            });
        }
    }
    static encodeSweepToken(token, amountMinimum, recipient, feeOptions) {
        recipient = (0, index_js_2.validateAndParseAddress)(recipient);
        if (feeOptions) {
            const feeBips = this.encodeFeeBips(feeOptions.fee);
            const feeRecipient = (0, index_js_2.validateAndParseAddress)(feeOptions.recipient);
            return (0, viem_1.encodeFunctionData)({
                abi: index_js_1.peripheryPaymentsWithFeeAbi_sweepTokenWithFee,
                functionName: 'sweepTokenWithFee',
                args: [token.address, amountMinimum, recipient, feeBips, feeRecipient],
            });
        }
        else {
            return (0, viem_1.encodeFunctionData)({
                abi: index_js_1.peripheryPaymentsWithFeeAbi_sweepToken,
                functionName: 'sweepToken',
                args: [token.address, amountMinimum, recipient],
            });
        }
    }
    static encodeRefundETH() {
        return (0, viem_1.encodeFunctionData)({
            abi: index_js_1.peripheryPaymentsWithFeeAbi_refundETH,
            functionName: 'refundETH',
        });
    }
}
exports.Payments = Payments;
