// sushi/_cjs/evm/pool/sushiswap-v3/entities/Position.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Position = void 0;
const tiny_invariant_1 = __importDefault(require("tiny-invariant"));
const amount_js_1 = require("../../../../generic/currency/amount.js");
const constants_js_1 = require("../../../../generic/math/constants.js");
const percent_js_1 = require("../../../../generic/math/percent.js");
const encodeSqrtRatioX96_js_1 = require("../utils/encodeSqrtRatioX96.js");
const maxLiquidityForAmounts_js_1 = require("../utils/maxLiquidityForAmounts.js");
const priceTickConversions_js_1 = require("../utils/priceTickConversions.js");
const sqrtPriceMath_js_1 = require("../utils/sqrtPriceMath.js");
const tickMath_js_1 = require("../utils/tickMath.js");
const SushiSwapV3Pool_js_1 = require("./SushiSwapV3Pool.js");
class Position {
    constructor({ pool, liquidity, tickLower, tickUpper, }) {
        this._token0Amount = null;
        this._token1Amount = null;
        this._mintAmounts = null;
        (0, tiny_invariant_1.default)(tickLower < tickUpper, 'TICK_ORDER');
        (0, tiny_invariant_1.default)(tickLower >= tickMath_js_1.TickMath.MIN_TICK && tickLower % pool.tickSpacing === 0, 'TICK_LOWER');
        (0, tiny_invariant_1.default)(tickUpper <= tickMath_js_1.TickMath.MAX_TICK && tickUpper % pool.tickSpacing === 0, 'TICK_UPPER');
        this.pool = pool;
        this.tickLower = tickLower;
        this.tickUpper = tickUpper;
        this.liquidity = BigInt(liquidity.toString());
    }
    get token0PriceLower() {
        return (0, priceTickConversions_js_1.tickToPrice)(this.pool.token0, this.pool.token1, this.tickLower);
    }
    get token0PriceUpper() {
        return (0, priceTickConversions_js_1.tickToPrice)(this.pool.token0, this.pool.token1, this.tickUpper);
    }
    get amount0() {
        if (this._token0Amount === null) {
            if (this.pool.tickCurrent < this.tickLower) {
                this._token0Amount = new amount_js_1.Amount(this.pool.token0, sqrtPriceMath_js_1.SqrtPriceMath.getAmount0Delta(tickMath_js_1.TickMath.getSqrtRatioAtTick(this.tickLower), tickMath_js_1.TickMath.getSqrtRatioAtTick(this.tickUpper), this.liquidity, false));
            }
            else if (this.pool.tickCurrent < this.tickUpper) {
                this._token0Amount = new amount_js_1.Amount(this.pool.token0, sqrtPriceMath_js_1.SqrtPriceMath.getAmount0Delta(this.pool.sqrtRatioX96, tickMath_js_1.TickMath.getSqrtRatioAtTick(this.tickUpper), this.liquidity, false));
            }
            else {
                this._token0Amount = new amount_js_1.Amount(this.pool.token0, 0n);
            }
        }
        return this._token0Amount;
    }
    get amount1() {
        if (this._token1Amount === null) {
            if (this.pool.tickCurrent < this.tickLower) {
                this._token1Amount = new amount_js_1.Amount(this.pool.token1, 0n);
            }
            else if (this.pool.tickCurrent < this.tickUpper) {
                this._token1Amount = new amount_js_1.Amount(this.pool.token1, sqrtPriceMath_js_1.SqrtPriceMath.getAmount1Delta(tickMath_js_1.TickMath.getSqrtRatioAtTick(this.tickLower), this.pool.sqrtRatioX96, this.liquidity, false));
            }
            else {
                this._token1Amount = new amount_js_1.Amount(this.pool.token1, sqrtPriceMath_js_1.SqrtPriceMath.getAmount1Delta(tickMath_js_1.TickMath.getSqrtRatioAtTick(this.tickLower), tickMath_js_1.TickMath.getSqrtRatioAtTick(this.tickUpper), this.liquidity, false));
            }
        }
        return this._token1Amount;
    }
    ratiosAfterSlippage(slippageTolerance) {
        const priceLower = this.pool.token0Price.asFraction.mul(new percent_js_1.Percent(1).sub(slippageTolerance));
        const priceUpper = this.pool.token0Price.asFraction.mul(slippageTolerance.add(1));
        let sqrtRatioX96Lower = (0, encodeSqrtRatioX96_js_1.encodeSqrtRatioX96)(priceLower.numerator, priceLower.denominator);
        if (sqrtRatioX96Lower <= tickMath_js_1.TickMath.MIN_SQRT_RATIO) {
            sqrtRatioX96Lower = tickMath_js_1.TickMath.MIN_SQRT_RATIO + 1n;
        }
        let sqrtRatioX96Upper = (0, encodeSqrtRatioX96_js_1.encodeSqrtRatioX96)(priceUpper.numerator, priceUpper.denominator);
        if (sqrtRatioX96Upper >= tickMath_js_1.TickMath.MAX_SQRT_RATIO) {
            sqrtRatioX96Upper = tickMath_js_1.TickMath.MAX_SQRT_RATIO - 1n;
        }
        return {
            sqrtRatioX96Lower,
            sqrtRatioX96Upper,
        };
    }
    mintAmountsWithSlippage(slippageTolerance) {
        const { sqrtRatioX96Upper, sqrtRatioX96Lower } = this.ratiosAfterSlippage(slippageTolerance);
        const poolLower = new SushiSwapV3Pool_js_1.SushiSwapV3Pool(this.pool.token0, this.pool.token1, this.pool.fee, sqrtRatioX96Lower, 0, tickMath_js_1.TickMath.getTickAtSqrtRatio(sqrtRatioX96Lower));
        const poolUpper = new SushiSwapV3Pool_js_1.SushiSwapV3Pool(this.pool.token0, this.pool.token1, this.pool.fee, sqrtRatioX96Upper, 0, tickMath_js_1.TickMath.getTickAtSqrtRatio(sqrtRatioX96Upper));
        const positionThatWillBeCreated = Position.fromAmounts({
            pool: this.pool,
            tickLower: this.tickLower,
            tickUpper: this.tickUpper,
            ...this.mintAmounts,
            useFullPrecision: false,
        });
        const { amount0 } = new Position({
            pool: poolUpper,
            liquidity: positionThatWillBeCreated.liquidity,
            tickLower: this.tickLower,
            tickUpper: this.tickUpper,
        }).mintAmounts;
        const { amount1 } = new Position({
            pool: poolLower,
            liquidity: positionThatWillBeCreated.liquidity,
            tickLower: this.tickLower,
            tickUpper: this.tickUpper,
        }).mintAmounts;
        return { amount0, amount1 };
    }
    burnAmountsWithSlippage(slippageTolerance) {
        const { sqrtRatioX96Upper, sqrtRatioX96Lower } = this.ratiosAfterSlippage(slippageTolerance);
        const poolLower = new SushiSwapV3Pool_js_1.SushiSwapV3Pool(this.pool.token0, this.pool.token1, this.pool.fee, sqrtRatioX96Lower, 0, tickMath_js_1.TickMath.getTickAtSqrtRatio(sqrtRatioX96Lower));
        const poolUpper = new SushiSwapV3Pool_js_1.SushiSwapV3Pool(this.pool.token0, this.pool.token1, this.pool.fee, sqrtRatioX96Upper, 0, tickMath_js_1.TickMath.getTickAtSqrtRatio(sqrtRatioX96Upper));
        const amount0 = new Position({
            pool: poolUpper,
            liquidity: this.liquidity,
            tickLower: this.tickLower,
            tickUpper: this.tickUpper,
        }).amount0;
        const amount1 = new Position({
            pool: poolLower,
            liquidity: this.liquidity,
            tickLower: this.tickLower,
            tickUpper: this.tickUpper,
        }).amount1;
        return { amount0: amount0.amount, amount1: amount1.amount };
    }
    get mintAmounts() {
        if (this._mintAmounts === null) {
            if (this.pool.tickCurrent < this.tickLower) {
                return {
                    amount0: sqrtPriceMath_js_1.SqrtPriceMath.getAmount0Delta(tickMath_js_1.TickMath.getSqrtRatioAtTick(this.tickLower), tickMath_js_1.TickMath.getSqrtRatioAtTick(this.tickUpper), this.liquidity, true),
                    amount1: 0n,
                };
            }
            else if (this.pool.tickCurrent < this.tickUpper) {
                return {
                    amount0: sqrtPriceMath_js_1.SqrtPriceMath.getAmount0Delta(this.pool.sqrtRatioX96, tickMath_js_1.TickMath.getSqrtRatioAtTick(this.tickUpper), this.liquidity, true),
                    amount1: sqrtPriceMath_js_1.SqrtPriceMath.getAmount1Delta(tickMath_js_1.TickMath.getSqrtRatioAtTick(this.tickLower), this.pool.sqrtRatioX96, this.liquidity, true),
                };
            }
            else {
                return {
                    amount0: 0n,
                    amount1: sqrtPriceMath_js_1.SqrtPriceMath.getAmount1Delta(tickMath_js_1.TickMath.getSqrtRatioAtTick(this.tickLower), tickMath_js_1.TickMath.getSqrtRatioAtTick(this.tickUpper), this.liquidity, true),
                };
            }
        }
        return this._mintAmounts;
    }
    static fromAmounts({ pool, tickLower, tickUpper, amount0, amount1, useFullPrecision, }) {
        const sqrtRatioAX96 = tickMath_js_1.TickMath.getSqrtRatioAtTick(tickLower);
        const sqrtRatioBX96 = tickMath_js_1.TickMath.getSqrtRatioAtTick(tickUpper);
        return new Position({
            pool,
            tickLower,
            tickUpper,
            liquidity: (0, maxLiquidityForAmounts_js_1.maxLiquidityForAmounts)(pool.sqrtRatioX96, sqrtRatioAX96, sqrtRatioBX96, amount0, amount1, useFullPrecision),
        });
    }
    static fromAmount0({ pool, tickLower, tickUpper, amount0, useFullPrecision, }) {
        return Position.fromAmounts({
            pool,
            tickLower,
            tickUpper,
            amount0,
            amount1: constants_js_1.MAX_UINT256,
            useFullPrecision,
        });
    }
    static fromAmount1({ pool, tickLower, tickUpper, amount1, }) {
        return Position.fromAmounts({
            pool,
            tickLower,
            tickUpper,
            amount0: constants_js_1.MAX_UINT256,
            amount1,
            useFullPrecision: true,
        });
    }
}
exports.Position = Position;
