// sushi/_cjs/evm/pool/sushiswap-v3/entities/SelfPermit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelfPermit = void 0;
const viem_1 = require("viem");
const index_js_1 = require("../../../abi/selfPermitAbi/index.js");
function isAllowedPermit(permitOptions) {
    return 'nonce' in permitOptions;
}
class SelfPermit {
    constructor() { }
    static encodePermit(token, options) {
        if (isAllowedPermit(options)) {
            return (0, viem_1.encodeFunctionData)({
                abi: index_js_1.selfPermitAbi_selfPermitAllowed,
                functionName: 'selfPermitAllowed',
                args: [
                    token.address,
                    BigInt(options.nonce),
                    BigInt(options.expiry),
                    options.v,
                    options.r,
                    options.s,
                ],
            });
        }
        return (0, viem_1.encodeFunctionData)({
            abi: index_js_1.selfPermitAbi_selfPermit,
            functionName: 'selfPermit',
            args: [
                token.address,
                BigInt(options.amount),
                BigInt(options.deadline),
                options.v,
                options.r,
                options.s,
            ],
        });
    }
}
exports.SelfPermit = SelfPermit;
