// sushi/_cjs/evm/pool/sushiswap-v3/entities/SushiSwapV3Pool.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SushiSwapV3Pool = void 0;
const tiny_invariant_1 = __importDefault(require("tiny-invariant"));
const amount_js_1 = require("../../../../generic/currency/amount.js");
const price_js_1 = require("../../../../generic/currency/price.js");
const index_js_1 = require("../../../config/index.js");
const internalConstants_js_1 = require("../internalConstants.js");
const computePoolAddress_js_1 = require("../utils/computePoolAddress.js");
const liquidityMath_js_1 = require("../utils/liquidityMath.js");
const swapMath_js_1 = require("../utils/swapMath.js");
const tickMath_js_1 = require("../utils/tickMath.js");
const TickDataProvider_js_1 = require("./TickDataProvider.js");
const TickListDataProvider_js_1 = require("./TickListDataProvider.js");
const NO_TICK_DATA_PROVIDER_DEFAULT = new TickDataProvider_js_1.NoTickDataProvider();
class SushiSwapV3Pool {
    static getAddress(tokenA, tokenB, fee, initCodeHashManualOverride, factoryAddressOverride) {
        return (0, computePoolAddress_js_1.computeSushiSwapV3PoolAddress)({
            factoryAddress: factoryAddressOverride ??
                index_js_1.SUSHISWAP_V3_FACTORY_ADDRESS[tokenA.chainId],
            fee,
            tokenA,
            tokenB,
            initCodeHashManualOverride,
        });
    }
    constructor(tokenA, tokenB, fee, sqrtRatioX96, liquidity, tickCurrent, ticks = NO_TICK_DATA_PROVIDER_DEFAULT) {
        (0, tiny_invariant_1.default)(Number.isInteger(fee) && fee < 1_000_000, 'FEE');
        const tickCurrentSqrtRatioX96 = tickMath_js_1.TickMath.getSqrtRatioAtTick(tickCurrent);
        const nextTickSqrtRatioX96 = tickMath_js_1.TickMath.getSqrtRatioAtTick(tickCurrent + 1);
        (0, tiny_invariant_1.default)(BigInt(sqrtRatioX96.toString()) >= tickCurrentSqrtRatioX96 &&
            BigInt(sqrtRatioX96.toString()) <= nextTickSqrtRatioX96, 'PRICE_BOUNDS');
        [this.token0, this.token1] = tokenA.sortsBefore(tokenB)
            ? [tokenA, tokenB]
            : [tokenB, tokenA];
        this.fee = fee;
        this.sqrtRatioX96 = BigInt(sqrtRatioX96.toString());
        this.liquidity = BigInt(liquidity.toString());
        this.tickCurrent = tickCurrent;
        this.tickDataProvider = Array.isArray(ticks)
            ? new TickListDataProvider_js_1.TickListDataProvider(ticks, index_js_1.TICK_SPACINGS[fee])
            : ticks;
    }
    involvesToken(token) {
        return token.isSame(this.token0) || token.isSame(this.token1);
    }
    get token0Price() {
        if (!this._token0Price) {
            this._token0Price = new price_js_1.Price({
                base: this.token0,
                quote: this.token1,
                denominator: internalConstants_js_1.Q192,
                numerator: this.sqrtRatioX96 * this.sqrtRatioX96,
            });
        }
        return this._token0Price;
    }
    get token1Price() {
        if (!this._token1Price) {
            this._token1Price = new price_js_1.Price({
                base: this.token1,
                quote: this.token0,
                denominator: this.sqrtRatioX96 * this.sqrtRatioX96,
                numerator: internalConstants_js_1.Q192,
            });
        }
        return this._token1Price;
    }
    priceOf(token) {
        (0, tiny_invariant_1.default)(this.involvesToken(token), 'TOKEN');
        return token.isSame(this.token0) ? this.token0Price : this.token1Price;
    }
    get chainId() {
        return this.token0.chainId;
    }
    async getOutputAmount(inputAmount, sqrtPriceLimitX96) {
        (0, tiny_invariant_1.default)(this.involvesToken(inputAmount.currency), 'TOKEN');
        const zeroForOne = inputAmount.currency.isSame(this.token0);
        const { amountCalculated: outputAmount, sqrtRatioX96, liquidity, tickCurrent, } = await this.swap(zeroForOne, inputAmount.amount, sqrtPriceLimitX96);
        const outputToken = zeroForOne ? this.token1 : this.token0;
        return [
            new amount_js_1.Amount(outputToken, outputAmount * -1n),
            new SushiSwapV3Pool(this.token0, this.token1, this.fee, sqrtRatioX96, liquidity, tickCurrent, this.tickDataProvider),
        ];
    }
    async getInputAmount(outputAmount, sqrtPriceLimitX96) {
        (0, tiny_invariant_1.default)(outputAmount.currency.type === 'token' &&
            this.involvesToken(outputAmount.currency), 'TOKEN');
        const zeroForOne = outputAmount.currency.isSame(this.token1);
        const { amountCalculated: inputAmount, sqrtRatioX96, liquidity, tickCurrent, } = await this.swap(zeroForOne, outputAmount.amount * -1n, sqrtPriceLimitX96);
        const inputToken = zeroForOne ? this.token0 : this.token1;
        return [
            new amount_js_1.Amount(inputToken, inputAmount),
            new SushiSwapV3Pool(this.token0, this.token1, this.fee, sqrtRatioX96, liquidity, tickCurrent, this.tickDataProvider),
        ];
    }
    async swap(zeroForOne, amountSpecified, sqrtPriceLimitX96) {
        if (!sqrtPriceLimitX96)
            sqrtPriceLimitX96 = zeroForOne
                ? tickMath_js_1.TickMath.MIN_SQRT_RATIO + 1n
                : tickMath_js_1.TickMath.MAX_SQRT_RATIO - 1n;
        if (zeroForOne) {
            (0, tiny_invariant_1.default)(sqrtPriceLimitX96 > tickMath_js_1.TickMath.MIN_SQRT_RATIO, 'RATIO_MIN');
            (0, tiny_invariant_1.default)(sqrtPriceLimitX96 < this.sqrtRatioX96, 'RATIO_CURRENT');
        }
        else {
            (0, tiny_invariant_1.default)(sqrtPriceLimitX96 < tickMath_js_1.TickMath.MAX_SQRT_RATIO, 'RATIO_MAX');
            (0, tiny_invariant_1.default)(sqrtPriceLimitX96 > this.sqrtRatioX96, 'RATIO_CURRENT');
        }
        const exactInput = amountSpecified >= 0n;
        const state = {
            amountSpecifiedRemaining: amountSpecified,
            amountCalculated: 0n,
            sqrtPriceX96: this.sqrtRatioX96,
            tick: this.tickCurrent,
            liquidity: this.liquidity,
        };
        while (state.amountSpecifiedRemaining !== 0n &&
            state.sqrtPriceX96 !== sqrtPriceLimitX96) {
            const step = {};
            step.sqrtPriceStartX96 = state.sqrtPriceX96;
            [step.tickNext, step.initialized] =
                await this.tickDataProvider.nextInitializedTickWithinOneWord(state.tick, zeroForOne, this.tickSpacing);
            if (step.tickNext < tickMath_js_1.TickMath.MIN_TICK) {
                step.tickNext = tickMath_js_1.TickMath.MIN_TICK;
            }
            else if (step.tickNext > tickMath_js_1.TickMath.MAX_TICK) {
                step.tickNext = tickMath_js_1.TickMath.MAX_TICK;
            }
            step.sqrtPriceNextX96 = tickMath_js_1.TickMath.getSqrtRatioAtTick(step.tickNext);
            [state.sqrtPriceX96, step.amountIn, step.amountOut, step.feeAmount] =
                swapMath_js_1.SwapMath.computeSwapStep(state.sqrtPriceX96, (zeroForOne
                    ? step.sqrtPriceNextX96 < sqrtPriceLimitX96
                    : step.sqrtPriceNextX96 > sqrtPriceLimitX96)
                    ? sqrtPriceLimitX96
                    : step.sqrtPriceNextX96, state.liquidity, state.amountSpecifiedRemaining, this.fee);
            if (exactInput) {
                state.amountSpecifiedRemaining =
                    state.amountSpecifiedRemaining - (step.amountIn + step.feeAmount);
                state.amountCalculated = state.amountCalculated - step.amountOut;
            }
            else {
                state.amountSpecifiedRemaining =
                    state.amountSpecifiedRemaining + step.amountOut;
                state.amountCalculated =
                    state.amountCalculated + (step.amountIn + step.feeAmount);
            }
            if (state.sqrtPriceX96 === step.sqrtPriceNextX96) {
                if (step.initialized) {
                    let liquidityNet = BigInt((await this.tickDataProvider.getTick(step.tickNext)).liquidityNet.toString());
                    if (zeroForOne)
                        liquidityNet = liquidityNet * -1n;
                    state.liquidity = liquidityMath_js_1.LiquidityMath.addDelta(state.liquidity, liquidityNet);
                }
                state.tick = zeroForOne ? step.tickNext - 1 : step.tickNext;
            }
            else if (state.sqrtPriceX96 !== step.sqrtPriceStartX96) {
                state.tick = tickMath_js_1.TickMath.getTickAtSqrtRatio(state.sqrtPriceX96);
            }
        }
        return {
            amountCalculated: state.amountCalculated,
            sqrtRatioX96: state.sqrtPriceX96,
            liquidity: state.liquidity,
            tickCurrent: state.tick,
        };
    }
    get tickSpacing() {
        return index_js_1.TICK_SPACINGS[this.fee];
    }
}
exports.SushiSwapV3Pool = SushiSwapV3Pool;
