// sushi/_cjs/tvm/chain/chains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TvmChainId = exports.tvmChainIds = exports.tvmChains = void 0;
exports.isTvmChainId = isTvmChainId;
exports.getTvmChainById = getTvmChainById;
exports.isTvmMainnetChainId = isTvmMainnetChainId;
exports.isTvmChainKey = isTvmChainKey;
exports.getTvmChainByKey = getTvmChainByKey;
const define_chain_js_1 = require("./define-chain.js");
const tron = (0, define_chain_js_1.defineTvmChain)({
    chainId: -2,
    name: 'Tron',
    shortName: 'trx',
    key: 'tron',
    blockExplorers: {
        default: { name: 'Tronscan', url: 'https://tronscan.io' },
    },
    netType: 'mainnet',
});
exports.tvmChains = [tron];
exports.tvmChainIds = exports.tvmChains.map((chain) => chain.chainId);
exports.TvmChainId = new Proxy({}, {
    get: (_, key) => {
        const _key = key.toLowerCase().replace(/_/g, '-');
        const chain = exports.tvmChains.find((chain) => chain.key === _key);
        if (!chain)
            return undefined;
        return chain.chainId;
    },
});
function isTvmChainId(chainId) {
    return exports.tvmChains.some((chain) => chain.chainId === chainId);
}
function getTvmChainById(chainId) {
    return exports.tvmChains.find((chain) => chain.chainId === chainId);
}
function isTvmMainnetChainId(chainId) {
    return exports.tvmChains.some((chain) => chain.chainId === chainId && chain.netType === 'mainnet');
}
function isTvmChainKey(key) {
    return exports.tvmChains.some((chain) => chain.key === key);
}
function getTvmChainByKey(key) {
    return exports.tvmChains.find((chain) => chain.key === key);
}
