// @coinbase/onchainkit/dist/wallet/components/ConnectWallet.js
'use client';
import { jsx, jsxs, Fragment } from "react/jsx-runtime";
import { useContext, useState, useMemo, useCallback, useEffect } from "react";
import { cn, pressable, text } from "../../styles/theme.js";
import { IdentityProvider } from "../../identity/components/IdentityProvider.js";
import { Avatar } from "../../identity/components/Avatar.js";
import "../../core/network/attestations.js";
import "../../identity/utils/easSupportedChains.js";
import "@tanstack/react-query";
import "../../identity/constants.js";
import "../../styles/constants.js";
import { useOnchainKit } from "../../useOnchainKit.js";
import "viem";
import { useAccount, useConnect } from "wagmi";
import { Name } from "../../identity/components/Name.js";
import "viem/chains";
import "viem/ens";
import { useAnalytics } from "../../core/analytics/hooks/useAnalytics.js";
import { WalletEvent } from "../../core/analytics/types.js";
import { Spinner } from "../../internal/components/Spinner.js";
import { WalletModal } from "./WalletModal.js";
import { WalletContext, WalletProvider, useWalletContext } from "./WalletProvider.js";
import { MiniKitContext } from "../../minikit/MiniKitProvider.js";
import { Button } from "../../ui/Button.js";
import { IfInMiniApp } from "../../minikit/components/IfInMiniApp.js";
import "@farcaster/miniapp-sdk";
import { useMiniKit } from "../../minikit/hooks/useMiniKit.js";
function MiniAppDefaultChildren() {
  const {
    context
  } = useMiniKit();
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Avatar, { className: "ock:h-6 ock:w-6", name: context == null ? void 0 : context.user.displayName, avatar: context == null ? void 0 : context.user.pfpUrl }),
    /* @__PURE__ */ jsx(Name, { name: context == null ? void 0 : context.user.displayName })
  ] });
}
function ConnectWalletContent({
  children = /* @__PURE__ */ jsx(IfInMiniApp, { fallback: /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Avatar, { className: "ock:h-6 ock:w-6" }),
    /* @__PURE__ */ jsx(Name, {})
  ] }), children: /* @__PURE__ */ jsx(MiniAppDefaultChildren, {}) }),
  className,
  onConnect,
  disconnectedLabel = "Connect Wallet",
  render
}) {
  var _a;
  const {
    config = {
      wallet: {
        display: void 0
      }
    }
  } = useOnchainKit();
  const walletContext = useWalletContext();
  const {
    isConnectModalOpen,
    setIsConnectModalOpen,
    isSubComponentOpen,
    setIsSubComponentOpen,
    handleClose
  } = walletContext;
  const {
    address: accountAddress,
    status,
    connector: accountConnector
  } = useAccount();
  const {
    connectors,
    connect,
    status: connectStatus
  } = useConnect();
  const {
    sendAnalytics
  } = useAnalytics();
  const [hasClickedConnect, setHasClickedConnect] = useState(false);
  const miniKitContext = useContext(MiniKitContext);
  const isWalletModalEnabled = useMemo(() => {
    var _a2;
    return ((_a2 = config == null ? void 0 : config.wallet) == null ? void 0 : _a2.display) === "modal" && !(miniKitContext == null ? void 0 : miniKitContext.context);
  }, [(_a = config == null ? void 0 : config.wallet) == null ? void 0 : _a.display, miniKitContext == null ? void 0 : miniKitContext.context]);
  const connector = accountConnector || connectors[0];
  const isLoading = connectStatus === "pending" || status === "connecting";
  const handleToggle = useCallback(() => {
    if (isSubComponentOpen) {
      handleClose();
    } else {
      setIsSubComponentOpen(true);
    }
  }, [isSubComponentOpen, handleClose, setIsSubComponentOpen]);
  const handleCloseConnectModal = useCallback(() => {
    setIsConnectModalOpen(false);
  }, [setIsConnectModalOpen]);
  const handleOpenConnectModal = useCallback(() => {
    setIsConnectModalOpen(true);
    setHasClickedConnect(true);
  }, [setIsConnectModalOpen]);
  const handleAnalyticsSuccess = useCallback(() => {
    if (!accountAddress || !connector) return;
    sendAnalytics(WalletEvent.ConnectSuccess, {
      address: accountAddress,
      walletProvider: connector == null ? void 0 : connector.name
    });
  }, [sendAnalytics, connector, accountAddress]);
  const handleAnalyticsError = useCallback((errorMessage, component) => {
    const walletProvider = connector == null ? void 0 : connector.name;
    sendAnalytics(WalletEvent.ConnectError, {
      error: errorMessage,
      metadata: {
        connector: walletProvider,
        component
      }
    });
  }, [sendAnalytics, connector]);
  useEffect(() => {
    if (status !== "connected") return;
    if (hasClickedConnect && onConnect) {
      onConnect();
      setHasClickedConnect(false);
    }
    handleAnalyticsSuccess();
  }, [status, hasClickedConnect, onConnect, accountAddress, connector, handleAnalyticsSuccess]);
  const handleConnectClick = useCallback(() => {
    if (isWalletModalEnabled) {
      handleOpenConnectModal();
      setHasClickedConnect(true);
      sendAnalytics(WalletEvent.ConnectInitiated, {
        component: "WalletModal"
      });
      return;
    }
    sendAnalytics(WalletEvent.ConnectInitiated, {
      component: "ConnectWallet"
    });
    connect({
      connector
    }, {
      onSuccess: () => {
        onConnect == null ? void 0 : onConnect();
        handleAnalyticsSuccess();
      },
      onError: (error) => {
        handleAnalyticsError(error.message, "ConnectWallet");
      }
    });
  }, [isWalletModalEnabled, connect, connector, handleAnalyticsError, handleAnalyticsSuccess, handleOpenConnectModal, onConnect, sendAnalytics]);
  const buttonContent = useMemo(() => {
    if (isLoading) return /* @__PURE__ */ jsx(Spinner, {});
    if (status === "disconnected") return disconnectedLabel;
    return /* @__PURE__ */ jsx("div", { className: "ock:flex ock:items-center ock:justify-center ock:gap-2", children });
  }, [isLoading, status, disconnectedLabel, children]);
  if (render) {
    return /* @__PURE__ */ jsx(ConnectWalletRenderHandler, { label: buttonContent, onClick: handleConnectClick, isLoading, render, isWalletModalEnabled });
  }
  if (status === "disconnected") {
    return /* @__PURE__ */ jsxs("div", { className: "ock:flex", "data-testid": "ockConnectWallet_Container", children: [
      /* @__PURE__ */ jsx(Button, { type: "button", "data-testid": "ockConnectButton", className: cn("ock:inline-flex ock:min-w-[153px]", className), onClick: handleConnectClick, children: buttonContent }),
      isWalletModalEnabled && /* @__PURE__ */ jsx(WalletModal, { isOpen: isConnectModalOpen, onClose: handleCloseConnectModal })
    ] });
  }
  if (isLoading) {
    return /* @__PURE__ */ jsx("div", { className: "ock:flex", "data-testid": "ockConnectWallet_Container", children: /* @__PURE__ */ jsx("button", { type: "button", "data-testid": "ockConnectAccountButtonInner", className: cn(pressable.primary, text.headline, "ock:text-ock-foreground-inverse", "ock:inline-flex ock:min-w-[153px] ock:items-center ock:justify-center ock:rounded-xl ock:px-4 ock:py-3", pressable.disabled, className), disabled: true, children: buttonContent }) });
  }
  return /* @__PURE__ */ jsx(IdentityProvider, { address: accountAddress, children: /* @__PURE__ */ jsx("div", { className: "ock:flex ock:gap-4", "data-testid": "ockConnectWallet_Container", children: /* @__PURE__ */ jsx("button", { type: "button", "data-testid": "ockConnectWallet_Connected", className: cn(pressable.secondary, "ock:rounded-ock-default", "ock:text-ock-foreground", "ock:px-4 ock:py-3", isSubComponentOpen && "ock:bg-ock-secondary-active ock:hover:bg-ock-secondary-active", className), onClick: handleToggle, children: /* @__PURE__ */ jsx("div", { className: "ock:flex ock:items-center ock:justify-center ock:gap-2", children }) }) }) });
}
function ConnectWalletRenderHandler({
  label,
  onClick,
  isLoading,
  render,
  isWalletModalEnabled
}) {
  const walletContext = useWalletContext();
  const {
    isConnectModalOpen,
    setIsConnectModalOpen
  } = walletContext;
  const {
    status,
    address
  } = useAccount();
  if (status === "disconnected") {
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      render({
        label,
        onClick,
        context: walletContext,
        status: "disconnected",
        isLoading
      }),
      isWalletModalEnabled && /* @__PURE__ */ jsx(WalletModal, { isOpen: isConnectModalOpen, onClose: () => setIsConnectModalOpen(false) })
    ] });
  }
  if (isLoading) {
    return render({
      label: /* @__PURE__ */ jsx(Spinner, {}),
      onClick,
      context: walletContext,
      status: "connecting",
      isLoading
    });
  }
  return /* @__PURE__ */ jsx(IdentityProvider, { address, children: render({
    label,
    onClick,
    context: walletContext,
    status: "connected",
    isLoading
  }) });
}
function ConnectWallet(props) {
  const walletContext = useContext(WalletContext);
  if (!walletContext) {
    return /* @__PURE__ */ jsx(WalletProvider, { children: /* @__PURE__ */ jsx(ConnectWalletContent, { ...props }) });
  }
  return /* @__PURE__ */ jsx(ConnectWalletContent, { ...props });
}
export {
  ConnectWallet
};
