// @coinbase/onchainkit/dist/wallet/components/WalletAdvancedTokenHoldings.js
'use client';
import { jsx, jsxs } from "react/jsx-runtime";
import { cn, text } from "../../styles/theme.js";
import { TokenImage } from "../../token/components/TokenImage.js";
import "../../token/components/TokenRow.js";
import "react";
import "../../styles/constants.js";
import "@radix-ui/react-dropdown-menu";
import "../../internal/components/LayerConfigProvider.js";
import "../../token/components/TokenSelectButton.js";
import { formatUnits } from "viem";
import { useWalletContext } from "./WalletProvider.js";
import { usePortfolio } from "../hooks/usePortfolio.js";
import { RequestContext } from "../../core/network/constants.js";
import { useAccount } from "wagmi";
function WalletAdvancedTokenHoldings({
  classNames
}) {
  const {
    address
  } = useAccount();
  const {
    animations
  } = useWalletContext();
  const {
    data: portfolioData,
    isFetching: isFetchingPortfolioData
  } = usePortfolio({
    address
  }, RequestContext.Wallet);
  const tokenBalances = portfolioData == null ? void 0 : portfolioData.tokenBalances;
  if (isFetchingPortfolioData || !tokenBalances || tokenBalances.length === 0) {
    return /* @__PURE__ */ jsx("div", { "data-testid": "ockWalletAdvanced_LoadingPlaceholder", className: "ock:mx-auto ock:my-2 ock:h-44 ock:w-80 ock:px-4 ock:py-3" });
  }
  return /* @__PURE__ */ jsx("div", { className: cn("ock:flex ock:flex-col ock:items-center ock:gap-4", "ock:my-2 ock:h-44 ock:max-h-44", "ock:scrollbar-hidden ock:w-88 ock:overflow-y-auto ock:px-4 ock:py-3", animations.content, classNames == null ? void 0 : classNames.container), "data-testid": "ockWalletAdvanced_TokenHoldings", children: tokenBalances.map((tokenBalance, index) => /* @__PURE__ */ jsx(TokenDetails, { token: {
    address: tokenBalance.address,
    chainId: tokenBalance.chainId,
    decimals: tokenBalance.decimals,
    image: tokenBalance.image,
    name: tokenBalance.name,
    symbol: tokenBalance.symbol
  }, balance: Number(formatUnits(BigInt(tokenBalance.cryptoBalance), tokenBalance.decimals)), valueInFiat: Number(tokenBalance.fiatBalance), classNames: classNames == null ? void 0 : classNames.tokenDetails }, `${tokenBalance.address}-${index}`)) });
}
function TokenDetails({
  token,
  balance,
  valueInFiat,
  classNames,
  tokenImageSize = 32
}) {
  var _a;
  const formattedBalance = new Intl.NumberFormat("en-US", {
    minimumFractionDigits: 2,
    maximumFractionDigits: 5
  }).format(balance);
  const formattedValueInFiat = new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD"
  }).format(valueInFiat);
  return /* @__PURE__ */ jsxs("div", { className: cn("ock:flex ock:w-full ock:flex-row ock:items-center ock:justify-between", classNames == null ? void 0 : classNames.container), children: [
    /* @__PURE__ */ jsxs("div", { "data-testid": "ockWalletAdvanced_TokenDetails_TokenImage", className: cn("ock:flex ock:flex-row ock:items-center ock:gap-2", classNames == null ? void 0 : classNames.tokenImage), children: [
      /* @__PURE__ */ jsx(TokenImage, { token, size: tokenImageSize }),
      /* @__PURE__ */ jsxs("div", { className: "ock:flex ock:flex-col", children: [
        /* @__PURE__ */ jsx("span", { className: cn(text.label1, "ock:text-ock-foreground", "ock:max-w-52 ock:overflow-hidden ock:text-ellipsis ock:whitespace-nowrap ock:text-left", classNames == null ? void 0 : classNames.tokenName), children: (_a = token.name) == null ? void 0 : _a.trim() }),
        /* @__PURE__ */ jsx("span", { className: cn(text.legal, "ock:text-ock-foreground-muted", classNames == null ? void 0 : classNames.tokenBalance), children: `${formattedBalance} ${token.symbol}` })
      ] })
    ] }),
    /* @__PURE__ */ jsx("span", { className: cn(text.label2, "ock:text-ock-foreground-muted", classNames == null ? void 0 : classNames.fiatValue), children: formattedValueInFiat })
  ] });
}
export {
  WalletAdvancedTokenHoldings
};
