// @coinbase/onchainkit/dist/wallet/components/WalletDropdownContent.js
import { jsx } from "react/jsx-runtime";
import { Sheet } from "../../internal/components/Sheet.js";
import { zIndex } from "../../styles/constants.js";
import { cn, text, border } from "../../styles/theme.js";
import { useCallback, useMemo } from "react";
import { WALLET_ADVANCED_DEFAULT_SWAPPABLE_TOKENS } from "../constants.js";
import { WalletAdvancedQrReceive } from "./WalletAdvancedQrReceive.js";
import { WalletAdvancedSwap } from "./WalletAdvancedSwap.js";
import { useWalletContext } from "./WalletProvider.js";
import { Send } from "./wallet-advanced-send/components/Send.js";
import { RequestContext } from "../../core/network/constants.js";
import { usePortfolio } from "../hooks/usePortfolio.js";
import { useAccount } from "wagmi";
function WalletDropdownContent({
  children,
  swappableTokens,
  classNames
}) {
  const {
    isSubComponentOpen,
    setIsSubComponentOpen,
    isSubComponentClosing,
    setIsSubComponentClosing,
    breakpoint,
    activeFeature,
    animations
  } = useWalletContext();
  const {
    address
  } = useAccount();
  const {
    data: portfolioData
  } = usePortfolio({
    address,
    enabled: Boolean(activeFeature)
  }, RequestContext.Wallet);
  const tokenBalances = portfolioData == null ? void 0 : portfolioData.tokenBalances;
  const handleBottomSheetClose = useCallback(() => {
    setIsSubComponentOpen(false);
  }, [setIsSubComponentOpen]);
  const handleAnimationEnd = useCallback(() => {
    if (isSubComponentClosing) {
      setIsSubComponentOpen(false);
      setIsSubComponentClosing(false);
    }
  }, [isSubComponentClosing, setIsSubComponentOpen, setIsSubComponentClosing]);
  const content = useMemo(() => {
    if (activeFeature === "send") {
      return /* @__PURE__ */ jsx(ContentWrapper, { children: /* @__PURE__ */ jsx(Send, { className: "ock:border-none" }) });
    }
    if (activeFeature === "qr") {
      return /* @__PURE__ */ jsx(ContentWrapper, { children: /* @__PURE__ */ jsx(WalletAdvancedQrReceive, { classNames: classNames == null ? void 0 : classNames.qr }) });
    }
    if (activeFeature === "swap") {
      return /* @__PURE__ */ jsx(ContentWrapper, { children: /* @__PURE__ */ jsx(WalletAdvancedSwap, { title: /* @__PURE__ */ jsx("div", { className: cn(text.headline, "ock:w-full ock:text-center ock:text-base"), children: "Swap" }), to: swappableTokens ?? WALLET_ADVANCED_DEFAULT_SWAPPABLE_TOKENS, from: (tokenBalances == null ? void 0 : tokenBalances.map((token) => ({
        address: token.address,
        chainId: token.chainId,
        symbol: token.symbol,
        decimals: token.decimals,
        image: token.image,
        name: token.name
      }))) ?? [], classNames: classNames == null ? void 0 : classNames.swap }) });
    }
    return /* @__PURE__ */ jsx(ContentWrapper, { children });
  }, [activeFeature, swappableTokens, tokenBalances, children, classNames]);
  if (breakpoint === "sm") {
    return /* @__PURE__ */ jsx(Sheet, { isOpen: isSubComponentOpen, onClose: handleBottomSheetClose, className: classNames == null ? void 0 : classNames.container, side: "bottom", title: "Wallet", description: "Wallet menu", children: /* @__PURE__ */ jsx("div", { className: "ock:flex ock:h-full ock:w-full ock:flex-col ock:items-center ock:justify-center", children: content }) });
  }
  if (!isSubComponentOpen && !isSubComponentClosing) {
    return null;
  }
  return /* @__PURE__ */ jsx("div", { "data-testid": "ockWalletDropdownContent", className: cn("ock:bg-ock-background", "ock:rounded-ock-default", border.lineDefault, zIndex.dropdown, "ock:my-1.5 ock:h-auto ock:w-full", "ock:flex ock:justify-center", "ock:overflow-hidden", animations.container, classNames == null ? void 0 : classNames.container), onAnimationEnd: handleAnimationEnd, children: content });
}
function ContentWrapper({
  children,
  className
}) {
  return /* @__PURE__ */ jsx("div", { className: cn("ock:flex ock:flex-col ock:justify-between", "ock:min-w-80 ock:max-h-120", className), children });
}
export {
  WalletDropdownContent
};
