// @coinbase/onchainkit/dist/wallet/components/WalletDropdown.js
'use client';
import { jsx, jsxs, Fragment } from "react/jsx-runtime";
import { Address } from "../../identity/components/Address.js";
import { Avatar } from "../../identity/components/Avatar.js";
import "../../core/network/attestations.js";
import "../../identity/utils/easSupportedChains.js";
import "@tanstack/react-query";
import "../../identity/constants.js";
import "../../styles/constants.js";
import { cn } from "../../styles/theme.js";
import { useEffect } from "react";
import "../../useOnchainKit.js";
import "../../identity/components/IdentityProvider.js";
import { EthBalance } from "../../identity/components/EthBalance.js";
import { Identity } from "../../identity/components/Identity.js";
import { Name } from "../../identity/components/Name.js";
import "viem/chains";
import "viem";
import "viem/ens";
import { useFloating, autoUpdate, offset, flip, shift, FloatingPortal } from "@floating-ui/react";
import { WalletDropdownContent } from "./WalletDropdownContent.js";
import { WalletDropdownDisconnect } from "./WalletDropdownDisconnect.js";
import { WalletDropdownLink } from "./WalletDropdownLink.js";
import { useWalletContext } from "./WalletProvider.js";
import { useAccount } from "wagmi";
import { useOutsideClick } from "../../internal/hooks/useOutsideClick.js";
const defaultWalletDropdownChildren = /* @__PURE__ */ jsxs(Fragment, { children: [
  /* @__PURE__ */ jsxs(Identity, { className: "ock:px-4 ock:pt-3 ock:pb-2", children: [
    /* @__PURE__ */ jsx(Avatar, {}),
    /* @__PURE__ */ jsx(Name, {}),
    /* @__PURE__ */ jsx(Address, { className: "text-ock-foreground-muted" }),
    /* @__PURE__ */ jsx(EthBalance, {})
  ] }, "wallet-dd-identity"),
  /* @__PURE__ */ jsx(WalletDropdownLink, { icon: "wallet", href: "https://keys.coinbase.com", target: "_blank", children: "Wallet" }, "wallet-dd-link"),
  /* @__PURE__ */ jsx(WalletDropdownDisconnect, {}, "wallet-dd-disconnect")
] });
function WalletDropdown({
  children,
  className,
  classNames,
  swappableTokens
}) {
  const {
    breakpoint,
    isSubComponentOpen,
    showSubComponentAbove,
    alignSubComponentRight,
    connectRef,
    handleClose
  } = useWalletContext();
  const {
    address
  } = useAccount();
  const {
    refs,
    floatingStyles
  } = useFloating({
    open: isSubComponentOpen,
    placement: showSubComponentAbove ? alignSubComponentRight ? "top-end" : "top-start" : alignSubComponentRight ? "bottom-end" : "bottom-start",
    middleware: [offset(6), flip({
      fallbackPlacements: ["top-start", "top-end", "bottom-start", "bottom-end"]
    }), shift({
      padding: 8
    })],
    whileElementsMounted: autoUpdate
  });
  useOutsideClick(refs.floating, handleClose);
  useEffect(() => {
    if (connectRef == null ? void 0 : connectRef.current) {
      refs.setReference(connectRef.current);
    }
  }, [connectRef, refs]);
  if (!address || !breakpoint || !isSubComponentOpen) {
    return null;
  }
  return /* @__PURE__ */ jsx(FloatingPortal, { children: /* @__PURE__ */ jsx("div", { ref: refs.setFloating, style: floatingStyles, "data-testid": "ockWalletDropdown", className: cn("ock:z-50", className), children: /* @__PURE__ */ jsx(WalletDropdownContent, { classNames, swappableTokens, children: children || defaultWalletDropdownChildren }) }) });
}
export {
  WalletDropdown
};
