// @coinbase/onchainkit/dist/wallet/components/WalletAdvancedAddressDetails.js
'use client';
import { jsx, jsxs } from "react/jsx-runtime";
import { useState, useCallback } from "react";
import { cn, text, pressable } from "../../styles/theme.js";
import "../../identity/components/IdentityProvider.js";
import { Avatar } from "../../identity/components/Avatar.js";
import "../../core/network/attestations.js";
import "../../identity/utils/easSupportedChains.js";
import "@tanstack/react-query";
import "../../identity/constants.js";
import { zIndex } from "../../styles/constants.js";
import { useOnchainKit } from "../../useOnchainKit.js";
import "viem";
import { useAccount } from "wagmi";
import { Name } from "../../identity/components/Name.js";
import "viem/chains";
import "viem/ens";
import { Spinner } from "../../internal/components/Spinner.js";
import { useWalletContext } from "./WalletProvider.js";
import { usePortfolio } from "../hooks/usePortfolio.js";
import { RequestContext } from "../../core/network/constants.js";
function WalletAdvancedAddressDetails({
  classNames
}) {
  const {
    animations
  } = useWalletContext();
  const {
    address
  } = useAccount();
  const {
    chain
  } = useOnchainKit();
  const [copyText, setCopyText] = useState("Copy");
  const handleCopyAddress = useCallback(async () => {
    try {
      await navigator.clipboard.writeText(String(address));
      setCopyText("Copied");
      setTimeout(() => setCopyText("Copy"), 2e3);
    } catch (err) {
      console.error("Failed to copy address:", err);
      setCopyText("Failed to copy");
      setTimeout(() => setCopyText("Copy"), 2e3);
    }
  }, [address]);
  if (!address || !chain) {
    return /* @__PURE__ */ jsx("div", { className: "ock:mt-1 ock:h-28 ock:w-10 ock:px-4 ock:py-3" });
  }
  return /* @__PURE__ */ jsxs("div", { "data-testid": "ockWalletAdvanced_AddressDetails", className: cn("ock:mt-2 ock:flex ock:w-88 ock:flex-col ock:items-center ock:justify-center ock:px-4 ock:py-3", "ock:text-ock-foreground", text.body, animations.content, classNames == null ? void 0 : classNames.container), children: [
    /* @__PURE__ */ jsx(Avatar, { address, chain, className: cn("ock:pointer-events-none ock:h-10 ock:w-10", classNames == null ? void 0 : classNames.avatar) }),
    /* @__PURE__ */ jsxs("div", { className: "ock:group ock:relative ock:mt-2 ock:text-base", children: [
      /* @__PURE__ */ jsx("button", { type: "button", onClick: handleCopyAddress, "data-testid": "ockWalletAdvanced_NameButton", children: /* @__PURE__ */ jsx(Name, { address, chain, className: cn("ock:hover:text-ock-foreground-muted ock:active:text-ock-primary", classNames == null ? void 0 : classNames.nameButton) }) }),
      /* @__PURE__ */ jsx("button", { type: "button", onClick: handleCopyAddress, className: cn(pressable.alternate, text.legal, "ock:text-ock-foreground", "ock:border-ock-background", "ock:rounded-ock-default", zIndex.tooltip, "ock:absolute ock:top-full ock:right-0 ock:mt-0.5 ock:px-1.5 ock:py-0.5 ock:opacity-0 ock:transition-opacity ock:group-hover:opacity-100"), "aria-live": "polite", "data-testid": "ockWalletAdvanced_NameTooltip", children: copyText })
    ] }),
    /* @__PURE__ */ jsx(AddressBalanceInFiat, { className: classNames == null ? void 0 : classNames.fiatBalance })
  ] });
}
function AddressBalanceInFiat({
  className
}) {
  const {
    address
  } = useAccount();
  const {
    data: portfolioData,
    isFetching: isFetchingPortfolioData
  } = usePortfolio({
    address
  }, RequestContext.Wallet);
  const portfolioFiatValue = portfolioData == null ? void 0 : portfolioData.portfolioBalanceInUsd;
  const formattedValueInFiat = new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD"
  }).format(Number(portfolioFiatValue));
  if (isFetchingPortfolioData) {
    return /* @__PURE__ */ jsx("div", { className: "ock:mt-1 ock:h-8 ock:w-full", children: /* @__PURE__ */ jsx(Spinner, {}) });
  }
  if (portfolioFiatValue === null || portfolioFiatValue === void 0) {
    return null;
  }
  return /* @__PURE__ */ jsx("div", { className: cn(text.title1, "ock:mt-1 ock:font-normal", className), "data-testid": "ockWalletAdvanced_AddressBalance", children: formattedValueInFiat });
}
export {
  WalletAdvancedAddressDetails
};
