// @coinbase/onchainkit/dist/wallet/components/WalletAdvancedTransactionActions.js
'use client';
import { jsx, jsxs } from "react/jsx-runtime";
import { useAnalytics } from "../../core/analytics/hooks/useAnalytics.js";
import { WalletEvent, WalletOption } from "../../core/analytics/types.js";
import { Skeleton } from "../../internal/components/Skeleton.js";
import { addSvgForeground } from "../../internal/svg/addForegroundSvg.js";
import { arrowUpRightSvg } from "../../internal/svg/arrowUpRightSvg.js";
import { toggleSvg } from "../../internal/svg/toggleSvg.js";
import { cn, text, pressable } from "../../styles/theme.js";
import { useOnchainKit } from "../../useOnchainKit.js";
import { useCallback } from "react";
import { useWalletContext } from "./WalletProvider.js";
import { RequestContext } from "../../core/network/constants.js";
import { usePortfolio } from "../hooks/usePortfolio.js";
import { useAccount } from "wagmi";
function WalletAdvancedTransactionActions({
  classNames
}) {
  const {
    setActiveFeature,
    animations
  } = useWalletContext();
  const {
    address
  } = useAccount();
  const {
    chain
  } = useOnchainKit();
  const {
    projectId
  } = useOnchainKit();
  const {
    sendAnalytics
  } = useAnalytics();
  const {
    isFetching: isFetchingPortfolioData
  } = usePortfolio({
    address
  }, RequestContext.Wallet);
  const handleAnalyticsOptionSelected = useCallback((option) => {
    sendAnalytics(WalletEvent.OptionSelected, {
      option
    });
  }, [sendAnalytics]);
  const handleBuy = useCallback(() => {
    handleAnalyticsOptionSelected(WalletOption.Buy);
    if (!projectId || !address || !(chain == null ? void 0 : chain.name)) {
      return;
    }
    const url = new URL("https://pay.coinbase.com/buy/select-asset");
    const params = new URLSearchParams({
      appId: projectId,
      destinationWallets: JSON.stringify([{
        address,
        blockchains: [chain.name.toLowerCase()]
      }]),
      defaultAsset: "USDC",
      defaultPaymentMethod: "CRYPTO_ACCOUNT",
      presetFiatAmount: "25"
    });
    window.open(`${url}?${params}`, "popup", "width=400,height=600,scrollbars=yes");
  }, [address, chain == null ? void 0 : chain.name, projectId, handleAnalyticsOptionSelected]);
  const handleSend = useCallback(() => {
    handleAnalyticsOptionSelected(WalletOption.Send);
    setActiveFeature("send");
  }, [handleAnalyticsOptionSelected, setActiveFeature]);
  const handleSwap = useCallback(() => {
    handleAnalyticsOptionSelected(WalletOption.Swap);
    setActiveFeature("swap");
  }, [setActiveFeature, handleAnalyticsOptionSelected]);
  if (isFetchingPortfolioData) {
    return /* @__PURE__ */ jsx(Skeleton, { className: "ock:mx-auto ock:my-3 ock:h-16 ock:w-80 ock:px-4 ock:py-3" });
  }
  return /* @__PURE__ */ jsxs("div", { "data-testid": "ockWalletAdvanced_TransactionActions", className: cn("ock:my-3 ock:flex ock:w-88 ock:flex-row ock:justify-between ock:gap-2 ock:px-4 ock:py-3", animations.content, classNames == null ? void 0 : classNames.container), children: [
    /* @__PURE__ */ jsx(WalletAdvancedTransactionAction, { icon: addSvgForeground, label: "Buy", action: handleBuy, classNames: classNames == null ? void 0 : classNames.leftAction }),
    /* @__PURE__ */ jsx(WalletAdvancedTransactionAction, { icon: arrowUpRightSvg, label: "Send", action: handleSend, classNames: classNames == null ? void 0 : classNames.middleAction }),
    /* @__PURE__ */ jsx(WalletAdvancedTransactionAction, { icon: toggleSvg, label: "Swap", action: handleSwap, classNames: classNames == null ? void 0 : classNames.rightAction })
  ] });
}
function WalletAdvancedTransactionAction({
  icon,
  label,
  action,
  classNames
}) {
  return /* @__PURE__ */ jsxs("button", { type: "button", className: cn("ock:flex ock:flex-col ock:items-center ock:justify-center ock:gap-2 ock:pt-2.5 ock:pb-2", "ock:h-16 ock:flex-1", "ock:rounded-ock-default", pressable.alternate, classNames == null ? void 0 : classNames.container), onClick: action, "aria-label": label, children: [
    /* @__PURE__ */ jsx("span", { className: cn("ock:flex ock:h-4 ock:w-4 ock:flex-col ock:items-center ock:justify-center", classNames == null ? void 0 : classNames.icon), children: icon }),
    /* @__PURE__ */ jsx("span", { className: cn(text.label2, "ock:text-ock-foreground", "ock:flex ock:flex-col ock:justify-center", classNames == null ? void 0 : classNames.label), children: label })
  ] });
}
export {
  WalletAdvancedTransactionActions
};
