// sushi/_cjs/evm/config/tokens/tokens/USDC.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STARGATE_USDC = exports.STARGATE_USDC_ADDRESS = exports.USDC = exports.USDC_ADDRESS = void 0;
const index_js_1 = require("../../../chain/index.js");
const token_js_1 = require("../../../currency/token.js");
const address_map_to_token_map_js_1 = require("../address-map-to-token-map.js");
const axlUSDC_js_1 = require("./axlUSDC.js");
exports.USDC_ADDRESS = {
    [index_js_1.EvmChainId.ETHEREUM]: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
    [index_js_1.EvmChainId.SEPOLIA]: '0x1c7d4b196cb0c7b01d743fbc6116a902379c7238',
    [index_js_1.EvmChainId.POLYGON]: '0x3c499c542cef5e3811e1192ce70d8cc03d5c3359',
    [index_js_1.EvmChainId.BSC]: '0x8ac76a51cc950d9822d68b83fe1ad97b32cd580d',
    [index_js_1.EvmChainId.HARMONY]: '0x985458e523db3d53125813ed68c274899e9dfab4',
    [index_js_1.EvmChainId.GNOSIS]: '0xddafbb505ad214d7b80b1f830fccc89b60fb7a83',
    [index_js_1.EvmChainId.ARBITRUM]: '0xaf88d065e77c8cc2239327c5edb3a432268e5831',
    [index_js_1.EvmChainId.AVALANCHE]: '0xb97ef9ef8734c71904d8002f8b6bc66dd9c48a6e',
    [index_js_1.EvmChainId.CELO]: '0xceba9300f2b948710d2653dd7b07f33a8b32118c',
    [index_js_1.EvmChainId.OPTIMISM]: '0x0b2c639c533813f4aa9d7837caf62653d097ff85',
    [index_js_1.EvmChainId.METIS]: '0xea32a96608495e54156ae48931a7c20f0dcc1a21',
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0x750ba8b76187092b0d1e87e28daaf484d1b5273b',
    [index_js_1.EvmChainId.BOBA]: '0x66a2a913e447d6b4bf33efbec43aaef87890fbbc',
    [index_js_1.EvmChainId.BOBA_BNB]: '0x9f98f9f312d23d078061962837042b8918e6aff2',
    [index_js_1.EvmChainId.BTTC]: '0xae17940943ba9440540940db0f1877f101d39e8b',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0xa8ce8aee21bc2a48a5ef670afcc9274c7bbbc035',
    [index_js_1.EvmChainId.THUNDERCORE]: '0x22e89898a04eaf43379beb70bf4e38b1faf8a31e',
    [index_js_1.EvmChainId.CORE]: '0xa4151b2b3e269645181dccf2d426ce75fcbdeca9',
    [index_js_1.EvmChainId.HAQQ]: axlUSDC_js_1.axlUSDC_ADDRESS[index_js_1.EvmChainId.HAQQ],
    [index_js_1.EvmChainId.ZKSYNC_ERA]: '0x3355df6d4c9c3035724fd0e3914de96a5a83aaf4',
    [index_js_1.EvmChainId.LINEA]: '0x176211869ca2b568f2a7d4ee941e073a821ee1ff',
    [index_js_1.EvmChainId.BASE]: '0x833589fcd6edb6e08f4c7c32d4f71b54bda02913',
    [index_js_1.EvmChainId.SCROLL]: '0x06efdbff2a14a7c8e15944d1f4a48f9f95f663a4',
    [index_js_1.EvmChainId.FILECOIN]: axlUSDC_js_1.axlUSDC_ADDRESS[index_js_1.EvmChainId.FILECOIN],
    [index_js_1.EvmChainId.ZETACHAIN]: '0x0cbe0df132a6c6b4a2974fa1b7fb953cf0cc798a',
    [index_js_1.EvmChainId.CRONOS]: '0xc21223249ca28397b4b6541dffaecc539bff0c59',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0x5f795bb52dac3085f578f4877d450e2929d2f13d',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0x1bda44fda023f2af8280a16fd1b01d1a493ba6c4',
    [index_js_1.EvmChainId.MANTLE]: '0x09bc4e0d864854c6afb6eb9a9cdf58ac190d0df9',
    [index_js_1.EvmChainId.MANTA]: '0xb73603c5d87fa094b7314c74ace2e64d165016fb',
    [index_js_1.EvmChainId.MODE]: '0xd988097fb8612cc24eec14542bc03424c656005f',
    [index_js_1.EvmChainId.TAIKO]: '0x07d83526730c7438048d55a4fc0b850e2aab6f0b',
    [index_js_1.EvmChainId.ZKLINK]: '0x1a1a3b2ff016332e866787b311fcb63928464509',
    [index_js_1.EvmChainId.SONIC]: '0x29219dd400f2bf60e5a23d13be72b486d4038894',
    [index_js_1.EvmChainId.HEMI]: '0xad11a8beb98bbf61dbb1aa0f6d6f2ecd87b35afa',
    [index_js_1.EvmChainId.KATANA]: '0x203a662b0bd271a6ed5a60edfbd04bfce608fd36',
    [index_js_1.EvmChainId.BERACHAIN]: '0x549943e04f40284185054145c6e4e9568c1d3241',
};
exports.USDC = {
    ...(0, address_map_to_token_map_js_1.addressMapToTokenMap)({
        decimals: 6,
        symbol: 'USDC',
        name: 'USD Coin',
    }, exports.USDC_ADDRESS),
    [index_js_1.EvmChainId.BSC]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.BSC,
        address: exports.USDC_ADDRESS[index_js_1.EvmChainId.BSC],
        decimals: 18,
        symbol: 'USDC',
        name: 'USD Coin',
    }),
    [index_js_1.EvmChainId.BOBA_BNB]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.BOBA_BNB,
        address: exports.USDC_ADDRESS[index_js_1.EvmChainId.BOBA_BNB],
        decimals: 18,
        symbol: 'USDC',
        name: 'USD Coin',
    }),
    [index_js_1.EvmChainId.ROOTSTOCK]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.ROOTSTOCK,
        address: exports.USDC_ADDRESS[index_js_1.EvmChainId.ROOTSTOCK],
        decimals: 18,
        symbol: 'rUSDC',
        name: 'rUSDC',
    }),
};
exports.STARGATE_USDC_ADDRESS = {
    [index_js_1.EvmChainId.FANTOM]: '0x28a92dde19d9989f39a49905d7c9c2fac7799bdf',
};
exports.STARGATE_USDC = (0, address_map_to_token_map_js_1.addressMapToTokenMap)({
    decimals: 6,
    symbol: 'stgUSDC',
    name: 'Stargate USDC',
    origin: 'stargate',
}, exports.STARGATE_USDC_ADDRESS);
