// sushi/_cjs/evm/config/tokens/tokens/USDT.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STARGATE_USDT = exports.STARGATE_USDT_ADDRESS = exports.USDT = exports.USDT_ADDRESS = void 0;
const index_js_1 = require("../../../chain/index.js");
const token_js_1 = require("../../../currency/token.js");
const address_map_to_token_map_js_1 = require("../address-map-to-token-map.js");
const axlUSDT_js_1 = require("./axlUSDT.js");
exports.USDT_ADDRESS = {
    [index_js_1.EvmChainId.ETHEREUM]: '0xdac17f958d2ee523a2206206994597c13d831ec7',
    [index_js_1.EvmChainId.POLYGON]: '0xc2132d05d31c914a87c6611c10748aeb04b58e8f',
    [index_js_1.EvmChainId.BASE]: '0xfde4c96c8593536e31f229ea8f37b2ada2699bb2',
    [index_js_1.EvmChainId.BSC]: '0x55d398326f99059ff775485246999027b3197955',
    [index_js_1.EvmChainId.HARMONY]: '0x3c2b8be99c50593081eaa2a724f0b8285f5aba8f',
    [index_js_1.EvmChainId.GNOSIS]: '0x4ecaba5870353805a9f068101a40e0f32ed605c6',
    [index_js_1.EvmChainId.ARBITRUM]: '0xfd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9',
    [index_js_1.EvmChainId.AVALANCHE]: '0x9702230a8ea53601f5cd2dc00fdbc13d4df4a8c7',
    [index_js_1.EvmChainId.CELO]: '0x88eec49252c8cbc039dcdb394c0c2ba2f1637ea0',
    [index_js_1.EvmChainId.OPTIMISM]: '0x94b008aa00579c1307b0ef2c499ad98a8ce58e58',
    [index_js_1.EvmChainId.KAVA]: '0x919c1c267bc06a7039e03fcc2ef738525769109c',
    [index_js_1.EvmChainId.METIS]: '0xbb06dca3ae6887fabf931640f67cab3e3a16f4dc',
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0xed9d63a96c27f87b07115b56b2e3572827f21646',
    [index_js_1.EvmChainId.BOBA]: '0x5de1677344d3cb0d7d465c10b72a8f60699c062d',
    [index_js_1.EvmChainId.BOBA_BNB]: '0x1e633dcd0d3d349126983d58988051f7c62c543d',
    [index_js_1.EvmChainId.BTTC]: '0xe887512ab8bc60bcc9224e1c3b5be68e26048b8b',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0x1e4a5963abfd975d8c9021ce480b42188849d41d',
    [index_js_1.EvmChainId.THUNDERCORE]: '0x4f3c8e20942461e2c3bdd8311ac57b0c222f2b82',
    [index_js_1.EvmChainId.CORE]: '0x900101d06a7426441ae63e9ab3b9b0f63be145f1',
    [index_js_1.EvmChainId.HAQQ]: axlUSDT_js_1.axlUSDT_ADDRESS[index_js_1.EvmChainId.HAQQ],
    [index_js_1.EvmChainId.ZKSYNC_ERA]: '0x493257fd37edb34451f62edf8d2a0c418852ba4c',
    [index_js_1.EvmChainId.SCROLL]: '0xf55bec9cafdbe8730f096aa55dad6d22d44099df',
    [index_js_1.EvmChainId.ZETACHAIN]: '0x7c8dda80bbbe1254a7aacf3219ebe1481c6e01d7',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0x1c0491e3396ad6a35f061c62387a95d7218fc515',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0xef213441a85df4d7acbdae0cf78004e1e486bb96',
    [index_js_1.EvmChainId.MANTA]: '0xf417f5a458ec102b90352f697d6e2ac3a3d2851f',
    [index_js_1.EvmChainId.MODE]: '0xf0f161fda2712db8b566946122a5af183995e2ed',
    [index_js_1.EvmChainId.TAIKO]: '0x2def195713cf4a606b49d07e520e22c17899a736',
    [index_js_1.EvmChainId.ZKLINK]: '0x2f8a25ac62179b31d62d7f80884ae57464699059',
    [index_js_1.EvmChainId.HEMI]: '0xbb0d083fb1be0a9f6157ec484b6c79e0a4e31c2e',
    [index_js_1.EvmChainId.LINEA]: '0xa219439258ca9da29e9cc4ce5596924745e12b93',
    [index_js_1.EvmChainId.MANTLE]: '0x201eba5cc46d216ce6dc03f6a759e8e766e956ae',
    [index_js_1.EvmChainId.KATANA]: '0x2dca96907fde857dd3d816880a0df407eeb2d2f2',
    [index_js_1.EvmChainId.HYPEREVM]: '0xb8ce59fc3717ada4c02eadf9682a9e934f625ebb',
    [index_js_1.EvmChainId.BERACHAIN]: '0x779ded0c9e1022225f8e0630b35a9b54be713736',
    [index_js_1.EvmChainId.PLASMA]: '0xb8ce59fc3717ada4c02eadf9682a9e934f625ebb',
};
exports.USDT = {
    ...(0, address_map_to_token_map_js_1.addressMapToTokenMap)({
        decimals: 6,
        symbol: 'USDT',
        name: 'Tether USD',
    }, exports.USDT_ADDRESS),
    [index_js_1.EvmChainId.BSC]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.BSC,
        address: exports.USDT_ADDRESS[index_js_1.EvmChainId.BSC],
        decimals: 18,
        symbol: 'USDT',
        name: 'Tether USD',
    }),
    [index_js_1.EvmChainId.BOBA_BNB]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.BOBA_BNB,
        address: exports.USDT_ADDRESS[index_js_1.EvmChainId.BOBA_BNB],
        decimals: 18,
        symbol: 'USDT',
        name: 'Tether USD',
    }),
    [index_js_1.EvmChainId.ROOTSTOCK]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.ROOTSTOCK,
        address: exports.USDT_ADDRESS[index_js_1.EvmChainId.ROOTSTOCK],
        decimals: 18,
        symbol: 'rUSDT',
        name: 'rUSDT',
    }),
};
exports.STARGATE_USDT_ADDRESS = {
    [index_js_1.EvmChainId.FANTOM]: '0xcc1b99ddac1a33c201a742a1851662e87bc7f22c',
    [index_js_1.EvmChainId.KAVA]: '0xaad094f6a75a14417d39f04e690fc216f080a41a',
};
exports.STARGATE_USDT = {
    [index_js_1.EvmChainId.FANTOM]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.FANTOM,
        address: exports.STARGATE_USDT_ADDRESS[index_js_1.EvmChainId.FANTOM],
        decimals: 6,
        symbol: 'stgUSDT',
        name: 'Stargate USDT',
        origin: 'stargate',
    }),
    [index_js_1.EvmChainId.KAVA]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.KAVA,
        address: exports.STARGATE_USDT_ADDRESS[index_js_1.EvmChainId.KAVA],
        decimals: 6,
        symbol: 'stgUSDT',
        name: 'Stargate USDT',
        origin: 'stargate',
    }),
};
