// opensea-js/src/typechain/contracts/TransferHelper.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "./common";

export type TransferHelperItemStruct = {
  itemType: BigNumberish;
  token: AddressLike;
  identifier: BigNumberish;
  amount: BigNumberish;
  recipient: AddressLike;
};

export type TransferHelperItemStructOutput = [
  itemType: bigint,
  token: string,
  identifier: bigint,
  amount: bigint,
  recipient: string
] & {
  itemType: bigint;
  token: string;
  identifier: bigint;
  amount: bigint;
  recipient: string;
};

export interface TransferHelperInterface extends Interface {
  getFunction(nameOrSignature: "bulkTransfer"): FunctionFragment;

  encodeFunctionData(
    functionFragment: "bulkTransfer",
    values: [TransferHelperItemStruct[], BytesLike]
  ): string;

  decodeFunctionResult(
    functionFragment: "bulkTransfer",
    data: BytesLike
  ): Result;
}

export interface TransferHelper extends BaseContract {
  connect(runner?: ContractRunner | null): TransferHelper;
  waitForDeployment(): Promise<this>;

  interface: TransferHelperInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  bulkTransfer: TypedContractMethod<
    [items: TransferHelperItemStruct[], conduitKey: BytesLike],
    [string],
    "nonpayable"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "bulkTransfer"
  ): TypedContractMethod<
    [items: TransferHelperItemStruct[], conduitKey: BytesLike],
    [string],
    "nonpayable"
  >;

  filters: {};
}
