// sushi/_cjs/mvm/chain/chains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MvmChainId = exports.mvmChainIds = exports.mvmChains = void 0;
exports.isMvmChainId = isMvmChainId;
exports.getMvmChainById = getMvmChainById;
exports.isMvmMainnetChainId = isMvmMainnetChainId;
exports.isMvmChainKey = isMvmChainKey;
exports.getMvmChainByKey = getMvmChainByKey;
const define_chain_js_1 = require("./define-chain.js");
const aptos = (0, define_chain_js_1.defineMvmChain)({
    chainId: -1,
    key: 'aptos',
    shortName: 'apt',
    name: 'Aptos',
    blockExplorers: {
        default: { name: 'Aptos Explorer', url: 'https://explorer.aptoslabs.com' },
    },
    netType: 'mainnet',
});
exports.mvmChains = [aptos];
exports.mvmChainIds = exports.mvmChains.map((chain) => chain.chainId);
exports.MvmChainId = new Proxy({}, {
    get: (_, key) => {
        const _key = key.toLowerCase().replace(/_/g, '-');
        const chain = exports.mvmChains.find((chain) => chain.key === _key);
        if (!chain)
            return undefined;
        return chain.chainId;
    },
});
function isMvmChainId(chainId) {
    return exports.mvmChains.some((chain) => chain.chainId === chainId);
}
function getMvmChainById(chainId) {
    return exports.mvmChains.find((chain) => chain.chainId === chainId);
}
function isMvmMainnetChainId(chainId) {
    return exports.mvmChains.some((chain) => chain.chainId === chainId && chain.netType === 'mainnet');
}
function isMvmChainKey(key) {
    return exports.mvmChains.some((chain) => chain.key === key);
}
function getMvmChainByKey(key) {
    return exports.mvmChains.find((chain) => chain.key === key);
}
