// @coinbase/agentkit/dist/action-providers/zora/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateCoinSchema = void 0;
const zod_1 = require("zod");
exports.CreateCoinSchema = zod_1.z
    .object({
    name: zod_1.z.string().describe("The name of the coin to create"),
    symbol: zod_1.z.string().describe("The symbol of the coin to create"),
    description: zod_1.z.string().describe("The description of the coin"),
    image: zod_1.z.string().describe("Local image file path or URI (ipfs:// or https://)"),
    category: zod_1.z
        .string()
        .default("social")
        .describe("The category of the coin, optional, defaults to 'social'"),
    payoutRecipient: zod_1.z
        .string()
        .regex(/^0x[a-fA-F0-9]{40}$/, "Invalid Ethereum address format")
        .optional()
        .describe("The address that will receive creator earnings, defaults to wallet address"),
    platformReferrer: zod_1.z
        .string()
        .regex(/^0x[a-fA-F0-9]{40}$/, "Invalid Ethereum address format")
        .optional()
        .describe("The address that will receive platform referrer fees, optional"),
    currency: zod_1.z
        .enum(["ZORA", "ETH"])
        .default("ZORA")
        .describe("Currency to be used for the trading pair, optional, defaults to 'ZORA'."),
})
    .strip()
    .describe("Instructions for creating a new coin on Zora");
