// sushi/_cjs/evm/abi/nonfungiblePositionManagerAbi/nonfungiblePositionManagerAbi_decreaseLiquidity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nonfungiblePositionManagerAbi_decreaseLiquidity = void 0;
exports.nonfungiblePositionManagerAbi_decreaseLiquidity = [
    {
        inputs: [
            {
                components: [
                    { internalType: 'uint256', name: 'tokenId', type: 'uint256' },
                    { internalType: 'uint128', name: 'liquidity', type: 'uint128' },
                    { internalType: 'uint256', name: 'amount0Min', type: 'uint256' },
                    { internalType: 'uint256', name: 'amount1Min', type: 'uint256' },
                    { internalType: 'uint256', name: 'deadline', type: 'uint256' },
                ],
                internalType: 'struct INonfungiblePositionManager.DecreaseLiquidityParams',
                name: 'params',
                type: 'tuple',
            },
        ],
        name: 'decreaseLiquidity',
        outputs: [
            { internalType: 'uint256', name: 'amount0', type: 'uint256' },
            { internalType: 'uint256', name: 'amount1', type: 'uint256' },
        ],
        stateMutability: 'payable',
        type: 'function',
    },
];
