// sushi/_cjs/evm/abi/nonfungiblePositionManagerAbi/nonfungiblePositionManagerAbi_safeTransferFrom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nonfungiblePositionManagerAbi_safeTransferFrom = void 0;
exports.nonfungiblePositionManagerAbi_safeTransferFrom = [
    {
        inputs: [
            {
                internalType: 'address',
                name: 'from',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'to',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'tokenId',
                type: 'uint256',
            },
        ],
        name: 'safeTransferFrom',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address', name: 'from', type: 'address' },
            { internalType: 'address', name: 'to', type: 'address' },
            { internalType: 'uint256', name: 'tokenId', type: 'uint256' },
            { internalType: 'bytes', name: '_data', type: 'bytes' },
        ],
        name: 'safeTransferFrom',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
