// sushi/_cjs/generic/math/fraction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fraction = void 0;
const tiny_invariant_1 = __importDefault(require("tiny-invariant"));
const number_js_1 = require("../format/number.js");
class Fraction {
    constructor(arg) {
        if (typeof arg === 'number') {
            (0, tiny_invariant_1.default)(Number.isInteger(arg) && arg >= 0, 'Fraction must be a non-negative integer');
            this.numerator = BigInt(arg);
            this.denominator = 1n;
            return;
        }
        this.numerator = BigInt(arg.numerator);
        this.denominator = BigInt(arg.denominator ?? 1n);
    }
    get quotient() {
        return this.numerator / this.denominator;
    }
    get remainder() {
        return this.numerator % this.denominator;
    }
    invert() {
        (0, tiny_invariant_1.default)(this.numerator !== 0n, 'Cannot invert a fraction with a numerator of 0');
        return new Fraction({
            numerator: this.denominator,
            denominator: this.numerator,
        });
    }
    add(other) {
        if (other instanceof Fraction) {
            return new Fraction({
                numerator: this.numerator * other.denominator +
                    other.numerator * this.denominator,
                denominator: this.denominator * other.denominator,
            });
        }
        return new Fraction({
            numerator: this.numerator + this.denominator * BigInt(other),
            denominator: this.denominator,
        });
    }
    sub(other) {
        if (other instanceof Fraction) {
            return new Fraction({
                numerator: this.numerator * other.denominator -
                    other.numerator * this.denominator,
                denominator: this.denominator * other.denominator,
            });
        }
        return new Fraction({
            numerator: this.numerator - this.denominator * BigInt(other),
            denominator: this.denominator,
        });
    }
    mul(other) {
        if (other instanceof Fraction) {
            return new Fraction({
                numerator: this.numerator * other.numerator,
                denominator: this.denominator * other.denominator,
            });
        }
        return new Fraction({
            numerator: this.numerator * BigInt(other),
            denominator: this.denominator,
        });
    }
    div(other) {
        if (other instanceof Fraction) {
            return new Fraction({
                numerator: this.numerator * other.denominator,
                denominator: this.denominator * other.numerator,
            });
        }
        return new Fraction({
            numerator: this.numerator,
            denominator: this.denominator * BigInt(other),
        });
    }
    compare(other, type) {
        let a;
        let b;
        if (other instanceof Fraction) {
            a = this.numerator * other.denominator;
            b = other.numerator * this.denominator;
        }
        else {
            a = this.numerator;
            b = BigInt(other) * this.denominator;
        }
        switch (type) {
            case '<':
                return a < b;
            case '<=':
                return a <= b;
            case '=':
                return a === b;
            case '>=':
                return a >= b;
            case '>':
                return a > b;
            case '!=':
                return a !== b;
            default:
                throw new Error(`Invalid comparison type: ${type}`);
        }
    }
    lt(other) {
        return this.compare(other, '<');
    }
    lte(other) {
        return this.compare(other, '<=');
    }
    eq(other) {
        return this.compare(other, '=');
    }
    gte(other) {
        return this.compare(other, '>=');
    }
    gt(other) {
        return this.compare(other, '>');
    }
    neq(other) {
        return this.compare(other, '!=');
    }
    toNumber() {
        return Number(this.numerator) / Number(this.denominator);
    }
    toString(args) {
        return (0, number_js_1.numberToFixed)(this.toNumber(), args);
    }
    toSignificant(significantDigits) {
        return this.toNumber().toPrecision(significantDigits);
    }
    toJSON() {
        return {
            n: this.numerator.toString(),
            d: this.denominator.toString(),
        };
    }
    get asFraction() {
        return new Fraction({
            numerator: this.numerator,
            denominator: this.denominator,
        });
    }
}
exports.Fraction = Fraction;
