// sushi/_cjs/generic/math/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sqrt = exports.numberToBigInt = exports.minimum = exports.maximum = exports.difference = exports.abs = void 0;
var abs_js_1 = require("./abs.js");
Object.defineProperty(exports, "abs", { enumerable: true, get: function () { return abs_js_1.abs; } });
__exportStar(require("./constants.js"), exports);
var difference_js_1 = require("./difference.js");
Object.defineProperty(exports, "difference", { enumerable: true, get: function () { return difference_js_1.difference; } });
__exportStar(require("./fraction.js"), exports);
var maximum_js_1 = require("./maximum.js");
Object.defineProperty(exports, "maximum", { enumerable: true, get: function () { return maximum_js_1.maximum; } });
var minimum_js_1 = require("./minimum.js");
Object.defineProperty(exports, "minimum", { enumerable: true, get: function () { return minimum_js_1.minimum; } });
var number_to_bigint_js_1 = require("./number-to-bigint.js");
Object.defineProperty(exports, "numberToBigInt", { enumerable: true, get: function () { return number_to_bigint_js_1.numberToBigInt; } });
__exportStar(require("./percent.js"), exports);
var sqrt_js_1 = require("./sqrt.js");
Object.defineProperty(exports, "sqrt", { enumerable: true, get: function () { return sqrt_js_1.sqrt; } });
