// @coinbase/onchainkit/dist/fund/hooks/useFundCardFundingUrl.js
import { useMemo } from "react";
import { useFundContext } from "../components/FundCardProvider.js";
import { getOnrampBuyUrl } from "../utils/getOnrampBuyUrl.js";
const useFundCardFundingUrl = () => {
  const {
    selectedPaymentMethod,
    selectedInputType,
    fundAmountFiat,
    fundAmountCrypto,
    currency,
    sessionToken
  } = useFundContext();
  return useMemo(() => {
    if (!sessionToken) {
      return void 0;
    }
    const fundAmount = selectedInputType === "fiat" ? fundAmountFiat : fundAmountCrypto;
    return getOnrampBuyUrl({
      sessionToken,
      presetFiatAmount: selectedInputType === "fiat" ? Number(fundAmount) : void 0,
      presetCryptoAmount: selectedInputType === "crypto" ? Number(fundAmount) : void 0,
      defaultPaymentMethod: selectedPaymentMethod == null ? void 0 : selectedPaymentMethod.id,
      fiatCurrency: currency,
      originComponentName: "FundCard"
    });
  }, [fundAmountFiat, fundAmountCrypto, selectedPaymentMethod, selectedInputType, sessionToken, currency]);
};
export {
  useFundCardFundingUrl
};
