// @coinbase/onchainkit/dist/checkout/constants.js
const GENERAL_CHECKOUT_ERROR_MESSAGE = "CHECKOUT_ERROR";
const GENERIC_ERROR_MESSAGE = "Something went wrong. Please try again.";
const NO_CONTRACTS_ERROR = "Contracts are not available";
const NO_CONNECTED_ADDRESS_ERROR = "No connected address";
const CHECKOUT_UNSUPPORTED_CHAIN_ERROR_MESSAGE = "UNSUPPORTED_CHAIN";
const CHECKOUT_TOO_MANY_REQUESTS_ERROR_MESSAGE = "CHECKOUT_TOO_MANY_REQUESTS_ERROR";
const CHECKOUT_INSUFFICIENT_BALANCE_ERROR = "User has insufficient balance";
const CHECKOUT_INVALID_CHARGE_ERROR_MESSAGE = "CHECKOUT_INVALID_CHARGE_ERROR";
const CHECKOUT_INVALID_PARAMETER_ERROR_MESSAGE = "CHECKOUT_INVALID_PARAMETER_ERROR";
const UNCAUGHT_CHECKOUT_ERROR_MESSAGE = "UNCAUGHT_CHECKOUT_ERROR";
const USER_REJECTED_ERROR = "Request denied.";
const CheckoutErrorCode = {
  INSUFFICIENT_BALANCE: "insufficient_balance",
  GENERIC_ERROR: "generic_error",
  UNEXPECTED_ERROR: "unexpected_error",
  USER_REJECTED_ERROR: "user_rejected"
};
const CHECKOUT_LIFECYCLE_STATUS = {
  FETCHING_DATA: "fetchingData",
  INIT: "init",
  PENDING: "pending",
  READY: "ready",
  SUCCESS: "success",
  ERROR: "error"
};
const USDC_ADDRESS_BASE = "0x833589fcd6edb6e08f4c7c32d4f71b54bda02913";
const CONTRACT_METHODS = {
  APPROVE: "approve",
  BALANCE_OF: "balanceOf",
  TRANSFER_TOKEN_PRE_APPROVED: "transferTokenPreApproved"
};
const COMMERCE_ABI = [{
  type: "function",
  name: "transferTokenPreApproved",
  inputs: [{
    name: "_intent",
    type: "tuple",
    components: [{
      name: "recipientAmount",
      type: "uint256"
    }, {
      name: "deadline",
      type: "uint256"
    }, {
      name: "recipient",
      type: "address"
    }, {
      name: "recipientCurrency",
      type: "address"
    }, {
      name: "refundDestination",
      type: "address"
    }, {
      name: "feeAmount",
      type: "uint256"
    }, {
      name: "id",
      type: "bytes16"
    }, {
      name: "operator",
      type: "address"
    }, {
      name: "signature",
      type: "bytes"
    }, {
      name: "prefix",
      type: "bytes"
    }]
  }],
  outputs: [],
  stateMutability: "nonpayable"
}];
export {
  CHECKOUT_INSUFFICIENT_BALANCE_ERROR,
  CHECKOUT_INVALID_CHARGE_ERROR_MESSAGE,
  CHECKOUT_INVALID_PARAMETER_ERROR_MESSAGE,
  CHECKOUT_LIFECYCLE_STATUS,
  CHECKOUT_TOO_MANY_REQUESTS_ERROR_MESSAGE,
  CHECKOUT_UNSUPPORTED_CHAIN_ERROR_MESSAGE,
  COMMERCE_ABI,
  CONTRACT_METHODS,
  CheckoutErrorCode,
  GENERAL_CHECKOUT_ERROR_MESSAGE,
  GENERIC_ERROR_MESSAGE,
  NO_CONNECTED_ADDRESS_ERROR,
  NO_CONTRACTS_ERROR,
  UNCAUGHT_CHECKOUT_ERROR_MESSAGE,
  USDC_ADDRESS_BASE,
  USER_REJECTED_ERROR
};
