// sushi/_esm/mvm/chain/chains.js
import { defineMvmChain } from './define-chain.js';
const aptos = defineMvmChain({
    chainId: -1,
    key: 'aptos',
    shortName: 'apt',
    name: 'Aptos',
    blockExplorers: {
        default: { name: 'Aptos Explorer', url: 'https://explorer.aptoslabs.com' },
    },
    netType: 'mainnet',
});
export const mvmChains = [aptos];
// MvmChainId
export const mvmChainIds = /* @__PURE__ */ mvmChains.map((chain) => chain.chainId);
export const MvmChainId = /* @__PURE__ */ new Proxy({}, {
    get: (_, key) => {
        const _key = key.toLowerCase().replace(/_/g, '-');
        const chain = mvmChains.find((chain) => chain.key === _key);
        if (!chain)
            return undefined;
        return chain.chainId;
    },
});
export function isMvmChainId(chainId) {
    return mvmChains.some((chain) => chain.chainId === chainId);
}
export function getMvmChainById(chainId) {
    return mvmChains.find((chain) => chain.chainId === chainId);
}
export function isMvmMainnetChainId(chainId) {
    return mvmChains.some((chain) => chain.chainId === chainId && chain.netType === 'mainnet');
}
export function isMvmChainKey(key) {
    return mvmChains.some((chain) => chain.key === key);
}
export function getMvmChainByKey(key) {
    return mvmChains.find((chain) => chain.key === key);
}
