// sushi/_cjs/generic/currency/amount.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Amount = void 0;
exports.serializedAmountSchema = serializedAmountSchema;
const utils_1 = require("viem/utils");
const z = __importStar(require("zod"));
const number_js_1 = require("../format/number.js");
const fraction_js_1 = require("../math/fraction.js");
class Amount {
    constructor(currency, amount) {
        this.currency = currency;
        this.amount = BigInt(amount);
    }
    static fromHuman(currency, human) {
        const amt = (0, utils_1.parseUnits)(String(human), currency.decimals);
        return new Amount(currency, amt);
    }
    static tryFromHuman(currency, human) {
        try {
            return Amount.fromHuman(currency, human);
        }
        catch {
            return undefined;
        }
    }
    wrap() {
        return new Amount(this.currency.wrap(), this.amount);
    }
    static getRawAmount(_ref, other) {
        if (other instanceof Amount) {
            return other.amount;
        }
        return BigInt(other);
    }
    add(other) {
        let add;
        if (other instanceof fraction_js_1.Fraction) {
            add = (other.numerator * Amount.SCALE) / other.denominator / Amount.SCALE;
        }
        else {
            add = Amount.getRawAmount(this, other);
        }
        return new Amount(this.currency, this.amount + add);
    }
    addHuman(other) {
        const addend = (0, utils_1.parseUnits)(String(other), this.currency.decimals);
        return new Amount(this.currency, this.amount + addend);
    }
    sub(other) {
        let sub;
        if (other instanceof fraction_js_1.Fraction) {
            sub = (other.numerator * Amount.SCALE) / other.denominator / Amount.SCALE;
        }
        else {
            sub = Amount.getRawAmount(this, other);
        }
        return new Amount(this.currency, this.amount - sub);
    }
    subHuman(other) {
        const sub = (0, utils_1.parseUnits)(String(other), this.currency.decimals);
        return new Amount(this.currency, this.amount - sub);
    }
    mul(other) {
        if (other instanceof fraction_js_1.Fraction) {
            return new Amount(this.currency, (this.amount * other.numerator) / other.denominator);
        }
        const mul = Amount.getRawAmount(this, other);
        return new Amount(this.currency, this.amount * mul);
    }
    mulHuman(other) {
        const mul = (0, utils_1.parseUnits)(String(other), Amount.SCALE_DECIMALS);
        const result = (this.amount * mul) / Amount.SCALE;
        return new Amount(this.currency, result);
    }
    div(divisor) {
        const d = (0, utils_1.parseUnits)(String(divisor), Amount.SCALE_DECIMALS);
        if (d === 0n)
            throw new Error('Cannot divide by zero');
        const result = (this.amount * Amount.SCALE) / d;
        return new Amount(this.currency, result);
    }
    divToFraction(other) {
        if (other instanceof fraction_js_1.Fraction) {
            return new fraction_js_1.Fraction({
                numerator: this.amount * other.denominator,
                denominator: other.numerator,
            });
        }
        const div = Amount.getRawAmount(this, other);
        return new fraction_js_1.Fraction({ numerator: this.amount, denominator: div });
    }
    gt(other) {
        const cmp = Amount.getRawAmount(this, other);
        return this.amount > cmp;
    }
    gte(other) {
        const cmp = Amount.getRawAmount(this, other);
        return this.amount >= cmp;
    }
    lt(other) {
        const cmp = Amount.getRawAmount(this, other);
        return this.amount < cmp;
    }
    lte(other) {
        const cmp = Amount.getRawAmount(this, other);
        return this.amount <= cmp;
    }
    eq(other) {
        const cmp = Amount.getRawAmount(this, other);
        return this.amount === cmp;
    }
    toJSON() {
        return {
            currency: this.currency.toJSON(),
            amount: this.amount.toString(),
        };
    }
    toString(args = {
        maxFixed: this.currency.decimals,
    }) {
        const str = (0, utils_1.formatUnits)(this.amount, this.currency.decimals);
        return (0, number_js_1.numberToFixed)(str, args);
    }
    toSignificant(significantDigits = 5) {
        return this.toString({
            significant: significantDigits,
        });
    }
}
exports.Amount = Amount;
Amount.SCALE_DECIMALS = 27;
Amount.SCALE = 10n ** BigInt(Amount.SCALE_DECIMALS);
function serializedAmountSchema(currencySchema) {
    return z.object({
        currency: currencySchema,
        amount: z.string(),
    });
}
