// sushi/_cjs/generic/currency/price.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Price = void 0;
const number_js_1 = require("../format/number.js");
const fraction_js_1 = require("../math/fraction.js");
const amount_js_1 = require("./amount.js");
class Price extends fraction_js_1.Fraction {
    constructor(args) {
        if ('baseAmount' in args && 'quoteAmount' in args) {
            const { baseAmount, quoteAmount } = args;
            super({ numerator: quoteAmount.amount, denominator: baseAmount.amount });
            this.base = baseAmount.currency;
            this.quote = quoteAmount.currency;
            return;
        }
        super({ numerator: args.numerator, denominator: args.denominator || 1n });
        this.base = args.base;
        this.quote = args.quote;
    }
    get scalar() {
        return new fraction_js_1.Fraction({
            numerator: 10n ** BigInt(this.base.decimals),
            denominator: 10n ** BigInt(this.quote.decimals),
        });
    }
    getQuote(baseAmount) {
        const quoteAmount = super.mul(baseAmount.amount).toNumber();
        return new amount_js_1.Amount(this.quote, BigInt(Math.floor(quoteAmount)));
    }
    invert() {
        return new Price({
            base: this.quote,
            quote: this.base,
            numerator: this.denominator,
            denominator: this.numerator,
        });
    }
    toNumber() {
        return super.mul(this.scalar).toNumber();
    }
    toString(args) {
        return (0, number_js_1.numberToFixed)(this.toNumber(), args);
    }
    toSignificant(significantDigits = 5) {
        return this.toString({
            significant: significantDigits,
        });
    }
}
exports.Price = Price;
