// @coinbase/onchainkit/dist/OnchainKitProviderBoundary.js
import { jsx } from "react/jsx-runtime";
import { Component } from "react";
import { sendAnalyticsPayload } from "./core/analytics/utils/analyticsService.js";
import { ErrorEvent } from "./core/analytics/types.js";
class OnchainKitProviderBoundary extends Component {
  constructor() {
    super(...arguments);
    this.state = {
      error: null
    };
  }
  static getDerivedStateFromError(error) {
    return {
      error
    };
  }
  componentDidCatch(error, errorInfo) {
    console.error("Uncaught error:", error, errorInfo);
    sendAnalyticsPayload(ErrorEvent.ComponentError, {
      component: "OnchainKitProviderBoundary",
      error: error.message,
      metadata: {
        message: error.message,
        stack: errorInfo.componentStack
      }
    });
  }
  render() {
    if (this.state.error) {
      if (this.props.fallback) {
        const Fallback = this.props.fallback;
        return /* @__PURE__ */ jsx(Fallback, { error: this.state.error });
      }
      return /* @__PURE__ */ jsx("h1", { children: "Sorry, we had an unhandled error" });
    }
    return this.props.children;
  }
}
export {
  OnchainKitProviderBoundary as default
};
