// @coinbase/onchainkit/dist/OnchainKitProvider.js
'use client';
import { jsx } from "react/jsx-runtime";
import { setOnchainKitConfig } from "./core/OnchainKitConfig.js";
import { useLayoutEffect, useContext, useEffect, useMemo } from "react";
import { DefaultOnchainKitProviders } from "./DefaultOnchainKitProviders.js";
import OnchainKitProviderBoundary from "./OnchainKitProviderBoundary.js";
import { DEFAULT_PRIVACY_URL, DEFAULT_TERMS_URL } from "./core/constants.js";
import { generateUUIDWithInsecureFallback } from "./internal/utils/crypto.js";
import { OnchainKitContext } from "./useOnchainKit.js";
import { useThemeRoot } from "./internal/hooks/useTheme.js";
import { clientMetaManager } from "./core/clientMeta/clientMetaManager.js";
import { MiniKitContext, MiniKitProvider } from "./minikit/MiniKitProvider.js";
import { useSessionStorage } from "usehooks-ts";
function OnchainKitProvider({
  analytics,
  apiKey,
  chain,
  children,
  config,
  projectId,
  rpcUrl,
  miniKit = {
    enabled: false
  },
  defaultPublicClients
}) {
  var _a, _b, _c;
  const [sessionId] = useSessionStorage("ock-session-id", generateUUIDWithInsecureFallback());
  const theme = useThemeRoot({
    theme: (_a = config == null ? void 0 : config.appearance) == null ? void 0 : _a.theme,
    mode: (_b = config == null ? void 0 : config.appearance) == null ? void 0 : _b.mode
  });
  useLayoutEffect(() => {
    document.documentElement.setAttribute("data-ock-theme", theme);
  }, [theme]);
  const isMiniKit = !!((_c = useContext(MiniKitContext)) == null ? void 0 : _c.__isMiniKit);
  useEffect(() => {
    if (clientMetaManager.isInitialized()) return;
    clientMetaManager.init({
      isMiniKit
    });
  }, [isMiniKit]);
  const value = useMemo(() => {
    var _a2, _b2, _c2, _d, _e, _f, _g, _h, _i, _j, _k, _l, _m, _n;
    const defaultPaymasterUrl = apiKey ? `https://api.developer.coinbase.com/rpc/v1/${chain.name.replace(" ", "-").toLowerCase()}/${apiKey}` : null;
    const onchainKitConfig = {
      apiKey: apiKey ?? null,
      chain,
      config: {
        analytics: analytics ?? true,
        analyticsUrl: (config == null ? void 0 : config.analyticsUrl) ?? null,
        appearance: {
          name: ((_a2 = config == null ? void 0 : config.appearance) == null ? void 0 : _a2.name) ?? "Dapp",
          logo: ((_b2 = config == null ? void 0 : config.appearance) == null ? void 0 : _b2.logo) ?? "",
          mode: ((_c2 = config == null ? void 0 : config.appearance) == null ? void 0 : _c2.mode) ?? "auto",
          theme: ((_d = config == null ? void 0 : config.appearance) == null ? void 0 : _d.theme) ?? "default"
        },
        paymaster: (config == null ? void 0 : config.paymaster) || defaultPaymasterUrl,
        wallet: {
          display: ((_e = config == null ? void 0 : config.wallet) == null ? void 0 : _e.display) ?? "classic",
          preference: ((_f = config == null ? void 0 : config.wallet) == null ? void 0 : _f.preference) ?? "all",
          termsUrl: ((_g = config == null ? void 0 : config.wallet) == null ? void 0 : _g.termsUrl) || DEFAULT_TERMS_URL,
          privacyUrl: ((_h = config == null ? void 0 : config.wallet) == null ? void 0 : _h.privacyUrl) || DEFAULT_PRIVACY_URL,
          supportedWallets: {
            rabby: ((_j = (_i = config == null ? void 0 : config.wallet) == null ? void 0 : _i.supportedWallets) == null ? void 0 : _j.rabby) ?? false,
            trust: ((_l = (_k = config == null ? void 0 : config.wallet) == null ? void 0 : _k.supportedWallets) == null ? void 0 : _l.trust) ?? false,
            frame: ((_n = (_m = config == null ? void 0 : config.wallet) == null ? void 0 : _m.supportedWallets) == null ? void 0 : _n.frame) ?? false
          }
        }
      },
      projectId: projectId ?? null,
      rpcUrl: rpcUrl ?? null,
      sessionId,
      defaultPublicClients,
      miniKit
    };
    setOnchainKitConfig(onchainKitConfig);
    return onchainKitConfig;
  }, [analytics, apiKey, chain, config, projectId, rpcUrl, sessionId, defaultPublicClients, miniKit]);
  return /* @__PURE__ */ jsx(OnchainKitContext.Provider, { value, children: /* @__PURE__ */ jsx(DefaultOnchainKitProviders, { children: /* @__PURE__ */ jsx(MiniKitProvider, { enabled: miniKit.enabled, notificationProxyUrl: miniKit.notificationProxyUrl, children: /* @__PURE__ */ jsx(OnchainKitProviderBoundary, { children }) }) }) });
}
export {
  OnchainKitProvider
};
