// sushi/_esm/evm/pool/sushiswap-v2/sushiswap-v2-pool.js
import invariant from 'tiny-invariant';
import { Amount } from '../../../generic/currency/amount.js';
import { Price } from '../../../generic/currency/price.js';
import { _997, _1000, FIVE, ONE, ZERO, } from '../../../generic/math/constants.js';
import { sqrt } from '../../../generic/math/sqrt.js';
import { getEvmChainById } from '../../chain/index.js';
import { isSushiSwapV2ChainId, SUSHISWAP_V2_FACTORY_ADDRESS, } from '../../config/features/sushiswap-v2.js';
import { EvmToken } from '../../currency/index.js';
import { InsufficientInputAmountError, InsufficientReservesError, } from '../../dex/index.js';
import { computeSushiSwapV2PoolAddress } from './compute-sushiswap-v2-pool-address.js';
import { sushiSwapV2PoolSchema } from './zod.js';
export class SushiSwapV2Pool {
    static getAddress(tokenA, tokenB) {
        if (!isSushiSwapV2ChainId(tokenA.chainId)) {
            throw new Error(`ChainId Error: SushiSwapV2 is not available on ${getEvmChainById(tokenA.chainId).name}`);
        }
        return computeSushiSwapV2PoolAddress({
            factoryAddress: SUSHISWAP_V2_FACTORY_ADDRESS[tokenA.chainId],
            tokenA,
            tokenB,
        });
    }
    constructor(amountA, amountB) {
        this.swapGasCost = 60000n;
        this.minLiquidity = 1000n;
        const amounts = (amountA.currency.sortsBefore(amountB.currency) // does safety checks
            ? [amountA, amountB]
            : [amountB, amountA]);
        this.liquidityToken = new EvmToken({
            chainId: amounts[0].currency.chainId,
            address: SushiSwapV2Pool.getAddress(amounts[0].currency, amounts[1].currency),
            decimals: 18,
            symbol: 'UNI-V2',
            name: 'Uniswap V2',
        });
        this.tokenAmounts = amounts;
    }
    /**
     * Returns true if the token is either token0 or token1
     * @param token to check
     */
    involvesToken(token) {
        return token.isSame(this.token0) || token.isSame(this.token1);
    }
    /**
     * Returns the current mid price of the pair in terms of token0, i.e. the ratio of reserve1 to reserve0
     */
    get token0Price() {
        const result = this.tokenAmounts[1].divToFraction(this.tokenAmounts[0].amount);
        return new Price({
            base: this.token0,
            quote: this.token1,
            denominator: result.denominator,
            numerator: result.numerator,
        });
    }
    /**
     * Returns the current mid price of the pair in terms of token1, i.e. the ratio of reserve0 to reserve1
     */
    get token1Price() {
        const result = this.tokenAmounts[0].divToFraction(this.tokenAmounts[1]);
        return new Price({
            base: this.token1,
            quote: this.token0,
            denominator: result.denominator,
            numerator: result.numerator,
        });
    }
    /**
     * Return the price of the given token in terms of the other token in the pair.
     * @param token token to return price of
     */
    priceOf(token) {
        invariant(this.involvesToken(token), 'TOKEN');
        return token.isSame(this.token0) ? this.token0Price : this.token1Price;
    }
    /**
     * Returns the chain ID of the tokens in the pair.
     */
    get chainId() {
        return this.token0.chainId;
    }
    get token0() {
        return this.tokenAmounts[0].currency;
    }
    get token1() {
        return this.tokenAmounts[1].currency;
    }
    get reserve0() {
        return this.tokenAmounts[0];
    }
    get reserve1() {
        return this.tokenAmounts[1];
    }
    reserveOf(token) {
        invariant(this.involvesToken(token), 'TOKEN');
        return token.isSame(this.token0) ? this.reserve0 : this.reserve1;
    }
    getOutputAmount(inputAmount) {
        invariant(this.involvesToken(inputAmount.currency), 'TOKEN');
        if (this.reserve0.amount === ZERO || this.reserve1.amount === ZERO) {
            throw new InsufficientReservesError();
        }
        const inputReserve = this.reserveOf(inputAmount.currency);
        const outputReserve = this.reserveOf(inputAmount.currency.isSame(this.token0) ? this.token1 : this.token0);
        const inputAmountWithFee = inputAmount.amount * _997;
        const numerator = inputAmountWithFee * outputReserve.amount;
        const denominator = inputReserve.amount * _1000 + inputAmountWithFee;
        const outputAmount = new Amount(inputAmount.currency.isSame(this.token0) ? this.token1 : this.token0, numerator / denominator);
        if (outputAmount.amount === ZERO) {
            throw new InsufficientInputAmountError();
        }
        return [
            outputAmount,
            new SushiSwapV2Pool(inputReserve.add(inputAmount), outputReserve.sub(outputAmount)),
        ];
    }
    getInputAmount(outputAmount) {
        invariant(this.involvesToken(outputAmount.currency), 'TOKEN');
        if (this.reserve0.amount === ZERO ||
            this.reserve1.amount === ZERO ||
            outputAmount.amount >= this.reserveOf(outputAmount.currency).amount) {
            throw new InsufficientReservesError();
        }
        const outputReserve = this.reserveOf(outputAmount.currency);
        const inputReserve = this.reserveOf(outputAmount.currency.isSame(this.token0) ? this.token1 : this.token0);
        const numerator = inputReserve.amount * outputAmount.amount * _1000;
        const denominator = (outputReserve.amount - outputAmount.amount) * _997;
        const inputAmount = new Amount(outputAmount.currency.isSame(this.token0) ? this.token1 : this.token0, numerator / denominator + ONE);
        return [
            inputAmount,
            new SushiSwapV2Pool(inputReserve.add(inputAmount), outputReserve.sub(outputAmount)),
        ];
    }
    getLiquidityMinted(totalSupply, tokenAmountA, tokenAmountB) {
        invariant(totalSupply.currency.isSame(this.liquidityToken), 'LIQUIDITY');
        const tokenAmounts = (tokenAmountA.currency.sortsBefore(tokenAmountB.currency) // does safety checks
            ? [tokenAmountA, tokenAmountB]
            : [tokenAmountB, tokenAmountA]);
        invariant(tokenAmounts[0].currency.isSame(this.token0) &&
            tokenAmounts[1].currency.isSame(this.token1), 'TOKEN');
        let liquidity;
        if (totalSupply.amount === ZERO) {
            liquidity =
                sqrt(tokenAmounts[0].amount * tokenAmounts[1].amount) -
                    this.minLiquidity;
        }
        else {
            const amount0 = (tokenAmounts[0].amount * totalSupply.amount) / this.reserve0.amount;
            const amount1 = (tokenAmounts[1].amount * totalSupply.amount) / this.reserve1.amount;
            liquidity = amount0 <= amount1 ? amount0 : amount1;
        }
        if (liquidity <= ZERO) {
            throw new InsufficientInputAmountError();
        }
        return new Amount(this.liquidityToken, liquidity);
    }
    getLiquidityValue(token, totalSupply, liquidity, feeOn = false, kLast) {
        invariant(this.involvesToken(token), 'TOKEN');
        invariant(totalSupply.currency.isSame(this.liquidityToken), 'TOTAL_SUPPLY');
        invariant(liquidity.currency.isSame(this.liquidityToken), 'LIQUIDITY');
        invariant(liquidity.amount <= totalSupply.amount, 'LIQUIDITY');
        let totalSupplyAdjusted;
        if (!feeOn) {
            totalSupplyAdjusted = totalSupply;
        }
        else {
            invariant(typeof kLast !== 'undefined', 'K_LAST');
            const kLastParsed = BigInt(typeof kLast === 'bigint' ? kLast.toString() : kLast);
            if (kLastParsed !== ZERO) {
                const rootK = sqrt(this.reserve0.amount * this.reserve1.amount);
                const rootKLast = sqrt(kLastParsed);
                if (rootK > rootKLast) {
                    const numerator = totalSupply.amount * (rootK - rootKLast);
                    const denominator = rootK * FIVE + rootKLast;
                    const feeLiquidity = numerator / denominator;
                    totalSupplyAdjusted = totalSupply.add(new Amount(this.liquidityToken, feeLiquidity));
                }
                else {
                    totalSupplyAdjusted = totalSupply;
                }
            }
            else {
                totalSupplyAdjusted = totalSupply;
            }
        }
        return new Amount(token, (liquidity.amount * this.reserveOf(token).amount) /
            totalSupplyAdjusted.amount);
    }
    serialize() {
        return sushiSwapV2PoolSchema.parse({
            reserve0: this.tokenAmounts[0].toJSON(),
            reserve1: this.tokenAmounts[1].toJSON(),
        });
    }
    static deserialize(pair) {
        const tokenA = new EvmToken(pair.reserve0.currency);
        const tokenB = new EvmToken(pair.reserve1.currency);
        const amountA = new Amount(tokenA, pair.reserve0.amount);
        const amountB = new Amount(tokenB, pair.reserve1.amount);
        return new SushiSwapV2Pool(amountA, amountB);
    }
}
