// sushi/_esm/kvm/currency/token.js
import * as z from 'zod';
import { Token } from '../../generic/currency/token.js';
import { isKvmChainId } from '../chain/chains.js';
export function isKvmTokenAddress(address) {
    if (address === 'coin')
        return true;
    // `${string}.${string}` - namespace.name - string parts can include alphanumeric characters, underscores, and dashes
    return !!address.match(/^([A-Za-z0-9_-]+)\.([A-Za-z0-9_-]+)$/);
}
export class KvmToken extends Token {
    constructor({ origin, ...rest }) {
        super(rest);
        this.origin = origin;
    }
    wrap() {
        return this;
    }
    toJSON() {
        return {
            chainId: this.chainId,
            address: this.address,
            symbol: this.symbol,
            name: this.name,
            decimals: this.decimals,
            type: this.type,
            metadata: this.metadata,
        };
    }
    static fromJSON(data) {
        return new this(data);
    }
}
export const serializedKvmTokenSchema = ({ metadata, } = {}) => z.object({
    chainId: z.number().int().refine(isKvmChainId),
    address: z.string().refine(isKvmTokenAddress),
    symbol: z.string(),
    name: z.string(),
    decimals: z.number().int().nonnegative(),
    type: z.literal('token'),
    metadata: (metadata ||
        z.record(z.unknown()).optional().default({})),
});
