// @zoralabs/protocol-deployments/src/generated/sparks.ts
export const chainConfigs = {
  1: {
    NONFUNGIBLE_POSITION_MANAGER: "0xC36442b4a4522E871399CD717aBDD847Ab11FE88",
    UNISWAP_SWAP_ROUTER: "0x68b3465833fb72A70ecDF485E0e4C7bD8665Fc45",
    PROXY_ADMIN: "0xDB392f4391462d60B8B4413ef72018Ab595Af9D0",
    ZORA_RECIPIENT: "0xd1d1D4e36117aB794ec5d4c78cBD3a8904E691D0",
    WETH: "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
  },
  10: {
    NONFUNGIBLE_POSITION_MANAGER: "0xC36442b4a4522E871399CD717aBDD847Ab11FE88",
    UNISWAP_SWAP_ROUTER: "0x68b3465833fb72A70ecDF485E0e4C7bD8665Fc45",
    PROXY_ADMIN: "0x4c7f7b6067fac9a737ecf2ca1a733fc85dd65a2b",
    ZORA_RECIPIENT: "0x7A810DCd0f8d83B20212326813Db6EF7E9FD030c",
    WETH: "0x4200000000000000000000000000000000000006",
  },
  8453: {
    NONFUNGIBLE_POSITION_MANAGER: "0x03a520b32C04BF3bEEf7BEb72E919cf822Ed34f1",
    UNISWAP_SWAP_ROUTER: "0x2626664c2603336E57B271c5C0b26F421741e481",
    UNISWAP_V3_FACTORY: "0x33128a8fC17869897dcE68Ed026d694621f6FDfD",
    UNISWAP_QUOTER_V2: "0x3d4e44Eb1374240CE5F1B871ab261CD16335B76a",
    DOPPLER_AIRLOCK: "0x660eAaEdEBc968f8f3694354FA8EC0b4c5Ba8D12",
    PROXY_ADMIN: "0x004d6611884B4A661749B64b2ADc78505c3e1AB3",
    ZORA_RECIPIENT: "0x7bf90111Ad7C22bec9E9dFf8A01A44713CC1b1B6",
    WETH: "0x4200000000000000000000000000000000000006",
    UNISWAP_V4_POOL_MANAGER: "0x498581ff718922c3f8e6a244956af099b2652b2b",
    UNISWAP_V4_POSITION_MANAGER: "0x7c5f5a4bbd8fd63184577525326123b519429bdc",
    UNISWAP_PERMIT2: "0x000000000022D473030F116dDEE9F6B43aC78BA3",
    UNISWAP_UNIVERSAL_ROUTER: "0x6ff5693b99212da76ad316178a184ab56d299b43",
  },
  42161: {
    NONFUNGIBLE_POSITION_MANAGER: "0xC36442b4a4522E871399CD717aBDD847Ab11FE88",
    UNISWAP_SWAP_ROUTER: "0x68b3465833fb72A70ecDF485E0e4C7bD8665Fc45",
    PROXY_ADMIN: "0xF7DafC329C93D84267c0E7B146C0bD68807f6A03",
    ZORA_RECIPIENT: "0x10Fb02A68C3EA8bD439A6f515dcE07CC049f5c94",
    WETH: "0x82af49447d8a07e3bd95bd0d56f35241523fbab1",
  },
  81457: {
    NONFUNGIBLE_POSITION_MANAGER: "0xB218e4f7cF0533d4696fDfC419A0023D33345F28",
    UNISWAP_SWAP_ROUTER: "0x549FEB8c9bd4c12Ad2AB27022dA12492aC452B66",
    PROXY_ADMIN: "0x5b297B1b87f8De28C9fA7AFe183Db9F9e6295523",
    ZORA_RECIPIENT: "0x72E6394056f9152D3a41e5A25De383c191D16134",
    WETH: "0x4300000000000000000000000000000000000004",
  },
  84532: {
    NONFUNGIBLE_POSITION_MANAGER: "0x27F971cb582BF9E50F397e4d29a5C7A34f11faA2",
    UNISWAP_SWAP_ROUTER: "0x94cC0AaC535CCDB3C01d6787D6413C739ae12bc4",
    UNISWAP_V3_FACTORY: "0x4752ba5DBc23f44D87826276BF6Fd6b1C372aD24",
    UNISWAP_QUOTER_V2: "0xC5290058841028F1614F3A6F0F5816cAd0df5E27",
    DOPPLER_AIRLOCK: "0xa24E35a5d71d02a59b41E7c93567626302da1958",
    PROXY_ADMIN: "0x5F14C23983c9e0840Dc60dA880349622f0785420",
    ZORA_RECIPIENT: "0x5F14C23983c9e0840Dc60dA880349622f0785420",
    WETH: "0x4200000000000000000000000000000000000006",
    UNISWAP_V4_POOL_MANAGER: "0x05E73354cFDd6745C338b50BcFDfA3Aa6fA03408",
    UNISWAP_V4_POSITION_MANAGER: "0x4b2c77d209d3405f41a037ec6c77f7f5b8e2ca80",
    UNISWAP_PERMIT2: "0x000000000022D473030F116dDEE9F6B43aC78BA3",
    UNISWAP_UNIVERSAL_ROUTER: "0x492E6456D9528771018DeB9E87ef7750EF184104",
  },
  421614: {
    NONFUNGIBLE_POSITION_MANAGER: "0x6b2937Bde17889EDCf8fbD8dE31C3C2a70Bc4d65",
    UNISWAP_SWAP_ROUTER: "0x101F443B4d1b059569D643917553c771E1b9663E",
    PROXY_ADMIN: "0x256537b56007b32BabEB44E868EC8AA5cBF03D30",
    ZORA_RECIPIENT: "0x256537b56007b32BabEB44E868EC8AA5cBF03D30",
    WETH: "0xc556bae1e86b2ae9c22ea5e036b07e55e7596074",
  },
  7777777: {
    NONFUNGIBLE_POSITION_MANAGER: "0xbC91e8DfA3fF18De43853372A3d7dfe585137D78",
    UNISWAP_SWAP_ROUTER: "0x7De04c96BE5159c3b5CeffC82aa176dc81281557",
    PROXY_ADMIN: "0xdEA20c96253dc2d64897D2b8d27A8d935dE74955",
    ZORA_RECIPIENT: "0xEcfc2ee50409E459c554a2b0376F882Ce916D853",
    WETH: "0x4200000000000000000000000000000000000006",
  },
  11155111: {
    NONFUNGIBLE_POSITION_MANAGER: "0x1238536071E1c677A632429e3655c799b22cDA52",
    UNISWAP_SWAP_ROUTER: "0x3bFA4769FB09eefC5a80d6E87c3B9C650f7Ae48E",
    PROXY_ADMIN: "0xCE9F2e8EaFa11637F8A1CB60AE8AaC601Ae30f2D",
    ZORA_RECIPIENT: "0xCE9F2e8EaFa11637F8A1CB60AE8AaC601Ae30f2D",
    WETH: "0x7b79995e5f793A07Bc00c21412e50Ecae098E7f9",
  },
  11155420: {
    NONFUNGIBLE_POSITION_MANAGER: "0xdA75cEf1C93078e8b736FCA5D5a30adb97C8957d",
    UNISWAP_SWAP_ROUTER: "0x94cC0AaC535CCDB3C01d6787D6413C739ae12bc4",
    PROXY_ADMIN: "0xFA3748b2dcF94a9CBdb5087333E9F093476e8389",
    ZORA_RECIPIENT: "0xFA3748b2dcF94a9CBdb5087333E9F093476e8389",
    WETH: "0x4200000000000000000000000000000000000006",
  },
  999999999: {
    NONFUNGIBLE_POSITION_MANAGER: "0xB8458EaAe43292e3c1F7994EFd016bd653d23c20",
    UNISWAP_SWAP_ROUTER: "0x6B36d761981d82B1e07cF3c4daF4cB4615c4850a",
    PROXY_ADMIN: "0xdae22ce69Afcb7f4bc37D32E267645722949DE0E",
    ZORA_RECIPIENT: "0xdae22ce69Afcb7f4bc37D32E267645722949DE0E",
    WETH: "0x4200000000000000000000000000000000000006",
  },
};
export const addresses = {
  10: {
    MINTS_ETH_UNWRAPPER_AND_CALLER:
      "0x0000000000000000000000000000000000000000",
    SPARKS_MANAGER_IMPL: "0x0000000000000000000000000000000000000000",
    SPARKS_MANAGER_IMPL_VERSION: "",
    SPONSORED_SPARKS_SPENDER: "0x35ca784918bf11692708c1D530691704AAcEA95E",
    SPONSORED_SPARKS_SPENDER_VERSION: "2.0.0",
  },
  8453: {
    MINTS_1155: "0x0000000000000000000000000000000000000000",
    MINTS_ETH_UNWRAPPER_AND_CALLER:
      "0x0000000000000000000000000000000000000000",
    MINTS_MANAGER: "0x0000000000000000000000000000000000000000",
    SPARKS_1155: "0x7777777b3eA6C126942BB14dD5C3C11D365C385D",
    SPARKS_MANAGER: "0x77777779acd6a96C0c57272112921A0b833c38fD",
    SPARKS_MANAGER_IMPL: "0x122648369E390009F18c1B2b9B100AfB09B982fc",
    SPARKS_MANAGER_IMPL_VERSION: "0.2.4",
    SPONSORED_SPARKS_SPENDER: "0x6958A8866A40118e2CC21ab4752353835321b5e6",
    SPONSORED_SPARKS_SPENDER_VERSION: "2.0.0",
  },
  42161: {
    MINTS_ETH_UNWRAPPER_AND_CALLER:
      "0x0000000000000000000000000000000000000000",
    SPARKS_MANAGER_IMPL: "0x0000000000000000000000000000000000000000",
    SPARKS_MANAGER_IMPL_VERSION: "",
    SPONSORED_SPARKS_SPENDER: "0x2fBdBc34B6015e7b40638179Aa05a2D2267452c7",
    SPONSORED_SPARKS_SPENDER_VERSION: "2.0.0",
  },
  81457: {
    MINTS_ETH_UNWRAPPER_AND_CALLER:
      "0x0000000000000000000000000000000000000000",
    SPARKS_MANAGER_IMPL: "0x0000000000000000000000000000000000000000",
    SPARKS_MANAGER_IMPL_VERSION: "",
    SPONSORED_SPARKS_SPENDER: "0x786F0f6aE196715Cd6d65693f06DA57426ac8463",
    SPONSORED_SPARKS_SPENDER_VERSION: "2.0.0",
  },
  84532: {
    MINTS_1155: "0x0000000000000000000000000000000000000000",
    MINTS_ETH_UNWRAPPER_AND_CALLER:
      "0x0000000000000000000000000000000000000000",
    MINTS_MANAGER: "0x0000000000000000000000000000000000000000",
    SPARKS_1155: "0x7777777b3eA6C126942BB14dD5C3C11D365C385D",
    SPARKS_MANAGER: "0x77777779acd6a96C0c57272112921A0b833c38fD",
    SPARKS_MANAGER_IMPL: "0x122648369E390009F18c1B2b9B100AfB09B982fc",
    SPARKS_MANAGER_IMPL_VERSION: "0.2.4",
    SPONSORED_SPARKS_SPENDER: "0x23150f60E233F519A01C735C29c1591f9E8353dE",
    SPONSORED_SPARKS_SPENDER_VERSION: "2.0.0",
  },
  7777777: {
    MINTS_ETH_UNWRAPPER_AND_CALLER:
      "0xb0994EB9520C98C97e1F3953a5964535C2bd271A",
    SPARKS_MANAGER_IMPL: "0xFB854463800d09B89682746c37A3AE696C92193D",
    SPARKS_MANAGER_IMPL_VERSION: "0.1.4",
    SPARKS_MANAGER: "0x77777779acd6a96C0c57272112921A0b833c38fD",
    SPARKS_1155: "0x7777777b3eA6C126942BB14dD5C3C11D365C385D",
    MINTS_MANAGER: "0x77777770cA269366c7208aFcF36FE2C6F7f7608B",
    MINTS_1155: "0x7777777d57c1C6e472fa379b7b3B6c6ba3835073",
    SPONSORED_SPARKS_SPENDER: "0x362Ef86acAD2FAa66A28Ee8d1F5a835ac0BBD11d",
    SPONSORED_SPARKS_SPENDER_VERSION: "2.0.0",
  },
  999999999: {
    MINTS_ETH_UNWRAPPER_AND_CALLER:
      "0xb0994EB9520C98C97e1F3953a5964535C2bd271A",
    SPARKS_MANAGER_IMPL: "0xf7D8d07267E840aE466E5f2fe8b089F4DE79c7da",
    SPARKS_MANAGER_IMPL_VERSION: "0.1.4",
    SPARKS_MANAGER: "0x77777779acd6a96C0c57272112921A0b833c38fD",
    SPARKS_1155: "0x7777777b3eA6C126942BB14dD5C3C11D365C385D",
    MINTS_MANAGER: "0x77777770cA269366c7208aFcF36FE2C6F7f7608B",
    MINTS_1155: "0x7777777d57c1C6e472fa379b7b3B6c6ba3835073",
    SPONSORED_SPARKS_SPENDER: "0x3A98313fB00918a81d98f9986aED089365428293",
    SPONSORED_SPARKS_SPENDER_VERSION: "2.0.0",
  },
};
