// @coinbase/agentkit/dist/action-providers/x402/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NETWORK_MAPPINGS = exports.SOLANA_USDC_ADDRESSES = exports.SUPPORTED_NETWORKS = exports.DEFAULT_FACILITATOR = exports.KNOWN_FACILITATORS = void 0;
/**
 * Known facilitator registry
 */
exports.KNOWN_FACILITATORS = {
    cdp: "https://api.cdp.coinbase.com/platform/v2/x402",
    payai: "https://facilitator.payai.network",
};
exports.DEFAULT_FACILITATOR = "cdp";
/**
 * Supported networks for x402 payment protocol
 */
exports.SUPPORTED_NETWORKS = [
    "base-mainnet",
    "base-sepolia",
    "solana-mainnet",
    "solana-devnet",
];
/**
 * USDC token addresses for Solana networks
 */
exports.SOLANA_USDC_ADDRESSES = {
    "solana-devnet": "4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU",
    "solana-mainnet": "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
};
/**
 * Network mapping from internal network ID to both v1 and v2 (CAIP-2) formats.
 * Used for filtering discovery results that may contain either format.
 */
exports.NETWORK_MAPPINGS = {
    "base-mainnet": ["base", "eip155:8453"],
    "base-sepolia": ["base-sepolia", "eip155:84532"],
    "solana-mainnet": ["solana", "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp"],
    "solana-devnet": ["solana-devnet", "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1"],
};
