// sushi/_cjs/evm/abi/sushiV3PoolAbi/sushiV3PoolAbi_ticks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sushiV3PoolAbi_ticks = void 0;
exports.sushiV3PoolAbi_ticks = [
    {
        inputs: [{ internalType: 'int24', name: '', type: 'int24' }],
        name: 'ticks',
        outputs: [
            { internalType: 'uint128', name: 'liquidityGross', type: 'uint128' },
            { internalType: 'int128', name: 'liquidityNet', type: 'int128' },
            {
                internalType: 'uint256',
                name: 'feeGrowthOutside0X128',
                type: 'uint256',
            },
            {
                internalType: 'uint256',
                name: 'feeGrowthOutside1X128',
                type: 'uint256',
            },
            { internalType: 'int56', name: 'tickCumulativeOutside', type: 'int56' },
            {
                internalType: 'uint160',
                name: 'secondsPerLiquidityOutsideX128',
                type: 'uint160',
            },
            { internalType: 'uint32', name: 'secondsOutside', type: 'uint32' },
            { internalType: 'bool', name: 'initialized', type: 'bool' },
        ],
        stateMutability: 'view',
        type: 'function',
    },
];
