// @coinbase/onchainkit/dist/nft/components/mint/NFTMintButton.js
import { jsx, jsxs, Fragment } from "react/jsx-runtime";
import { Spinner } from "../../../internal/components/Spinner.js";
import { useNFTLifecycleContext } from "../NFTLifecycleProvider.js";
import { useNFTContext } from "../NFTProvider.js";
import { useMintAnalytics } from "../../hooks/useMintAnalytics.js";
import { cn, text } from "../../../styles/theme.js";
import { Transaction } from "../../../transaction/components/Transaction.js";
import { TransactionButton } from "../../../transaction/components/TransactionButton.js";
import { TransactionSponsor } from "../../../transaction/components/TransactionSponsor.js";
import { TransactionStatus } from "../../../transaction/components/TransactionStatus.js";
import { TransactionStatusAction } from "../../../transaction/components/TransactionStatusAction.js";
import { TransactionStatusLabel } from "../../../transaction/components/TransactionStatusLabel.js";
import { useState, useCallback, useEffect, useMemo } from "react";
import "@radix-ui/react-toast";
import "../../../transaction/components/TransactionProvider.js";
import { useChainId, useAccount } from "wagmi";
import "wagmi/experimental";
import "../../../core/network/getChainExplorer.js";
import { ConnectWallet } from "../../../wallet/components/ConnectWallet.js";
import "../../../identity/components/IdentityProvider.js";
import "viem";
import "viem/chains";
import "viem/ens";
import "../../../identity/constants.js";
import "@tanstack/react-query";
import "../../../core/network/attestations.js";
import "../../../identity/utils/easSupportedChains.js";
import "../../../styles/constants.js";
import "../../../useOnchainKit.js";
import "@floating-ui/react";
import "@radix-ui/react-dialog";
import "../../../token/constants.js";
import "../../../internal/components/QrCode/useLogo.js";
import "qrcode";
import "../../../wallet/components/WalletProvider.js";
import "@farcaster/miniapp-sdk";
import "../../../internal/components/TextInput.js";
import "../../../token/components/TokenRow.js";
import "@radix-ui/react-dropdown-menu";
import "../../../internal/components/LayerConfigProvider.js";
import "../../../token/components/TokenSelectButton.js";
import "../../../swap/components/SwapProvider.js";
import "@radix-ui/react-popover";
import "../../../wallet/components/wallet-advanced-send/components/SendAmountInputTypeSwitch.js";
import "../../../wallet/components/wallet-advanced-send/components/SendProvider.js";
import "../../../fund/components/FundCardProvider.js";
import "../../../fund/components/FundCardPaymentMethodSelectRow.js";
import "../../../fund/components/FundCardPaymentMethodSelectorToggle.js";
import "../../../internal/svg/trustWalletSvg.js";
import "wagmi/connectors";
import "viem/account-abstraction";
function NFTMintButton({
  className,
  label = "Mint",
  disabled
}) {
  const chainId = useChainId();
  const {
    address
  } = useAccount();
  const {
    contractAddress,
    tokenId,
    network,
    isEligibleToMint,
    buildMintTransaction,
    isSponsored,
    quantity,
    name
  } = useNFTContext();
  const {
    updateLifecycleStatus
  } = useNFTLifecycleContext();
  const [callData, setCallData] = useState([]);
  const [mintError, setMintError] = useState(null);
  const {
    setTransactionState
  } = useMintAnalytics();
  const handleTransactionError = useCallback((error) => {
    updateLifecycleStatus({
      statusName: "error",
      statusData: {
        error: "Error building mint transaction",
        code: "NmNBc01",
        // NFT module NFTMintButton component 01 error
        message: error
      }
    });
    setMintError(error);
  }, [updateLifecycleStatus]);
  const fetchTransactions = useCallback(async () => {
    if (name && address && buildMintTransaction && isEligibleToMint) {
      try {
        setCallData([]);
        setMintError(null);
        setTransactionState("buildingTransaction");
        const mintTransaction = await buildMintTransaction({
          takerAddress: address,
          contractAddress,
          tokenId,
          network,
          quantity
        });
        setCallData(mintTransaction);
      } catch (error) {
        handleTransactionError(error);
      }
    } else {
      setCallData([]);
    }
  }, [address, buildMintTransaction, contractAddress, handleTransactionError, isEligibleToMint, name, network, quantity, setTransactionState, tokenId]);
  useEffect(() => {
    fetchTransactions();
  }, [fetchTransactions]);
  const handleOnStatus = useCallback((transactionStatus) => {
    setTransactionState(transactionStatus.statusName);
    if (transactionStatus.statusName === "transactionPending") {
      updateLifecycleStatus({
        statusName: "transactionPending"
      });
    }
    if (transactionStatus.statusName === "transactionLegacyExecuted" || transactionStatus.statusName === "success" || transactionStatus.statusName === "error") {
      updateLifecycleStatus(transactionStatus);
    }
  }, [updateLifecycleStatus, setTransactionState]);
  const transactionButtonLabel = useMemo(() => {
    if (isEligibleToMint === false || mintError) {
      return "Minting not available";
    }
    if (callData.length === 0) {
      return /* @__PURE__ */ jsx(Spinner, {});
    }
    return label;
  }, [callData, isEligibleToMint, label, mintError]);
  if (!buildMintTransaction) {
    return null;
  }
  if (!address) {
    return /* @__PURE__ */ jsx(ConnectWallet, { className: cn("ock:w-full", className) });
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsxs(Transaction, { isSponsored, chainId, calls: callData, onStatus: handleOnStatus, children: [
      /* @__PURE__ */ jsx(TransactionButton, { className, text: transactionButtonLabel, disabled: disabled || transactionButtonLabel !== label }),
      !mintError && /* @__PURE__ */ jsx(TransactionSponsor, {}),
      /* @__PURE__ */ jsxs(TransactionStatus, { children: [
        /* @__PURE__ */ jsx(TransactionStatusLabel, {}),
        /* @__PURE__ */ jsx(TransactionStatusAction, {})
      ] })
    ] }),
    mintError && /* @__PURE__ */ jsx("div", { className: cn(text.label2, "ock:text-ock-foreground-muted", "ock:pb-2"), children: mintError })
  ] });
}
export {
  NFTMintButton
};
