// @coinbase/onchainkit/dist/swap/components/SwapSettingsSlippageInput.js
'use client';
import { jsxs, jsx } from "react/jsx-runtime";
import { useAnalytics } from "../../core/analytics/hooks/useAnalytics.js";
import { SwapEvent } from "../../core/analytics/types.js";
import { useState, useCallback, useId } from "react";
import { cn, text, pressable } from "../../styles/theme.js";
import { useSwapContext } from "./SwapProvider.js";
import { TextInput } from "../../internal/components/TextInput.js";
import { SLIPPAGE_SETTINGS } from "../constants.js";
function SwapSettingsSlippageInput({
  className,
  render
}) {
  const {
    sendAnalytics
  } = useAnalytics();
  const {
    config: {
      maxSlippage: defaultMaxSlippage
    },
    updateLifecycleStatus,
    lifecycleStatus
  } = useSwapContext();
  const [slippageSetting, setSlippageSetting] = useState(lifecycleStatus.statusData.maxSlippage === defaultMaxSlippage ? SLIPPAGE_SETTINGS.AUTO : SLIPPAGE_SETTINGS.CUSTOM);
  const handleAnalyticsSlippageChange = useCallback((previousSlippage, newSlippage) => {
    sendAnalytics(SwapEvent.SlippageChanged, {
      previousSlippage,
      slippage: newSlippage
    });
  }, [sendAnalytics]);
  const updateSlippage = useCallback((newSlippage) => {
    const currentSlippage = lifecycleStatus.statusData.maxSlippage;
    if (newSlippage !== currentSlippage) {
      handleAnalyticsSlippageChange(currentSlippage, newSlippage);
      if (slippageSetting === SLIPPAGE_SETTINGS.AUTO) {
        setSlippageSetting(SLIPPAGE_SETTINGS.CUSTOM);
      }
      updateLifecycleStatus({
        statusName: "slippageChange",
        statusData: {
          maxSlippage: newSlippage
        }
      });
    }
  }, [lifecycleStatus.statusData.maxSlippage, updateLifecycleStatus, handleAnalyticsSlippageChange, slippageSetting]);
  const handleSlippageChange = useCallback((newSlippage) => {
    const parsedSlippage = Number.parseFloat(newSlippage);
    const isValidNumber = !Number.isNaN(parsedSlippage);
    updateSlippage(isValidNumber ? parsedSlippage : 0);
  }, [updateSlippage]);
  const handleSlippageSettingChange = useCallback((setting) => {
    setSlippageSetting(setting);
    if (setting === SLIPPAGE_SETTINGS.AUTO) {
      updateSlippage(defaultMaxSlippage);
    }
  }, [defaultMaxSlippage, updateSlippage]);
  const slippageInputId = useId();
  if (render) {
    return render({
      slippageSetting,
      setSlippageSetting: handleSlippageSettingChange,
      setSlippageValue: updateSlippage
    });
  }
  return /* @__PURE__ */ jsxs("section", { className: cn("ock:bg-ock-background", "ock:border-ock-background-active", "ock:rounded-ock-default", "ock:flex ock:items-center ock:gap-2 ock:flex-grow ock:max-sm:pt-4", className), children: [
    /* @__PURE__ */ jsxs("fieldset", { className: cn("ock:bg-ock-background", "ock:border-ock-background-active", "ock:rounded-ock-default", "ock:flex ock:h-9 ock:flex-1 ock:rounded-xl ock:border ock:p-1"), children: [
      /* @__PURE__ */ jsx("legend", { className: "ock:sr-only", children: "Slippage Setting" }),
      Object.values(SLIPPAGE_SETTINGS).map((setting) => /* @__PURE__ */ jsx("button", { type: "button", "aria-current": slippageSetting === setting ? "true" : void 0, className: cn(
        pressable.default,
        "ock:text-ock-foreground",
        text.label1,
        "ock:rounded-ock-inner",
        "ock:flex-1 ock:px-3 ock:py-1 ock:transition-colors",
        // Highlight the button if it is selected
        slippageSetting === setting ? cn("bg-ock-background-inverse", "text-ock-primary", "shadow-ock-default") : "ock:text-ock-foreground-muted"
      ), onClick: () => handleSlippageSettingChange(setting), children: setting }, setting))
    ] }),
    /* @__PURE__ */ jsxs("div", { className: cn("ock:bg-ock-background", "ock:border-ock-background-active", "ock:rounded-ock-default", "ock:flex ock:h-9 ock:w-24 ock:items-center ock:justify-between ock:border ock:px-2 ock:py-1", slippageSetting === SLIPPAGE_SETTINGS.AUTO && "ock:opacity-50"), children: [
      /* @__PURE__ */ jsx("label", { htmlFor: slippageInputId, className: "ock:sr-only", children: "Slippage Percentage" }),
      /* @__PURE__ */ jsx(TextInput, { id: slippageInputId, type: "text", value: String(lifecycleStatus.statusData.maxSlippage), onChange: handleSlippageChange, placeholder: "0", disabled: slippageSetting === SLIPPAGE_SETTINGS.AUTO, className: cn("ock:text-ock-foreground ock:truncate", text.label2, "ock:w-full ock:flex-grow ock:bg-transparent ock:pl-1 ock:font-normal ock:leading-6 ock:focus:outline-none", slippageSetting === SLIPPAGE_SETTINGS.AUTO && "ock:cursor-not-allowed") }),
      /* @__PURE__ */ jsx("span", { className: cn("ock:bg-ock-background", "ock:text-ock-foreground", text.label2, "ock:ml-1 ock:flex-shrink-0 ock:font-normal ock:leading-6"), children: "%" })
    ] })
  ] });
}
export {
  SwapSettingsSlippageInput
};
