// @coinbase/onchainkit/dist/identity/index.js
import { Address } from "./components/Address.js";
import { Avatar } from "./components/Avatar.js";
import { Badge } from "./components/Badge.js";
import { EthBalance } from "./components/EthBalance.js";
import { Identity } from "./components/Identity.js";
import { Name } from "./components/Name.js";
import { Socials } from "./components/Socials.js";
import { isBasename } from "./utils/isBasename.js";
import { IdentityCard } from "./components/IdentityCard.js";
import { getAddress } from "./utils/getAddress.js";
import { getAddresses } from "./utils/getAddresses.js";
import { getAttestations } from "./utils/getAttestations.js";
import { getAvatar } from "./utils/getAvatar.js";
import { getAvatars } from "./utils/getAvatars.js";
import { getName } from "./utils/getName.js";
import { getNames } from "./utils/getNames.js";
import { useAddress } from "./hooks/useAddress.js";
import { useAddresses } from "./hooks/useAddresses.js";
import { useAttestations } from "./hooks/useAttestations.js";
import { useAvatar } from "./hooks/useAvatar.js";
import { useAvatars } from "./hooks/useAvatars.js";
import { useName } from "./hooks/useName.js";
import { useNames } from "./hooks/useNames.js";
export {
  Address,
  Avatar,
  Badge,
  EthBalance,
  Identity,
  IdentityCard,
  Name,
  Socials,
  getAddress,
  getAddresses,
  getAttestations,
  getAvatar,
  getAvatars,
  getName,
  getNames,
  isBasename,
  useAddress,
  useAddresses,
  useAttestations,
  useAvatar,
  useAvatars,
  useName,
  useNames
};
