// clanker-sdk/dist/cli/utils/output.js
// src/cli/utils/style.ts
var isColorSupported = process.env.NO_COLOR === void 0 && process.env.FORCE_COLOR !== "0" && process.env.TERM !== "dumb" && (process.env.FORCE_COLOR !== void 0 || process.stdout.isTTY === true);
var ESC = "\x1B[";
var RST = `${ESC}0m`;
function wrap(code, text) {
  if (!isColorSupported) return text;
  return `${ESC}${code}m${text}${RST}`;
}
var bold = (s) => wrap("1", s);
var dim = (s) => wrap("2", s);
var red = (s) => wrap("31", s);
var green = (s) => wrap("32", s);
var yellow = (s) => wrap("33", s);
var cyan = (s) => wrap("36", s);
var gray = (s) => wrap("90", s);
var brightCyan = (s) => wrap("96", s);
function gradient(text, from, to) {
  if (!isColorSupported) return text;
  const chars = [...text];
  const total = chars.length;
  if (total === 0) return "";
  return chars.map((ch, i) => {
    if (ch === " ") return ch;
    const t = total > 1 ? i / (total - 1) : 0;
    const r = Math.round(from[0] + (to[0] - from[0]) * t);
    const g = Math.round(from[1] + (to[1] - from[1]) * t);
    const b = Math.round(from[2] + (to[2] - from[2]) * t);
    return `${ESC}38;2;${r};${g};${b}m${ch}`;
  }).join("") + RST;
}
var BRAND_FROM = [6, 182, 212];
var BRAND_TO = [168, 85, 247];
var SYM = {
  check: isColorSupported ? "\u2714" : "+",
  cross: isColorSupported ? "\u2716" : "x",
  arrow: isColorSupported ? "\u25B8" : ">",
  dot: isColorSupported ? "\u25CF" : "*",
  star: isColorSupported ? "\u2605" : "*",
  info: isColorSupported ? "\u2139" : "i",
  warn: isColorSupported ? "\u26A0" : "!",
  rocket: isColorSupported ? "\u{1F680}" : "^"
};
var BOX = {
  tl: "\u256D",
  tr: "\u256E",
  bl: "\u2570",
  br: "\u256F",
  h: "\u2500",
  v: "\u2502",
  lt: "\u251C",
  rt: "\u2524"
};
function stripAnsi(str) {
  return str.replace(/\x1b\[[0-9;]*m/g, "");
}
function drawBox(lines, options) {
  const color = options?.color || cyan;
  const pad = options?.padding ?? 1;
  const contentWidth = Math.max(
    ...lines.map((l) => stripAnsi(l).length),
    options?.title ? stripAnsi(options.title).length : 0,
    20
  );
  const innerWidth = contentWidth + pad * 2;
  const out = [];
  out.push(color(`${BOX.tl}${BOX.h.repeat(innerWidth)}${BOX.tr}`));
  if (options?.title) {
    const titleLen = stripAnsi(options.title).length;
    const titlePad = innerWidth - titleLen - pad;
    out.push(
      `${color(BOX.v)}${" ".repeat(pad)}${options.title}${" ".repeat(Math.max(0, titlePad))}${color(BOX.v)}`
    );
    out.push(color(`${BOX.lt}${BOX.h.repeat(innerWidth)}${BOX.rt}`));
  }
  for (const line of lines) {
    const lineLen = stripAnsi(line).length;
    const linePad = innerWidth - lineLen - pad;
    out.push(
      `${color(BOX.v)}${" ".repeat(pad)}${line}${" ".repeat(Math.max(0, linePad))}${color(BOX.v)}`
    );
  }
  out.push(color(`${BOX.bl}${BOX.h.repeat(innerWidth)}${BOX.br}`));
  return out.join("\n");
}
function separator(width = 50) {
  if (!isColorSupported) return BOX.h.repeat(width);
  return dim(gray(BOX.h.repeat(width)));
}

// src/cli/utils/output.ts
var BANNER_ART = [
  "   ____ _       _    _   _ _  _______ ____  ",
  "  / ___| |     / \\  | \\ | | |/ / ____|  _ \\ ",
  " | |   | |    / _ \\ |  \\| | ' /|  _| | |_) |",
  " | |___| |___/ ___ \\| |\\  | . \\| |___|  _ < ",
  "  \\____|_____/_/   \\_\\_| \\_|_|\\_\\_____|_| \\_\\"
];
function getBanner(version) {
  const lines = [""];
  for (const line of BANNER_ART) {
    lines.push(`  ${gradient(line, BRAND_FROM, BRAND_TO)}`);
  }
  lines.push("");
  lines.push(`    ${dim(gray("Deploy tokens on EVM chains"))}  ${dim(cyan(`v${version}`))}`);
  lines.push(`  ${separator(48)}`);
  lines.push("");
  return lines.join("\n");
}
function printBanner(version) {
  console.log(getBanner(version));
}
function blockExplorerUrl(chain, hash, type = "tx") {
  const explorer = chain.blockExplorers?.default?.url;
  if (!explorer) return hash;
  return `${explorer}/${type}/${hash}`;
}
function printResult(data, jsonMode) {
  if (jsonMode) {
    console.log(JSON.stringify(data, bigintReplacer, 2));
    return;
  }
  const lines = Object.entries(data).map(([key, value]) => {
    return `${cyan(key)} ${dim(BOX.v)} ${formatValue(value)}`;
  });
  console.log("");
  console.log(drawBox(lines, { color: cyan }));
  console.log("");
}
function printSuccess(message, jsonMode, data) {
  if (jsonMode) {
    console.log(JSON.stringify({ success: true, message, ...data }, bigintReplacer, 2));
    return;
  }
  const title = `${green(bold(SYM.check))} ${green(bold(message))}`;
  if (data) {
    const lines = Object.entries(data).map(([key, value]) => {
      return `${brightCyan(key)} ${dim(BOX.v)} ${formatValue(value)}`;
    });
    console.log("");
    console.log(drawBox(lines, { title, color: green }));
    console.log("");
  } else {
    console.log(`
  ${title}
`);
  }
}
function printError(error, jsonMode) {
  const message = error instanceof Error ? error.message : String(error);
  if (jsonMode) {
    console.error(JSON.stringify({ success: false, error: message }));
    return;
  }
  console.error(`
  ${red(bold(SYM.cross))} ${red(bold("Error:"))} ${message}
`);
}
function printStep(message, jsonMode) {
  if (jsonMode) return;
  console.log(`  ${cyan(SYM.arrow)} ${message}`);
}
function printInfo(label, value, jsonMode) {
  if (jsonMode) return;
  console.log(`  ${dim(gray(SYM.dot))} ${dim(label)} ${value}`);
}
function printWarning(message, jsonMode) {
  if (jsonMode) return;
  console.log(`  ${yellow(SYM.warn)}  ${yellow(message)}`);
}
function printKeyValue(data, jsonMode, title) {
  if (jsonMode) {
    console.log(JSON.stringify(data, bigintReplacer, 2));
    return;
  }
  const lines = [];
  for (const [key, value] of Object.entries(data)) {
    if (typeof value === "object" && value !== null && !Array.isArray(value)) {
      lines.push(bold(cyan(key)));
      for (const [subKey, subVal] of Object.entries(value)) {
        lines.push(`  ${gray(subKey)} ${dim(BOX.v)} ${formatValue(subVal)}`);
      }
    } else {
      lines.push(`${cyan(key)} ${dim(BOX.v)} ${formatValue(value)}`);
    }
  }
  console.log("");
  console.log(drawBox(lines, { color: cyan, title }));
  console.log("");
}
function formatValue(value) {
  if (typeof value === "bigint") return value.toString();
  if (typeof value === "object" && value !== null) return JSON.stringify(value, bigintReplacer);
  return String(value);
}
function bigintReplacer(_key, value) {
  return typeof value === "bigint" ? value.toString() : value;
}
export {
  blockExplorerUrl,
  getBanner,
  printBanner,
  printError,
  printInfo,
  printKeyValue,
  printResult,
  printStep,
  printSuccess,
  printWarning
};
