// clanker-sdk/dist/cli/utils/style.js
// src/cli/utils/style.ts
var isColorSupported = process.env.NO_COLOR === void 0 && process.env.FORCE_COLOR !== "0" && process.env.TERM !== "dumb" && (process.env.FORCE_COLOR !== void 0 || process.stdout.isTTY === true);
var ESC = "\x1B[";
var RST = `${ESC}0m`;
function wrap(code, text) {
  if (!isColorSupported) return text;
  return `${ESC}${code}m${text}${RST}`;
}
var bold = (s) => wrap("1", s);
var dim = (s) => wrap("2", s);
var italic = (s) => wrap("3", s);
var underline = (s) => wrap("4", s);
var red = (s) => wrap("31", s);
var green = (s) => wrap("32", s);
var yellow = (s) => wrap("33", s);
var blue = (s) => wrap("34", s);
var magenta = (s) => wrap("35", s);
var cyan = (s) => wrap("36", s);
var white = (s) => wrap("37", s);
var gray = (s) => wrap("90", s);
var brightCyan = (s) => wrap("96", s);
var brightGreen = (s) => wrap("92", s);
var brightMagenta = (s) => wrap("95", s);
function rgb(r, g, b, text) {
  if (!isColorSupported) return text;
  return `${ESC}38;2;${r};${g};${b}m${text}${RST}`;
}
function gradient(text, from, to) {
  if (!isColorSupported) return text;
  const chars = [...text];
  const total = chars.length;
  if (total === 0) return "";
  return chars.map((ch, i) => {
    if (ch === " ") return ch;
    const t = total > 1 ? i / (total - 1) : 0;
    const r = Math.round(from[0] + (to[0] - from[0]) * t);
    const g = Math.round(from[1] + (to[1] - from[1]) * t);
    const b = Math.round(from[2] + (to[2] - from[2]) * t);
    return `${ESC}38;2;${r};${g};${b}m${ch}`;
  }).join("") + RST;
}
var BRAND_FROM = [6, 182, 212];
var BRAND_TO = [168, 85, 247];
function brandGradient(text) {
  return gradient(text, BRAND_FROM, BRAND_TO);
}
var SYM = {
  check: isColorSupported ? "\u2714" : "+",
  cross: isColorSupported ? "\u2716" : "x",
  arrow: isColorSupported ? "\u25B8" : ">",
  dot: isColorSupported ? "\u25CF" : "*",
  star: isColorSupported ? "\u2605" : "*",
  info: isColorSupported ? "\u2139" : "i",
  warn: isColorSupported ? "\u26A0" : "!",
  rocket: isColorSupported ? "\u{1F680}" : "^"
};
var BOX = {
  tl: "\u256D",
  tr: "\u256E",
  bl: "\u2570",
  br: "\u256F",
  h: "\u2500",
  v: "\u2502",
  lt: "\u251C",
  rt: "\u2524"
};
function stripAnsi(str) {
  return str.replace(/\x1b\[[0-9;]*m/g, "");
}
function drawBox(lines, options) {
  const color = options?.color || cyan;
  const pad = options?.padding ?? 1;
  const contentWidth = Math.max(
    ...lines.map((l) => stripAnsi(l).length),
    options?.title ? stripAnsi(options.title).length : 0,
    20
  );
  const innerWidth = contentWidth + pad * 2;
  const out = [];
  out.push(color(`${BOX.tl}${BOX.h.repeat(innerWidth)}${BOX.tr}`));
  if (options?.title) {
    const titleLen = stripAnsi(options.title).length;
    const titlePad = innerWidth - titleLen - pad;
    out.push(
      `${color(BOX.v)}${" ".repeat(pad)}${options.title}${" ".repeat(Math.max(0, titlePad))}${color(BOX.v)}`
    );
    out.push(color(`${BOX.lt}${BOX.h.repeat(innerWidth)}${BOX.rt}`));
  }
  for (const line of lines) {
    const lineLen = stripAnsi(line).length;
    const linePad = innerWidth - lineLen - pad;
    out.push(
      `${color(BOX.v)}${" ".repeat(pad)}${line}${" ".repeat(Math.max(0, linePad))}${color(BOX.v)}`
    );
  }
  out.push(color(`${BOX.bl}${BOX.h.repeat(innerWidth)}${BOX.br}`));
  return out.join("\n");
}
function separator(width = 50) {
  if (!isColorSupported) return BOX.h.repeat(width);
  return dim(gray(BOX.h.repeat(width)));
}
export {
  BOX,
  BRAND_FROM,
  BRAND_TO,
  SYM,
  blue,
  bold,
  brandGradient,
  brightCyan,
  brightGreen,
  brightMagenta,
  cyan,
  dim,
  drawBox,
  gradient,
  gray,
  green,
  italic,
  magenta,
  red,
  rgb,
  separator,
  stripAnsi,
  underline,
  white,
  yellow
};
