// clanker-sdk/dist/cli/utils/wallet.js
// src/cli/utils/wallet.ts
import {
  createPublicClient,
  createWalletClient,
  http
} from "viem";
import { privateKeyToAccount } from "viem/accounts";

// src/cli/utils/chains.ts
import { arbitrum, base, baseSepolia, bsc, mainnet, unichain } from "viem/chains";

// src/utils/chains/monad.ts
import { defineChain } from "viem";
var monad = /* @__PURE__ */ defineChain({
  id: 143,
  name: "Monad",
  blockTime: 400,
  nativeCurrency: {
    name: "MON",
    symbol: "MON",
    decimals: 18
  },
  rpcUrls: {
    default: {
      http: ["https://rpc.monad.xyz"]
    }
  },
  blockExplorers: {
    default: {
      name: "Monad explorer",
      url: "TODO"
    }
  },
  contracts: {},
  testnet: false
});

// src/cli/utils/chains.ts
var CHAIN_MAP = {
  base,
  "base-sepolia": baseSepolia,
  arbitrum,
  ethereum: mainnet,
  bsc,
  unichain,
  monad
};
var CHAIN_NAMES = Object.keys(CHAIN_MAP);
function resolveChain(name) {
  const chain = CHAIN_MAP[name];
  if (!chain) {
    throw new Error(`Unknown chain "${name}". Supported: ${CHAIN_NAMES.join(", ")}`);
  }
  return chain;
}

// src/cli/utils/config.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
var CONFIG_DIR = join(homedir(), ".clanker");
var CONFIG_PATH = join(CONFIG_DIR, "config.json");
function loadConfig() {
  try {
    if (!existsSync(CONFIG_PATH)) return {};
    return JSON.parse(readFileSync(CONFIG_PATH, "utf-8"));
  } catch {
    return {};
  }
}

// src/cli/utils/wallet.ts
function resolvePrivateKey(opts) {
  const key = opts.privateKey || process.env.PRIVATE_KEY || loadConfig().privateKey;
  if (!key) {
    throw new Error(
      "Private key required. Run `clanker setup`, pass --private-key, or set PRIVATE_KEY env var."
    );
  }
  return key;
}
function resolveRpcUrl(opts, chainName) {
  if (opts.rpc) return opts.rpc;
  const config = loadConfig();
  return config.rpc?.[chainName];
}
function resolveClients(opts) {
  const chainName = opts.chain || loadConfig().defaultChain || "base";
  const chain = resolveChain(chainName);
  const rpcUrl = resolveRpcUrl(opts, chainName);
  const transport = rpcUrl ? http(rpcUrl) : http();
  const account = privateKeyToAccount(resolvePrivateKey(opts));
  const walletClient = createWalletClient({ account, chain, transport });
  const publicClient = createPublicClient({ chain, transport });
  return { chain, account, walletClient, publicClient };
}
function resolvePublicClient(opts) {
  const chainName = opts.chain || loadConfig().defaultChain || "base";
  const chain = resolveChain(chainName);
  const rpcUrl = resolveRpcUrl(opts, chainName);
  const transport = rpcUrl ? http(rpcUrl) : http();
  const publicClient = createPublicClient({ chain, transport });
  return { chain, publicClient };
}
export {
  resolveClients,
  resolvePublicClient
};
