// sushi/_esm/tvm/chain/chains.js
import { defineTvmChain } from './define-chain.js';
const tron = defineTvmChain({
    chainId: -2,
    name: 'Tron',
    shortName: 'trx',
    key: 'tron',
    blockExplorers: {
        default: { name: 'Tronscan', url: 'https://tronscan.io' },
    },
    netType: 'mainnet',
});
export const tvmChains = [tron];
// TvmChainId
export const tvmChainIds = /* @__PURE__ */ tvmChains.map((chain) => chain.chainId);
export const TvmChainId = /* @__PURE__ */ new Proxy({}, {
    get: (_, key) => {
        const _key = key.toLowerCase().replace(/_/g, '-');
        const chain = tvmChains.find((chain) => chain.key === _key);
        if (!chain)
            return undefined;
        return chain.chainId;
    },
});
export function isTvmChainId(chainId) {
    return tvmChains.some((chain) => chain.chainId === chainId);
}
export function getTvmChainById(chainId) {
    return tvmChains.find((chain) => chain.chainId === chainId);
}
export function isTvmMainnetChainId(chainId) {
    return tvmChains.some((chain) => chain.chainId === chainId && chain.netType === 'mainnet');
}
export function isTvmChainKey(key) {
    return tvmChains.some((chain) => chain.key === key);
}
export function getTvmChainByKey(key) {
    return tvmChains.find((chain) => chain.key === key);
}
