// @coinbase/onchainkit/dist/minikit/MiniKitProvider.js
'use client';
import { jsx } from "react/jsx-runtime";
import sdk from "@farcaster/miniapp-sdk";
import { createContext, useState, useCallback, useEffect, useMemo } from "react";
import { AutoConnect } from "./components/AutoConnect.js";
const MiniKitContext = createContext({
  enabled: false,
  context: null,
  updateClientContext: () => {
  },
  notificationProxyUrl: "",
  __isMiniKit: false
});
function MiniKitProviderContent({
  children,
  notificationProxyUrl = "/api/notify",
  autoConnect = true
}) {
  const [context, setContext] = useState(null);
  const updateClientContext = useCallback(({
    details,
    miniAppAdded
  }) => {
    setContext((prevContext) => {
      if (!prevContext) {
        return null;
      }
      return {
        ...prevContext,
        client: {
          ...prevContext.client,
          notificationDetails: details ?? void 0,
          added: miniAppAdded ?? prevContext.client.added
        }
      };
    });
  }, []);
  useEffect(() => {
    sdk.on("miniAppAdded", ({
      notificationDetails
    }) => {
      if (notificationDetails) {
        updateClientContext({
          details: notificationDetails,
          miniAppAdded: true
        });
      }
    });
    sdk.on("miniAppAddRejected", ({
      reason
    }) => {
      console.error("MiniApp add rejected", reason);
    });
    sdk.on("miniAppRemoved", () => {
      updateClientContext({
        details: void 0,
        miniAppAdded: false
      });
    });
    sdk.on("notificationsEnabled", ({
      notificationDetails
    }) => {
      updateClientContext({
        details: notificationDetails
      });
    });
    sdk.on("notificationsDisabled", () => {
      updateClientContext({
        details: void 0
      });
    });
    async function fetchContext() {
      try {
        const context2 = await sdk.context;
        setContext(context2);
      } catch (error) {
        console.error("Error fetching context:", error);
      }
    }
    fetchContext();
    return () => {
      sdk.removeAllListeners();
    };
  }, [updateClientContext]);
  const value = useMemo(() => {
    return {
      enabled: true,
      context,
      updateClientContext,
      notificationProxyUrl,
      __isMiniKit: true
    };
  }, [updateClientContext, notificationProxyUrl, context]);
  return /* @__PURE__ */ jsx(MiniKitContext.Provider, { value, children: /* @__PURE__ */ jsx(AutoConnect, { enabled: autoConnect, children }) });
}
function MiniKitProvider({
  children,
  notificationProxyUrl,
  enabled,
  autoConnect
}) {
  if (!enabled) {
    return children;
  }
  return /* @__PURE__ */ jsx(MiniKitProviderContent, { notificationProxyUrl, autoConnect, children });
}
export {
  MiniKitContext,
  MiniKitProvider
};
