// sushi/_cjs/tvm/abi/routerAbi/routerAbi_swapTokensForExactETH.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.routerAbi_swapTokensForExactETH = void 0;
exports.routerAbi_swapTokensForExactETH = {
    inputs: [
        {
            internalType: 'uint256',
            name: 'amountOut',
            type: 'uint256',
        },
        {
            internalType: 'uint256',
            name: 'amountInMax',
            type: 'uint256',
        },
        {
            internalType: 'address[]',
            name: 'path',
            type: 'address[]',
        },
        {
            internalType: 'address',
            name: 'to',
            type: 'address',
        },
        {
            internalType: 'uint256',
            name: 'deadline',
            type: 'uint256',
        },
    ],
    name: 'swapTokensForExactETH',
    outputs: [
        {
            internalType: 'uint256[]',
            name: 'amounts',
            type: 'uint256[]',
        },
    ],
    stateMutability: 'nonpayable',
    type: 'function',
};
