// sushi/_cjs/tvm/abi/routerAbi/routerAbi_addLiquidityETH.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.routerAbi_addLiquidityETH = void 0;
exports.routerAbi_addLiquidityETH = {
    inputs: [
        {
            internalType: 'address',
            name: 'token',
            type: 'address',
        },
        {
            internalType: 'uint256',
            name: 'amountTokenDesired',
            type: 'uint256',
        },
        {
            internalType: 'uint256',
            name: 'amountTokenMin',
            type: 'uint256',
        },
        {
            internalType: 'uint256',
            name: 'amountETHMin',
            type: 'uint256',
        },
        {
            internalType: 'address',
            name: 'to',
            type: 'address',
        },
        {
            internalType: 'uint256',
            name: 'deadline',
            type: 'uint256',
        },
    ],
    name: 'addLiquidityETH',
    outputs: [
        {
            internalType: 'uint256',
            name: 'amountToken',
            type: 'uint256',
        },
        {
            internalType: 'uint256',
            name: 'amountETH',
            type: 'uint256',
        },
        {
            internalType: 'uint256',
            name: 'liquidity',
            type: 'uint256',
        },
    ],
    stateMutability: 'payable',
    type: 'function',
};
