// sushi/_cjs/evm/pool/sushiswap-v3/utils/priceTickConversions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tickToPrice = tickToPrice;
exports.priceToClosestTick = priceToClosestTick;
const price_js_1 = require("../../../../generic/currency/price.js");
const internalConstants_js_1 = require("../internalConstants.js");
const encodeSqrtRatioX96_js_1 = require("./encodeSqrtRatioX96.js");
const tickMath_js_1 = require("./tickMath.js");
function tickToPrice(baseToken, quoteToken, tick) {
    const sqrtRatioX96 = tickMath_js_1.TickMath.getSqrtRatioAtTick(tick);
    const ratioX192 = sqrtRatioX96 * sqrtRatioX96;
    return baseToken.sortsBefore(quoteToken)
        ? new price_js_1.Price({
            base: baseToken,
            quote: quoteToken,
            denominator: internalConstants_js_1.Q192,
            numerator: ratioX192,
        })
        : new price_js_1.Price({
            base: baseToken,
            quote: quoteToken,
            denominator: ratioX192,
            numerator: internalConstants_js_1.Q192,
        });
}
function priceToClosestTick(price) {
    const sorted = price.base.sortsBefore(price.quote);
    const sqrtRatioX96 = sorted
        ? (0, encodeSqrtRatioX96_js_1.encodeSqrtRatioX96)(price.numerator, price.denominator)
        : (0, encodeSqrtRatioX96_js_1.encodeSqrtRatioX96)(price.denominator, price.numerator);
    let tick = tickMath_js_1.TickMath.getTickAtSqrtRatio(sqrtRatioX96);
    const nextTickPrice = tickToPrice(price.base, price.quote, tick + 1);
    if (sorted) {
        if (!price.lt(nextTickPrice)) {
            tick++;
        }
    }
    else {
        if (!price.gt(nextTickPrice)) {
            tick++;
        }
    }
    return tick;
}
