// sushi/_esm/evm/config/subgraphs/get-subgraph-url.js
import { getDecentralizedHostByDeploymentId, getDecentralizedHostBySubgraphId, } from './hosts.js';
function isIn(key, obj) {
    return key in obj;
}
// Double-wrap to allow of automatic DECENTRALIZED and OTHER type inference
function getSubgraphUrlWrapperWrapper(subgraphs) {
    // CHAINS is the union of related chainIds
    // COMPLETENESS tells us if the list of subgraphs is complete or partial for the given chainIds
    function getSubgraphUrlWrapper() {
        // Handles the actual url lookup
        function getSubgraphUrl(chainId, 
        // Require the decentralized key only if the subgraph is on decentralized network
        ...rest) {
            const [config] = rest;
            if (isIn(chainId, subgraphs.decentralizedIds)) {
                const _config = config;
                const entry = subgraphs.decentralizedIds[chainId];
                if (entry.type === 'deploymentId') {
                    return `${getDecentralizedHostByDeploymentId(_config.decentralizedKey)}/${subgraphs.decentralizedIds[chainId].id}`;
                }
                if (entry.type === 'subgraphId') {
                    return `${getDecentralizedHostBySubgraphId(_config.decentralizedKey)}/${subgraphs.decentralizedIds[chainId].id}`;
                }
                throw new Error(`Invalid entry type: ${entry.type}`);
            }
            if (isIn(chainId, subgraphs.otherUrls)) {
                return subgraphs.otherUrls[chainId];
            }
            return undefined;
        }
        return getSubgraphUrl;
    }
    return getSubgraphUrlWrapper;
}
export { getSubgraphUrlWrapperWrapper as getSubgraphUrlWrapper };
export function wrapAsIdType(entries, type) {
    const result = {};
    for (const chainId in entries) {
        result[chainId] = {
            type: type,
            id: entries[chainId],
        };
    }
    return result;
}
