// sushi/_esm/generic/currency/amount.js
import { formatUnits, parseUnits } from 'viem/utils';
import * as z from 'zod';
import { numberToFixed } from '../format/number.js';
import { Fraction } from '../math/fraction.js';
/**
 * Represents an amount of a particular currency.
 */
export class Amount {
    constructor(currency, amount) {
        this.currency = currency;
        this.amount = BigInt(amount);
    }
    /**
     * Creates an Amount from a human-readable value, e.g. "1.5".
     * Just a wrapper around viem's `parseUnits` to handle the currency's decimals.
     */
    static fromHuman(currency, human) {
        const amt = parseUnits(String(human), currency.decimals);
        return new Amount(currency, amt);
    }
    /**
     * Creates an Amount from a human-readable value, e.g. "1.5".
     */
    static tryFromHuman(currency, human) {
        try {
            return Amount.fromHuman(currency, human);
        }
        catch {
            return undefined;
        }
    }
    /**
     * Wraps the amount's currency into a Token.
     */
    wrap() {
        return new Amount(this.currency.wrap(), this.amount);
    }
    static getRawAmount(_ref, other) {
        if (other instanceof Amount) {
            return other.amount;
        }
        return BigInt(other);
    }
    /**
     * Adds another Amount or raw value.
     */
    add(other) {
        let add;
        if (other instanceof Fraction) {
            add = (other.numerator * Amount.SCALE) / other.denominator / Amount.SCALE;
        }
        else {
            add = Amount.getRawAmount(this, other);
        }
        return new Amount(this.currency, this.amount + add);
    }
    /**
     * Adds a human-readable value, e.g. "0.5".
     */
    addHuman(other) {
        const addend = parseUnits(String(other), this.currency.decimals);
        return new Amount(this.currency, this.amount + addend);
    }
    /**
     * Subtracts another Amount or raw value.
     */
    sub(other) {
        let sub;
        if (other instanceof Fraction) {
            sub = (other.numerator * Amount.SCALE) / other.denominator / Amount.SCALE;
        }
        else {
            sub = Amount.getRawAmount(this, other);
        }
        return new Amount(this.currency, this.amount - sub);
    }
    /**
     * Subtracts a human-readable value, e.g. "0.5".
     */
    subHuman(other) {
        const sub = parseUnits(String(other), this.currency.decimals);
        return new Amount(this.currency, this.amount - sub);
    }
    /**
     * Multiplies this amount by another amount or raw value.
     */
    mul(other) {
        if (other instanceof Fraction) {
            return new Amount(this.currency, (this.amount * other.numerator) / other.denominator);
        }
        const mul = Amount.getRawAmount(this, other);
        return new Amount(this.currency, this.amount * mul);
    }
    /**
     * Multiplies this amount by a human-readable multiplier.
     */
    mulHuman(other) {
        const mul = parseUnits(String(other), Amount.SCALE_DECIMALS);
        const result = (this.amount * mul) / Amount.SCALE;
        return new Amount(this.currency, result);
    }
    /**
     * Divides this amount by a divisor.
     * @throws Error if divisor is zero.
     * @returns a new Amount with the result.
     */
    div(divisor) {
        const d = parseUnits(String(divisor), Amount.SCALE_DECIMALS);
        if (d === 0n)
            throw new Error('Cannot divide by zero');
        const result = (this.amount * Amount.SCALE) / d;
        return new Amount(this.currency, result);
    }
    /**
     * Divides this amount by another Amount or raw value.
     * @returns a {@link Fraction} for a precise fractional result.
     */
    divToFraction(other) {
        if (other instanceof Fraction) {
            return new Fraction({
                numerator: this.amount * other.denominator,
                denominator: other.numerator,
            });
        }
        const div = Amount.getRawAmount(this, other);
        return new Fraction({ numerator: this.amount, denominator: div });
    }
    /**
     * Returns true if this amount > other.
     */
    gt(other) {
        const cmp = Amount.getRawAmount(this, other);
        return this.amount > cmp;
    }
    /**
     * Returns true if this amount >= other.
     */
    gte(other) {
        const cmp = Amount.getRawAmount(this, other);
        return this.amount >= cmp;
    }
    /**
     * Returns true if this amount < other.
     */
    lt(other) {
        const cmp = Amount.getRawAmount(this, other);
        return this.amount < cmp;
    }
    /**
     * Returns true if this amount <= other.
     */
    lte(other) {
        const cmp = Amount.getRawAmount(this, other);
        return this.amount <= cmp;
    }
    /**
     * Returns true if this amount == other.
     */
    eq(other) {
        const cmp = Amount.getRawAmount(this, other);
        return this.amount === cmp;
    }
    toJSON() {
        return {
            currency: this.currency.toJSON(),
            amount: this.amount.toString(),
        };
    }
    /**
     * @param args.maxFixed - The maximum number of fixed decimal places to display, only if relevant, eg. "0" for 0, "1.23" for 1.23, etc.
     * @param args.fixed - The number of fixed decimal places to display, always, eg. "0.00" for 0, "1.23" for 1.23, etc.
     * @returns Formatted string representation of the amount
     */
    toString(args = {
        maxFixed: this.currency.decimals,
    }) {
        const str = formatUnits(this.amount, this.currency.decimals);
        return numberToFixed(str, args);
    }
    toSignificant(significantDigits = 5) {
        return this.toString({
            significant: significantDigits,
        });
    }
}
Amount.SCALE_DECIMALS = 27;
Amount.SCALE = 10n ** BigInt(Amount.SCALE_DECIMALS);
export function serializedAmountSchema(currencySchema) {
    return z.object({
        currency: currencySchema,
        amount: z.string(),
    });
}
