// sushi/_esm/generic/currency/price.js
import { numberToFixed } from '../format/number.js';
import { Fraction } from '../math/fraction.js';
import { Amount } from './amount.js';
export class Price extends Fraction {
    constructor(args) {
        if ('baseAmount' in args && 'quoteAmount' in args) {
            const { baseAmount, quoteAmount } = args;
            super({ numerator: quoteAmount.amount, denominator: baseAmount.amount });
            this.base = baseAmount.currency;
            this.quote = quoteAmount.currency;
            return;
        }
        super({ numerator: args.numerator, denominator: args.denominator || 1n });
        this.base = args.base;
        this.quote = args.quote;
    }
    get scalar() {
        return new Fraction({
            numerator: 10n ** BigInt(this.base.decimals),
            denominator: 10n ** BigInt(this.quote.decimals),
        });
    }
    /**
     *
     * @param baseAmount The amount of the base currency to convert
     * @returns The equivalent amount in the quote currency
     */
    getQuote(baseAmount) {
        const quoteAmount = super.mul(baseAmount.amount).toNumber();
        return new Amount(this.quote, BigInt(Math.floor(quoteAmount)));
    }
    invert() {
        return new Price({
            base: this.quote,
            quote: this.base,
            numerator: this.denominator,
            denominator: this.numerator,
        });
    }
    toNumber() {
        return super.mul(this.scalar).toNumber();
    }
    toString(args) {
        return numberToFixed(this.toNumber(), args);
    }
    toSignificant(significantDigits = 5) {
        return this.toString({
            significant: significantDigits,
        });
    }
}
