// sushi/_esm/evm/api/quote.js
import { z } from 'zod';
import { version } from '../../version.js';
import { szevm } from '../validate/zod.js';
import { RouteStatus } from './types.js';
function quoteResponseSchema(visualize) {
    const tokenSchema = z.object({
        address: szevm.address(),
        decimals: z.number(),
        symbol: z.string(),
        name: z.string(),
    });
    const baseSuccessPartial = z
        .object({
        status: z.literal(RouteStatus.Success).or(z.literal(RouteStatus.Partial)),
        tokens: z.array(tokenSchema),
        tokenFrom: z.number(),
        tokenTo: z.number(),
        swapPrice: z.number(),
        priceImpact: z.number(),
        amountIn: z.string(),
        assumedAmountOut: z.string(),
    })
        .transform((data) => ({
        ...data,
        tokenFrom: data.tokens[data.tokenFrom],
        tokenTo: data.tokens[data.tokenTo],
    }));
    const baseNoWay = z.object({
        status: z.literal(RouteStatus.NoWay),
    });
    const baseSchema = baseSuccessPartial.or(baseNoWay);
    const baseVisualizeSchema = baseSchema
        .and(z.object({
        visualization: z.object({
            liquidityProviders: z.array(z.string()),
            nodes: z.array(tokenSchema),
            links: z.array(z.object({
                source: z.number(),
                target: z.number(),
                liquidityProvider: z.number(),
                amountIn: z.string(),
                amountOut: z.string(),
                value: z.number(),
            })),
        }),
    }))
        .or(baseNoWay);
    return (visualize ? baseVisualizeSchema : baseSchema);
}
export async function getQuote(params, options) {
    // TODO: VALIDATE PARAMS
    const url = new URL(`quote/v7/${params.chainId}`, params.baseUrl ?? 'https://api.sushi.com');
    url.searchParams.append('tokenIn', params.tokenIn);
    url.searchParams.append('tokenOut', params.tokenOut);
    url.searchParams.append('amount', params.amount.toString());
    url.searchParams.append('maxSlippage', params.maxSlippage.toString());
    if (params.maxPriceImpact) {
        url.searchParams.append('maxPriceImpact', params.maxPriceImpact.toString());
    }
    if (typeof params.fee === 'bigint' &&
        params.fee > 0n &&
        params.feeReceiver !== undefined) {
        url.searchParams.append('fee', params.fee.toString());
        url.searchParams.append('feeReceiver', params.feeReceiver);
        if (params.feeBy !== undefined) {
            url.searchParams.append('feeBy', params.feeBy);
        }
    }
    if (params?.visualize) {
        url.searchParams.append('visualize', params.visualize.toString());
    }
    if (params.referrer) {
        url.searchParams.append('referrer', params.referrer);
    }
    else {
        url.searchParams.append('referrer', `sushi-sdk/${version}`);
    }
    const res = await fetch(url.toString(), options);
    if (!res.ok) {
        throw new Error(`Failed to fetch quote: ${await res.text()}`);
    }
    return quoteResponseSchema(params.visualize).parse(await res.json());
}
