// @coinbase/onchainkit/dist/core/analytics/types.js
const WalletEvent = {
  /** Wallet connection fails */
  ConnectError: "walletConnectError",
  /** User clicks connect wallet button */
  ConnectInitiated: "walletConnectInitiated",
  /** Wallet successfully connected */
  ConnectSuccess: "walletConnectSuccess",
  /** User disconnects wallet */
  Disconnect: "walletDisconnect",
  /** User selects a wallet option */
  OptionSelected: "walletOptionSelected",
  /** User cancels wallet connection */
  ConnectCanceled: "walletConnectCanceled"
};
const WalletOption = {
  Buy: "buy",
  Explorer: "explorer",
  QR: "qr",
  Refresh: "refresh",
  Send: "send",
  Swap: "swap"
};
const SwapEvent = {
  SlippageChanged: "swapSlippageChanged",
  TokenSelected: "swapTokenSelected",
  SwapSuccess: "swapSuccess",
  SwapInitiated: "swapInitiated",
  SwapFailure: "swapFailure",
  SwapCanceled: "swapCanceled"
};
const BuyOption = {
  APPLE_PAY: "apple_pay",
  COINBASE: "coinbase_account",
  DEBIT_CARD: "debit_card",
  ETH: "eth",
  USDC: "usdc"
};
const BuyEvent = {
  BuyFailure: "buyFailure",
  BuyInitiated: "buyInitiated",
  BuyOptionSelected: "buyOptionSelected",
  BuySuccess: "buySuccess",
  BuyCanceled: "buyCanceled"
};
const CheckoutEvent = {
  CheckoutFailure: "checkoutFailure",
  CheckoutInitiated: "checkoutInitiated",
  CheckoutSuccess: "checkoutSuccess",
  CheckoutCanceled: "checkoutCanceled"
};
const MintEvent = {
  MintFailure: "mintFailure",
  MintInitiated: "mintInitiated",
  MintQuantityChanged: "mintQuantityChanged",
  MintSuccess: "mintSuccess",
  MintCanceled: "mintCanceled"
};
const TransactionEvent = {
  TransactionFailure: "transactionFailure",
  TransactionInitiated: "transactionInitiated",
  TransactionSuccess: "transactionSuccess",
  TransactionCanceled: "transactionCanceled"
};
const FundEvent = {
  FundAmountChanged: "fundAmountChanged",
  FundFailure: "fundFailure",
  FundInitiated: "fundInitiated",
  FundOptionSelected: "fundOptionSelected",
  FundSuccess: "fundSuccess",
  FundCanceled: "fundCanceled"
};
const EarnEvent = {
  EarnDepositInitiated: "earnDepositInitiated",
  EarnDepositSuccess: "earnDepositSuccess",
  EarnDepositFailure: "earnDepositFailure",
  EarnDepositCanceled: "earnDepositCanceled",
  EarnWithdrawInitiated: "earnWithdrawInitiated",
  EarnWithdrawSuccess: "earnWithdrawSuccess",
  EarnWithdrawFailure: "earnWithdrawFailure",
  EarnWithdrawCanceled: "earnWithdrawCanceled"
};
const AppchainEvent = {
  AppchainBridgeDepositInitiated: "appchainBridgeDepositInitiated",
  AppchainBridgeDepositSuccess: "appchainBridgeDepositSuccess",
  AppchainBridgeDepositFailure: "appchainBridgeDepositFailure",
  AppchainBridgeWithdrawInitiated: "appchainBridgeWithdrawInitiated",
  AppchainBridgeWithdrawSuccess: "appchainBridgeWithdrawSuccess",
  AppchainBridgeWithdrawFailure: "appchainBridgeWithdrawFailure",
  AppchainBridgeWaitForClaimFailure: "appchainBridgeWaitForClaimFailure",
  AppchainBridgeClaimSuccess: "appchainBridgeClaimSuccess",
  AppchainBridgeClaimFailure: "appchainBridgeClaimFailure"
};
const ErrorEvent = {
  ComponentError: "componentError"
};
export {
  AppchainEvent,
  BuyEvent,
  BuyOption,
  CheckoutEvent,
  EarnEvent,
  ErrorEvent,
  FundEvent,
  MintEvent,
  SwapEvent,
  TransactionEvent,
  WalletEvent,
  WalletOption
};
