// sushi/_cjs/evm/config/tokens/token-maps/default-bases.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EVM_DEFAULT_BASES = void 0;
const index_js_1 = require("../../../chain/index.js");
const index_js_2 = require("../../../config/tokens/tokens/index.js");
const native_js_1 = require("../../../currency/native.js");
const token_js_1 = require("../../../currency/token.js");
const HONEY_js_1 = require("../tokens/berachain/HONEY.js");
const HEMI_js_1 = require("../tokens/hemi/HEMI.js");
const wrapped_native_js_1 = require("../wrapped-native.js");
exports.EVM_DEFAULT_BASES = {
    [index_js_1.EvmChainId.ETHEREUM]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.ETHEREUM),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.ETHEREUM],
        index_js_2.SUSHI[index_js_1.EvmChainId.ETHEREUM],
        index_js_2.WBTC[index_js_1.EvmChainId.ETHEREUM],
        index_js_2.USDC[index_js_1.EvmChainId.ETHEREUM],
        index_js_2.USDT[index_js_1.EvmChainId.ETHEREUM],
        index_js_2.DAI[index_js_1.EvmChainId.ETHEREUM],
    ],
    [index_js_1.EvmChainId.SEPOLIA]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.SEPOLIA),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.SEPOLIA],
        index_js_2.USDC[index_js_1.EvmChainId.SEPOLIA],
    ],
    [index_js_1.EvmChainId.GNOSIS]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.GNOSIS),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.GNOSIS],
        index_js_2.GNO[index_js_1.EvmChainId.GNOSIS],
        index_js_2.WETH9[index_js_1.EvmChainId.GNOSIS],
        index_js_2.USDC[index_js_1.EvmChainId.GNOSIS],
        index_js_2.USDT[index_js_1.EvmChainId.GNOSIS],
        index_js_2.DAI[index_js_1.EvmChainId.GNOSIS],
    ],
    [index_js_1.EvmChainId.BSC]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.BSC),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.BSC],
        index_js_2.WETH9[index_js_1.EvmChainId.BSC],
        index_js_2.USDC[index_js_1.EvmChainId.BSC],
        index_js_2.USDT[index_js_1.EvmChainId.BSC],
        index_js_2.DAI[index_js_1.EvmChainId.BSC],
    ],
    [index_js_1.EvmChainId.POLYGON]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.POLYGON),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.POLYGON],
        index_js_2.WBTC[index_js_1.EvmChainId.POLYGON],
        index_js_2.WETH9[index_js_1.EvmChainId.POLYGON],
        index_js_2.USDC[index_js_1.EvmChainId.POLYGON],
        index_js_2.USDT[index_js_1.EvmChainId.POLYGON],
        index_js_2.DAI[index_js_1.EvmChainId.POLYGON],
    ],
    [index_js_1.EvmChainId.AVALANCHE]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.AVALANCHE),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.AVALANCHE],
        index_js_2.WETH9[index_js_1.EvmChainId.AVALANCHE],
        index_js_2.WBTC[index_js_1.EvmChainId.AVALANCHE],
        index_js_2.USDC[index_js_1.EvmChainId.AVALANCHE],
        index_js_2.USDT[index_js_1.EvmChainId.AVALANCHE],
        index_js_2.DAI[index_js_1.EvmChainId.AVALANCHE],
        index_js_2.MIM[index_js_1.EvmChainId.AVALANCHE],
        index_js_2.FRAX[index_js_1.EvmChainId.AVALANCHE],
    ],
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.ARBITRUM_NOVA),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.ARBITRUM_NOVA],
        index_js_2.ARB[index_js_1.EvmChainId.ARBITRUM_NOVA],
        index_js_2.WBTC[index_js_1.EvmChainId.ARBITRUM_NOVA],
        index_js_2.USDC[index_js_1.EvmChainId.ARBITRUM_NOVA],
        index_js_2.USDT[index_js_1.EvmChainId.ARBITRUM_NOVA],
        index_js_2.DAI[index_js_1.EvmChainId.ARBITRUM_NOVA],
    ],
    [index_js_1.EvmChainId.BOBA]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.BOBA),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.BOBA],
        index_js_2.USDC[index_js_1.EvmChainId.BOBA],
        index_js_2.USDT[index_js_1.EvmChainId.BOBA],
        index_js_2.DAI[index_js_1.EvmChainId.BOBA],
        index_js_2.FRAX[index_js_1.EvmChainId.BOBA],
        index_js_2.WBTC[index_js_1.EvmChainId.BOBA],
    ],
    [index_js_1.EvmChainId.FANTOM]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.FANTOM),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.FANTOM],
        index_js_2.axlUSDC[index_js_1.EvmChainId.FANTOM],
        new token_js_1.EvmToken({
            chainId: index_js_1.EvmChainId.FANTOM,
            address: '0x28a92dde19d9989f39a49905d7c9c2fac7799bdf',
            decimals: 6,
            symbol: 'USDC',
            name: 'USD Coin',
        }),
        new token_js_1.EvmToken({
            chainId: index_js_1.EvmChainId.FANTOM,
            address: '0xcc1b99ddac1a33c201a742a1851662e87bc7f22c',
            decimals: 6,
            symbol: 'USDT',
            name: 'Tether USD',
        }),
        new token_js_1.EvmToken({
            chainId: index_js_1.EvmChainId.FANTOM,
            address: '0x695921034f0387eac4e11620ee91b1b15a6a09fe',
            decimals: 18,
            symbol: 'WETH',
            name: 'Wrapped Ether',
        }),
        new token_js_1.EvmToken({
            chainId: index_js_1.EvmChainId.FANTOM,
            address: '0xf1648c50d2863f780c57849d812b4b7686031a3d',
            decimals: 18,
            symbol: 'WBTC',
            name: 'Wrapped BTC',
        }),
        index_js_2.axlUSDC[index_js_1.EvmChainId.FANTOM],
        index_js_2.MIM[index_js_1.EvmChainId.FANTOM],
    ],
    [index_js_1.EvmChainId.ARBITRUM]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.ARBITRUM),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.ARBITRUM],
        index_js_2.ARB[index_js_1.EvmChainId.ARBITRUM],
        index_js_2.WBTC[index_js_1.EvmChainId.ARBITRUM],
        index_js_2.USDC[index_js_1.EvmChainId.ARBITRUM],
        index_js_2.USDT[index_js_1.EvmChainId.ARBITRUM],
        index_js_2.DAI[index_js_1.EvmChainId.ARBITRUM],
        index_js_2.MIM[index_js_1.EvmChainId.ARBITRUM],
    ],
    [index_js_1.EvmChainId.ARBITRUM_SEPOLIA]: [],
    [index_js_1.EvmChainId.HARMONY]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.HARMONY),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.HARMONY],
        index_js_2.WETH9[index_js_1.EvmChainId.HARMONY],
        index_js_2.USDC[index_js_1.EvmChainId.HARMONY],
        index_js_2.USDT[index_js_1.EvmChainId.HARMONY],
        index_js_2.DAI[index_js_1.EvmChainId.HARMONY],
    ],
    [index_js_1.EvmChainId.CELO]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.CELO),
        index_js_2.WETH9[index_js_1.EvmChainId.CELO],
        index_js_2.USDC[index_js_1.EvmChainId.CELO],
        index_js_2.USDT[index_js_1.EvmChainId.CELO],
        index_js_2.DAI[index_js_1.EvmChainId.CELO],
    ],
    [index_js_1.EvmChainId.OPTIMISM]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.OPTIMISM),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.OPTIMISM],
        index_js_2.WBTC[index_js_1.EvmChainId.OPTIMISM],
        index_js_2.USDC[index_js_1.EvmChainId.OPTIMISM],
        index_js_2.USDT[index_js_1.EvmChainId.OPTIMISM],
        index_js_2.DAI[index_js_1.EvmChainId.OPTIMISM],
    ],
    [index_js_1.EvmChainId.KAVA]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.KAVA),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.KAVA],
        index_js_2.axlWBTC[index_js_1.EvmChainId.KAVA],
        new token_js_1.EvmToken({
            chainId: index_js_1.EvmChainId.KAVA,
            address: '0x2dfd4de5ae386cd3f4fc8e2cb39240852e47f5e8',
            decimals: 18,
            symbol: 'WETH',
            name: 'Wrapped Ether',
        }),
        index_js_2.axlUSDC[index_js_1.EvmChainId.KAVA],
        index_js_2.USDT[index_js_1.EvmChainId.KAVA],
    ],
    [index_js_1.EvmChainId.METIS]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.METIS),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.METIS],
        index_js_2.WBTC[index_js_1.EvmChainId.METIS],
        index_js_2.WETH9[index_js_1.EvmChainId.METIS],
        index_js_2.USDC[index_js_1.EvmChainId.METIS],
        index_js_2.USDT[index_js_1.EvmChainId.METIS],
        index_js_2.DAI[index_js_1.EvmChainId.METIS],
    ],
    [index_js_1.EvmChainId.BOBA_BNB]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.BOBA_BNB),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.BOBA_BNB],
        new token_js_1.EvmToken({
            chainId: index_js_1.EvmChainId.BOBA_BNB,
            symbol: 'BNB',
            name: 'Binance Coin',
            decimals: 18,
            address: '0x4200000000000000000000000000000000000023',
        }),
        index_js_2.USDC[index_js_1.EvmChainId.BOBA_BNB],
        index_js_2.USDT[index_js_1.EvmChainId.BOBA_BNB],
    ],
    [index_js_1.EvmChainId.BTTC]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.BTTC),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.BTTC],
        index_js_2.WETH9[index_js_1.EvmChainId.BTTC],
        index_js_2.USDC[index_js_1.EvmChainId.BTTC],
        index_js_2.USDT[index_js_1.EvmChainId.BTTC],
    ],
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.POLYGON_ZKEVM),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.POLYGON_ZKEVM],
        index_js_2.MATIC[index_js_1.EvmChainId.POLYGON_ZKEVM],
        index_js_2.USDC[index_js_1.EvmChainId.POLYGON_ZKEVM],
        index_js_2.USDT[index_js_1.EvmChainId.POLYGON_ZKEVM],
        index_js_2.DAI[index_js_1.EvmChainId.POLYGON_ZKEVM],
        index_js_2.WBTC[index_js_1.EvmChainId.POLYGON_ZKEVM],
    ],
    [index_js_1.EvmChainId.THUNDERCORE]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.THUNDERCORE),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.THUNDERCORE],
        index_js_2.WETH9[index_js_1.EvmChainId.THUNDERCORE],
        index_js_2.USDC[index_js_1.EvmChainId.THUNDERCORE],
        index_js_2.USDT[index_js_1.EvmChainId.THUNDERCORE],
        index_js_2.WBTC[index_js_1.EvmChainId.THUNDERCORE],
    ],
    [index_js_1.EvmChainId.HAQQ]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.HAQQ),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.HAQQ],
        index_js_2.axlETH[index_js_1.EvmChainId.HAQQ],
        index_js_2.axlWBTC[index_js_1.EvmChainId.HAQQ],
        index_js_2.axlUSDC[index_js_1.EvmChainId.HAQQ],
        index_js_2.axlUSDT[index_js_1.EvmChainId.HAQQ],
        index_js_2.axlDAI[index_js_1.EvmChainId.HAQQ],
    ],
    [index_js_1.EvmChainId.CORE]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.CORE),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.CORE],
        index_js_2.WETH9[index_js_1.EvmChainId.CORE],
        index_js_2.USDC[index_js_1.EvmChainId.CORE],
        index_js_2.USDT[index_js_1.EvmChainId.CORE],
    ],
    [index_js_1.EvmChainId.ZKSYNC_ERA]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.ZKSYNC_ERA),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.ZKSYNC_ERA],
        index_js_2.WBTC[index_js_1.EvmChainId.ZKSYNC_ERA],
        index_js_2.USDC[index_js_1.EvmChainId.ZKSYNC_ERA],
    ],
    [index_js_1.EvmChainId.LINEA]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.LINEA),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.LINEA],
        index_js_2.DAI[index_js_1.EvmChainId.LINEA],
        index_js_2.USDC[index_js_1.EvmChainId.LINEA],
    ],
    [index_js_1.EvmChainId.BASE]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.BASE),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.BASE],
        index_js_2.DAI[index_js_1.EvmChainId.BASE],
        new token_js_1.EvmToken({
            chainId: index_js_1.EvmChainId.BASE,
            symbol: 'USDbC',
            name: 'USD Base Coin',
            decimals: 6,
            address: '0xd9aaec86b65d86f6a7b5b1b0c42ffa531710b6ca',
        }),
        index_js_2.USDC[index_js_1.EvmChainId.BASE],
        new token_js_1.EvmToken({
            chainId: index_js_1.EvmChainId.BASE,
            symbol: 'OX',
            name: 'OX Coin',
            decimals: 18,
            address: '0xba0dda8762c24da9487f5fa026a9b64b695a07ea',
        }),
    ],
    [index_js_1.EvmChainId.SCROLL]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.SCROLL),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.SCROLL],
        index_js_2.USDC[index_js_1.EvmChainId.SCROLL],
        index_js_2.USDT[index_js_1.EvmChainId.SCROLL],
        index_js_2.WBTC[index_js_1.EvmChainId.SCROLL],
    ],
    [index_js_1.EvmChainId.FILECOIN]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.FILECOIN),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.FILECOIN],
        index_js_2.USDC[index_js_1.EvmChainId.FILECOIN],
        index_js_2.DAI[index_js_1.EvmChainId.FILECOIN],
    ],
    [index_js_1.EvmChainId.ZETACHAIN]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.ZETACHAIN),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.ZETACHAIN],
        index_js_2.ZETA_USDC_ETH[index_js_1.EvmChainId.ZETACHAIN],
        index_js_2.ZETA_USDT_ETH[index_js_1.EvmChainId.ZETACHAIN],
        index_js_2.WETH9[index_js_1.EvmChainId.ZETACHAIN],
    ],
    [index_js_1.EvmChainId.CRONOS]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.CRONOS),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.CRONOS],
        index_js_2.WETH9[index_js_1.EvmChainId.CRONOS],
        index_js_2.WBTC[index_js_1.EvmChainId.CRONOS],
        index_js_2.USDC[index_js_1.EvmChainId.CRONOS],
    ],
    [index_js_1.EvmChainId.BLAST]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.BLAST),
        index_js_2.USDB[index_js_1.EvmChainId.BLAST],
    ],
    [index_js_1.EvmChainId.SKALE_EUROPA]: [
        index_js_2.SKL[index_js_1.EvmChainId.SKALE_EUROPA],
        index_js_2.USDC[index_js_1.EvmChainId.SKALE_EUROPA],
        index_js_2.WETH9[index_js_1.EvmChainId.SKALE_EUROPA],
    ],
    [index_js_1.EvmChainId.ROOTSTOCK]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.ROOTSTOCK),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.ROOTSTOCK],
        index_js_2.USDT0[index_js_1.EvmChainId.ROOTSTOCK],
        new token_js_1.EvmToken({
            chainId: index_js_1.EvmChainId.ROOTSTOCK,
            address: '0x1d931bf8656d795e50ef6d639562c5bd8ac2b78f',
            decimals: 18,
            symbol: 'ETHs',
            name: 'ETHs',
        }),
        index_js_2.SolvBTC[index_js_1.EvmChainId.ROOTSTOCK],
        index_js_2.USDT[index_js_1.EvmChainId.ROOTSTOCK],
        index_js_2.RBTC[index_js_1.EvmChainId.ROOTSTOCK],
    ],
    [index_js_1.EvmChainId.MANTLE]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.MANTLE),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.MANTLE],
        index_js_2.WETH9[index_js_1.EvmChainId.MANTLE],
        index_js_2.USDC[index_js_1.EvmChainId.MANTLE],
    ],
    [index_js_1.EvmChainId.MANTA]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.MANTA),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.MANTA],
        index_js_2.USDC[index_js_1.EvmChainId.MANTA],
        index_js_2.USDT[index_js_1.EvmChainId.MANTA],
        index_js_2.WBTC[index_js_1.EvmChainId.MANTA],
    ],
    [index_js_1.EvmChainId.MODE]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.MODE),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.MODE],
        index_js_2.USDC[index_js_1.EvmChainId.MODE],
        index_js_2.USDT[index_js_1.EvmChainId.MODE],
        index_js_2.WBTC[index_js_1.EvmChainId.MODE],
    ],
    [index_js_1.EvmChainId.TAIKO]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.TAIKO),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.TAIKO],
        index_js_2.USDC[index_js_1.EvmChainId.TAIKO],
        index_js_2.USDT[index_js_1.EvmChainId.TAIKO],
        index_js_2.DAI[index_js_1.EvmChainId.TAIKO],
        index_js_2.WBTC[index_js_1.EvmChainId.TAIKO],
    ],
    [index_js_1.EvmChainId.ZKLINK]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.ZKLINK),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.ZKLINK],
        index_js_2.USDC[index_js_1.EvmChainId.ZKLINK],
        index_js_2.USDT[index_js_1.EvmChainId.ZKLINK],
        index_js_2.DAI[index_js_1.EvmChainId.ZKLINK],
        index_js_2.WBTC[index_js_1.EvmChainId.ZKLINK],
    ],
    [index_js_1.EvmChainId.APE]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.APE),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.APE],
        index_js_2.APE_USD[index_js_1.EvmChainId.APE],
    ],
    [index_js_1.EvmChainId.SONIC]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.SONIC),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.SONIC],
        index_js_2.USDC[index_js_1.EvmChainId.SONIC],
        index_js_2.WETH9[index_js_1.EvmChainId.SONIC],
    ],
    [index_js_1.EvmChainId.HEMI]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.HEMI),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.HEMI],
        index_js_2.USDC[index_js_1.EvmChainId.HEMI],
        index_js_2.USDT[index_js_1.EvmChainId.HEMI],
        index_js_2.WBTC[index_js_1.EvmChainId.HEMI],
        HEMI_js_1.HEMI_HEMI[index_js_1.EvmChainId.HEMI],
    ],
    [index_js_1.EvmChainId.TATARA]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.TATARA),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.TATARA],
        index_js_2.AUSD[index_js_1.EvmChainId.TATARA],
    ],
    [index_js_1.EvmChainId.KATANA]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.KATANA),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.KATANA],
        index_js_2.USDC[index_js_1.EvmChainId.KATANA],
        index_js_2.USDT[index_js_1.EvmChainId.KATANA],
        index_js_2.AUSD[index_js_1.EvmChainId.KATANA],
    ],
    [index_js_1.EvmChainId.HYPEREVM]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.HYPEREVM),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.HYPEREVM],
        index_js_2.USDT[index_js_1.EvmChainId.HYPEREVM],
    ],
    [index_js_1.EvmChainId.BERACHAIN]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.BERACHAIN),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.BERACHAIN],
        index_js_2.USDT[index_js_1.EvmChainId.BERACHAIN],
        index_js_2.USDC[index_js_1.EvmChainId.BERACHAIN],
        HONEY_js_1.BERACHAIN_HONEY[index_js_1.EvmChainId.BERACHAIN],
    ],
    [index_js_1.EvmChainId.PLASMA]: [
        native_js_1.EvmNative.fromChainId(index_js_1.EvmChainId.PLASMA),
        wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.PLASMA],
        index_js_2.USDT[index_js_1.EvmChainId.PLASMA],
        index_js_2.WETH9[index_js_1.EvmChainId.PLASMA],
    ],
};
