// sushi/_cjs/evm/config/tokens/token-maps/default-quote.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultQuoteCurrency = exports.defaultCurrency = void 0;
const index_js_1 = require("../../../chain/index.js");
const index_js_2 = require("../../../config/tokens/tokens/index.js");
const native_js_1 = require("../../../currency/native.js");
const token_js_1 = require("../../../currency/token.js");
const wrapped_native_js_1 = require("../wrapped-native.js");
exports.defaultCurrency = {
    ...Object.fromEntries(index_js_1.evmChainIds.map((chainId) => [chainId, native_js_1.EvmNative.fromChainId(chainId)])),
    [index_js_1.EvmChainId.SKALE_EUROPA]: index_js_2.WETH9[index_js_1.EvmChainId.SKALE_EUROPA],
};
exports.defaultQuoteCurrency = {
    [index_js_1.EvmChainId.ETHEREUM]: index_js_2.SUSHI[index_js_1.EvmChainId.ETHEREUM],
    [index_js_1.EvmChainId.SEPOLIA]: index_js_2.USDC[index_js_1.EvmChainId.SEPOLIA],
    [index_js_1.EvmChainId.POLYGON]: index_js_2.SUSHI[index_js_1.EvmChainId.POLYGON],
    [index_js_1.EvmChainId.FANTOM]: index_js_2.axlUSDC[index_js_1.EvmChainId.FANTOM],
    [index_js_1.EvmChainId.GNOSIS]: index_js_2.GNO[index_js_1.EvmChainId.GNOSIS],
    [index_js_1.EvmChainId.BSC]: index_js_2.BUSD[index_js_1.EvmChainId.BSC],
    [index_js_1.EvmChainId.ARBITRUM]: index_js_2.ARB[index_js_1.EvmChainId.ARBITRUM],
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: index_js_2.ARB[index_js_1.EvmChainId.ARBITRUM_NOVA],
    [index_js_1.EvmChainId.ARBITRUM_SEPOLIA]: wrapped_native_js_1.WNATIVE[index_js_1.EvmChainId.ARBITRUM_SEPOLIA],
    [index_js_1.EvmChainId.AVALANCHE]: index_js_2.USDC[index_js_1.EvmChainId.AVALANCHE],
    [index_js_1.EvmChainId.HARMONY]: index_js_2.USDC[index_js_1.EvmChainId.HARMONY],
    [index_js_1.EvmChainId.CELO]: index_js_2.USDC[index_js_1.EvmChainId.CELO],
    [index_js_1.EvmChainId.OPTIMISM]: index_js_2.OP[index_js_1.EvmChainId.OPTIMISM],
    [index_js_1.EvmChainId.KAVA]: index_js_2.axlUSDC[index_js_1.EvmChainId.KAVA],
    [index_js_1.EvmChainId.METIS]: index_js_2.USDC[index_js_1.EvmChainId.METIS],
    [index_js_1.EvmChainId.BOBA]: index_js_2.USDC[index_js_1.EvmChainId.BOBA],
    [index_js_1.EvmChainId.BOBA_BNB]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.BOBA_BNB,
        address: '0x4200000000000000000000000000000000000023',
        decimals: 18,
        symbol: 'BNB',
        name: 'Binance Coin',
    }),
    [index_js_1.EvmChainId.BTTC]: index_js_2.USDC[index_js_1.EvmChainId.BTTC],
    [index_js_1.EvmChainId.THUNDERCORE]: index_js_2.USDC[index_js_1.EvmChainId.THUNDERCORE],
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: index_js_2.USDC[index_js_1.EvmChainId.POLYGON_ZKEVM],
    [index_js_1.EvmChainId.HAQQ]: index_js_2.USDC[index_js_1.EvmChainId.HAQQ],
    [index_js_1.EvmChainId.CORE]: index_js_2.USDC[index_js_1.EvmChainId.CORE],
    [index_js_1.EvmChainId.ZKSYNC_ERA]: index_js_2.USDC[index_js_1.EvmChainId.ZKSYNC_ERA],
    [index_js_1.EvmChainId.LINEA]: index_js_2.axlUSDC[index_js_1.EvmChainId.LINEA],
    [index_js_1.EvmChainId.BASE]: index_js_2.USDC[index_js_1.EvmChainId.BASE],
    [index_js_1.EvmChainId.SCROLL]: index_js_2.USDC[index_js_1.EvmChainId.SCROLL],
    [index_js_1.EvmChainId.FILECOIN]: index_js_2.axlUSDC[index_js_1.EvmChainId.FILECOIN],
    [index_js_1.EvmChainId.ZETACHAIN]: index_js_2.WETH9[index_js_1.EvmChainId.ZETACHAIN],
    [index_js_1.EvmChainId.CRONOS]: index_js_2.WETH9[index_js_1.EvmChainId.CRONOS],
    [index_js_1.EvmChainId.BLAST]: index_js_2.USDB[index_js_1.EvmChainId.BLAST],
    [index_js_1.EvmChainId.SKALE_EUROPA]: index_js_2.USDC[index_js_1.EvmChainId.SKALE_EUROPA],
    [index_js_1.EvmChainId.ROOTSTOCK]: index_js_2.USDT0[index_js_1.EvmChainId.ROOTSTOCK],
    [index_js_1.EvmChainId.MANTLE]: index_js_2.WETH9[index_js_1.EvmChainId.MANTLE],
    [index_js_1.EvmChainId.MANTA]: index_js_2.USDC[index_js_1.EvmChainId.MANTA],
    [index_js_1.EvmChainId.MODE]: index_js_2.USDC[index_js_1.EvmChainId.MODE],
    [index_js_1.EvmChainId.TAIKO]: index_js_2.USDC[index_js_1.EvmChainId.TAIKO],
    [index_js_1.EvmChainId.ZKLINK]: index_js_2.USDC[index_js_1.EvmChainId.ZKLINK],
    [index_js_1.EvmChainId.APE]: index_js_2.APE_USD[index_js_1.EvmChainId.APE],
    [index_js_1.EvmChainId.SONIC]: index_js_2.WETH9[index_js_1.EvmChainId.SONIC],
    [index_js_1.EvmChainId.HEMI]: index_js_2.USDC[index_js_1.EvmChainId.HEMI],
    [index_js_1.EvmChainId.TATARA]: index_js_2.AUSD[index_js_1.EvmChainId.TATARA],
    [index_js_1.EvmChainId.KATANA]: index_js_2.USDC[index_js_1.EvmChainId.KATANA],
    [index_js_1.EvmChainId.HYPEREVM]: index_js_2.USDT[index_js_1.EvmChainId.HYPEREVM],
    [index_js_1.EvmChainId.BERACHAIN]: index_js_2.USDT[index_js_1.EvmChainId.BERACHAIN],
    [index_js_1.EvmChainId.PLASMA]: index_js_2.USDT[index_js_1.EvmChainId.PLASMA],
};
