// sushi/_cjs/evm/config/tokens/token-maps/short-name-to-currency.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.currencyFromShortCurrencyName = exports.isShortCurrencyName = exports.isShortCurrencyNameSupported = void 0;
const chains_js_1 = require("../../../chain/chains.js");
const index_js_1 = require("../../../config/tokens/tokens/index.js");
const native_js_1 = require("../../../currency/native.js");
const token_js_1 = require("../../../currency/token.js");
const wrapped_native_js_1 = require("../wrapped-native.js");
const CHAIN_ID_SHORT_CURRENCY_NAME_TO_CURRENCY = {
    [chains_js_1.EvmChainId.ARBITRUM]: {
        ETH: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.ARBITRUM),
        WETH: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.ARBITRUM],
        WBTC: index_js_1.WBTC[chains_js_1.EvmChainId.ARBITRUM],
        USDC: index_js_1.USDC[chains_js_1.EvmChainId.ARBITRUM],
        USDT: index_js_1.USDT[chains_js_1.EvmChainId.ARBITRUM],
        DAI: index_js_1.DAI[chains_js_1.EvmChainId.ARBITRUM],
        FRAX: index_js_1.FRAX[chains_js_1.EvmChainId.ARBITRUM],
        MIM: index_js_1.MIM[chains_js_1.EvmChainId.ARBITRUM],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.ARBITRUM],
        MAI: index_js_1.MAI[chains_js_1.EvmChainId.ARBITRUM],
        UNI: index_js_1.UNI[chains_js_1.EvmChainId.ARBITRUM],
        AGEUR: index_js_1.AGEUR[chains_js_1.EvmChainId.ARBITRUM],
    },
    [chains_js_1.EvmChainId.ARBITRUM_NOVA]: {
        ETH: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.ARBITRUM_NOVA),
        WETH: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.ARBITRUM_NOVA],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.ARBITRUM_NOVA],
    },
    [chains_js_1.EvmChainId.AVALANCHE]: {
        AVAX: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.AVALANCHE),
        WAVAX: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.AVALANCHE],
        ETH: index_js_1.WETH9[chains_js_1.EvmChainId.AVALANCHE],
        WETH: index_js_1.WETH9[chains_js_1.EvmChainId.AVALANCHE],
        WBTC: index_js_1.WBTC[chains_js_1.EvmChainId.AVALANCHE],
        USDC: index_js_1.USDC[chains_js_1.EvmChainId.AVALANCHE],
        USDT: index_js_1.USDT[chains_js_1.EvmChainId.AVALANCHE],
        DAI: index_js_1.DAI[chains_js_1.EvmChainId.AVALANCHE],
        FRAX: index_js_1.FRAX[chains_js_1.EvmChainId.AVALANCHE],
        MIM: index_js_1.MIM[chains_js_1.EvmChainId.AVALANCHE],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.AVALANCHE],
        MAI: index_js_1.MAI[chains_js_1.EvmChainId.AVALANCHE],
        UNI: index_js_1.UNI[chains_js_1.EvmChainId.AVALANCHE],
        AGEUR: index_js_1.AGEUR[chains_js_1.EvmChainId.AVALANCHE],
    },
    [chains_js_1.EvmChainId.BOBA]: {
        ETH: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.BOBA),
        WETH: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.BOBA],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.BOBA],
    },
    [chains_js_1.EvmChainId.BOBA_BNB]: {
        BOBA: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.BOBA_BNB),
        WBOBA: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.BOBA_BNB],
        USDC: index_js_1.USDC[chains_js_1.EvmChainId.BOBA_BNB],
        BNB: new token_js_1.EvmToken({
            chainId: chains_js_1.EvmChainId.BOBA_BNB,
            symbol: 'BNB',
            name: 'Binance Coin',
            decimals: 18,
            address: '0x4200000000000000000000000000000000000023',
        }),
    },
    [chains_js_1.EvmChainId.BSC]: {
        BNB: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.BSC),
        WBNB: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.BSC],
        ETH: index_js_1.WETH9[chains_js_1.EvmChainId.BSC],
        WETH: index_js_1.WETH9[chains_js_1.EvmChainId.BSC],
        USDC: index_js_1.USDC[chains_js_1.EvmChainId.BSC],
        USDT: index_js_1.USDT[chains_js_1.EvmChainId.BSC],
        DAI: index_js_1.DAI[chains_js_1.EvmChainId.BSC],
        FRAX: index_js_1.FRAX[chains_js_1.EvmChainId.BSC],
        MIM: index_js_1.MIM[chains_js_1.EvmChainId.BSC],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.BSC],
        MAI: index_js_1.MAI[chains_js_1.EvmChainId.BSC],
        UNI: index_js_1.UNI[chains_js_1.EvmChainId.BSC],
        AGEUR: index_js_1.AGEUR[chains_js_1.EvmChainId.BSC],
    },
    [chains_js_1.EvmChainId.BTTC]: {
        BTT: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.BTTC),
        WBTT: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.BTTC],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.BTTC],
    },
    [chains_js_1.EvmChainId.CELO]: {
        CELO: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.CELO),
        WCELO: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.CELO],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.CELO],
        AGEUR: index_js_1.AGEUR[chains_js_1.EvmChainId.CELO],
    },
    [chains_js_1.EvmChainId.ETHEREUM]: {
        ETH: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.ETHEREUM),
        WETH: index_js_1.WETH9[chains_js_1.EvmChainId.ETHEREUM],
        WBTC: index_js_1.WBTC[chains_js_1.EvmChainId.ETHEREUM],
        USDC: index_js_1.USDC[chains_js_1.EvmChainId.ETHEREUM],
        USDT: index_js_1.USDT[chains_js_1.EvmChainId.ETHEREUM],
        DAI: index_js_1.DAI[chains_js_1.EvmChainId.ETHEREUM],
        FRAX: index_js_1.FRAX[chains_js_1.EvmChainId.ETHEREUM],
        MIM: index_js_1.MIM[chains_js_1.EvmChainId.ETHEREUM],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.ETHEREUM],
        MAI: index_js_1.MAI[chains_js_1.EvmChainId.ETHEREUM],
        UNI: index_js_1.UNI[chains_js_1.EvmChainId.ETHEREUM],
        LUSD: index_js_1.LUSD[chains_js_1.EvmChainId.ETHEREUM],
        AGEUR: index_js_1.AGEUR[chains_js_1.EvmChainId.ETHEREUM],
    },
    [chains_js_1.EvmChainId.FANTOM]: {
        FTM: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.FANTOM),
        WFTM: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.FANTOM],
        FRAX: index_js_1.FRAX[chains_js_1.EvmChainId.FANTOM],
        MIM: index_js_1.MIM[chains_js_1.EvmChainId.FANTOM],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.FANTOM],
        MAI: index_js_1.MAI[chains_js_1.EvmChainId.FANTOM],
    },
    [chains_js_1.EvmChainId.GNOSIS]: {
        XDAI: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.GNOSIS),
        WXDAI: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.GNOSIS],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.GNOSIS],
        AGEUR: index_js_1.AGEUR[chains_js_1.EvmChainId.GNOSIS],
    },
    [chains_js_1.EvmChainId.KAVA]: {
        KAVA: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.KAVA),
        WKAVA: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.KAVA],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.KAVA],
    },
    [chains_js_1.EvmChainId.METIS]: {
        METIS: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.METIS),
        WMETIS: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.METIS],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.METIS],
    },
    [chains_js_1.EvmChainId.OPTIMISM]: {
        ETH: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.OPTIMISM),
        WETH: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.OPTIMISM],
        USDC: index_js_1.USDC[chains_js_1.EvmChainId.OPTIMISM],
        USDT: index_js_1.USDT[chains_js_1.EvmChainId.OPTIMISM],
        OP: index_js_1.OP[chains_js_1.EvmChainId.OPTIMISM],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.OPTIMISM],
        AGEUR: index_js_1.AGEUR[chains_js_1.EvmChainId.OPTIMISM],
    },
    [chains_js_1.EvmChainId.POLYGON]: {
        MATIC: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.POLYGON),
        WMATIC: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.POLYGON],
        ETH: index_js_1.WETH9[chains_js_1.EvmChainId.POLYGON],
        WETH: index_js_1.WETH9[chains_js_1.EvmChainId.POLYGON],
        WBTC: index_js_1.WBTC[chains_js_1.EvmChainId.POLYGON],
        USDC: index_js_1.USDC[chains_js_1.EvmChainId.POLYGON],
        USDT: index_js_1.USDT[chains_js_1.EvmChainId.POLYGON],
        DAI: index_js_1.DAI[chains_js_1.EvmChainId.POLYGON],
        FRAX: index_js_1.FRAX[chains_js_1.EvmChainId.POLYGON],
        MIM: index_js_1.MIM[chains_js_1.EvmChainId.POLYGON],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.POLYGON],
        MAI: index_js_1.MAI[chains_js_1.EvmChainId.POLYGON],
        UNI: index_js_1.UNI[chains_js_1.EvmChainId.POLYGON],
        AGEUR: index_js_1.AGEUR[chains_js_1.EvmChainId.POLYGON],
    },
    [chains_js_1.EvmChainId.HARMONY]: {
        ONE: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.HARMONY),
        WONE: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.HARMONY],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.HARMONY],
    },
    [chains_js_1.EvmChainId.THUNDERCORE]: {
        NATIVE: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.THUNDERCORE),
        WNATIVE: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.THUNDERCORE],
        SUSHI: index_js_1.SUSHI[chains_js_1.EvmChainId.THUNDERCORE],
    },
    [chains_js_1.EvmChainId.POLYGON_ZKEVM]: {
        NATIVE: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.POLYGON_ZKEVM),
        WNATIVE: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.POLYGON_ZKEVM],
        ETH: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.POLYGON_ZKEVM),
        WETH: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.POLYGON_ZKEVM],
    },
    [chains_js_1.EvmChainId.CORE]: {
        NATIVE: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.CORE),
        WNATIVE: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.CORE],
        ETH: index_js_1.WETH9[chains_js_1.EvmChainId.CORE],
        WETH: index_js_1.WETH9[chains_js_1.EvmChainId.CORE],
    },
    [chains_js_1.EvmChainId.HAQQ]: {
        NATIVE: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.HAQQ),
        WNATIVE: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.HAQQ],
        ETH: index_js_1.WETH9[chains_js_1.EvmChainId.HAQQ],
        WETH: index_js_1.WETH9[chains_js_1.EvmChainId.HAQQ],
    },
    [chains_js_1.EvmChainId.ZKSYNC_ERA]: {
        NATIVE: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.ZKSYNC_ERA),
        WNATIVE: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.ZKSYNC_ERA],
        ETH: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.ZKSYNC_ERA),
        WETH: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.ZKSYNC_ERA],
    },
    [chains_js_1.EvmChainId.LINEA]: {
        NATIVE: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.LINEA),
        WNATIVE: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.LINEA],
        ETH: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.LINEA),
        WETH: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.LINEA],
    },
    [chains_js_1.EvmChainId.BASE]: {
        NATIVE: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.BASE),
        WNATIVE: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.BASE],
        ETH: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.BASE),
        WETH: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.BASE],
    },
    [chains_js_1.EvmChainId.SCROLL]: {
        NATIVE: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.SCROLL),
        WNATIVE: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.SCROLL],
        ETH: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.SCROLL),
        WETH: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.SCROLL],
    },
    [chains_js_1.EvmChainId.FILECOIN]: {
        NATIVE: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.FILECOIN),
        WNATIVE: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.FILECOIN],
    },
    [chains_js_1.EvmChainId.ZETACHAIN]: {
        NATIVE: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.ZETACHAIN),
        WNATIVE: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.ZETACHAIN],
    },
    [chains_js_1.EvmChainId.BLAST]: {
        NATIVE: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.BLAST),
        WNATIVE: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.BLAST],
        USDB: index_js_1.USDB[chains_js_1.EvmChainId.BLAST],
    },
    [chains_js_1.EvmChainId.SKALE_EUROPA]: {
        SKL: index_js_1.SKL,
        ETH: index_js_1.WETH9[chains_js_1.EvmChainId.SKALE_EUROPA],
        USDC: index_js_1.USDC[chains_js_1.EvmChainId.SKALE_EUROPA],
    },
    [chains_js_1.EvmChainId.ROOTSTOCK]: {
        NATIVE: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.ROOTSTOCK),
        WNATIVE: wrapped_native_js_1.WNATIVE[chains_js_1.EvmChainId.ROOTSTOCK],
        WBTC: native_js_1.EvmNative.fromChainId(chains_js_1.EvmChainId.ROOTSTOCK),
        WETH: index_js_1.WETH9[chains_js_1.EvmChainId.ROOTSTOCK],
    },
};
const isShortCurrencyNameSupported = (chainId) => chainId in CHAIN_ID_SHORT_CURRENCY_NAME_TO_CURRENCY;
exports.isShortCurrencyNameSupported = isShortCurrencyNameSupported;
const isShortCurrencyName = (chainId, shortCurrencyName) => {
    return ((0, exports.isShortCurrencyNameSupported)(chainId) &&
        shortCurrencyName in CHAIN_ID_SHORT_CURRENCY_NAME_TO_CURRENCY[chainId]);
};
exports.isShortCurrencyName = isShortCurrencyName;
const currencyFromShortCurrencyName = (chainId, shortCurrencyName) => {
    if (!(0, exports.isShortCurrencyNameSupported)(chainId))
        throw new Error(`Unsupported chain id ${chainId} for short currency name ${shortCurrencyName}`);
    if (!(shortCurrencyName in CHAIN_ID_SHORT_CURRENCY_NAME_TO_CURRENCY[chainId]))
        throw new Error(`Unsupported short currency name ${shortCurrencyName} on chain ${chainId}`);
    return CHAIN_ID_SHORT_CURRENCY_NAME_TO_CURRENCY[chainId][shortCurrencyName];
};
exports.currencyFromShortCurrencyName = currencyFromShortCurrencyName;
