// sushi/_cjs/evm/config/subgraphs/get-subgraph-url.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSubgraphUrlWrapper = getSubgraphUrlWrapperWrapper;
exports.wrapAsIdType = wrapAsIdType;
const hosts_js_1 = require("./hosts.js");
function isIn(key, obj) {
    return key in obj;
}
function getSubgraphUrlWrapperWrapper(subgraphs) {
    function getSubgraphUrlWrapper() {
        function getSubgraphUrl(chainId, ...rest) {
            const [config] = rest;
            if (isIn(chainId, subgraphs.decentralizedIds)) {
                const _config = config;
                const entry = subgraphs.decentralizedIds[chainId];
                if (entry.type === 'deploymentId') {
                    return `${(0, hosts_js_1.getDecentralizedHostByDeploymentId)(_config.decentralizedKey)}/${subgraphs.decentralizedIds[chainId].id}`;
                }
                if (entry.type === 'subgraphId') {
                    return `${(0, hosts_js_1.getDecentralizedHostBySubgraphId)(_config.decentralizedKey)}/${subgraphs.decentralizedIds[chainId].id}`;
                }
                throw new Error(`Invalid entry type: ${entry.type}`);
            }
            if (isIn(chainId, subgraphs.otherUrls)) {
                return subgraphs.otherUrls[chainId];
            }
            return undefined;
        }
        return getSubgraphUrl;
    }
    return getSubgraphUrlWrapper;
}
function wrapAsIdType(entries, type) {
    const result = {};
    for (const chainId in entries) {
        result[chainId] = {
            type: type,
            id: entries[chainId],
        };
    }
    return result;
}
