// sushi/_cjs/evm/config/subgraphs/hosts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUSHI_DATA_API_HOST = exports.getDecentralizedHostByDeploymentId = exports.getDecentralizedHostBySubgraphId = exports.SUSHI_DOMAIN_RESTRICTED_API_KEY = exports.SKALE_HOST = exports.METIS_0XGRAPH_HOST = exports.WAGMI_KAVA_HOST = exports.WAGMI_METIS_HOST = exports.GOLDSKY_COMMUNITY_HOST = exports.SUSHI_DEDICATED_GOLDSKY_HOST = exports.SUSHI_GOLDSKY_HOST = exports.PCS_STUDIO_HOST = exports.HAQQ_HOST = exports.CORE_HOST = exports.THUNDERCORE_HOST = exports.STUDIO_HOST = exports.PENDING_METIS_HOST = exports.METIS_HOST = exports.PENDING_KAVA_HOST = exports.KAVA_HOST = exports.PENDING_GRAPH_HOST = exports.GRAPH_HOST = void 0;
exports.GRAPH_HOST = 'api.thegraph.com/subgraphs/name';
exports.PENDING_GRAPH_HOST = 'api.thegraph.com/subgraphs/id';
exports.KAVA_HOST = 'pvt.graph.kava.io/subgraphs/name';
exports.PENDING_KAVA_HOST = 'pvt.graph.kava.io/subgraphs/id';
exports.METIS_HOST = 'andromeda.thegraph.metis.io/subgraphs/name';
exports.PENDING_METIS_HOST = 'andromeda.thegraph.metis.io/subgraphs/id';
exports.STUDIO_HOST = 'api.studio.thegraph.com/query/32073';
exports.THUNDERCORE_HOST = 'graph-node.thundercore.com/subgraphs/name';
exports.CORE_HOST = 'thegraph.coredao.org/subgraphs/name';
exports.HAQQ_HOST = 'haqq.graph.p2p.org/subgraphs/name';
exports.PCS_STUDIO_HOST = 'api.studio.thegraph.com/query/45376';
exports.SUSHI_GOLDSKY_HOST = 'api.goldsky.com/api/public/project_cls39ugcfyhbq01xl9tsf6g38/subgraphs';
exports.SUSHI_DEDICATED_GOLDSKY_HOST = 'api.goldsky.com/api/public/project_clslspm3c0knv01wvgfb2fqyq/subgraphs';
exports.GOLDSKY_COMMUNITY_HOST = 'api.goldsky.com/api/public/project_cl8ylkiw00krx0hvza0qw17vn/subgraphs';
exports.WAGMI_METIS_HOST = 'metis.graph.wagmi.com/subgraphs/name';
exports.WAGMI_KAVA_HOST = 'kava.graph.wagmi.com/subgraphs/name';
exports.METIS_0XGRAPH_HOST = 'metisapi.0xgraph.xyz/api/public/fc1ae952-7a36-44ac-9e9b-f46d70cedf7d/subgraphs';
exports.SKALE_HOST = 'elated-tan-skat-graph.skalenodes.com:8000/subgraphs/name';
exports.SUSHI_DOMAIN_RESTRICTED_API_KEY = '5d5d00365d2b8f675e12952d6eb5b9b0';
const getDecentralizedHostBySubgraphId = (key) => `gateway-arbitrum.network.thegraph.com/api/${key}/subgraphs/id`;
exports.getDecentralizedHostBySubgraphId = getDecentralizedHostBySubgraphId;
const getDecentralizedHostByDeploymentId = (key) => `gateway-arbitrum.network.thegraph.com/api/${key}/deployments/id`;
exports.getDecentralizedHostByDeploymentId = getDecentralizedHostByDeploymentId;
exports.SUSHI_DATA_API_HOST = 'https://production.data-gcp.sushi.com';
