// opensea-js/lib/sdk/orders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrdersManager = void 0;
const ethers_1 = require("ethers");
const constants_1 = require("../constants");
const utils_1 = require("../orders/utils");
const types_1 = require("../types");
const utils_2 = require("../utils/utils");
/**
 * Manager for order building and creation operations.
 * Handles listing creation, offer creation, and collection offers.
 */
class OrdersManager {
    constructor(context, getPriceParametersCallback) {
        this.context = context;
        this.getPriceParametersCallback = getPriceParametersCallback;
    }
    getAmountWithBasisPointsApplied(amount, basisPoints) {
        return ((amount * basisPoints) / constants_1.INVERSE_BASIS_POINT).toString();
    }
    isNotMarketplaceFee(fee) {
        return (fee.recipient.toLowerCase() !==
            (0, utils_2.getFeeRecipient)(this.context.chain).toLowerCase());
    }
    getNFTItems(nfts, quantities = []) {
        return nfts.map((nft, index) => ({
            itemType: (0, utils_2.getAssetItemType)(nft.token_standard.toUpperCase()),
            token: (0, utils_2.getAddressAfterRemappingSharedStorefrontAddressToLazyMintAdapterAddress)(nft.contract),
            identifier: nft.identifier ?? undefined,
            amount: quantities[index].toString() ?? "1",
        }));
    }
    async getFees({ collection, seller, paymentTokenAddress, startAmount, endAmount, includeOptionalCreatorFees = false, isPrivateListing = false, }) {
        let collectionFees = includeOptionalCreatorFees
            ? collection.fees
            : collection.fees.filter((fee) => fee.required);
        if (isPrivateListing) {
            collectionFees = collectionFees.filter((fee) => this.isNotMarketplaceFee(fee));
        }
        const collectionFeesBasisPoints = (0, utils_2.totalBasisPointsForFees)(collectionFees);
        const sellerBasisPoints = constants_1.INVERSE_BASIS_POINT - collectionFeesBasisPoints;
        const getConsiderationItem = (basisPoints, recipient) => {
            return {
                token: paymentTokenAddress,
                amount: this.getAmountWithBasisPointsApplied(startAmount, basisPoints),
                endAmount: this.getAmountWithBasisPointsApplied(endAmount ?? startAmount, basisPoints),
                recipient,
            };
        };
        const considerationItems = [];
        if (seller) {
            considerationItems.push(getConsiderationItem(sellerBasisPoints, seller));
        }
        if (collectionFeesBasisPoints > 0) {
            for (const fee of collectionFees) {
                considerationItems.push(getConsiderationItem((0, utils_2.basisPointsForFee)(fee), fee.recipient));
            }
        }
        return considerationItems;
    }
    /**
     * Build listing order without submitting to API
     * @param options Listing parameters
     * @returns OrderWithCounter ready for API submission or onchain validation
     */
    async _buildListingOrder({ asset, accountAddress, startAmount, endAmount, quantity = 1, domain, salt, listingTime, expirationTime, paymentTokenAddress = (0, utils_2.getListingPaymentToken)(this.context.chain), buyerAddress, englishAuction, includeOptionalCreatorFees = false, zone = ethers_1.ZeroAddress, }) {
        await this.context.requireAccountIsAvailable(accountAddress);
        const { nft } = await this.context.api.getNFT(asset.tokenAddress, asset.tokenId);
        const offerAssetItems = this.getNFTItems([nft], [BigInt(quantity ?? 1)]);
        if (englishAuction) {
            throw new Error("English auctions are no longer supported on OpenSea");
        }
        if (englishAuction && paymentTokenAddress == ethers_1.ethers.ZeroAddress) {
            throw new Error(`English auctions must use wrapped ETH or an ERC-20 token.`);
        }
        const { basePrice, endPrice } = await this.getPriceParametersCallback(types_1.OrderSide.LISTING, paymentTokenAddress, expirationTime ?? (0, utils_2.getMaxOrderExpirationTimestamp)(), startAmount, endAmount ?? undefined);
        const collection = await this.context.api.getCollection(nft.collection);
        const considerationFeeItems = await this.getFees({
            collection,
            seller: accountAddress,
            paymentTokenAddress,
            startAmount: basePrice,
            endAmount: endPrice,
            includeOptionalCreatorFees,
            isPrivateListing: !!buyerAddress,
        });
        if (buyerAddress) {
            const { getPrivateListingConsiderations } = await Promise.resolve().then(() => require("../orders/privateListings"));
            considerationFeeItems.push(...getPrivateListingConsiderations(offerAssetItems, buyerAddress));
        }
        if (englishAuction) {
            zone = constants_1.ENGLISH_AUCTION_ZONE_MAINNETS;
        }
        else if (collection.requiredZone) {
            zone = collection.requiredZone;
        }
        const { executeAllActions } = await this.context.seaport.createOrder({
            offer: offerAssetItems,
            consideration: considerationFeeItems,
            startTime: listingTime?.toString(),
            endTime: expirationTime?.toString() ??
                (0, utils_2.getMaxOrderExpirationTimestamp)().toString(),
            zone,
            domain,
            salt: BigInt(salt ?? 0).toString(),
            restrictedByZone: zone !== ethers_1.ZeroAddress,
            allowPartialFills: englishAuction ? false : true,
        }, accountAddress);
        return executeAllActions();
    }
    /**
     * Build listing order components without submitting to API
     * @param options Listing parameters
     * @returns OrderComponents ready for onchain validation
     */
    async buildListingOrderComponents({ asset, accountAddress, startAmount, endAmount, quantity = 1, domain, salt, listingTime, expirationTime, paymentTokenAddress = (0, utils_2.getListingPaymentToken)(this.context.chain), buyerAddress, englishAuction, includeOptionalCreatorFees = false, zone = ethers_1.ZeroAddress, }) {
        const order = await this._buildListingOrder({
            asset,
            accountAddress,
            startAmount,
            endAmount,
            quantity,
            domain,
            salt,
            listingTime,
            expirationTime,
            paymentTokenAddress,
            buyerAddress,
            englishAuction,
            includeOptionalCreatorFees,
            zone,
        });
        return order.parameters;
    }
    /**
     * Build offer order without submitting to API
     * @param options Offer parameters
     * @returns OrderWithCounter ready for API submission or onchain validation
     */
    async _buildOfferOrder({ asset, accountAddress, startAmount, quantity = 1, domain, salt, expirationTime, paymentTokenAddress = (0, utils_2.getOfferPaymentToken)(this.context.chain), zone = (0, utils_2.getSignedZone)(this.context.chain), }) {
        await this.context.requireAccountIsAvailable(accountAddress);
        const { nft } = await this.context.api.getNFT(asset.tokenAddress, asset.tokenId);
        const considerationAssetItems = this.getNFTItems([nft], [BigInt(quantity ?? 1)]);
        const { basePrice } = await this.getPriceParametersCallback(types_1.OrderSide.OFFER, paymentTokenAddress, expirationTime ?? (0, utils_2.getMaxOrderExpirationTimestamp)(), startAmount);
        const collection = await this.context.api.getCollection(nft.collection);
        const considerationFeeItems = await this.getFees({
            collection,
            paymentTokenAddress,
            startAmount: basePrice,
        });
        if (collection.requiredZone) {
            zone = collection.requiredZone;
        }
        const { executeAllActions } = await this.context.seaport.createOrder({
            offer: [
                {
                    token: paymentTokenAddress,
                    amount: basePrice.toString(),
                },
            ],
            consideration: [...considerationAssetItems, ...considerationFeeItems],
            endTime: expirationTime !== undefined
                ? BigInt(expirationTime).toString()
                : (0, utils_2.getMaxOrderExpirationTimestamp)().toString(),
            zone,
            domain,
            salt: BigInt(salt ?? 0).toString(),
            restrictedByZone: zone !== ethers_1.ZeroAddress,
            allowPartialFills: true,
        }, accountAddress);
        return executeAllActions();
    }
    /**
     * Build offer order components without submitting to API
     * @param options Offer parameters
     * @returns OrderComponents ready for onchain validation
     */
    async buildOfferOrderComponents({ asset, accountAddress, startAmount, quantity = 1, domain, salt, expirationTime, paymentTokenAddress = (0, utils_2.getOfferPaymentToken)(this.context.chain), zone = (0, utils_2.getSignedZone)(this.context.chain), }) {
        const order = await this._buildOfferOrder({
            asset,
            accountAddress,
            startAmount,
            quantity,
            domain,
            salt,
            expirationTime,
            paymentTokenAddress,
            zone,
        });
        return order.parameters;
    }
    /**
     * Create and submit an offer on an asset.
     * @param options
     * @param options.asset The asset to trade. tokenAddress and tokenId must be defined.
     * @param options.accountAddress Address of the wallet making the offer.
     * @param options.startAmount Value of the offer in units, not base units e.g. not wei, of the payment token (or WETH if no payment token address specified)
     * @param options.quantity The number of assets to bid for (if fungible or semi-fungible). Defaults to 1.
     * @param options.domain An optional domain to be hashed and included in the first four bytes of the random salt.
     * @param options.salt Arbitrary salt. If not passed in, a random salt will be generated with the first four bytes being the domain hash or empty.
     * @param options.expirationTime Expiration time for the order, in UTC seconds
     * @param options.paymentTokenAddress ERC20 address for the payment token in the order. If unspecified, defaults to WETH
     * @param options.zone The zone to use for the order. If unspecified, defaults to the chain's signed zone for order protection.
     *
     * @returns The {@link OrderV2} that was created.
     *
     * @throws Error if the asset does not contain a token id.
     * @throws Error if the accountAddress is not available through wallet or provider.
     * @throws Error if the startAmount is not greater than 0.
     * @throws Error if paymentTokenAddress is not WETH on anything other than Ethereum mainnet.
     */
    async createOffer({ asset, accountAddress, startAmount, quantity = 1, domain, salt, expirationTime, paymentTokenAddress = (0, utils_2.getOfferPaymentToken)(this.context.chain), zone = (0, utils_2.getSignedZone)(this.context.chain), }) {
        const order = await this._buildOfferOrder({
            asset,
            accountAddress,
            startAmount,
            quantity,
            domain,
            salt,
            expirationTime,
            paymentTokenAddress,
            zone,
        });
        return this.context.api.postOrder(order, {
            protocol: "seaport",
            protocolAddress: utils_1.DEFAULT_SEAPORT_CONTRACT_ADDRESS,
            side: types_1.OrderSide.OFFER,
        });
    }
    /**
     * Create and submit a listing for an asset.
     * @param options
     * @param options.asset The asset to trade. tokenAddress and tokenId must be defined.
     * @param options.accountAddress  Address of the wallet making the listing
     * @param options.startAmount Value of the listing at the start of the auction in units, not base units e.g. not wei, of the payment token (or WETH if no payment token address specified)
     * @param options.endAmount Value of the listing at the end of the auction. If specified, price will change linearly between startAmount and endAmount as time progresses.
     * @param options.quantity The number of assets to list (if fungible or semi-fungible). Defaults to 1.
     * @param options.domain An optional domain to be hashed and included in the first four bytes of the random salt. This can be used for on-chain order attribution to assist with analytics.
     * @param options.salt Arbitrary salt. If not passed in, a random salt will be generated with the first four bytes being the domain hash or empty.
     * @param options.listingTime Optional time when the order will become fulfillable, in UTC seconds. Undefined means it will start now.
     * @param options.expirationTime Expiration time for the order, in UTC seconds.
     * @param options.paymentTokenAddress ERC20 address for the payment token in the order. If unspecified, defaults to ETH
     * @param options.buyerAddress Optional address that's allowed to purchase this item. If specified, no other address will be able to take the order, unless its value is the null address.
     * @param options.englishAuction If true, the order will be listed as an English auction.
     * @param options.includeOptionalCreatorFees If true, optional creator fees will be included in the listing. Default: false.
     * @param options.zone The zone to use for the order. For order protection, pass SIGNED_ZONE. If unspecified, defaults to no zone.
     * @returns The {@link OrderV2} that was created.
     *
     * @throws Error if the asset does not contain a token id.
     * @throws Error if the accountAddress is not available through wallet or provider.
     * @throws Error if the startAmount is not greater than 0.
     * @throws Error if paymentTokenAddress is not WETH on anything other than Ethereum mainnet.
     */
    async createListing({ asset, accountAddress, startAmount, endAmount, quantity = 1, domain, salt, listingTime, expirationTime, paymentTokenAddress = (0, utils_2.getListingPaymentToken)(this.context.chain), buyerAddress, englishAuction, includeOptionalCreatorFees = false, zone = ethers_1.ZeroAddress, }) {
        const order = await this._buildListingOrder({
            asset,
            accountAddress,
            startAmount,
            endAmount,
            quantity,
            domain,
            salt,
            listingTime,
            expirationTime,
            paymentTokenAddress,
            buyerAddress,
            englishAuction,
            includeOptionalCreatorFees,
            zone,
        });
        return this.context.api.postOrder(order, {
            protocol: "seaport",
            protocolAddress: utils_1.DEFAULT_SEAPORT_CONTRACT_ADDRESS,
            side: types_1.OrderSide.LISTING,
        });
    }
    /**
     * Create and submit a collection offer.
     * @param options
     * @param options.collectionSlug Identifier for the collection.
     * @param options.accountAddress Address of the wallet making the offer.
     * @param options.amount Value of the offer in units, not base units e.g. not wei, of the payment token (or WETH if no payment token address specified).
     * @param options.quantity The number of assets to bid for (if fungible or semi-fungible).
     * @param options.domain An optional domain to be hashed and included in the first four bytes of the random salt. This can be used for on-chain order attribution to assist with analytics.
     * @param options.salt Arbitrary salt. If not passed in, a random salt will be generated with the first four bytes being the domain hash or empty.
     * @param options.expirationTime Expiration time for the order, in UTC seconds.
     * @param options.paymentTokenAddress ERC20 address for the payment token in the order. If unspecified, defaults to WETH.
     * @param options.offerProtectionEnabled Build the offer on OpenSea's signed zone to provide offer protections from receiving an item which is disabled from trading.
     * @param options.traitType If defined, the trait name to create the collection offer for.
     * @param options.traitValue If defined, the trait value to create the collection offer for.
     * @returns The {@link CollectionOffer} that was created.
     */
    async createCollectionOffer({ collectionSlug, accountAddress, amount, quantity, domain, salt, expirationTime, paymentTokenAddress = (0, utils_2.getOfferPaymentToken)(this.context.chain), offerProtectionEnabled = true, traitType, traitValue, }) {
        await this.context.requireAccountIsAvailable(accountAddress);
        const collection = await this.context.api.getCollection(collectionSlug);
        const buildOfferResult = await this.context.api.buildOffer(accountAddress, quantity, collectionSlug, offerProtectionEnabled, traitType, traitValue);
        const item = buildOfferResult.partialParameters.consideration[0];
        const convertedConsiderationItem = {
            itemType: item.itemType,
            token: item.token,
            identifier: item.identifierOrCriteria,
            amount: item.startAmount,
        };
        const { basePrice } = await this.getPriceParametersCallback(types_1.OrderSide.LISTING, paymentTokenAddress, expirationTime ?? (0, utils_2.getMaxOrderExpirationTimestamp)(), amount);
        const considerationFeeItems = await this.getFees({
            collection,
            paymentTokenAddress,
            startAmount: basePrice,
            endAmount: basePrice,
        });
        const considerationItems = [
            convertedConsiderationItem,
            ...considerationFeeItems,
        ];
        const payload = {
            offerer: accountAddress,
            offer: [
                {
                    token: paymentTokenAddress,
                    amount: basePrice.toString(),
                },
            ],
            consideration: considerationItems,
            endTime: expirationTime?.toString() ??
                (0, utils_2.getMaxOrderExpirationTimestamp)().toString(),
            zone: buildOfferResult.partialParameters.zone,
            domain,
            salt: BigInt(salt ?? 0).toString(),
            restrictedByZone: true,
            allowPartialFills: true,
        };
        const { executeAllActions } = await this.context.seaport.createOrder(payload, accountAddress);
        const order = await executeAllActions();
        return this.context.api.postCollectionOffer(order, collectionSlug, traitType, traitValue);
    }
}
exports.OrdersManager = OrdersManager;
