// sushi/_esm/kvm/chain/chains.js
import { defineKvmChain } from './define-chain.js';
const kadena = defineKvmChain({
    chainId: -3,
    key: 'kadena',
    shortName: 'kda',
    name: 'Kadena',
    blockExplorers: {
        default: {
            name: 'Kadena Explorer',
            url: 'https://explorer.chainweb.com/mainnet',
        },
    },
    netType: 'mainnet',
});
export const kvmChains = [kadena];
export const kvmChainIds = /* @__PURE__ */ kvmChains.map((chain) => chain.chainId);
export const KvmChainId = /* @__PURE__ */ new Proxy({}, {
    get: (_, key) => {
        const _key = key.toLowerCase().replace(/_/g, '-');
        const chain = kvmChains.find((chain) => chain.key === _key);
        if (!chain)
            return undefined;
        return chain.chainId;
    },
});
export function isKvmChainId(chainId) {
    return kvmChains.some((chain) => chain.chainId === chainId);
}
export function getKvmChainById(chainId) {
    return kvmChains.find((chain) => chain.chainId === chainId);
}
export function isKvmMainnetChainId(chainId) {
    return kvmChains.some((chain) => chain.chainId === chainId && chain.netType === 'mainnet');
}
export function isKvmChainKey(key) {
    return kvmChains.some((chain) => chain.key === key);
}
export function getKvmChainByKey(key) {
    return kvmChains.find((chain) => chain.key === key);
}
