// @coinbase/onchainkit/dist/fund/components/FundButton.js
'use client';
import { jsx, jsxs, Fragment } from "react/jsx-runtime";
import { useCallback, useMemo } from "react";
import { prefixClassName, pressable, cn, text } from "../../styles/theme.js";
import { useAccount } from "wagmi";
import { useAnalytics } from "../../core/analytics/hooks/useAnalytics.js";
import { FundEvent } from "../../core/analytics/types.js";
import { usePopupMonitor } from "../../buy/hooks/usePopupMonitor.js";
import { ErrorSvg } from "../../internal/svg/errorSvg.js";
import { openPopup } from "../../internal/utils/openPopup.js";
import { Spinner } from "../../internal/components/Spinner.js";
import { AddSvg } from "../../internal/svg/addSvg.js";
import { SuccessSvg } from "../../internal/svg/successSvg.js";
import { ConnectWallet } from "../../wallet/components/ConnectWallet.js";
import { useGetFundingUrl } from "../hooks/useGetFundingUrl.js";
import { getFundingPopupSize } from "../utils/getFundingPopupSize.js";
function FundButton({
  className,
  children,
  disabled = false,
  fundingUrl,
  openIn = "popup",
  popupSize = "md",
  target,
  state: buttonState = "default",
  fiatCurrency = "USD",
  onPopupClose,
  onClick,
  render,
  sessionToken
}) {
  const fallbackFundingUrl = useGetFundingUrl({
    fiatCurrency,
    originComponentName: "FundButton",
    sessionToken
  });
  const {
    address
  } = useAccount();
  const fundingUrlToRender = fundingUrl ?? fallbackFundingUrl;
  const isDisabled = disabled || !fundingUrlToRender;
  const {
    startPopupMonitor
  } = usePopupMonitor(onPopupClose);
  const {
    sendAnalytics
  } = useAnalytics();
  const handleAnalyticsInitiated = useCallback(() => {
    sendAnalytics(FundEvent.FundInitiated, {
      currency: fiatCurrency
    });
  }, [sendAnalytics, fiatCurrency]);
  const handleAnalyticsFailure = useCallback((error) => {
    sendAnalytics(FundEvent.FundFailure, {
      error,
      metadata: {
        currency: fiatCurrency
      }
    });
  }, [sendAnalytics, fiatCurrency]);
  const handleClick = useCallback((e) => {
    e.preventDefault();
    if (fundingUrlToRender && openIn === "tab") {
      handleAnalyticsInitiated();
      onClick == null ? void 0 : onClick();
      window.open(fundingUrlToRender, "_blank");
      return;
    }
    if (fundingUrlToRender) {
      handleAnalyticsInitiated();
      onClick == null ? void 0 : onClick();
      const {
        height,
        width
      } = getFundingPopupSize(popupSize, fundingUrlToRender);
      const popupWindow = openPopup({
        url: fundingUrlToRender,
        height,
        width,
        target
      });
      if (popupWindow) {
        startPopupMonitor(popupWindow);
      } else {
        handleAnalyticsFailure("Failed to open funding popup");
      }
    }
  }, [fundingUrlToRender, openIn, handleAnalyticsInitiated, onClick, popupSize, target, startPopupMonitor, handleAnalyticsFailure]);
  const buttonColorClass = useMemo(() => {
    if (buttonState === "error") {
      return prefixClassName("bg-ock-error");
    }
    return pressable.primary;
  }, [buttonState]);
  const classNames = cn(buttonColorClass, "rounded-ock-default px-4 py-3 inline-flex items-center justify-center space-x-2", {
    [pressable.disabled]: isDisabled
  }, text.headline, "text-ock-foreground-inverse", className);
  const buttonIcon = useMemo(() => {
    switch (buttonState) {
      case "loading":
        return "";
      case "success":
        return /* @__PURE__ */ jsx(SuccessSvg, { className: "ock:fill-[#F9FAFB]" });
      case "error":
        return /* @__PURE__ */ jsx(ErrorSvg, { className: "ock:fill-[#F9FAFB]" });
      default:
        return /* @__PURE__ */ jsx(AddSvg, {});
    }
  }, [buttonState]);
  const buttonTextContent = useMemo(() => {
    switch (buttonState) {
      case "loading":
        return "";
      case "success":
        return "Success";
      case "error":
        return "Something went wrong";
      default:
        return "Fund";
    }
  }, [buttonState]);
  const buttonContent = useMemo(() => {
    if (buttonState === "loading") {
      return /* @__PURE__ */ jsx(Spinner, {});
    }
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      buttonIcon && /* @__PURE__ */ jsx("span", { "data-testid": "ockFundButtonIcon", className: "ock:flex ock:h-6 ock:items-center", children: buttonIcon }),
      /* @__PURE__ */ jsx("span", { "data-testid": "ockFundButtonTextContent", children: buttonTextContent })
    ] });
  }, [buttonState, buttonIcon, buttonTextContent]);
  if (!address) {
    return /* @__PURE__ */ jsx(ConnectWallet, { className: cn("ock:w-full", className) });
  }
  if (render) {
    return render({
      status: buttonState,
      onClick: handleClick,
      isDisabled
    });
  }
  return /* @__PURE__ */ jsx("button", { className: classNames, onClick: handleClick, type: "button", disabled: isDisabled, "data-testid": "ockFundButton", children: children ?? buttonContent });
}
export {
  FundButton
};
