// @coinbase/onchainkit/dist/minikit/components/SafeArea.js
'use client';
import { jsx } from "react/jsx-runtime";
import { useMemo, useLayoutEffect, isValidElement, Children, cloneElement } from "react";
import { useIsInMiniApp } from "../hooks/useIsInMiniApp.js";
import { useMiniKit } from "../hooks/useMiniKit.js";
function SafeArea({
  children,
  asChild
}) {
  const {
    context
  } = useMiniKit();
  const {
    isInMiniApp
  } = useIsInMiniApp();
  const safeAreaInsets = useMemo(() => {
    var _a;
    const {
      top,
      bottom,
      left,
      right
    } = ((_a = context == null ? void 0 : context.client) == null ? void 0 : _a.safeAreaInsets) || {
      top: 0,
      bottom: 0,
      left: 0,
      right: 0
    };
    return {
      top,
      bottom,
      left,
      right
    };
  }, [context]);
  useLayoutEffect(() => {
    if (!isInMiniApp) {
      return;
    }
    Object.entries(safeAreaInsets).forEach(([key, value]) => {
      document.documentElement.style.setProperty(`--ock-minikit-safe-area-inset-${key}`, `${value}px`);
    });
  }, [safeAreaInsets, isInMiniApp]);
  const paddingStyles = useMemo(() => {
    return {
      paddingTop: "var(--ock-minikit-safe-area-inset-top, 0px)",
      paddingRight: "var(--ock-minikit-safe-area-inset-right, 0px)",
      paddingBottom: "var(--ock-minikit-safe-area-inset-bottom, 0px)",
      paddingLeft: "var(--ock-minikit-safe-area-inset-left, 0px)"
    };
  }, []);
  if (!isInMiniApp) {
    return children;
  }
  if (!children) {
    return null;
  }
  if (asChild) {
    if (!isValidElement(children)) {
      console.warn("SafeArea: children is not a valid element. Returning children as is.");
      return children;
    }
    const onlyChild = Children.only(children);
    const mergedStyle = {
      ...onlyChild.props.style,
      ...paddingStyles
    };
    return cloneElement(onlyChild, {
      style: mergedStyle
    });
  }
  return /* @__PURE__ */ jsx("div", { style: paddingStyles, children });
}
export {
  SafeArea
};
