// sushi/_cjs/evm/pool/sushiswap-v2/sushiswap-v2-pool.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SushiSwapV2Pool = void 0;
const tiny_invariant_1 = __importDefault(require("tiny-invariant"));
const amount_js_1 = require("../../../generic/currency/amount.js");
const price_js_1 = require("../../../generic/currency/price.js");
const constants_js_1 = require("../../../generic/math/constants.js");
const sqrt_js_1 = require("../../../generic/math/sqrt.js");
const index_js_1 = require("../../chain/index.js");
const sushiswap_v2_js_1 = require("../../config/features/sushiswap-v2.js");
const index_js_2 = require("../../currency/index.js");
const index_js_3 = require("../../dex/index.js");
const compute_sushiswap_v2_pool_address_js_1 = require("./compute-sushiswap-v2-pool-address.js");
const zod_js_1 = require("./zod.js");
class SushiSwapV2Pool {
    static getAddress(tokenA, tokenB) {
        if (!(0, sushiswap_v2_js_1.isSushiSwapV2ChainId)(tokenA.chainId)) {
            throw new Error(`ChainId Error: SushiSwapV2 is not available on ${(0, index_js_1.getEvmChainById)(tokenA.chainId).name}`);
        }
        return (0, compute_sushiswap_v2_pool_address_js_1.computeSushiSwapV2PoolAddress)({
            factoryAddress: sushiswap_v2_js_1.SUSHISWAP_V2_FACTORY_ADDRESS[tokenA.chainId],
            tokenA,
            tokenB,
        });
    }
    constructor(amountA, amountB) {
        this.swapGasCost = 60000n;
        this.minLiquidity = 1000n;
        const amounts = (amountA.currency.sortsBefore(amountB.currency)
            ? [amountA, amountB]
            : [amountB, amountA]);
        this.liquidityToken = new index_js_2.EvmToken({
            chainId: amounts[0].currency.chainId,
            address: SushiSwapV2Pool.getAddress(amounts[0].currency, amounts[1].currency),
            decimals: 18,
            symbol: 'UNI-V2',
            name: 'Uniswap V2',
        });
        this.tokenAmounts = amounts;
    }
    involvesToken(token) {
        return token.isSame(this.token0) || token.isSame(this.token1);
    }
    get token0Price() {
        const result = this.tokenAmounts[1].divToFraction(this.tokenAmounts[0].amount);
        return new price_js_1.Price({
            base: this.token0,
            quote: this.token1,
            denominator: result.denominator,
            numerator: result.numerator,
        });
    }
    get token1Price() {
        const result = this.tokenAmounts[0].divToFraction(this.tokenAmounts[1]);
        return new price_js_1.Price({
            base: this.token1,
            quote: this.token0,
            denominator: result.denominator,
            numerator: result.numerator,
        });
    }
    priceOf(token) {
        (0, tiny_invariant_1.default)(this.involvesToken(token), 'TOKEN');
        return token.isSame(this.token0) ? this.token0Price : this.token1Price;
    }
    get chainId() {
        return this.token0.chainId;
    }
    get token0() {
        return this.tokenAmounts[0].currency;
    }
    get token1() {
        return this.tokenAmounts[1].currency;
    }
    get reserve0() {
        return this.tokenAmounts[0];
    }
    get reserve1() {
        return this.tokenAmounts[1];
    }
    reserveOf(token) {
        (0, tiny_invariant_1.default)(this.involvesToken(token), 'TOKEN');
        return token.isSame(this.token0) ? this.reserve0 : this.reserve1;
    }
    getOutputAmount(inputAmount) {
        (0, tiny_invariant_1.default)(this.involvesToken(inputAmount.currency), 'TOKEN');
        if (this.reserve0.amount === constants_js_1.ZERO || this.reserve1.amount === constants_js_1.ZERO) {
            throw new index_js_3.InsufficientReservesError();
        }
        const inputReserve = this.reserveOf(inputAmount.currency);
        const outputReserve = this.reserveOf(inputAmount.currency.isSame(this.token0) ? this.token1 : this.token0);
        const inputAmountWithFee = inputAmount.amount * constants_js_1._997;
        const numerator = inputAmountWithFee * outputReserve.amount;
        const denominator = inputReserve.amount * constants_js_1._1000 + inputAmountWithFee;
        const outputAmount = new amount_js_1.Amount(inputAmount.currency.isSame(this.token0) ? this.token1 : this.token0, numerator / denominator);
        if (outputAmount.amount === constants_js_1.ZERO) {
            throw new index_js_3.InsufficientInputAmountError();
        }
        return [
            outputAmount,
            new SushiSwapV2Pool(inputReserve.add(inputAmount), outputReserve.sub(outputAmount)),
        ];
    }
    getInputAmount(outputAmount) {
        (0, tiny_invariant_1.default)(this.involvesToken(outputAmount.currency), 'TOKEN');
        if (this.reserve0.amount === constants_js_1.ZERO ||
            this.reserve1.amount === constants_js_1.ZERO ||
            outputAmount.amount >= this.reserveOf(outputAmount.currency).amount) {
            throw new index_js_3.InsufficientReservesError();
        }
        const outputReserve = this.reserveOf(outputAmount.currency);
        const inputReserve = this.reserveOf(outputAmount.currency.isSame(this.token0) ? this.token1 : this.token0);
        const numerator = inputReserve.amount * outputAmount.amount * constants_js_1._1000;
        const denominator = (outputReserve.amount - outputAmount.amount) * constants_js_1._997;
        const inputAmount = new amount_js_1.Amount(outputAmount.currency.isSame(this.token0) ? this.token1 : this.token0, numerator / denominator + constants_js_1.ONE);
        return [
            inputAmount,
            new SushiSwapV2Pool(inputReserve.add(inputAmount), outputReserve.sub(outputAmount)),
        ];
    }
    getLiquidityMinted(totalSupply, tokenAmountA, tokenAmountB) {
        (0, tiny_invariant_1.default)(totalSupply.currency.isSame(this.liquidityToken), 'LIQUIDITY');
        const tokenAmounts = (tokenAmountA.currency.sortsBefore(tokenAmountB.currency)
            ? [tokenAmountA, tokenAmountB]
            : [tokenAmountB, tokenAmountA]);
        (0, tiny_invariant_1.default)(tokenAmounts[0].currency.isSame(this.token0) &&
            tokenAmounts[1].currency.isSame(this.token1), 'TOKEN');
        let liquidity;
        if (totalSupply.amount === constants_js_1.ZERO) {
            liquidity =
                (0, sqrt_js_1.sqrt)(tokenAmounts[0].amount * tokenAmounts[1].amount) -
                    this.minLiquidity;
        }
        else {
            const amount0 = (tokenAmounts[0].amount * totalSupply.amount) / this.reserve0.amount;
            const amount1 = (tokenAmounts[1].amount * totalSupply.amount) / this.reserve1.amount;
            liquidity = amount0 <= amount1 ? amount0 : amount1;
        }
        if (liquidity <= constants_js_1.ZERO) {
            throw new index_js_3.InsufficientInputAmountError();
        }
        return new amount_js_1.Amount(this.liquidityToken, liquidity);
    }
    getLiquidityValue(token, totalSupply, liquidity, feeOn = false, kLast) {
        (0, tiny_invariant_1.default)(this.involvesToken(token), 'TOKEN');
        (0, tiny_invariant_1.default)(totalSupply.currency.isSame(this.liquidityToken), 'TOTAL_SUPPLY');
        (0, tiny_invariant_1.default)(liquidity.currency.isSame(this.liquidityToken), 'LIQUIDITY');
        (0, tiny_invariant_1.default)(liquidity.amount <= totalSupply.amount, 'LIQUIDITY');
        let totalSupplyAdjusted;
        if (!feeOn) {
            totalSupplyAdjusted = totalSupply;
        }
        else {
            (0, tiny_invariant_1.default)(typeof kLast !== 'undefined', 'K_LAST');
            const kLastParsed = BigInt(typeof kLast === 'bigint' ? kLast.toString() : kLast);
            if (kLastParsed !== constants_js_1.ZERO) {
                const rootK = (0, sqrt_js_1.sqrt)(this.reserve0.amount * this.reserve1.amount);
                const rootKLast = (0, sqrt_js_1.sqrt)(kLastParsed);
                if (rootK > rootKLast) {
                    const numerator = totalSupply.amount * (rootK - rootKLast);
                    const denominator = rootK * constants_js_1.FIVE + rootKLast;
                    const feeLiquidity = numerator / denominator;
                    totalSupplyAdjusted = totalSupply.add(new amount_js_1.Amount(this.liquidityToken, feeLiquidity));
                }
                else {
                    totalSupplyAdjusted = totalSupply;
                }
            }
            else {
                totalSupplyAdjusted = totalSupply;
            }
        }
        return new amount_js_1.Amount(token, (liquidity.amount * this.reserveOf(token).amount) /
            totalSupplyAdjusted.amount);
    }
    serialize() {
        return zod_js_1.sushiSwapV2PoolSchema.parse({
            reserve0: this.tokenAmounts[0].toJSON(),
            reserve1: this.tokenAmounts[1].toJSON(),
        });
    }
    static deserialize(pair) {
        const tokenA = new index_js_2.EvmToken(pair.reserve0.currency);
        const tokenB = new index_js_2.EvmToken(pair.reserve1.currency);
        const amountA = new amount_js_1.Amount(tokenA, pair.reserve0.amount);
        const amountB = new amount_js_1.Amount(tokenB, pair.reserve1.amount);
        return new SushiSwapV2Pool(amountA, amountB);
    }
}
exports.SushiSwapV2Pool = SushiSwapV2Pool;
