// sushi/_cjs/evm/config/tokens/wrapped-native.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isWNativeSupported = exports.WNATIVE = exports.WNATIVE_ADDRESS = void 0;
exports.isWrapOrUnwrap = isWrapOrUnwrap;
const viem_1 = require("viem");
const index_js_1 = require("../../chain/index.js");
const token_js_1 = require("../../currency/token.js");
const WETH9_js_1 = require("./tokens/WETH9.js");
const WRBTC_js_1 = require("./tokens/WRBTC.js");
exports.WNATIVE_ADDRESS = {
    [index_js_1.EvmChainId.ETHEREUM]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.ETHEREUM],
    [index_js_1.EvmChainId.SEPOLIA]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.SEPOLIA],
    [index_js_1.EvmChainId.OPTIMISM]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.OPTIMISM],
    [index_js_1.EvmChainId.ARBITRUM]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.ARBITRUM],
    [index_js_1.EvmChainId.ARBITRUM_SEPOLIA]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.ARBITRUM_SEPOLIA],
    [index_js_1.EvmChainId.FANTOM]: '0x21be370d5312f44cb42ce377bc9b8a0cef1a4c83',
    [index_js_1.EvmChainId.POLYGON]: '0x0d500b1d8e8ef31e21c99d1db9a6444d3adf1270',
    [index_js_1.EvmChainId.GNOSIS]: '0xe91d153e0b41518a2ce8dd3d7944fa863463a97d',
    [index_js_1.EvmChainId.BSC]: '0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c',
    [index_js_1.EvmChainId.AVALANCHE]: '0xb31f66aa3c1e785363f0875a1b74e27b85fd66c7',
    [index_js_1.EvmChainId.HARMONY]: '0xcf664087a5bb0237a0bad6742852ec6c8d69a27a',
    [index_js_1.EvmChainId.CELO]: '0x471ece3750da237f93b8e339c536989b8978a438',
    [index_js_1.EvmChainId.KAVA]: '0xc86c7c0efbd6a49b35e8714c5f59d99de09a225b',
    [index_js_1.EvmChainId.METIS]: '0x75cb093e4d61d2a2e65d8e0bbb01de8d89b53481',
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.ARBITRUM_NOVA],
    [index_js_1.EvmChainId.BOBA]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.BOBA],
    [index_js_1.EvmChainId.BOBA_BNB]: '0xc58aad327d6d58d979882601ba8dda0685b505ea',
    [index_js_1.EvmChainId.BTTC]: '0x23181f21dea5936e24163ffaba4ea3b316b57f3c',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.POLYGON_ZKEVM],
    [index_js_1.EvmChainId.THUNDERCORE]: '0x413cefea29f2d07b8f2acfa69d92466b9535f717',
    [index_js_1.EvmChainId.FILECOIN]: '0x60e1773636cf5e4a227d9ac24f20feca034ee25a',
    [index_js_1.EvmChainId.HAQQ]: '0xec8cc083787c6e5218d86f9ff5f28d4cc377ac54',
    [index_js_1.EvmChainId.CORE]: '0x40375c92d9faf44d2f9db9bd9ba41a3317a2404f',
    [index_js_1.EvmChainId.ZKSYNC_ERA]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.ZKSYNC_ERA],
    [index_js_1.EvmChainId.LINEA]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.LINEA],
    [index_js_1.EvmChainId.BASE]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.BASE],
    [index_js_1.EvmChainId.SCROLL]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.SCROLL],
    [index_js_1.EvmChainId.ZETACHAIN]: '0x5f0b1a82749cb4e2278ec87f8bf6b618dc71a8bf',
    [index_js_1.EvmChainId.CRONOS]: '0x5c7f8a570d578ed84e63fdfa7b1ee72deae1ae23',
    [index_js_1.EvmChainId.BLAST]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.BLAST],
    [index_js_1.EvmChainId.SKALE_EUROPA]: viem_1.zeroAddress,
    [index_js_1.EvmChainId.ROOTSTOCK]: WRBTC_js_1.WRBTC_ADDRESS[index_js_1.EvmChainId.ROOTSTOCK],
    [index_js_1.EvmChainId.MANTLE]: '0x78c1b0c915c4faa5fffa6cabf0219da63d7f4cb8',
    [index_js_1.EvmChainId.MANTA]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.MANTA],
    [index_js_1.EvmChainId.MODE]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.MODE],
    [index_js_1.EvmChainId.TAIKO]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.TAIKO],
    [index_js_1.EvmChainId.ZKLINK]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.ZKLINK],
    [index_js_1.EvmChainId.APE]: '0x48b62137edfa95a428d35c09e44256a739f6b557',
    [index_js_1.EvmChainId.SONIC]: '0x039e2fb66102314ce7b64ce5ce3e5183bc94ad38',
    [index_js_1.EvmChainId.HEMI]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.HEMI],
    [index_js_1.EvmChainId.TATARA]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.TATARA],
    [index_js_1.EvmChainId.KATANA]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.KATANA],
    [index_js_1.EvmChainId.HYPEREVM]: WETH9_js_1.WETH9_ADDRESS[index_js_1.EvmChainId.HYPEREVM],
    [index_js_1.EvmChainId.BERACHAIN]: '0x6969696969696969696969696969696969696969',
    [index_js_1.EvmChainId.PLASMA]: '0x6100e367285b01f48d07953803a2d8dca5d19873',
};
exports.WNATIVE = {
    [index_js_1.EvmChainId.ETHEREUM]: WETH9_js_1.WETH9[index_js_1.EvmChainId.ETHEREUM],
    [index_js_1.EvmChainId.SEPOLIA]: WETH9_js_1.WETH9[index_js_1.EvmChainId.SEPOLIA],
    [index_js_1.EvmChainId.OPTIMISM]: WETH9_js_1.WETH9[index_js_1.EvmChainId.OPTIMISM],
    [index_js_1.EvmChainId.FANTOM]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.FANTOM,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.FANTOM],
        decimals: 18,
        symbol: 'WFTM',
        name: 'Wrapped FTM',
    }),
    [index_js_1.EvmChainId.POLYGON]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.POLYGON,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.POLYGON],
        decimals: 18,
        symbol: 'WPOL',
        name: 'Wrapped POL',
    }),
    [index_js_1.EvmChainId.GNOSIS]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.GNOSIS,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.GNOSIS],
        decimals: 18,
        symbol: 'WXDAI',
        name: 'Wrapped xDai',
    }),
    [index_js_1.EvmChainId.BSC]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.BSC,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.BSC],
        decimals: 18,
        symbol: 'WBNB',
        name: 'Wrapped BNB',
    }),
    [index_js_1.EvmChainId.ARBITRUM]: WETH9_js_1.WETH9[index_js_1.EvmChainId.ARBITRUM],
    [index_js_1.EvmChainId.ARBITRUM_SEPOLIA]: WETH9_js_1.WETH9[index_js_1.EvmChainId.ARBITRUM_SEPOLIA],
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: WETH9_js_1.WETH9[index_js_1.EvmChainId.ARBITRUM_NOVA],
    [index_js_1.EvmChainId.AVALANCHE]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.AVALANCHE,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.AVALANCHE],
        decimals: 18,
        symbol: 'WAVAX',
        name: 'Wrapped AVAX',
    }),
    [index_js_1.EvmChainId.HARMONY]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.HARMONY,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.HARMONY],
        decimals: 18,
        symbol: 'WONE',
        name: 'Wrapped ONE',
    }),
    [index_js_1.EvmChainId.CELO]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.CELO,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.CELO],
        decimals: 18,
        symbol: 'CELO',
        name: 'Celo',
    }),
    [index_js_1.EvmChainId.KAVA]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.KAVA,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.KAVA],
        decimals: 18,
        symbol: 'WKAVA',
        name: 'Wrapped Kava',
    }),
    [index_js_1.EvmChainId.METIS]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.METIS,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.METIS],
        decimals: 18,
        symbol: 'WMETIS',
        name: 'Wrapped Metis',
    }),
    [index_js_1.EvmChainId.BOBA]: WETH9_js_1.WETH9[index_js_1.EvmChainId.BOBA],
    [index_js_1.EvmChainId.BOBA_BNB]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.BOBA_BNB,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.BOBA_BNB],
        decimals: 18,
        symbol: 'WBOBA',
        name: 'Wrapped Boba',
    }),
    [index_js_1.EvmChainId.BTTC]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.BTTC,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.BTTC],
        decimals: 18,
        symbol: 'WBTT',
        name: 'Wrapped BitTorrent Token',
    }),
    [index_js_1.EvmChainId.THUNDERCORE]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.THUNDERCORE,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.THUNDERCORE],
        decimals: 18,
        symbol: 'WTT',
        name: 'Wrapped Thunder Token',
    }),
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: WETH9_js_1.WETH9[index_js_1.EvmChainId.POLYGON_ZKEVM],
    [index_js_1.EvmChainId.HAQQ]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.HAQQ,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.HAQQ],
        decimals: 18,
        symbol: 'WISLM',
        name: 'Wrapped Islamic Coin',
    }),
    [index_js_1.EvmChainId.CORE]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.CORE,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.CORE],
        decimals: 18,
        symbol: 'WCORE',
        name: 'Wrapped Core',
    }),
    [index_js_1.EvmChainId.ZKSYNC_ERA]: WETH9_js_1.WETH9[index_js_1.EvmChainId.ZKSYNC_ERA],
    [index_js_1.EvmChainId.LINEA]: WETH9_js_1.WETH9[index_js_1.EvmChainId.LINEA],
    [index_js_1.EvmChainId.BASE]: WETH9_js_1.WETH9[index_js_1.EvmChainId.BASE],
    [index_js_1.EvmChainId.SCROLL]: WETH9_js_1.WETH9[index_js_1.EvmChainId.SCROLL],
    [index_js_1.EvmChainId.FILECOIN]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.FILECOIN,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.FILECOIN],
        decimals: 18,
        symbol: 'WFIL',
        name: 'Wrapped FIL',
    }),
    [index_js_1.EvmChainId.ZETACHAIN]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.ZETACHAIN,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.ZETACHAIN],
        decimals: 18,
        symbol: 'WZETA',
        name: 'Wrapped ZETA',
    }),
    [index_js_1.EvmChainId.CRONOS]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.CRONOS,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.CRONOS],
        decimals: 18,
        symbol: 'WCRO',
        name: 'Wrapped CRO',
    }),
    [index_js_1.EvmChainId.BLAST]: WETH9_js_1.WETH9[index_js_1.EvmChainId.BLAST],
    [index_js_1.EvmChainId.SKALE_EUROPA]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.SKALE_EUROPA,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.SKALE_EUROPA],
        decimals: 0,
        symbol: 'WsFUEL',
        name: 'Wrapped sFUEL',
    }),
    [index_js_1.EvmChainId.ROOTSTOCK]: WRBTC_js_1.WRBTC[index_js_1.EvmChainId.ROOTSTOCK],
    [index_js_1.EvmChainId.MANTLE]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.MANTLE,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.MANTLE],
        decimals: 18,
        symbol: 'WMNT',
        name: 'Wrapped MNT',
    }),
    [index_js_1.EvmChainId.MANTA]: WETH9_js_1.WETH9[index_js_1.EvmChainId.MANTA],
    [index_js_1.EvmChainId.MODE]: WETH9_js_1.WETH9[index_js_1.EvmChainId.MODE],
    [index_js_1.EvmChainId.TAIKO]: WETH9_js_1.WETH9[index_js_1.EvmChainId.TAIKO],
    [index_js_1.EvmChainId.ZKLINK]: WETH9_js_1.WETH9[index_js_1.EvmChainId.ZKLINK],
    [index_js_1.EvmChainId.APE]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.APE,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.APE],
        decimals: 18,
        symbol: 'WAPE',
        name: 'Wrapped Ape',
    }),
    [index_js_1.EvmChainId.SONIC]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.SONIC,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.SONIC],
        decimals: 18,
        symbol: 'wS',
        name: 'Wrapped Sonic',
    }),
    [index_js_1.EvmChainId.HEMI]: WETH9_js_1.WETH9[index_js_1.EvmChainId.HEMI],
    [index_js_1.EvmChainId.TATARA]: WETH9_js_1.WETH9[index_js_1.EvmChainId.TATARA],
    [index_js_1.EvmChainId.KATANA]: WETH9_js_1.WETH9[index_js_1.EvmChainId.KATANA],
    [index_js_1.EvmChainId.HYPEREVM]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.HYPEREVM,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.HYPEREVM],
        decimals: 18,
        symbol: 'WHYPE',
        name: 'Wrapped HYPE',
    }),
    [index_js_1.EvmChainId.BERACHAIN]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.BERACHAIN,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.BERACHAIN],
        decimals: 18,
        symbol: 'WBERA',
        name: 'Wrapped BERA',
    }),
    [index_js_1.EvmChainId.PLASMA]: new token_js_1.EvmToken({
        chainId: index_js_1.EvmChainId.PLASMA,
        address: exports.WNATIVE_ADDRESS[index_js_1.EvmChainId.PLASMA],
        decimals: 18,
        symbol: 'WXPL',
        name: 'Wrapped XPL',
    }),
};
const isWNativeSupported = (chainId) => exports.WNATIVE_ADDRESS[chainId] !== viem_1.zeroAddress;
exports.isWNativeSupported = isWNativeSupported;
function isWrapOrUnwrap({ from, to, }) {
    if (from.type === 'native' && from.wrap().isSame(to)) {
        return true;
    }
    return to.type === 'native' && to.wrap().isSame(from);
}
