// sushi/_cjs/evm/api/quote.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getQuote = getQuote;
const zod_1 = require("zod");
const version_js_1 = require("../../version.js");
const zod_js_1 = require("../validate/zod.js");
const types_js_1 = require("./types.js");
function quoteResponseSchema(visualize) {
    const tokenSchema = zod_1.z.object({
        address: zod_js_1.szevm.address(),
        decimals: zod_1.z.number(),
        symbol: zod_1.z.string(),
        name: zod_1.z.string(),
    });
    const baseSuccessPartial = zod_1.z
        .object({
        status: zod_1.z.literal(types_js_1.RouteStatus.Success).or(zod_1.z.literal(types_js_1.RouteStatus.Partial)),
        tokens: zod_1.z.array(tokenSchema),
        tokenFrom: zod_1.z.number(),
        tokenTo: zod_1.z.number(),
        swapPrice: zod_1.z.number(),
        priceImpact: zod_1.z.number(),
        amountIn: zod_1.z.string(),
        assumedAmountOut: zod_1.z.string(),
    })
        .transform((data) => ({
        ...data,
        tokenFrom: data.tokens[data.tokenFrom],
        tokenTo: data.tokens[data.tokenTo],
    }));
    const baseNoWay = zod_1.z.object({
        status: zod_1.z.literal(types_js_1.RouteStatus.NoWay),
    });
    const baseSchema = baseSuccessPartial.or(baseNoWay);
    const baseVisualizeSchema = baseSchema
        .and(zod_1.z.object({
        visualization: zod_1.z.object({
            liquidityProviders: zod_1.z.array(zod_1.z.string()),
            nodes: zod_1.z.array(tokenSchema),
            links: zod_1.z.array(zod_1.z.object({
                source: zod_1.z.number(),
                target: zod_1.z.number(),
                liquidityProvider: zod_1.z.number(),
                amountIn: zod_1.z.string(),
                amountOut: zod_1.z.string(),
                value: zod_1.z.number(),
            })),
        }),
    }))
        .or(baseNoWay);
    return (visualize ? baseVisualizeSchema : baseSchema);
}
async function getQuote(params, options) {
    const url = new URL(`quote/v7/${params.chainId}`, params.baseUrl ?? 'https://api.sushi.com');
    url.searchParams.append('tokenIn', params.tokenIn);
    url.searchParams.append('tokenOut', params.tokenOut);
    url.searchParams.append('amount', params.amount.toString());
    url.searchParams.append('maxSlippage', params.maxSlippage.toString());
    if (params.maxPriceImpact) {
        url.searchParams.append('maxPriceImpact', params.maxPriceImpact.toString());
    }
    if (typeof params.fee === 'bigint' &&
        params.fee > 0n &&
        params.feeReceiver !== undefined) {
        url.searchParams.append('fee', params.fee.toString());
        url.searchParams.append('feeReceiver', params.feeReceiver);
        if (params.feeBy !== undefined) {
            url.searchParams.append('feeBy', params.feeBy);
        }
    }
    if (params?.visualize) {
        url.searchParams.append('visualize', params.visualize.toString());
    }
    if (params.referrer) {
        url.searchParams.append('referrer', params.referrer);
    }
    else {
        url.searchParams.append('referrer', `sushi-sdk/${version_js_1.version}`);
    }
    const res = await fetch(url.toString(), options);
    if (!res.ok) {
        throw new Error(`Failed to fetch quote: ${await res.text()}`);
    }
    return quoteResponseSchema(params.visualize).parse(await res.json());
}
