// sushi/_cjs/evm/api/swap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSwap = getSwap;
const zod_1 = require("zod");
const zod_js_1 = require("../../generic/validate/zod.js");
const version_js_1 = require("../../version.js");
const zod_js_2 = require("../validate/zod.js");
const types_js_1 = require("./types.js");
function swapResponseSchema(simulate) {
    const tokenSchema = zod_1.z.object({
        address: zod_js_2.szevm.address(),
        decimals: zod_1.z.number(),
        symbol: zod_1.z.string(),
        name: zod_1.z.string(),
    });
    const txSchema = zod_1.z.object({
        from: zod_js_2.szevm.address(),
        to: zod_js_2.szevm.address(),
        gasPrice: zod_1.z.number(),
        data: zod_js_1.sz.hex(),
        value: zod_1.z
            .string()
            .optional()
            .transform((value) => BigInt(value || 0)),
    });
    const baseSuccessPartial = zod_1.z
        .object({
        status: zod_1.z.literal(types_js_1.RouteStatus.Success).or(zod_1.z.literal(types_js_1.RouteStatus.Partial)),
        tokens: zod_1.z.array(tokenSchema),
        tokenFrom: zod_1.z.number(),
        tokenTo: zod_1.z.number(),
        swapPrice: zod_1.z.number(),
        priceImpact: zod_1.z.number(),
        amountIn: zod_1.z.string(),
        assumedAmountOut: zod_1.z.string(),
        tx: txSchema,
    })
        .transform((data) => ({
        ...data,
        tokenFrom: data.tokens[data.tokenFrom],
        tokenTo: data.tokens[data.tokenTo],
    }));
    const baseNoWay = zod_1.z.object({
        status: zod_1.z.literal(types_js_1.RouteStatus.NoWay),
    });
    const baseSchema = baseSuccessPartial.or(baseNoWay);
    const baseSimulateSchema = baseSchema
        .and(zod_1.z.object({ tx: txSchema.extend({ gas: zod_1.z.string() }) }))
        .or(baseNoWay);
    return (simulate ? baseSimulateSchema : baseSchema);
}
async function getSwap(params, options) {
    const url = new URL(`swap/v7/${params.chainId}`, params.baseUrl ?? 'https://api.sushi.com');
    url.searchParams.append('tokenIn', params.tokenIn);
    url.searchParams.append('tokenOut', params.tokenOut);
    url.searchParams.append('sender', params.sender);
    url.searchParams.append('amount', params.amount.toString());
    url.searchParams.append('maxSlippage', params.maxSlippage.toString());
    if (params.source) {
        url.searchParams.append('source', params.source);
    }
    if (params.maxPriceImpact) {
        url.searchParams.append('maxPriceImpact', params.maxPriceImpact.toString());
    }
    if (params.recipient) {
        url.searchParams.append('recipient', params.recipient);
    }
    if (typeof params.fee === 'bigint' &&
        params.fee > 0n &&
        params.feeReceiver !== undefined) {
        url.searchParams.append('fee', params.fee.toString());
        url.searchParams.append('feeReceiver', params.feeReceiver);
        if (params.feeBy !== undefined) {
            url.searchParams.append('feeBy', params.feeBy);
        }
    }
    if (params.referrer) {
        url.searchParams.append('referrer', params.referrer);
    }
    else {
        url.searchParams.append('referrer', `sushi-sdk/${version_js_1.version}`);
    }
    if (params.simulate !== undefined) {
        url.searchParams.append('simulate', params.simulate.toString());
        if (params.validate !== undefined) {
            url.searchParams.append('validate', params.validate.toString());
        }
    }
    if (params.apiKey !== undefined) {
        url.searchParams.append('apiKey', params.apiKey);
    }
    const res = await fetch(url.toString(), options);
    if (!res.ok) {
        throw new Error(`Failed to fetch swap: ${await res.text()}`);
    }
    return swapResponseSchema(params.simulate).parse(await res.json());
}
