// @coinbase/agentkit/dist/action-providers/truemarkets/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarketStatus = exports.YESNO_DECIMALS = exports.TYD_DECIMALS = exports.USDC_DECIMALS = exports.TYD_ADDRESS = exports.USDC_ADDRESS = exports.TruthMarketManager_ADDRESS = exports.UniswapV3PoolABI = exports.TruthMarketManagerABI = exports.TruthMarketABI = void 0;
const viem_1 = require("viem");
// Simplified ABIs
exports.TruthMarketABI = [
    (0, viem_1.parseAbiItem)("function marketQuestion() view returns (string)"),
    (0, viem_1.parseAbiItem)("function marketSource() view returns (string)"),
    (0, viem_1.parseAbiItem)("function getCurrentStatus() view returns (uint8)"),
    (0, viem_1.parseAbiItem)("function endOfTrading() view returns (uint256)"),
    (0, viem_1.parseAbiItem)("function getPoolAddresses() view returns (address yesPool, address noPool)"),
    (0, viem_1.parseAbiItem)("function paymentToken() view returns (address)"),
    (0, viem_1.parseAbiItem)("function additionalInfo() view returns (string)"),
    (0, viem_1.parseAbiItem)("function winningPosition() view returns (uint8)"),
];
exports.TruthMarketManagerABI = [
    (0, viem_1.parseAbiItem)("function numberOfActiveMarkets() view returns (uint256)"),
    (0, viem_1.parseAbiItem)("function getActiveMarketAddress(uint256) view returns (address)"),
    (0, viem_1.parseAbiItem)("function creatorAddress(address) view returns (address)"),
    (0, viem_1.parseAbiItem)("function resolverAddress(address) view returns (address)"),
];
exports.UniswapV3PoolABI = [
    (0, viem_1.parseAbiItem)("function liquidity() view returns (uint128)"),
    (0, viem_1.parseAbiItem)("function slot0() view returns (uint160 sqrtPriceX96, int24 tick, uint16 observationIndex, uint16 observationCardinality, uint16 observationCardinalityNext, uint8 feeProtocol, bool unlocked)"),
    (0, viem_1.parseAbiItem)("function token0() view returns (address)"),
    (0, viem_1.parseAbiItem)("function token1() view returns (address)"),
];
exports.TruthMarketManager_ADDRESS = "0x61A98Bef11867c69489B91f340fE545eEfc695d7";
exports.USDC_ADDRESS = "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913";
exports.TYD_ADDRESS = "0xb13CF163d916917d9cD6E836905cA5f12a1dEF4B";
exports.USDC_DECIMALS = 6;
exports.TYD_DECIMALS = 6;
exports.YESNO_DECIMALS = 18;
/**
 * Market status enum
 */
exports.MarketStatus = {
    Created: 0,
    OpenForResolution: 1,
    ResolutionProposed: 2,
    DisputeRaised: 3,
    SetByCouncil: 4,
    ResetByCouncil: 5,
    EscalatedDisputeRaised: 6,
    Finalized: 7,
};
