// sushi/_esm/tvm/currency/native.js
import * as z from 'zod';
import { Native } from '../../generic/currency/native.js';
import { isTvmChainId } from '../chain/chains.js';
import { NATIVE } from '../config/index.js';
import { WNATIVE } from '../config/tokens/wrapped-native.js';
import { TvmToken } from './token.js';
export class TvmNative extends Native {
    wrap() {
        return new TvmToken({
            ...WNATIVE[this.chainId],
            metadata: structuredClone(this.metadata),
        });
    }
    toJSON() {
        return {
            chainId: this.chainId,
            symbol: this.symbol,
            name: this.name,
            decimals: this.decimals,
            type: this.type,
            metadata: this.metadata,
        };
    }
    static fromJSON(data) {
        return new this(data);
    }
    static fromChainId(chainId) {
        return NATIVE[chainId];
    }
}
export const serializedTvmNativeSchema = ({ metadata, } = {}) => z.object({
    chainId: z.number().int().refine(isTvmChainId),
    symbol: z.string(),
    name: z.string(),
    decimals: z.number().int().nonnegative(),
    type: z.literal('native'),
    metadata: (metadata ||
        z.record(z.unknown()).optional().default({})),
});
