// sushi/_esm/tvm/currency/token.js
import * as z from 'zod';
import { Token } from '../../generic/currency/token.js';
import { isTvmChainId } from '../chain/chains.js';
export function isTvmAddress(address) {
    return !!address.match(/^T[1-9A-HJ-NP-Za-km-z]{33}$/);
}
export class TvmToken extends Token {
    constructor({ origin, address, ...rest }) {
        super({ address: address.toLowerCase(), ...rest });
        this.origin = origin;
    }
    wrap() {
        return this;
    }
    toJSON() {
        return {
            chainId: this.chainId,
            address: this.address,
            symbol: this.symbol,
            name: this.name,
            decimals: this.decimals,
            type: this.type,
            metadata: this.metadata,
        };
    }
    static fromJSON(data) {
        return new this(data);
    }
}
export const serializedTvmTokenSchema = ({ metadata, } = {}) => z.object({
    chainId: z.number().int().refine(isTvmChainId),
    address: z.string().refine(isTvmAddress),
    symbol: z.string(),
    name: z.string(),
    decimals: z.number().int().nonnegative(),
    type: z.literal('token'),
    metadata: (metadata ||
        z.record(z.unknown()).optional().default({})),
});
