// @coinbase/onchainkit/dist/identity/components/Avatar.js
'use client';
import { jsx, jsxs } from "react/jsx-runtime";
import { useIdentityContext } from "./IdentityProvider.js";
import { useAvatar } from "../hooks/useAvatar.js";
import { useName } from "../hooks/useName.js";
import { findComponent } from "../../internal/utils/findComponent.js";
import { useMemo, Children } from "react";
import { defaultAvatarSVG } from "../../internal/svg/defaultAvatarSVG.js";
import { defaultLoadingSVG } from "../../internal/svg/defaultLoadingSVG.js";
import { cn } from "../../styles/theme.js";
import { Badge } from "./Badge.js";
import { DisplayBadge } from "./DisplayBadge.js";
function Avatar({
  address = null,
  chain,
  className,
  defaultComponent,
  loadingComponent,
  children,
  ...props
}) {
  const {
    address: contextAddress
  } = useIdentityContext();
  const accountAddress = address ?? contextAddress;
  if (!accountAddress) {
    console.error("Avatar: an Ethereum address must be provided to the Identity or Avatar component.");
    return null;
  }
  return /* @__PURE__ */ jsx(AvatarContent, { address, chain, className, defaultComponent, loadingComponent, ...props, children });
}
function AvatarContent({
  address = null,
  chain,
  className,
  defaultComponent,
  loadingComponent,
  children,
  name: nameOverride,
  avatar: avatarOverride,
  ...props
}) {
  const {
    address: contextAddress,
    chain: contextChain
  } = useIdentityContext();
  const accountAddress = address ?? contextAddress;
  const accountChain = chain ?? contextChain;
  const {
    data: name,
    isLoading: isLoadingName
  } = useName({
    address: accountAddress,
    chain: accountChain
  });
  const {
    data: avatar,
    isLoading: isLoadingAvatar
  } = useAvatar({
    ensName: name ?? "",
    chain: accountChain
  }, {
    enabled: !!name
  });
  const {
    resolvedName,
    resolvedAvatar
  } = useMemo(() => {
    return {
      resolvedName: nameOverride ?? name ?? "",
      resolvedAvatar: avatarOverride ?? avatar ?? ""
    };
  }, [name, avatar, nameOverride, avatarOverride]);
  const badge = useMemo(() => {
    return Children.toArray(children).find(findComponent(Badge));
  }, [children]);
  const defaultAvatar = defaultComponent || defaultAvatarSVG;
  const loadingAvatar = loadingComponent || defaultLoadingSVG;
  if (isLoadingName || isLoadingAvatar) {
    return /* @__PURE__ */ jsx("div", { className: cn("ock:h-8 ock:w-8 ock:overflow-hidden ock:rounded-full", className), children: loadingAvatar });
  }
  return /* @__PURE__ */ jsx(AvatarRenderer, { avatar: resolvedAvatar, name: resolvedName, defaultAvatar, badge, accountAddress, className, ...props });
}
function AvatarRenderer({
  className,
  avatar,
  name,
  defaultAvatar,
  badge,
  accountAddress,
  ...rest
}) {
  return /* @__PURE__ */ jsxs("div", { className: "ock:relative", children: [
    /* @__PURE__ */ jsx("div", { "data-testid": "ockAvatar_ImageContainer", className: cn("ock:h-10 ock:w-10 ock:overflow-hidden ock:rounded-full", className), children: name && avatar ? /* @__PURE__ */ jsx("img", { className: "ock:min-h-full ock:min-w-full ock:object-cover", "data-testid": "ockAvatar_Image", loading: "lazy", width: "100%", height: "100%", decoding: "async", src: avatar, alt: name, ...rest }) : /* @__PURE__ */ jsx("div", { className: cn("ock:border-ock-background", "ock:h-full ock:w-full ock:border"), children: defaultAvatar }) }),
    badge && /* @__PURE__ */ jsx(DisplayBadge, { address: accountAddress, children: /* @__PURE__ */ jsx("div", { "data-testid": "ockAvatar_BadgeContainer", className: "ock:-bottom-0.5 ock:-right-0.5 ock:absolute ock:flex ock:h-[15px] ock:w-[15px] ock:items-center ock:justify-center ock:rounded-full ock:bg-transparent", children: /* @__PURE__ */ jsx("div", { className: "ock:flex ock:h-3 ock:w-3 ock:items-center ock:justify-center", children: badge }) }) })
  ] });
}
export {
  Avatar
};
