// @coinbase/onchainkit/dist/identity/components/Badge.js
'use client';
import { jsxs, jsx } from "react/jsx-runtime";
import { useAttestations } from "../hooks/useAttestations.js";
import { COINBASE_VERIFIED_ACCOUNT_SCHEMA_ID } from "../constants.js";
import { badgeSvg } from "../../internal/svg/badgeSvg.js";
import { zIndex } from "../../styles/constants.js";
import { cn, text, pressable } from "../../styles/theme.js";
import { useState, useMemo } from "react";
import { useOnchainKit } from "../../useOnchainKit.js";
import { useIdentityContext } from "./IdentityProvider.js";
function Badge({
  className,
  tooltip = false
}) {
  const [showTooltip, setShowTooltip] = useState(false);
  const {
    address,
    schemaId: contextSchemaId
  } = useIdentityContext();
  const {
    chain
  } = useOnchainKit();
  const attestations = useAttestations({
    address,
    chain,
    schemaId: tooltip ? contextSchemaId ?? COINBASE_VERIFIED_ACCOUNT_SCHEMA_ID : null
  });
  const displayText = useMemo(() => {
    if (!tooltip) {
      return null;
    }
    return typeof tooltip === "string" ? tooltip : extractAttestationName(attestations[0]);
  }, [tooltip, attestations]);
  const ariaLabel = useMemo(() => {
    if (displayText) {
      return `Verification badge: ${displayText}`;
    }
    return "Verification badge";
  }, [displayText]);
  const badgeSize = "12px";
  return /* @__PURE__ */ jsxs("div", { className: "ock:relative ock:inline-flex", "data-testid": "ockBadgeContainer", children: [
    /* @__PURE__ */ jsx("span", { className: cn("ock:bg-ock-primary", "ock:border-ock-background", "ock:rounded-ock-default", tooltip && "ock:cursor-pointer", className), style: {
      height: badgeSize,
      width: badgeSize,
      maxHeight: badgeSize,
      maxWidth: badgeSize
    }, "aria-label": ariaLabel, "data-testid": "ockBadge", ...tooltip && {
      onMouseEnter: () => setShowTooltip(true),
      onMouseLeave: () => setShowTooltip(false)
    }, children: badgeSvg }),
    showTooltip && tooltip && /* @__PURE__ */ jsxs("div", { className: cn("ock:rounded-ock-default", "ock:border-ock-background", pressable.alternate, text.legal, "ock:text-ock-foreground", zIndex.tooltip, "ock:-translate-x-1/2 ock:absolute ock:bottom-full ock:left-1/2 ock:mb-1 ock:transform", "ock:whitespace-nowrap ock:px-1.5 ock:py-0.5"), "data-testid": "ockBadgeTooltip", children: [
      displayText,
      /* @__PURE__ */ jsx("div", { className: cn("ock:-translate-x-1/2 ock:absolute ock:top-full ock:left-1/2 ock:transform") })
    ] })
  ] });
}
function extractAttestationName(attestation) {
  var _a;
  if (!(attestation == null ? void 0 : attestation.decodedDataJson)) {
    return "Verified";
  }
  try {
    const decodedData = JSON.parse(attestation.decodedDataJson);
    if (Array.isArray(decodedData) && ((_a = decodedData[0]) == null ? void 0 : _a.name)) {
      return decodedData[0].name;
    }
    const value = Object.values(decodedData)[0];
    if (typeof value === "string") {
      return value;
    }
    if (value && typeof value === "object" && "value" in value && typeof value.value === "string") {
      return value.value;
    }
  } catch {
  }
  return "Verified";
}
export {
  Badge
};
