// @coinbase/onchainkit/dist/identity/components/IdentityLayout.js
import { jsxs, jsx } from "react/jsx-runtime";
import { findComponent } from "../../internal/utils/findComponent.js";
import { cn } from "../../styles/theme.js";
import { useMemo, Children, cloneElement } from "react";
import { Address } from "./Address.js";
import { Avatar } from "./Avatar.js";
import { EthBalance } from "./EthBalance.js";
import { Name } from "./Name.js";
import { Socials } from "./Socials.js";
function IdentityLayout({
  children,
  className,
  hasCopyAddressOnClick
}) {
  const {
    avatar,
    name,
    address: addressComponent,
    ethBalance,
    socials
  } = useMemo(() => {
    const childrenArray = Children.toArray(children);
    const addressElement = childrenArray.find(findComponent(Address));
    return {
      avatar: childrenArray.find(findComponent(Avatar)),
      name: childrenArray.find(findComponent(Name)),
      address: addressElement ? cloneElement(addressElement, {
        hasCopyAddressOnClick
      }) : void 0,
      ethBalance: childrenArray.find(findComponent(EthBalance)),
      socials: childrenArray.find(findComponent(Socials))
    };
  }, [children, hasCopyAddressOnClick]);
  return /* @__PURE__ */ jsxs("div", { className: cn("ock:bg-ock-background", "ock:flex ock:flex-col ock:px-4 ock:py-1", className), "data-testid": "ockIdentityLayout_container", children: [
    /* @__PURE__ */ jsxs("div", { className: "ock:flex ock:items-center ock:space-x-3", children: [
      /* @__PURE__ */ jsx("div", { className: "ock:flex-shrink-0", children: avatar }),
      /* @__PURE__ */ jsxs("div", { className: "ock:flex ock:flex-col", children: [
        name,
        addressComponent && !ethBalance && addressComponent,
        !addressComponent && ethBalance && ethBalance,
        addressComponent && ethBalance && /* @__PURE__ */ jsxs("div", { className: "ock:flex ock:items-center ock:gap-1", children: [
          addressComponent,
          /* @__PURE__ */ jsx("span", { className: "text-ock-foreground-muted", children: "·" }),
          ethBalance
        ] })
      ] })
    ] }),
    socials
  ] });
}
export {
  IdentityLayout
};
