// @coinbase/onchainkit/dist/identity/components/Socials.js
'use client';
import { jsx } from "react/jsx-runtime";
import { useIdentityContext } from "./IdentityProvider.js";
import { useName } from "../hooks/useName.js";
import { useSocials } from "../hooks/useSocials.js";
import { cn } from "../../styles/theme.js";
import { SocialPlatformDetails } from "./SocialPlatformDetails.js";
function Socials({
  address,
  chain,
  className
}) {
  const {
    address: contextAddress,
    chain: contextChain
  } = useIdentityContext();
  const accountAddress = address ?? contextAddress;
  const accountChain = chain ?? contextChain;
  const {
    data: name,
    isLoading: isLoadingName
  } = useName({
    address: accountAddress,
    chain: accountChain
  });
  const {
    data: socials,
    isLoading: isLoadingSocials
  } = useSocials({
    ensName: name ?? "",
    chain: accountChain
  }, {
    enabled: !!name
  });
  if (!accountAddress) {
    console.error("Socials: an Ethereum address must be provided to the Socials component.");
    return null;
  }
  if (isLoadingName || isLoadingSocials) {
    return /* @__PURE__ */ jsx("span", { className });
  }
  if (!socials || Object.values(socials).every((value) => !value)) {
    return null;
  }
  return /* @__PURE__ */ jsx("div", { className: cn("ock:border-ock-background", "ock:mt-2 ock:w-full ock:pl-1", className), children: /* @__PURE__ */ jsx("div", { className: "left-4 flex space-x-2", children: Object.entries(socials).map(([platform, value]) => value && /* @__PURE__ */ jsx(SocialPlatformDetails, { platform, value }, platform)) }) });
}
export {
  Socials
};
