// @coinbase/onchainkit/dist/identity/components/SocialPlatformDetails.js
import { jsxs, jsx } from "react/jsx-runtime";
import { githubSvg } from "../../internal/svg/githubSvg.js";
import { twitterSvg } from "../../internal/svg/twitterSvg.js";
import { warpcastSvg } from "../../internal/svg/warpcastSvg.js";
import { websiteSvg } from "../../internal/svg/websiteSvg.js";
import { cn, pressable } from "../../styles/theme.js";
const PLATFORM_CONFIG = {
  twitter: {
    href: (value) => `https://x.com/${value}`,
    icon: twitterSvg,
    ariaLabel: (value) => `Visit ${value} on X (formerly Twitter)`
  },
  github: {
    href: (value) => `https://github.com/${value}`,
    icon: githubSvg,
    ariaLabel: (value) => `Visit ${value} on GitHub`
  },
  farcaster: {
    href: (value) => {
      const username = value.split("/").pop();
      return `https://warpcast.com/${username}`;
    },
    icon: warpcastSvg,
    ariaLabel: (value) => {
      const username = value.split("/").pop();
      return `Visit ${username} on Farcaster (Warpcast)`;
    }
  },
  website: {
    href: (value) => value,
    icon: websiteSvg,
    ariaLabel: (value) => `Visit ${value}`
  }
};
function SocialPlatformDetails({
  platform,
  value
}) {
  const config = PLATFORM_CONFIG[platform];
  const ariaLabel = config.ariaLabel(value);
  return /* @__PURE__ */ jsxs("a", { href: config.href(value), target: "_blank", rel: "noopener noreferrer", className: cn(pressable.default, "ock:rounded-ock-default", "ock:border-ock-background", "ock:flex ock:items-center ock:justify-center ock:p-2"), "data-testid": `ockSocials_${platform.charAt(0).toUpperCase() + platform.slice(1)}`, "aria-label": ariaLabel, children: [
    /* @__PURE__ */ jsx("span", { className: "ock:sr-only", children: platform }),
    /* @__PURE__ */ jsx("div", { className: cn("ock:flex ock:h-4 ock:w-4 ock:items-center ock:justify-center"), children: config.icon })
  ] });
}
export {
  PLATFORM_CONFIG,
  SocialPlatformDetails
};
