// sushi/_esm/mvm/currency/token.js
import * as z from 'zod';
import { Token } from '../../generic/currency/token.js';
import { isMvmChainId } from '../chain/chains.js';
export function isMvmAddress(address) {
    // 0x${string}::${string}::${string}
    return !!address.match(/^0x([^:]+)::([^:]+)::([^:]+)$/);
}
export class MvmToken extends Token {
    constructor({ address, ...rest }) {
        super({ address: address.toLowerCase(), ...rest });
    }
    wrap() {
        return this;
    }
    toJSON() {
        return {
            chainId: this.chainId,
            address: this.address,
            symbol: this.symbol,
            name: this.name,
            decimals: this.decimals,
            type: this.type,
            metadata: this.metadata,
        };
    }
    static fromJSON(data) {
        return new this(data);
    }
}
export const serializedMvmTokenSchema = ({ metadata, } = {}) => z.object({
    chainId: z.number().int().refine(isMvmChainId),
    address: z.string().refine(isMvmAddress),
    symbol: z.string(),
    name: z.string(),
    decimals: z.number().int().nonnegative(),
    type: z.literal('token'),
    metadata: (metadata ||
        z.record(z.unknown()).optional().default({})),
});
