// @coinbase/onchainkit/dist/wallet/components/wallet-advanced-send/components/RenderSendButton.js
import { jsx } from "react/jsx-runtime";
import { Spinner } from "../../../../internal/components/Spinner.js";
import { prefixClassName, cn, text, pressable } from "../../../../styles/theme.js";
import "react";
import { useOnchainKit } from "../../../../useOnchainKit.js";
import { useAccount } from "wagmi";
import "../../../../core/network/getChainExplorer.js";
import { useTransactionContext } from "../../../../transaction/components/TransactionProvider.js";
import "@radix-ui/react-toast";
import "wagmi/experimental";
import { defaultSendTxSuccessHandler } from "../utils/defaultSendTxSuccessHandler.js";
import { useWalletContext } from "../../WalletProvider.js";
function RenderSendButton({
  onSubmit,
  isDisabled,
  label
}) {
  const context = useTransactionContext();
  const {
    setActiveFeature
  } = useWalletContext();
  const {
    address
  } = useAccount();
  const {
    chain: senderChain
  } = useOnchainKit();
  const classNames = prefixClassName(cn(pressable.primary, "rounded-ock-default", "w-full rounded-xl", "px-4 py-3 font-medium leading-6", isDisabled && pressable.disabled, text.headline, "text-ock-foreground"));
  if (context.receipt) {
    return /* @__PURE__ */ jsx("button", { className: classNames, onClick: () => defaultSendTxSuccessHandler({
      transactionId: context == null ? void 0 : context.transactionId,
      transactionHash: context == null ? void 0 : context.transactionHash,
      senderChain: senderChain ?? void 0,
      address: address ?? void 0,
      onComplete: () => setActiveFeature(null)
    })(context.receipt), disabled: isDisabled, children: "View transaction" });
  }
  if (context.errorMessage) {
    return /* @__PURE__ */ jsx("button", { className: classNames, onClick: onSubmit, disabled: isDisabled, children: "Try again" });
  }
  if (context.isLoading) {
    return /* @__PURE__ */ jsx("button", { className: classNames, disabled: isDisabled, children: /* @__PURE__ */ jsx(Spinner, {}) });
  }
  return /* @__PURE__ */ jsx("button", { className: classNames, disabled: isDisabled, onClick: onSubmit, children: label });
}
export {
  RenderSendButton
};
