// @coinbase/onchainkit/dist/wallet/components/wallet-advanced-send/components/SendButton.js
'use client';
import { jsxs, jsx } from "react/jsx-runtime";
import { Transaction } from "../../../../transaction/components/Transaction.js";
import { TransactionButton } from "../../../../transaction/components/TransactionButton.js";
import { TransactionStatus } from "../../../../transaction/components/TransactionStatus.js";
import { TransactionStatusAction } from "../../../../transaction/components/TransactionStatusAction.js";
import { TransactionStatusLabel } from "../../../../transaction/components/TransactionStatusLabel.js";
import { useCallback } from "react";
import { parseUnits } from "viem";
import { base } from "viem/chains";
import { useWalletContext } from "../../WalletProvider.js";
import { useOnchainKit } from "../../../../useOnchainKit.js";
import { getSendCalldata } from "../utils/getSendCalldata.js";
import { useSendContext } from "./SendProvider.js";
import { RenderSendButton } from "./RenderSendButton.js";
function SendButton() {
  const {
    chain: senderChain
  } = useOnchainKit();
  const {
    isSponsored
  } = useWalletContext();
  const {
    recipientState,
    cryptoAmount: inputAmount,
    selectedToken,
    updateLifecycleStatus
  } = useSendContext();
  const {
    calldata,
    error
  } = getSendCalldata({
    recipientAddress: recipientState.address,
    token: selectedToken,
    amount: inputAmount
  });
  const disableSendButton = Boolean(error) || !validateAmountInput({
    inputAmount: inputAmount ?? "",
    balance: BigInt((selectedToken == null ? void 0 : selectedToken.cryptoBalance) ?? 0),
    selectedToken: selectedToken ?? void 0
  });
  const buttonLabel = getDefaultSendButtonLabel(inputAmount, selectedToken);
  const handleStatus = useCallback((status) => {
    const validStatuses = ["transactionPending", "transactionLegacyExecuted", "success", "error"];
    if (validStatuses.includes(status.statusName)) {
      updateLifecycleStatus(status);
    }
  }, [updateLifecycleStatus]);
  return /* @__PURE__ */ jsxs(Transaction, { isSponsored, chainId: (senderChain == null ? void 0 : senderChain.id) ?? base.id, calls: calldata ? [calldata] : [], onStatus: handleStatus, children: [
    /* @__PURE__ */ jsx(TransactionButton, { text: buttonLabel, render: (params) => /* @__PURE__ */ jsx(RenderSendButton, { ...params, label: buttonLabel }), disabled: disableSendButton }),
    /* @__PURE__ */ jsxs(TransactionStatus, { children: [
      /* @__PURE__ */ jsx(TransactionStatusLabel, {}),
      /* @__PURE__ */ jsx(TransactionStatusAction, {})
    ] })
  ] });
}
function getDefaultSendButtonLabel(cryptoAmount, selectedToken) {
  if (!cryptoAmount) {
    return "Input amount";
  }
  if (!selectedToken) {
    return "Select token";
  }
  if (parseUnits(cryptoAmount, selectedToken.decimals) > selectedToken.cryptoBalance) {
    return "Insufficient balance";
  }
  return "Continue";
}
function validateAmountInput({
  inputAmount,
  balance,
  selectedToken
}) {
  if (!inputAmount || !selectedToken || !balance) {
    return false;
  }
  const parsedCryptoAmount = parseUnits(inputAmount, selectedToken.decimals);
  return parsedCryptoAmount > 0n && parsedCryptoAmount <= balance;
}
export {
  SendButton
};
