// @vaultsfyi/sdk/dist/client.mjs
// src/client.ts
import { wrapFetchWithPayment } from "x402-fetch";

// src/errors.ts
var HttpResponseError = class extends Error {
};

// src/utils/generateQueryParams.ts
function generateQueryParams(queryParams) {
  if (!queryParams || Object.keys(queryParams).length === 0) return "";
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(queryParams)) {
    if (value == null) continue;
    if (Array.isArray(value)) {
      for (const item of value) {
        params.append(key, item);
      }
    } else {
      params.set(key, value.toString());
    }
  }
  return `?${params}`;
}

// src/client.ts
var SUPPORTED_NETWORKS = [
  "mainnet",
  "optimism",
  "arbitrum",
  "polygon",
  "gnosis",
  "base",
  "unichain",
  "swellchain",
  "celo",
  "worldchain",
  "berachain",
  "ink",
  "bsc",
  "hyperliquid",
  "plasma",
  "avalanche",
  "mega-eth",
  "monad",
  "etherlink"
];
var VaultsSdk = class {
  apiBaseUrl;
  auth;
  authHeaders;
  constructor(options, internalOptions) {
    this.auth = options;
    if ("apiKey" in this.auth) {
      this.authHeaders = { "x-api-key": this.auth.apiKey };
    } else {
      this.authHeaders = { "x-402-auth": "true" };
    }
    this.apiBaseUrl = internalOptions?.apiBaseUrl ?? "https://api.vaults.fyi";
  }
  async fetchData(endpoint, params, method = "GET") {
    const queryParams = generateQueryParams(params?.query);
    const body = params && "body" in params ? params.body : void 0;
    const stringifiedBody = body ? JSON.stringify(body) : void 0;
    const wrappedFetch = "apiKey" in this.auth ? fetch : wrapFetchWithPayment(fetch, this.auth.client);
    const response = await wrappedFetch(`${this.apiBaseUrl}${endpoint}${queryParams}`, {
      method,
      headers: {
        "Content-Type": "application/json",
        ...this.authHeaders
      },
      body: stringifiedBody
    });
    if (!response.ok) {
      const error = await response.json();
      throw new HttpResponseError(error.message);
    }
    return response.json();
  }
  async getAssets(params) {
    return await this.fetchData("/v2/assets", params);
  }
  async getBenchmarks(params) {
    return await this.fetchData(`/v2/benchmarks/${params.path.network}`, params);
  }
  async getHistoricalBenchmarks(params) {
    return await this.fetchData(
      `/v2/historical-benchmarks/${params.path.network}`,
      params
    );
  }
  async getVaultHistoricalData(params) {
    const { network, vaultAddress } = params.path;
    return await this.fetchData(
      `/v2/historical/${network}/${vaultAddress}`,
      params
    );
  }
  async getVaultHistoricalApy(params) {
    const { network, vaultAddress } = params.path;
    return await this.fetchData(
      `/v2/historical/${network}/${vaultAddress}/apy`,
      params
    );
  }
  async getVaultHistoricalTvl(params) {
    const { network, vaultAddress } = params.path;
    return await this.fetchData(
      `/v2/historical/${network}/${vaultAddress}/tvl`,
      params
    );
  }
  async getVaultHistoricalSharePrice(params) {
    const { network, vaultAddress } = params.path;
    return await this.fetchData(
      `/v2/historical/${network}/${vaultAddress}/sharePrice`,
      params
    );
  }
  async getPositions(params) {
    const { userAddress } = params.path;
    return await this.fetchData(
      `/v2/portfolio/positions/${userAddress}`,
      params
    );
  }
  async getPosition(params) {
    const { userAddress, network, vaultAddress } = params.path;
    return await this.fetchData(
      `/v2/portfolio/positions/${userAddress}/${network}/${vaultAddress}`,
      params
    );
  }
  async getDepositOptions(params) {
    const { userAddress } = params.path;
    return await this.fetchData(
      `/v2/portfolio/best-deposit-options/${userAddress}`,
      params
    );
  }
  async getAllVaults(params) {
    return await this.fetchData("/v2/detailed-vaults", params);
  }
  async getVault(params) {
    const { network, vaultAddress } = params.path;
    return await this.fetchData(
      `/v2/detailed-vaults/${network}/${vaultAddress}`,
      params
    );
  }
  async getIdleAssets(params) {
    const { userAddress } = params.path;
    return await this.fetchData(
      `/v2/portfolio/idle-assets/${userAddress}`,
      params
    );
  }
  async getTransactionsContext(params) {
    const { userAddress, network, vaultAddress } = params.path;
    return await this.fetchData(
      `/v2/transactions/context/${userAddress}/${network}/${vaultAddress}`,
      params
    );
  }
  async getActions(params) {
    const { action, userAddress, network, vaultAddress } = params.path;
    return await this.fetchData(
      `/v2/transactions/${action}/${userAddress}/${network}/${vaultAddress}`,
      params
    );
  }
  async getRewardsTransactionsContext(params) {
    const { userAddress } = params.path;
    return await this.fetchData(
      `/v2/transactions/rewards/context/${userAddress}`,
      params
    );
  }
  async getRewardsClaimActions(params) {
    const { userAddress } = params.path;
    return await this.fetchData(
      `/v2/transactions/rewards/claim/${userAddress}`,
      params
    );
  }
  async getUserVaultEvents(params) {
    const { userAddress, network, vaultAddress } = params.path;
    return await this.fetchData(
      `/v2/portfolio/events/${userAddress}/${network}/${vaultAddress}`,
      params
    );
  }
  async getUserVaultTotalReturns(params) {
    const { userAddress, network, vaultAddress } = params.path;
    return await this.fetchData(
      `/v2/portfolio/total-returns/${userAddress}/${network}/${vaultAddress}`,
      params
    );
  }
};
export {
  SUPPORTED_NETWORKS,
  VaultsSdk
};
