// @coinbase/onchainkit/dist/buy/components/BuyButton.js
'use client';
import { jsx } from "react/jsx-runtime";
import { useCallback, useMemo } from "react";
import { Spinner } from "../../internal/components/Spinner.js";
import { checkmarkSvg } from "../../internal/svg/checkmarkSvg.js";
import { CloseSvg } from "../../internal/svg/closeSvg.js";
import { cn, text, pressable } from "../../styles/theme.js";
import { ConnectWallet } from "../../wallet/components/ConnectWallet.js";
import "../../identity/components/IdentityProvider.js";
import "viem";
import "viem/chains";
import "viem/ens";
import "../../identity/constants.js";
import "@tanstack/react-query";
import "../../core/network/attestations.js";
import "../../identity/utils/easSupportedChains.js";
import "../../styles/constants.js";
import "../../useOnchainKit.js";
import "wagmi";
import "@floating-ui/react";
import "@radix-ui/react-dialog";
import "../../token/constants.js";
import "../../internal/components/QrCode/useLogo.js";
import "qrcode";
import "../../wallet/components/WalletProvider.js";
import "@farcaster/miniapp-sdk";
import "../../internal/components/TextInput.js";
import "../../token/components/TokenRow.js";
import "@radix-ui/react-dropdown-menu";
import "../../internal/components/LayerConfigProvider.js";
import "../../token/components/TokenSelectButton.js";
import "../../swap/components/SwapProvider.js";
import "@radix-ui/react-popover";
import "../../core/network/getChainExplorer.js";
import "@radix-ui/react-toast";
import "../../wallet/components/wallet-advanced-send/components/SendAmountInputTypeSwitch.js";
import "../../wallet/components/wallet-advanced-send/components/SendProvider.js";
import "../../transaction/components/TransactionProvider.js";
import "wagmi/experimental";
import "../../fund/components/FundCardProvider.js";
import "../../fund/components/FundCardPaymentMethodSelectRow.js";
import "../../fund/components/FundCardPaymentMethodSelectorToggle.js";
import "../../internal/svg/trustWalletSvg.js";
import "wagmi/connectors";
import "viem/account-abstraction";
import { useBuyContext } from "./BuyProvider.js";
function BuyButton() {
  const {
    address,
    disabled,
    setIsDropdownOpen,
    isDropdownOpen,
    from,
    fromETH,
    fromUSDC,
    to,
    lifecycleStatus: {
      statusName
    },
    updateLifecycleStatus
  } = useBuyContext();
  const isLoading = (to == null ? void 0 : to.loading) || (from == null ? void 0 : from.loading) || (fromETH == null ? void 0 : fromETH.loading) || (fromUSDC == null ? void 0 : fromUSDC.loading) || statusName === "transactionPending" || statusName === "transactionApproved";
  const isMissingRequiredField = !(to == null ? void 0 : to.amount) || !(to == null ? void 0 : to.token);
  const isDisabled = isLoading || disabled;
  const handleSubmit = useCallback(() => {
    if (isMissingRequiredField) {
      updateLifecycleStatus({
        statusName: "error",
        statusData: {
          code: "TmBPc05",
          error: "Missing required fields",
          message: "Complete the field to continue"
        }
      });
      return;
    }
    if (isDropdownOpen) {
      setIsDropdownOpen(false);
      return;
    }
    if (!isDropdownOpen) {
      setIsDropdownOpen(true);
    }
  }, [isMissingRequiredField, setIsDropdownOpen, isDropdownOpen, updateLifecycleStatus]);
  const buttonContent = useMemo(() => {
    if (statusName === "success") {
      return checkmarkSvg;
    }
    if (isDropdownOpen) {
      return /* @__PURE__ */ jsx(CloseSvg, { className: "ock:fill-ock-foreground-inverse" });
    }
    return "Buy";
  }, [statusName, isDropdownOpen]);
  if (!isDisabled && !address) {
    return /* @__PURE__ */ jsx(ConnectWallet, { disconnectedLabel: "Buy", className: "ock:h-12 ock:w-24 ock:min-w-24" });
  }
  return /* @__PURE__ */ jsx("button", { type: "button", className: cn("ock:bg-ock-primary", "ock:rounded-ock-default", "ock:flex ock:rounded-xl", "ock:h-12 ock:w-24 ock:items-center ock:justify-center ock:px-4 ock:py-3", isDisabled && pressable.disabled, text.headline), onClick: handleSubmit, "data-testid": "ockBuyButton_Button", disabled: isDisabled, children: isLoading ? /* @__PURE__ */ jsx(Spinner, {}) : /* @__PURE__ */ jsx("span", { className: cn(text.headline, "ock:text-ock-foreground-inverse"), children: buttonContent }) });
}
export {
  BuyButton
};
