// sushi/_cjs/evm/config/features/fee-collector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isProtocolFeeCollectorChainId = exports.PROTOCOL_FEE_COLLECTOR_ADDRESS = exports.isSurplusFeeCollectorChainId = exports.SURPLUS_FEE_COLLECTOR_ADDRESS = exports.isUIFeeCollectorChainId = exports.UI_FEE_COLLECTOR_ADDRESS = exports.FEE_COLLECTOR_CHAIN_IDS = void 0;
const index_js_1 = require("../../chain/index.js");
exports.FEE_COLLECTOR_CHAIN_IDS = [
    index_js_1.EvmChainId.ETHEREUM,
    index_js_1.EvmChainId.ARBITRUM,
    index_js_1.EvmChainId.OPTIMISM,
    index_js_1.EvmChainId.BASE,
    index_js_1.EvmChainId.POLYGON,
    index_js_1.EvmChainId.AVALANCHE,
    index_js_1.EvmChainId.BSC,
    index_js_1.EvmChainId.LINEA,
    index_js_1.EvmChainId.ARBITRUM_NOVA,
    index_js_1.EvmChainId.GNOSIS,
    index_js_1.EvmChainId.FANTOM,
    index_js_1.EvmChainId.BTTC,
    index_js_1.EvmChainId.CELO,
    index_js_1.EvmChainId.FILECOIN,
    index_js_1.EvmChainId.HAQQ,
    index_js_1.EvmChainId.KAVA,
    index_js_1.EvmChainId.METIS,
    index_js_1.EvmChainId.THUNDERCORE,
    index_js_1.EvmChainId.SCROLL,
    index_js_1.EvmChainId.ZETACHAIN,
    index_js_1.EvmChainId.POLYGON_ZKEVM,
    index_js_1.EvmChainId.HARMONY,
    index_js_1.EvmChainId.BOBA,
    index_js_1.EvmChainId.BOBA_BNB,
    index_js_1.EvmChainId.CORE,
    index_js_1.EvmChainId.CRONOS,
    index_js_1.EvmChainId.BLAST,
    index_js_1.EvmChainId.SKALE_EUROPA,
    index_js_1.EvmChainId.ROOTSTOCK,
    index_js_1.EvmChainId.MANTLE,
    index_js_1.EvmChainId.ZKSYNC_ERA,
    index_js_1.EvmChainId.MANTA,
    index_js_1.EvmChainId.MODE,
    index_js_1.EvmChainId.TAIKO,
    index_js_1.EvmChainId.ZKLINK,
    index_js_1.EvmChainId.APE,
    index_js_1.EvmChainId.SONIC,
    index_js_1.EvmChainId.HEMI,
    index_js_1.EvmChainId.KATANA,
    index_js_1.EvmChainId.HYPEREVM,
    index_js_1.EvmChainId.BERACHAIN,
    index_js_1.EvmChainId.PLASMA,
];
const isFeeCollectorChainId = (chainId) => exports.FEE_COLLECTOR_CHAIN_IDS.includes(chainId);
exports.UI_FEE_COLLECTOR_ADDRESS = {
    [index_js_1.EvmChainId.ETHEREUM]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.ARBITRUM]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.OPTIMISM]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.BASE]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.POLYGON]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.AVALANCHE]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.BSC]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.LINEA]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.GNOSIS]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.FANTOM]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.BTTC]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.CELO]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.FILECOIN]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.HAQQ]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.KAVA]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.METIS]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.THUNDERCORE]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.SCROLL]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.ZETACHAIN]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.HARMONY]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.BOBA]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.BOBA_BNB]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.CORE]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.CRONOS]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.BLAST]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.MANTLE]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.ZKSYNC_ERA]: '0xfbf5a439d64c0cc383a063f9e0cec6b7113931c4',
    [index_js_1.EvmChainId.MANTA]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.MODE]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.TAIKO]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.ZKLINK]: '0xfbf5a439d64c0cc383a063f9e0cec6b7113931c4',
    [index_js_1.EvmChainId.APE]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.SONIC]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.HEMI]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.KATANA]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.HYPEREVM]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.BERACHAIN]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
    [index_js_1.EvmChainId.PLASMA]: '0xde7259893af7cdbc9fd806c6ba61d22d581d5667',
};
exports.isUIFeeCollectorChainId = isFeeCollectorChainId;
exports.SURPLUS_FEE_COLLECTOR_ADDRESS = {
    [index_js_1.EvmChainId.ETHEREUM]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.ARBITRUM]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.OPTIMISM]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.BASE]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.POLYGON]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.AVALANCHE]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.BSC]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.LINEA]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.GNOSIS]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.FANTOM]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.BTTC]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.CELO]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.FILECOIN]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.HAQQ]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.KAVA]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.METIS]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.THUNDERCORE]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.SCROLL]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.ZETACHAIN]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.HARMONY]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.BOBA]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.BOBA_BNB]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.CORE]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.CRONOS]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.BLAST]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.MANTLE]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.ZKSYNC_ERA]: '0x199128c21573469750f3c77f98ab1b9a4010fd6b',
    [index_js_1.EvmChainId.MANTA]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.MODE]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.TAIKO]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.ZKLINK]: '0x199128c21573469750f3c77f98ab1b9a4010fd6b',
    [index_js_1.EvmChainId.APE]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.SONIC]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.HEMI]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.KATANA]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.HYPEREVM]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.BERACHAIN]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
    [index_js_1.EvmChainId.PLASMA]: '0x5c2e112783a6854653b4bc7dc22248d3e592559c',
};
exports.isSurplusFeeCollectorChainId = isFeeCollectorChainId;
exports.PROTOCOL_FEE_COLLECTOR_ADDRESS = {
    [index_js_1.EvmChainId.ETHEREUM]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.ARBITRUM]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.OPTIMISM]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.BASE]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.POLYGON]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.AVALANCHE]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.BSC]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.LINEA]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.GNOSIS]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.FANTOM]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.BTTC]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.CELO]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.FILECOIN]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.HAQQ]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.KAVA]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.METIS]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.THUNDERCORE]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.SCROLL]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.ZETACHAIN]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.HARMONY]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.BOBA]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.BOBA_BNB]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.CORE]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.CRONOS]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.BLAST]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.MANTLE]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.ZKSYNC_ERA]: '0xdfa3d8378280679db91448252b8a53fbb6cf5a9e',
    [index_js_1.EvmChainId.MANTA]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.MODE]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.TAIKO]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.ZKLINK]: '0xdfa3d8378280679db91448252b8a53fbb6cf5a9e',
    [index_js_1.EvmChainId.APE]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.SONIC]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.HEMI]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.KATANA]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.HYPEREVM]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.BERACHAIN]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
    [index_js_1.EvmChainId.PLASMA]: '0xdbeca8fb948c42634256609bce5a3768c9b3e9eb',
};
exports.isProtocolFeeCollectorChainId = isFeeCollectorChainId;
