// sushi/_cjs/evm/config/features/pancakeswap-v3.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PANCAKESWAP_V3_DEPLOYER_ADDRESS = exports.PANCAKESWAP_V3_FACTORY_ADDRESS = exports.PANCAKESWAP_V3_INIT_CODE_HASH = exports.isPancakeSwapV3ChainId = exports.PancakeSwapV3ChainIds = exports.PANCAKESWAP_V3_SUPPORTED_CHAIN_IDS = exports.PANCAKESWAP_V3_FEE_SPACING_MAP = exports.PancakeSwapV3FeeAmount = void 0;
const index_js_1 = require("../../chain/index.js");
const POOL_INIT_CODE_HASH = '0x6ce8eb472fa82df5469c6ab6d485f17c3ad13c8cd7af59b3d4a8026c5ce0f7e2';
var PancakeSwapV3FeeAmount;
(function (PancakeSwapV3FeeAmount) {
    PancakeSwapV3FeeAmount[PancakeSwapV3FeeAmount["LOWEST"] = 100] = "LOWEST";
    PancakeSwapV3FeeAmount[PancakeSwapV3FeeAmount["LOW"] = 500] = "LOW";
    PancakeSwapV3FeeAmount[PancakeSwapV3FeeAmount["MEDIUM"] = 2500] = "MEDIUM";
    PancakeSwapV3FeeAmount[PancakeSwapV3FeeAmount["HIGH"] = 10000] = "HIGH";
})(PancakeSwapV3FeeAmount || (exports.PancakeSwapV3FeeAmount = PancakeSwapV3FeeAmount = {}));
exports.PANCAKESWAP_V3_FEE_SPACING_MAP = {
    100: 1,
    500: 10,
    2500: 50,
    10_000: 200,
};
exports.PANCAKESWAP_V3_SUPPORTED_CHAIN_IDS = [
    index_js_1.EvmChainId.ARBITRUM,
    index_js_1.EvmChainId.BASE,
    index_js_1.EvmChainId.BSC,
    index_js_1.EvmChainId.ETHEREUM,
    index_js_1.EvmChainId.LINEA,
    index_js_1.EvmChainId.POLYGON_ZKEVM,
    index_js_1.EvmChainId.ZKSYNC_ERA,
];
exports.PancakeSwapV3ChainIds = exports.PANCAKESWAP_V3_SUPPORTED_CHAIN_IDS;
const isPancakeSwapV3ChainId = (chainId) => exports.PANCAKESWAP_V3_SUPPORTED_CHAIN_IDS.includes(chainId);
exports.isPancakeSwapV3ChainId = isPancakeSwapV3ChainId;
exports.PANCAKESWAP_V3_INIT_CODE_HASH = {
    [index_js_1.EvmChainId.ARBITRUM]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.BASE]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.BSC]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.ETHEREUM]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.LINEA]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.ZKSYNC_ERA]: '0x01001487a7c45b21c52a0bc0558bf48d897d14792f1d0cc82733c8271d069178',
};
exports.PANCAKESWAP_V3_FACTORY_ADDRESS = {
    [index_js_1.EvmChainId.ARBITRUM]: '0x0bfbcf9fa4f9c56b0f40a671ad40e0805a091865',
    [index_js_1.EvmChainId.BASE]: '0x0bfbcf9fa4f9c56b0f40a671ad40e0805a091865',
    [index_js_1.EvmChainId.BSC]: '0x0bfbcf9fa4f9c56b0f40a671ad40e0805a091865',
    [index_js_1.EvmChainId.ETHEREUM]: '0x0bfbcf9fa4f9c56b0f40a671ad40e0805a091865',
    [index_js_1.EvmChainId.LINEA]: '0x0bfbcf9fa4f9c56b0f40a671ad40e0805a091865',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0x0bfbcf9fa4f9c56b0f40a671ad40e0805a091865',
    [index_js_1.EvmChainId.ZKSYNC_ERA]: '0x1bb72e0cbbea93c08f535fc7856e0338d7f7a8ab',
};
exports.PANCAKESWAP_V3_DEPLOYER_ADDRESS = {
    [index_js_1.EvmChainId.ARBITRUM]: '0x41ff9aa7e16b8b1a8a8dc4f0efacd93d02d071c9',
    [index_js_1.EvmChainId.BASE]: '0x41ff9aa7e16b8b1a8a8dc4f0efacd93d02d071c9',
    [index_js_1.EvmChainId.BSC]: '0x41ff9aa7e16b8b1a8a8dc4f0efacd93d02d071c9',
    [index_js_1.EvmChainId.ETHEREUM]: '0x41ff9aa7e16b8b1a8a8dc4f0efacd93d02d071c9',
    [index_js_1.EvmChainId.LINEA]: '0x41ff9aa7e16b8b1a8a8dc4f0efacd93d02d071c9',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0x41ff9aa7e16b8b1a8a8dc4f0efacd93d02d071c9',
    [index_js_1.EvmChainId.ZKSYNC_ERA]: '0x7f71382044a6a62595d5d357fe75ca8199123ad6',
};
