// sushi/_cjs/evm/config/features/route-processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRouteProcessor9ChainId = exports.ROUTE_PROCESSOR_9_ADDRESS = exports.ROUTE_PROCESSOR_9_SUPPORTED_CHAIN_IDS = exports.isRouteProcessor8ChainId = exports.ROUTE_PROCESSOR_8_ADDRESS = exports.ROUTE_PROCESSOR_8_SUPPORTED_CHAIN_IDS = exports.isRouteProcessor7ChainId = exports.ROUTE_PROCESSOR_7_ADDRESS = exports.ROUTE_PROCESSOR_7_SUPPORTED_CHAIN_IDS = exports.isRouteProcessor6ChainId = exports.ROUTE_PROCESSOR_6_ADDRESS = exports.ROUTE_PROCESSOR_6_SUPPORTED_CHAIN_IDS = void 0;
const index_js_1 = require("../../chain/index.js");
exports.ROUTE_PROCESSOR_6_SUPPORTED_CHAIN_IDS = [
    index_js_1.EvmChainId.ETHEREUM,
    index_js_1.EvmChainId.ARBITRUM,
    index_js_1.EvmChainId.OPTIMISM,
    index_js_1.EvmChainId.BASE,
    index_js_1.EvmChainId.POLYGON,
    index_js_1.EvmChainId.AVALANCHE,
    index_js_1.EvmChainId.BSC,
    index_js_1.EvmChainId.LINEA,
    index_js_1.EvmChainId.ARBITRUM_NOVA,
    index_js_1.EvmChainId.GNOSIS,
    index_js_1.EvmChainId.FANTOM,
    index_js_1.EvmChainId.BTTC,
    index_js_1.EvmChainId.CELO,
    index_js_1.EvmChainId.FILECOIN,
    index_js_1.EvmChainId.HAQQ,
    index_js_1.EvmChainId.KAVA,
    index_js_1.EvmChainId.METIS,
    index_js_1.EvmChainId.THUNDERCORE,
    index_js_1.EvmChainId.SCROLL,
    index_js_1.EvmChainId.ZETACHAIN,
    index_js_1.EvmChainId.POLYGON_ZKEVM,
    index_js_1.EvmChainId.HARMONY,
    index_js_1.EvmChainId.BOBA,
    index_js_1.EvmChainId.BOBA_BNB,
    index_js_1.EvmChainId.CORE,
    index_js_1.EvmChainId.CRONOS,
    index_js_1.EvmChainId.BLAST,
    index_js_1.EvmChainId.SKALE_EUROPA,
    index_js_1.EvmChainId.ROOTSTOCK,
    index_js_1.EvmChainId.ZKSYNC_ERA,
    index_js_1.EvmChainId.MANTLE,
    index_js_1.EvmChainId.MANTA,
    index_js_1.EvmChainId.MODE,
    index_js_1.EvmChainId.TAIKO,
    index_js_1.EvmChainId.ZKLINK,
    index_js_1.EvmChainId.APE,
    index_js_1.EvmChainId.SONIC,
    index_js_1.EvmChainId.HEMI,
    index_js_1.EvmChainId.SEPOLIA,
];
exports.ROUTE_PROCESSOR_6_ADDRESS = {
    [index_js_1.EvmChainId.ETHEREUM]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.ARBITRUM]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.OPTIMISM]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.BASE]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.POLYGON]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.AVALANCHE]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.BSC]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.LINEA]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.GNOSIS]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.FANTOM]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.BTTC]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.CELO]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.FILECOIN]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.HAQQ]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.KAVA]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.METIS]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.THUNDERCORE]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.SCROLL]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.ZETACHAIN]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.HARMONY]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.BOBA]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.BOBA_BNB]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.CORE]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.CRONOS]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.BLAST]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.ZKSYNC_ERA]: '0xe6fd46600a97ce06703b58333b9c2399f4bf6fec',
    [index_js_1.EvmChainId.MANTLE]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.MANTA]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.MODE]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.TAIKO]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.ZKLINK]: '0xe6fd46600a97ce06703b58333b9c2399f4bf6fec',
    [index_js_1.EvmChainId.APE]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.SONIC]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.HEMI]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
    [index_js_1.EvmChainId.SEPOLIA]: '0x85cd07ea01423b1e937929b44e4ad8c40bbb5e71',
};
const isRouteProcessor6ChainId = (chainId) => exports.ROUTE_PROCESSOR_6_SUPPORTED_CHAIN_IDS.includes(chainId);
exports.isRouteProcessor6ChainId = isRouteProcessor6ChainId;
exports.ROUTE_PROCESSOR_7_SUPPORTED_CHAIN_IDS = [
    index_js_1.EvmChainId.ETHEREUM,
    index_js_1.EvmChainId.ARBITRUM,
    index_js_1.EvmChainId.OPTIMISM,
    index_js_1.EvmChainId.BASE,
    index_js_1.EvmChainId.POLYGON,
    index_js_1.EvmChainId.AVALANCHE,
    index_js_1.EvmChainId.BSC,
    index_js_1.EvmChainId.LINEA,
    index_js_1.EvmChainId.ARBITRUM_NOVA,
    index_js_1.EvmChainId.GNOSIS,
    index_js_1.EvmChainId.FANTOM,
    index_js_1.EvmChainId.BTTC,
    index_js_1.EvmChainId.CELO,
    index_js_1.EvmChainId.FILECOIN,
    index_js_1.EvmChainId.HAQQ,
    index_js_1.EvmChainId.KAVA,
    index_js_1.EvmChainId.METIS,
    index_js_1.EvmChainId.THUNDERCORE,
    index_js_1.EvmChainId.SCROLL,
    index_js_1.EvmChainId.ZETACHAIN,
    index_js_1.EvmChainId.POLYGON_ZKEVM,
    index_js_1.EvmChainId.HARMONY,
    index_js_1.EvmChainId.BOBA,
    index_js_1.EvmChainId.BOBA_BNB,
    index_js_1.EvmChainId.CORE,
    index_js_1.EvmChainId.CRONOS,
    index_js_1.EvmChainId.BLAST,
    index_js_1.EvmChainId.SKALE_EUROPA,
    index_js_1.EvmChainId.ROOTSTOCK,
    index_js_1.EvmChainId.ZKSYNC_ERA,
    index_js_1.EvmChainId.MANTLE,
    index_js_1.EvmChainId.MANTA,
    index_js_1.EvmChainId.MODE,
    index_js_1.EvmChainId.TAIKO,
    index_js_1.EvmChainId.ZKLINK,
    index_js_1.EvmChainId.APE,
    index_js_1.EvmChainId.SONIC,
    index_js_1.EvmChainId.HEMI,
    index_js_1.EvmChainId.SEPOLIA,
    index_js_1.EvmChainId.KATANA,
    index_js_1.EvmChainId.HYPEREVM,
    index_js_1.EvmChainId.BERACHAIN,
    index_js_1.EvmChainId.PLASMA,
];
exports.ROUTE_PROCESSOR_7_ADDRESS = {
    [index_js_1.EvmChainId.ETHEREUM]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.ARBITRUM]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.OPTIMISM]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.BASE]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.POLYGON]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.AVALANCHE]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.BSC]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.LINEA]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.GNOSIS]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.FANTOM]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.BTTC]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.CELO]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.FILECOIN]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.HAQQ]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.KAVA]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.METIS]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.THUNDERCORE]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.SCROLL]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.ZETACHAIN]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.HARMONY]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.BOBA]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.BOBA_BNB]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.CORE]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.CRONOS]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.BLAST]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.ZKSYNC_ERA]: '0xd7c94c8c61628911c41141289af1d0de17d7ac8d',
    [index_js_1.EvmChainId.MANTLE]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.MANTA]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.MODE]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.TAIKO]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.ZKLINK]: '0xd7c94c8c61628911c41141289af1d0de17d7ac8d',
    [index_js_1.EvmChainId.APE]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.SONIC]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.HEMI]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.SEPOLIA]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.KATANA]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.HYPEREVM]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.BERACHAIN]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
    [index_js_1.EvmChainId.PLASMA]: '0x3ced11c610556e5292fbc2e75d68c3899098c14c',
};
const isRouteProcessor7ChainId = (chainId) => exports.ROUTE_PROCESSOR_7_SUPPORTED_CHAIN_IDS.includes(chainId);
exports.isRouteProcessor7ChainId = isRouteProcessor7ChainId;
exports.ROUTE_PROCESSOR_8_SUPPORTED_CHAIN_IDS = [
    index_js_1.EvmChainId.ETHEREUM,
    index_js_1.EvmChainId.ARBITRUM,
    index_js_1.EvmChainId.OPTIMISM,
    index_js_1.EvmChainId.BASE,
    index_js_1.EvmChainId.POLYGON,
    index_js_1.EvmChainId.POLYGON_ZKEVM,
    index_js_1.EvmChainId.MANTLE,
    index_js_1.EvmChainId.KATANA,
    index_js_1.EvmChainId.HYPEREVM,
];
exports.ROUTE_PROCESSOR_8_ADDRESS = {
    [index_js_1.EvmChainId.ETHEREUM]: '0x2905d7e4d048d29954f81b02171dd313f457a4a4',
    [index_js_1.EvmChainId.ARBITRUM]: '0x2905d7e4d048d29954f81b02171dd313f457a4a4',
    [index_js_1.EvmChainId.OPTIMISM]: '0x2905d7e4d048d29954f81b02171dd313f457a4a4',
    [index_js_1.EvmChainId.BASE]: '0x2905d7e4d048d29954f81b02171dd313f457a4a4',
    [index_js_1.EvmChainId.POLYGON]: '0x2905d7e4d048d29954f81b02171dd313f457a4a4',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0x2905d7e4d048d29954f81b02171dd313f457a4a4',
    [index_js_1.EvmChainId.MANTLE]: '0x2905d7e4d048d29954f81b02171dd313f457a4a4',
    [index_js_1.EvmChainId.KATANA]: '0x2905d7e4d048d29954f81b02171dd313f457a4a4',
    [index_js_1.EvmChainId.HYPEREVM]: '0x2905d7e4d048d29954f81b02171dd313f457a4a4',
};
const isRouteProcessor8ChainId = (chainId) => exports.ROUTE_PROCESSOR_8_SUPPORTED_CHAIN_IDS.includes(chainId);
exports.isRouteProcessor8ChainId = isRouteProcessor8ChainId;
exports.ROUTE_PROCESSOR_9_SUPPORTED_CHAIN_IDS = [
    index_js_1.EvmChainId.ETHEREUM,
    index_js_1.EvmChainId.ARBITRUM,
    index_js_1.EvmChainId.OPTIMISM,
    index_js_1.EvmChainId.BASE,
    index_js_1.EvmChainId.POLYGON,
    index_js_1.EvmChainId.AVALANCHE,
    index_js_1.EvmChainId.BSC,
    index_js_1.EvmChainId.LINEA,
    index_js_1.EvmChainId.ARBITRUM_NOVA,
    index_js_1.EvmChainId.GNOSIS,
    index_js_1.EvmChainId.FANTOM,
    index_js_1.EvmChainId.BTTC,
    index_js_1.EvmChainId.CELO,
    index_js_1.EvmChainId.FILECOIN,
    index_js_1.EvmChainId.HAQQ,
    index_js_1.EvmChainId.KAVA,
    index_js_1.EvmChainId.METIS,
    index_js_1.EvmChainId.THUNDERCORE,
    index_js_1.EvmChainId.SCROLL,
    index_js_1.EvmChainId.ZETACHAIN,
    index_js_1.EvmChainId.POLYGON_ZKEVM,
    index_js_1.EvmChainId.HARMONY,
    index_js_1.EvmChainId.BOBA,
    index_js_1.EvmChainId.BOBA_BNB,
    index_js_1.EvmChainId.CORE,
    index_js_1.EvmChainId.CRONOS,
    index_js_1.EvmChainId.BLAST,
    index_js_1.EvmChainId.SKALE_EUROPA,
    index_js_1.EvmChainId.ROOTSTOCK,
    index_js_1.EvmChainId.ZKSYNC_ERA,
    index_js_1.EvmChainId.MANTLE,
    index_js_1.EvmChainId.MANTA,
    index_js_1.EvmChainId.MODE,
    index_js_1.EvmChainId.TAIKO,
    index_js_1.EvmChainId.ZKLINK,
    index_js_1.EvmChainId.APE,
    index_js_1.EvmChainId.SONIC,
    index_js_1.EvmChainId.HEMI,
    index_js_1.EvmChainId.KATANA,
    index_js_1.EvmChainId.HYPEREVM,
    index_js_1.EvmChainId.BERACHAIN,
    index_js_1.EvmChainId.SEPOLIA,
    index_js_1.EvmChainId.PLASMA,
];
exports.ROUTE_PROCESSOR_9_ADDRESS = {
    [index_js_1.EvmChainId.ETHEREUM]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.ARBITRUM]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.OPTIMISM]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.BASE]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.POLYGON]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.AVALANCHE]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.BSC]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.LINEA]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.GNOSIS]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.FANTOM]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.BTTC]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.CELO]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.FILECOIN]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.HAQQ]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.KAVA]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.METIS]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.THUNDERCORE]: '0x861255aef9aad3e268fb67a3c97afd490bff3d6b',
    [index_js_1.EvmChainId.SCROLL]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.ZETACHAIN]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.HARMONY]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.BOBA]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.BOBA_BNB]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.CORE]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.CRONOS]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.BLAST]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.ZKSYNC_ERA]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.MANTLE]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.MANTA]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.MODE]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.TAIKO]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.ZKLINK]: '0xfb7eedd827c4bed92b9b6c5159bee11301f3da8c',
    [index_js_1.EvmChainId.APE]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.SONIC]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.HEMI]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.KATANA]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.HYPEREVM]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.BERACHAIN]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.SEPOLIA]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
    [index_js_1.EvmChainId.PLASMA]: '0x3b0aa7d38bf3c103bf02d1de2e37568cbed3d6e8',
};
const isRouteProcessor9ChainId = (chainId) => exports.ROUTE_PROCESSOR_9_SUPPORTED_CHAIN_IDS.includes(chainId);
exports.isRouteProcessor9ChainId = isRouteProcessor9ChainId;
