// sushi/_cjs/evm/config/features/sushiswap-v3.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDeprecatedSushiSwapV3ChainId = exports.DeprecatedSushiSwapV3ChainIds = exports.DEPRECATED_SUSHISWAP_V3_CHAIN_IDS = exports.isSushiSwapV3ChainId = exports.SushiSwapV3ChainIds = exports.SUSHISWAP_V3_SUPPORTED_CHAIN_IDS = exports.SUSHISWAP_V3_TICK_LENS = exports.SUSHISWAP_V3_POSITION_MANAGER = exports.SUSHISWAP_V3_INIT_CODE_HASH = exports.SUSHISWAP_V3_FACTORY_ADDRESS = exports.TICK_SPACINGS = exports.SushiSwapV3FeeAmount = void 0;
const index_js_1 = require("../../chain/index.js");
const POOL_INIT_CODE_HASH = '0xe34f199b19b2b4f47f68442619d555527d244f78a3297ea89325f843f87b8b54';
var SushiSwapV3FeeAmount;
(function (SushiSwapV3FeeAmount) {
    SushiSwapV3FeeAmount[SushiSwapV3FeeAmount["LOWEST"] = 100] = "LOWEST";
    SushiSwapV3FeeAmount[SushiSwapV3FeeAmount["LOW"] = 500] = "LOW";
    SushiSwapV3FeeAmount[SushiSwapV3FeeAmount["MEDIUM"] = 3000] = "MEDIUM";
    SushiSwapV3FeeAmount[SushiSwapV3FeeAmount["HIGH"] = 10000] = "HIGH";
})(SushiSwapV3FeeAmount || (exports.SushiSwapV3FeeAmount = SushiSwapV3FeeAmount = {}));
exports.TICK_SPACINGS = {
    [SushiSwapV3FeeAmount.LOWEST]: 1,
    [SushiSwapV3FeeAmount.LOW]: 10,
    [SushiSwapV3FeeAmount.MEDIUM]: 60,
    [SushiSwapV3FeeAmount.HIGH]: 200,
};
exports.SUSHISWAP_V3_FACTORY_ADDRESS = {
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0xaa26771d497814e81d305c511efbb3ced90bf5bd',
    [index_js_1.EvmChainId.ARBITRUM]: '0x1af415a1eba07a4986a52b6f2e7de7003d82231e',
    [index_js_1.EvmChainId.AVALANCHE]: '0x3e603c14af37ebdad31709c4f848fc6ad5bec715',
    [index_js_1.EvmChainId.BSC]: '0x126555dd55a39328f69400d6ae4f782bd4c34abb',
    [index_js_1.EvmChainId.BTTC]: '0xbbde1d67297329148fe1ed5e6b00114842728e65',
    [index_js_1.EvmChainId.ETHEREUM]: '0xbaceb8ec6b9355dfc0269c18bac9d6e2bdc29c4f',
    [index_js_1.EvmChainId.FANTOM]: '0x7770978eed668a3ba661d51a773d3a992fc9ddcb',
    [index_js_1.EvmChainId.GNOSIS]: '0xf78031cbca409f2fb6876bdfdbc1b2df24cf9bef',
    [index_js_1.EvmChainId.KAVA]: '0x1e9b24073183d5c6b7ae5fb4b8f0b1dd83fdc77a',
    [index_js_1.EvmChainId.METIS]: '0x145d82bca93cca2ae057d1c6f26245d1b9522e6f',
    [index_js_1.EvmChainId.OPTIMISM]: '0x9c6522117e2ed1fe5bdb72bb0ed5e3f2bde7dbe0',
    [index_js_1.EvmChainId.POLYGON]: '0x917933899c6a5f8e37f31e19f92cdbff7e8ff0e2',
    [index_js_1.EvmChainId.BOBA]: '0x0be808376ecb75a5cf9bb6d237d16cd37893d904',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0x1b02da8cb0d097eb8d57a175b88c7d8b47997506',
    [index_js_1.EvmChainId.THUNDERCORE]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.HAQQ]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.CORE]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.LINEA]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.BASE]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.SCROLL]: '0x46b3fdf7b5cde91ac049936bf0bdb12c5d22202e',
    [index_js_1.EvmChainId.FILECOIN]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.ZETACHAIN]: '0xb45e53277a7e0f1d35f2a77160e91e25507f1763',
    [index_js_1.EvmChainId.BLAST]: '0x7680d4b43f3d1d54d6cfeeb2169463bfa7a6cf0d',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0x51d15889b66a2c919dbbd624d53b47a9e8fec4bb',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0x46b3fdf7b5cde91ac049936bf0bdb12c5d22202e',
    [index_js_1.EvmChainId.SONIC]: '0x46b3fdf7b5cde91ac049936bf0bdb12c5d22202e',
    [index_js_1.EvmChainId.HEMI]: '0xcdbcd51a5e8728e0af4895ce5771b7d17ff71959',
    [index_js_1.EvmChainId.TATARA]: '0x9b3336186a38e1b6c21955d112dbb0343ee061ee',
    [index_js_1.EvmChainId.KATANA]: '0x203e8740894c8955cb8950759876d7e7e45e04c1',
    [index_js_1.EvmChainId.SEPOLIA]: '0x1f2fcf1d036b375b384012e61d3aa33f8c256bbe',
};
exports.SUSHISWAP_V3_INIT_CODE_HASH = {
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.ARBITRUM]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.AVALANCHE]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.BSC]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.BTTC]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.ETHEREUM]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.FANTOM]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.GNOSIS]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.KAVA]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.METIS]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.OPTIMISM]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.POLYGON]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.BOBA]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.THUNDERCORE]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.HAQQ]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.CORE]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.LINEA]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.BASE]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.SCROLL]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.FILECOIN]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.ZETACHAIN]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.BLAST]: '0x8e13daee7f5a62e37e71bf852bcd44e7d16b90617ed2b17c24c2ee62411c5bae',
    [index_js_1.EvmChainId.SKALE_EUROPA]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.ROOTSTOCK]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.SONIC]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.HEMI]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.TATARA]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.KATANA]: '0xe040f12c7cee3904b78f24f8fc395629c2e69525c2815da7a659f7483e378ecb',
    [index_js_1.EvmChainId.SEPOLIA]: POOL_INIT_CODE_HASH,
};
exports.SUSHISWAP_V3_POSITION_MANAGER = {
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0x258f7e97149afd7d7f84fa63b10e4a3f0c38b788',
    [index_js_1.EvmChainId.ARBITRUM]: '0xf0cbce1942a68beb3d1b73f0dd86c8dcc363ef49',
    [index_js_1.EvmChainId.AVALANCHE]: '0x18350b048ab366ed601ffdbc669110ecb36016f3',
    [index_js_1.EvmChainId.BSC]: '0xf70c086618dcf2b1a461311275e00d6b722ef914',
    [index_js_1.EvmChainId.BTTC]: '0x57bffa72db682f7eb6c132dae03ff36bbeb0c459',
    [index_js_1.EvmChainId.ETHEREUM]: '0x2214a42d8e2a1d20635c2cb0664422c528b6a432',
    [index_js_1.EvmChainId.FANTOM]: '0x10c19390e1ac2fd6d0c3643a2320b0aba38e5baa',
    [index_js_1.EvmChainId.GNOSIS]: '0xab235da7f52d35fb4551afba11bfb56e18774a65',
    [index_js_1.EvmChainId.KAVA]: '0xbf3b71decbcefabb3210b9d8f18ec22e0556f5f0',
    [index_js_1.EvmChainId.METIS]: '0x630be2985674d31920babb4f96657960f131e7b1',
    [index_js_1.EvmChainId.OPTIMISM]: '0x1af415a1eba07a4986a52b6f2e7de7003d82231e',
    [index_js_1.EvmChainId.POLYGON]: '0xb7402ee99f0a008e461098ac3a27f4957df89a40',
    [index_js_1.EvmChainId.BOBA]: '0x1b9d177ccdea3c79b6c8f40761fc8dc9d0500eaa',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0xf4d73326c13a4fc5fd7a064217e12780e9bd62c3',
    [index_js_1.EvmChainId.THUNDERCORE]: '0xf4d73326c13a4fc5fd7a064217e12780e9bd62c3',
    [index_js_1.EvmChainId.HAQQ]: '0x80c7dd17b01855a6d2347444a0fcc36136a314de',
    [index_js_1.EvmChainId.CORE]: '0xf4d73326c13a4fc5fd7a064217e12780e9bd62c3',
    [index_js_1.EvmChainId.LINEA]: '0x80c7dd17b01855a6d2347444a0fcc36136a314de',
    [index_js_1.EvmChainId.BASE]: '0x80c7dd17b01855a6d2347444a0fcc36136a314de',
    [index_js_1.EvmChainId.SCROLL]: '0x0389879e0156033202c44bf784ac18fc02edee4f',
    [index_js_1.EvmChainId.FILECOIN]: '0xf4d73326c13a4fc5fd7a064217e12780e9bd62c3',
    [index_js_1.EvmChainId.ZETACHAIN]: '0xcdbcd51a5e8728e0af4895ce5771b7d17ff71959',
    [index_js_1.EvmChainId.BLAST]: '0x51edb3e5bce8618b77b60215f84ad3db14709051',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0xa4d2268f145cd5c835115441b877295c46f25b9b',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0x0389879e0156033202c44bf784ac18fc02edee4f',
    [index_js_1.EvmChainId.SONIC]: '0x0389879e0156033202c44bf784ac18fc02edee4f',
    [index_js_1.EvmChainId.HEMI]: '0xe43ca1dee3f0fc1e2df73a0745674545f11a59f5',
    [index_js_1.EvmChainId.TATARA]: '0x1400fefd6f9b897970f00df6237ff2b8b27dc82c',
    [index_js_1.EvmChainId.KATANA]: '0x2659c6085d26144117d904c46b48b6d180393d27',
    [index_js_1.EvmChainId.SEPOLIA]: '0x544ba588efd839d2692fc31ea991cd39993c135f',
};
exports.SUSHISWAP_V3_TICK_LENS = {
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0xf60e5f4a44a510742457d8064ffd360b12d8d9af',
    [index_js_1.EvmChainId.ARBITRUM]: '0x8516944e89f296eb6473d79aed1ba12088016c9e',
    [index_js_1.EvmChainId.AVALANCHE]: '0xddc1b5920723f774d2ec2c3c9355251a20819776',
    [index_js_1.EvmChainId.BSC]: '0x10c19390e1ac2fd6d0c3643a2320b0aba38e5baa',
    [index_js_1.EvmChainId.BTTC]: '0x1400fefd6f9b897970f00df6237ff2b8b27dc82c',
    [index_js_1.EvmChainId.ETHEREUM]: '0xfb70ad5a200d784e7901230e6875d91d5fa6b68c',
    [index_js_1.EvmChainId.FANTOM]: '0xd75f5369724b513b497101fb15211160c1d96550',
    [index_js_1.EvmChainId.GNOSIS]: '0xaa26771d497814e81d305c511efbb3ced90bf5bd',
    [index_js_1.EvmChainId.KAVA]: '0xa62ec622dba415aa94110739b1f951b1202cf322',
    [index_js_1.EvmChainId.METIS]: '0x078047150f8efa223b3d407f00e462e38f4b1b9c',
    [index_js_1.EvmChainId.OPTIMISM]: '0x0367a647a68f304f2a6e453c25033a4249d7f2c6',
    [index_js_1.EvmChainId.POLYGON]: '0x9fdea1412e50d78b25ace4f96d35801647fdf7da',
    [index_js_1.EvmChainId.BOBA]: '0x9b3ff703fa9c8b467f5886d7b61e61ba07a9b51c',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0x0be808376ecb75a5cf9bb6d237d16cd37893d904',
    [index_js_1.EvmChainId.THUNDERCORE]: '0x0be808376ecb75a5cf9bb6d237d16cd37893d904',
    [index_js_1.EvmChainId.HAQQ]: '0xf4d73326c13a4fc5fd7a064217e12780e9bd62c3',
    [index_js_1.EvmChainId.CORE]: '0x1b02da8cb0d097eb8d57a175b88c7d8b47997506',
    [index_js_1.EvmChainId.LINEA]: '0xf4d73326c13a4fc5fd7a064217e12780e9bd62c3',
    [index_js_1.EvmChainId.BASE]: '0xf4d73326c13a4fc5fd7a064217e12780e9bd62c3',
    [index_js_1.EvmChainId.SCROLL]: '0x1f2fcf1d036b375b384012e61d3aa33f8c256bbe',
    [index_js_1.EvmChainId.FILECOIN]: '0x1be211d8da40bc0ae8719c6663307bfc987b1d6c',
    [index_js_1.EvmChainId.ZETACHAIN]: '0xe43ca1dee3f0fc1e2df73a0745674545f11a59f5',
    [index_js_1.EvmChainId.BLAST]: '0x039e87ab90205f9d87c5b40d4b28e2be45da4a20',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0x65a7158be70a1270b01713910db7cbfe8d60ff4d',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0x33d91116e0370970444b0281ab117e161febfcdd',
    [index_js_1.EvmChainId.SONIC]: '0x33d91116e0370970444b0281ab117e161febfcdd',
    [index_js_1.EvmChainId.HEMI]: '0x1f2fcf1d036b375b384012e61d3aa33f8c256bbe',
    [index_js_1.EvmChainId.TATARA]: '0xb46e319390de313b8cc95ea5aa30c7bbfd79da94',
    [index_js_1.EvmChainId.KATANA]: '0x35dc3e13469e980c37b6f288bbb9822b1f9bd435',
    [index_js_1.EvmChainId.SEPOLIA]: '0xc3ec4e1511c6935ed2f92b9a61881a1b95bb1566',
};
exports.SUSHISWAP_V3_SUPPORTED_CHAIN_IDS = [
    index_js_1.EvmChainId.ARBITRUM_NOVA,
    index_js_1.EvmChainId.ARBITRUM,
    index_js_1.EvmChainId.AVALANCHE,
    index_js_1.EvmChainId.BSC,
    index_js_1.EvmChainId.BTTC,
    index_js_1.EvmChainId.ETHEREUM,
    index_js_1.EvmChainId.FANTOM,
    index_js_1.EvmChainId.GNOSIS,
    index_js_1.EvmChainId.OPTIMISM,
    index_js_1.EvmChainId.POLYGON,
    index_js_1.EvmChainId.BOBA,
    index_js_1.EvmChainId.THUNDERCORE,
    index_js_1.EvmChainId.HAQQ,
    index_js_1.EvmChainId.CORE,
    index_js_1.EvmChainId.LINEA,
    index_js_1.EvmChainId.BASE,
    index_js_1.EvmChainId.SCROLL,
    index_js_1.EvmChainId.KAVA,
    index_js_1.EvmChainId.METIS,
    index_js_1.EvmChainId.FILECOIN,
    index_js_1.EvmChainId.ZETACHAIN,
    index_js_1.EvmChainId.BLAST,
    index_js_1.EvmChainId.SKALE_EUROPA,
    index_js_1.EvmChainId.ROOTSTOCK,
    index_js_1.EvmChainId.SONIC,
    index_js_1.EvmChainId.HEMI,
    index_js_1.EvmChainId.TATARA,
    index_js_1.EvmChainId.KATANA,
    index_js_1.EvmChainId.SEPOLIA,
];
exports.SushiSwapV3ChainIds = exports.SUSHISWAP_V3_SUPPORTED_CHAIN_IDS;
const isSushiSwapV3ChainId = (chainId) => exports.SUSHISWAP_V3_SUPPORTED_CHAIN_IDS.includes(chainId);
exports.isSushiSwapV3ChainId = isSushiSwapV3ChainId;
exports.DEPRECATED_SUSHISWAP_V3_CHAIN_IDS = [
    index_js_1.EvmChainId.POLYGON_ZKEVM,
];
exports.DeprecatedSushiSwapV3ChainIds = exports.DEPRECATED_SUSHISWAP_V3_CHAIN_IDS;
const isDeprecatedSushiSwapV3ChainId = (chainId) => exports.DEPRECATED_SUSHISWAP_V3_CHAIN_IDS.includes(chainId);
exports.isDeprecatedSushiSwapV3ChainId = isDeprecatedSushiSwapV3ChainId;
