// sushi/_cjs/evm/config/features/sushiswap-v2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUSHISWAP_V2_ROUTER_ADDRESS = exports.SUSHISWAP_V2_FACTORY_ADDRESS = exports.SUSHISWAP_V2_INIT_CODE_HASH = exports.isDeprecatedSushiSwapV2ChainId = exports.DeprecatedSushiSwapV2ChainIds = exports.DEPRECATED_SUSHISWAP_V2_SUPPORTED_CHAIN_IDS = exports.isSushiSwapV2ChainId = exports.SushiSwapV2ChainIds = exports.SUSHISWAP_V2_SUPPORTED_CHAIN_IDS = void 0;
const index_js_1 = require("../../chain/index.js");
exports.SUSHISWAP_V2_SUPPORTED_CHAIN_IDS = [
    index_js_1.EvmChainId.ARBITRUM,
    index_js_1.EvmChainId.ARBITRUM_NOVA,
    index_js_1.EvmChainId.AVALANCHE,
    index_js_1.EvmChainId.BASE,
    index_js_1.EvmChainId.BOBA,
    index_js_1.EvmChainId.BOBA_BNB,
    index_js_1.EvmChainId.BSC,
    index_js_1.EvmChainId.CELO,
    index_js_1.EvmChainId.ETHEREUM,
    index_js_1.EvmChainId.FANTOM,
    index_js_1.EvmChainId.GNOSIS,
    index_js_1.EvmChainId.HAQQ,
    index_js_1.EvmChainId.POLYGON,
    index_js_1.EvmChainId.SCROLL,
    index_js_1.EvmChainId.KAVA,
    index_js_1.EvmChainId.METIS,
    index_js_1.EvmChainId.BTTC,
    index_js_1.EvmChainId.FILECOIN,
    index_js_1.EvmChainId.ZETACHAIN,
    index_js_1.EvmChainId.CORE,
    index_js_1.EvmChainId.THUNDERCORE,
    index_js_1.EvmChainId.HAQQ,
    index_js_1.EvmChainId.OPTIMISM,
    index_js_1.EvmChainId.LINEA,
    index_js_1.EvmChainId.BLAST,
    index_js_1.EvmChainId.SKALE_EUROPA,
    index_js_1.EvmChainId.ROOTSTOCK,
    index_js_1.EvmChainId.SONIC,
    index_js_1.EvmChainId.HEMI,
    index_js_1.EvmChainId.KATANA,
    index_js_1.EvmChainId.SEPOLIA,
    index_js_1.EvmChainId.TATARA,
];
exports.SushiSwapV2ChainIds = exports.SUSHISWAP_V2_SUPPORTED_CHAIN_IDS;
const isSushiSwapV2ChainId = (chainId) => exports.SUSHISWAP_V2_SUPPORTED_CHAIN_IDS.includes(chainId);
exports.isSushiSwapV2ChainId = isSushiSwapV2ChainId;
exports.DEPRECATED_SUSHISWAP_V2_SUPPORTED_CHAIN_IDS = [
    index_js_1.EvmChainId.POLYGON_ZKEVM,
    index_js_1.EvmChainId.HARMONY,
];
exports.DeprecatedSushiSwapV2ChainIds = exports.DEPRECATED_SUSHISWAP_V2_SUPPORTED_CHAIN_IDS;
const isDeprecatedSushiSwapV2ChainId = (chainId) => exports.DEPRECATED_SUSHISWAP_V2_SUPPORTED_CHAIN_IDS.includes(chainId);
exports.isDeprecatedSushiSwapV2ChainId = isDeprecatedSushiSwapV2ChainId;
exports.SUSHISWAP_V2_INIT_CODE_HASH = {
    [index_js_1.EvmChainId.ETHEREUM]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.FANTOM]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.POLYGON]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.GNOSIS]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.BSC]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.ARBITRUM]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.AVALANCHE]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.HARMONY]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.CELO]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.BOBA]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.BOBA_BNB]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.BASE]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.SCROLL]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.KAVA]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.METIS]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.BTTC]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.FILECOIN]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.ZETACHAIN]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.HAQQ]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.CORE]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.THUNDERCORE]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.OPTIMISM]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.LINEA]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.BLAST]: '0x0871b2842bc5ad89183710ec5587b7e7e285f1212e8960a4941335bab95cf6af',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.SONIC]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.HEMI]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.SEPOLIA]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.TATARA]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
    [index_js_1.EvmChainId.KATANA]: '0xe18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303',
};
exports.SUSHISWAP_V2_FACTORY_ADDRESS = {
    [index_js_1.EvmChainId.ETHEREUM]: '0xc0aee478e3658e2610c5f7a4a2e1777ce9e4f2ac',
    [index_js_1.EvmChainId.FANTOM]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.POLYGON]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.GNOSIS]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.BSC]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.ARBITRUM]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.AVALANCHE]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.HAQQ]: '0xb45e53277a7e0f1d35f2a77160e91e25507f1763',
    [index_js_1.EvmChainId.HARMONY]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.CELO]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.BOBA]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.BOBA_BNB]: '0xc35dadb65012ec5796536bd9864ed8773abc74c4',
    [index_js_1.EvmChainId.BASE]: '0x71524b4f93c58fcbf659783284e38825f0622859',
    [index_js_1.EvmChainId.SCROLL]: '0xb45e53277a7e0f1d35f2a77160e91e25507f1763',
    [index_js_1.EvmChainId.KAVA]: '0xd408a20f1213286fb3158a2bfbf5bffaca8bf269',
    [index_js_1.EvmChainId.METIS]: '0x580ed43f3bba06555785c81c2957efcca71f7483',
    [index_js_1.EvmChainId.BTTC]: '0xb45e53277a7e0f1d35f2a77160e91e25507f1763',
    [index_js_1.EvmChainId.FILECOIN]: '0x9b3336186a38e1b6c21955d112dbb0343ee061ee',
    [index_js_1.EvmChainId.ZETACHAIN]: '0x33d91116e0370970444b0281ab117e161febfcdd',
    [index_js_1.EvmChainId.CORE]: '0xb45e53277a7e0f1d35f2a77160e91e25507f1763',
    [index_js_1.EvmChainId.THUNDERCORE]: '0xb45e53277a7e0f1d35f2a77160e91e25507f1763',
    [index_js_1.EvmChainId.OPTIMISM]: '0xfbc12984689e5f15626bad03ad60160fe98b303c',
    [index_js_1.EvmChainId.LINEA]: '0xfbc12984689e5f15626bad03ad60160fe98b303c',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0xb45e53277a7e0f1d35f2a77160e91e25507f1763',
    [index_js_1.EvmChainId.BLAST]: '0x42fa929fc636e657ac568c0b5cf38e203b67ac2b',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0x1aaf6eb4f85f8775400c1b10e6bbbd98b2ff8483',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0xb45e53277a7e0f1d35f2a77160e91e25507f1763',
    [index_js_1.EvmChainId.SONIC]: '0xb45e53277a7e0f1d35f2a77160e91e25507f1763',
    [index_js_1.EvmChainId.HEMI]: '0x9b3336186a38e1b6c21955d112dbb0343ee061ee',
    [index_js_1.EvmChainId.SEPOLIA]: '0x734583f62bb6ace3c9ba9bd5a53143ca2ce8c55a',
    [index_js_1.EvmChainId.TATARA]: '0xb45e53277a7e0f1d35f2a77160e91e25507f1763',
    [index_js_1.EvmChainId.KATANA]: '0x72d111b4d6f31b38919ae39779f570b747d6acd9',
};
exports.SUSHISWAP_V2_ROUTER_ADDRESS = {
    [index_js_1.EvmChainId.ETHEREUM]: '0xd9e1ce17f2641f24ae83637ab66a2cca9c378b9f',
    [index_js_1.EvmChainId.FANTOM]: '0x1b02da8cb0d097eb8d57a175b88c7d8b47997506',
    [index_js_1.EvmChainId.POLYGON]: '0x1b02da8cb0d097eb8d57a175b88c7d8b47997506',
    [index_js_1.EvmChainId.GNOSIS]: '0x1b02da8cb0d097eb8d57a175b88c7d8b47997506',
    [index_js_1.EvmChainId.BSC]: '0x1b02da8cb0d097eb8d57a175b88c7d8b47997506',
    [index_js_1.EvmChainId.ARBITRUM]: '0x1b02da8cb0d097eb8d57a175b88c7d8b47997506',
    [index_js_1.EvmChainId.AVALANCHE]: '0x1b02da8cb0d097eb8d57a175b88c7d8b47997506',
    [index_js_1.EvmChainId.HARMONY]: '0x1b02da8cb0d097eb8d57a175b88c7d8b47997506',
    [index_js_1.EvmChainId.CELO]: '0xb45e53277a7e0f1d35f2a77160e91e25507f1763',
    [index_js_1.EvmChainId.ARBITRUM_NOVA]: '0x1b02da8cb0d097eb8d57a175b88c7d8b47997506',
    [index_js_1.EvmChainId.BOBA]: '0x1b02da8cb0d097eb8d57a175b88c7d8b47997506',
    [index_js_1.EvmChainId.BOBA_BNB]: '0x1b02da8cb0d097eb8d57a175b88c7d8b47997506',
    [index_js_1.EvmChainId.BASE]: '0x6bded42c6da8fbf0d2ba55b2fa120c5e0c8d7891',
    [index_js_1.EvmChainId.SCROLL]: '0x9b3336186a38e1b6c21955d112dbb0343ee061ee',
    [index_js_1.EvmChainId.KAVA]: '0x1719def1bf8422a777f2442bce704ac4fb20c7f0',
    [index_js_1.EvmChainId.METIS]: '0xbf3b71decbcefabb3210b9d8f18ec22e0556f5f0',
    [index_js_1.EvmChainId.BTTC]: '0x9b3336186a38e1b6c21955d112dbb0343ee061ee',
    [index_js_1.EvmChainId.FILECOIN]: '0x46b3fdf7b5cde91ac049936bf0bdb12c5d22202e',
    [index_js_1.EvmChainId.HAQQ]: '0x9b3336186a38e1b6c21955d112dbb0343ee061ee',
    [index_js_1.EvmChainId.ZETACHAIN]: '0x1f2fcf1d036b375b384012e61d3aa33f8c256bbe',
    [index_js_1.EvmChainId.CORE]: '0x9b3336186a38e1b6c21955d112dbb0343ee061ee',
    [index_js_1.EvmChainId.THUNDERCORE]: '0x9b3336186a38e1b6c21955d112dbb0343ee061ee',
    [index_js_1.EvmChainId.OPTIMISM]: '0x2abf469074dc0b54d793850807e6eb5faf2625b1',
    [index_js_1.EvmChainId.LINEA]: '0x2abf469074dc0b54d793850807e6eb5faf2625b1',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0x9b3336186a38e1b6c21955d112dbb0343ee061ee',
    [index_js_1.EvmChainId.BLAST]: '0x54cf3d259a06601b5bc45f61a16443ed5404dd64',
    [index_js_1.EvmChainId.SKALE_EUROPA]: '0x4cddf8d1473df386b926ec14b23bfbd566ce827a',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0x9b3336186a38e1b6c21955d112dbb0343ee061ee',
    [index_js_1.EvmChainId.SONIC]: '0x9b3336186a38e1b6c21955d112dbb0343ee061ee',
    [index_js_1.EvmChainId.HEMI]: '0x46b3fdf7b5cde91ac049936bf0bdb12c5d22202e',
    [index_js_1.EvmChainId.SEPOLIA]: '0xeabce3e74ef41fb40024a21cc2ee2f5ddc615791',
    [index_js_1.EvmChainId.TATARA]: '0x57bffa72db682f7eb6c132dae03ff36bbeb0c459',
    [index_js_1.EvmChainId.KATANA]: '0x69cc349932ae18ed406eeb917d79b9b3033fb68e',
};
