// sushi/_cjs/evm/config/features/uniswap-v3.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UNISWAP_V3_FACTORY_ADDRESS = exports.UNISWAP_V3_INIT_CODE_HASH = exports.isUniswapV3ChainId = exports.UniswapV3ChainIds = exports.UNISWAP_V3_SUPPORTED_CHAIN_IDS = exports.UNISWAP_V3_TICK_SPACINGS = exports.UniswapV3FeeAmount = void 0;
const index_js_1 = require("../../chain/index.js");
const POOL_INIT_CODE_HASH = '0xe34f199b19b2b4f47f68442619d555527d244f78a3297ea89325f843f87b8b54';
var UniswapV3FeeAmount;
(function (UniswapV3FeeAmount) {
    UniswapV3FeeAmount[UniswapV3FeeAmount["LOWEST"] = 100] = "LOWEST";
    UniswapV3FeeAmount[UniswapV3FeeAmount["LOW"] = 500] = "LOW";
    UniswapV3FeeAmount[UniswapV3FeeAmount["MEDIUM"] = 3000] = "MEDIUM";
    UniswapV3FeeAmount[UniswapV3FeeAmount["HIGH"] = 10000] = "HIGH";
})(UniswapV3FeeAmount || (exports.UniswapV3FeeAmount = UniswapV3FeeAmount = {}));
exports.UNISWAP_V3_TICK_SPACINGS = {
    [UniswapV3FeeAmount.LOWEST]: 1,
    [UniswapV3FeeAmount.LOW]: 10,
    [UniswapV3FeeAmount.MEDIUM]: 60,
    [UniswapV3FeeAmount.HIGH]: 200,
};
exports.UNISWAP_V3_SUPPORTED_CHAIN_IDS = [
    index_js_1.EvmChainId.ARBITRUM,
    index_js_1.EvmChainId.BSC,
    index_js_1.EvmChainId.CELO,
    index_js_1.EvmChainId.ETHEREUM,
    index_js_1.EvmChainId.OPTIMISM,
    index_js_1.EvmChainId.POLYGON,
    index_js_1.EvmChainId.BASE,
    index_js_1.EvmChainId.AVALANCHE,
    index_js_1.EvmChainId.SEPOLIA,
    index_js_1.EvmChainId.BOBA,
    index_js_1.EvmChainId.ROOTSTOCK,
    index_js_1.EvmChainId.SCROLL,
    index_js_1.EvmChainId.FILECOIN,
    index_js_1.EvmChainId.LINEA,
    index_js_1.EvmChainId.BLAST,
    index_js_1.EvmChainId.ZKSYNC_ERA,
    index_js_1.EvmChainId.POLYGON_ZKEVM,
    index_js_1.EvmChainId.MANTLE,
    index_js_1.EvmChainId.MANTA,
    index_js_1.EvmChainId.TAIKO,
    index_js_1.EvmChainId.GNOSIS,
    index_js_1.EvmChainId.SONIC,
    index_js_1.EvmChainId.HEMI,
    index_js_1.EvmChainId.PLASMA,
];
exports.UniswapV3ChainIds = exports.UNISWAP_V3_SUPPORTED_CHAIN_IDS;
const isUniswapV3ChainId = (chainId) => exports.UNISWAP_V3_SUPPORTED_CHAIN_IDS.includes(chainId);
exports.isUniswapV3ChainId = isUniswapV3ChainId;
exports.UNISWAP_V3_INIT_CODE_HASH = {
    [index_js_1.EvmChainId.ETHEREUM]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.POLYGON]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.ARBITRUM]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.OPTIMISM]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.BSC]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.CELO]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.BASE]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.AVALANCHE]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.SEPOLIA]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.SCROLL]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.ROOTSTOCK]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.FILECOIN]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.BOBA]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.LINEA]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.BLAST]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.ZKSYNC_ERA]: '0x010013f177ea1fcbc4520f9a3ca7cd2d1d77959e05aa66484027cb38e712aeed',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.MANTLE]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.MANTA]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.TAIKO]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.GNOSIS]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.SONIC]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.HEMI]: POOL_INIT_CODE_HASH,
    [index_js_1.EvmChainId.PLASMA]: POOL_INIT_CODE_HASH,
};
exports.UNISWAP_V3_FACTORY_ADDRESS = {
    [index_js_1.EvmChainId.ETHEREUM]: '0x1f98431c8ad98523631ae4a59f267346ea31f984',
    [index_js_1.EvmChainId.POLYGON]: '0x1f98431c8ad98523631ae4a59f267346ea31f984',
    [index_js_1.EvmChainId.ARBITRUM]: '0x1f98431c8ad98523631ae4a59f267346ea31f984',
    [index_js_1.EvmChainId.OPTIMISM]: '0x1f98431c8ad98523631ae4a59f267346ea31f984',
    [index_js_1.EvmChainId.BSC]: '0xdb1d10011ad0ff90774d0c6bb92e5c5c8b4461f7',
    [index_js_1.EvmChainId.CELO]: '0xafe208a311b21f13ef87e33a90049fc17a7acdec',
    [index_js_1.EvmChainId.BASE]: '0x33128a8fc17869897dce68ed026d694621f6fdfd',
    [index_js_1.EvmChainId.AVALANCHE]: '0x740b1c1de25031c31ff4fc9a62f554a55cdc1bad',
    [index_js_1.EvmChainId.SEPOLIA]: '0x0227628f3f023bb0b980b67d528571c95c6dac1c',
    [index_js_1.EvmChainId.SCROLL]: '0x70c62c8b8e801124a4aa81ce07b637a3e83cb919',
    [index_js_1.EvmChainId.ROOTSTOCK]: '0xaf37ec98a00fd63689cf3060bf3b6784e00cad82',
    [index_js_1.EvmChainId.FILECOIN]: '0xb4c47ed546fc31e26470a186ec2c5f19ef09ba41',
    [index_js_1.EvmChainId.BOBA]: '0xffcd7aed9c627e82a765c3247d562239507f6f1b',
    [index_js_1.EvmChainId.LINEA]: '0x31fafd4889fa1269f7a13a66ee0fb458f27d72a9',
    [index_js_1.EvmChainId.BLAST]: '0x792edade80af5fc680d96a2ed80a44247d2cf6fd',
    [index_js_1.EvmChainId.ZKSYNC_ERA]: '0x8fda5a7a8dca67bbcdd10f02fa0649a937215422',
    [index_js_1.EvmChainId.POLYGON_ZKEVM]: '0xff83c3c800fec21de45c5ec30b69ddd5ee60dfc2',
    [index_js_1.EvmChainId.MANTLE]: '0x0d922fb1bc191f64970ac40376643808b4b74df9',
    [index_js_1.EvmChainId.MANTA]: '0x06d830e15081f65923674268121ff57cc54e4e23',
    [index_js_1.EvmChainId.TAIKO]: '0x75fc67473a91335b5b8f8821277262a13b38c9b3',
    [index_js_1.EvmChainId.GNOSIS]: '0xe32f7dd7e3f098d518ff19a22d5f028e076489b1',
    [index_js_1.EvmChainId.SONIC]: '0xcb2436774c3e191c85056d248ef4260ce5f27a9d',
    [index_js_1.EvmChainId.HEMI]: '0x346239972d1fa486fc4a521031bc81bfb7d6e8a4',
    [index_js_1.EvmChainId.PLASMA]: '0xcb2436774c3e191c85056d248ef4260ce5f27a9d',
};
