// sushi/_esm/evm/currency/native.js
import * as z from 'zod';
import { Native } from '../../generic/currency/native.js';
import { getEvmChainById, isEvmChainId, } from '../chain/chains.js';
import { WNATIVE } from '../config/tokens/wrapped-native.js';
import { EvmToken } from './token.js';
export class EvmNative extends Native {
    wrap() {
        return new EvmToken({
            ...WNATIVE[this.chainId],
            metadata: structuredClone(this.metadata),
        });
    }
    toJSON() {
        return {
            chainId: this.chainId,
            symbol: this.symbol,
            name: this.name,
            decimals: this.decimals,
            type: this.type,
            metadata: this.metadata,
        };
    }
    static fromJSON(data) {
        return new this(data);
    }
    static fromChainId(chainId) {
        const chain = getEvmChainById(chainId);
        return new EvmNative({
            chainId,
            symbol: chain.viemChain.nativeCurrency.symbol,
            name: chain.viemChain.nativeCurrency.name,
            decimals: chain.viemChain.nativeCurrency.decimals,
        });
    }
}
export const serializedEvmNativeSchema = ({ metadata, } = {}) => z.object({
    chainId: z.number().int().refine(isEvmChainId),
    symbol: z.string(),
    name: z.string(),
    decimals: z.number().int().nonnegative(),
    type: z.literal('native'),
    metadata: (metadata ||
        z.record(z.unknown()).optional().default({})),
});
