// sushi/_esm/evm/currency/token.js
import invariant from 'tiny-invariant';
import { isAddress } from 'viem';
import * as z from 'zod';
import { Token } from '../../generic/currency/token.js';
import { isEvmChainId } from '../chain/chains.js';
export function isEvmAddress(address) {
    return isAddress(address, { strict: false });
}
export class EvmToken extends Token {
    constructor({ origin, address, ...rest }) {
        super({ address: address.toLowerCase(), ...rest });
        this.origin = origin;
    }
    sortsBefore(other) {
        invariant(this.chainId === other.chainId, 'Tokens must be on the same chain');
        invariant(this.address.toLowerCase() !== other.address.toLowerCase(), 'Tokens must have different addresses');
        return this.address.toLowerCase() < other.address.toLowerCase();
    }
    wrap() {
        return this;
    }
    toJSON() {
        return {
            chainId: this.chainId,
            address: this.address,
            symbol: this.symbol,
            name: this.name,
            decimals: this.decimals,
            type: this.type,
            metadata: this.metadata,
        };
    }
    static fromJSON(data) {
        return new this(data);
    }
}
export const serializedEvmTokenSchema = ({ metadata, } = {}) => z.object({
    chainId: z.number().int().refine(isEvmChainId),
    address: z.string().refine(isEvmAddress),
    symbol: z.string(),
    name: z.string(),
    decimals: z.number().int().nonnegative(),
    type: z.literal('token'),
    metadata: (metadata ||
        z.record(z.unknown()).optional().default({})),
});
