// sushi/_esm/evm/pool/sushiswap-v3/utils/sqrtPriceMath.js
import invariant from 'tiny-invariant';
import { MAX_UINT256 } from '../../../../generic/math/constants.js';
import { Q96 } from '../internalConstants.js';
import { FullMath } from './fullMath.js';
const MaxUint160 = 2n ** 160n - 1n;
function multiplyIn256(x, y) {
    const product = x * y;
    return product & MAX_UINT256;
}
function addIn256(x, y) {
    const sum = x + y;
    return sum & MAX_UINT256;
}
export class SqrtPriceMath {
    /**
     * Cannot be constructed.
     */
    constructor() { }
    static getAmount0Delta(sqrtRatioAX96, sqrtRatioBX96, liquidity, roundUp) {
        if (sqrtRatioAX96 > sqrtRatioBX96) {
            ;
            [sqrtRatioAX96, sqrtRatioBX96] = [sqrtRatioBX96, sqrtRatioAX96];
        }
        const numerator1 = liquidity << 96n;
        const numerator2 = sqrtRatioBX96 - sqrtRatioAX96;
        return roundUp
            ? FullMath.mulDivRoundingUp(FullMath.mulDivRoundingUp(numerator1, numerator2, sqrtRatioBX96), 1n, sqrtRatioAX96)
            : (numerator1 * numerator2) / sqrtRatioBX96 / sqrtRatioAX96;
    }
    static getAmount1Delta(sqrtRatioAX96, sqrtRatioBX96, liquidity, roundUp) {
        if (sqrtRatioAX96 > sqrtRatioBX96) {
            ;
            [sqrtRatioAX96, sqrtRatioBX96] = [sqrtRatioBX96, sqrtRatioAX96];
        }
        return roundUp
            ? FullMath.mulDivRoundingUp(liquidity, sqrtRatioBX96 - sqrtRatioAX96, Q96)
            : (liquidity * (sqrtRatioBX96 - sqrtRatioAX96)) / Q96;
    }
    static getNextSqrtPriceFromInput(sqrtPX96, liquidity, amountIn, zeroForOne) {
        invariant(sqrtPX96 > 0n);
        invariant(liquidity > 0n);
        return zeroForOne
            ? this.getNextSqrtPriceFromAmount0RoundingUp(sqrtPX96, liquidity, amountIn, true)
            : this.getNextSqrtPriceFromAmount1RoundingDown(sqrtPX96, liquidity, amountIn, true);
    }
    static getNextSqrtPriceFromOutput(sqrtPX96, liquidity, amountOut, zeroForOne) {
        invariant(sqrtPX96 > 0n);
        invariant(liquidity > 0n);
        return zeroForOne
            ? this.getNextSqrtPriceFromAmount1RoundingDown(sqrtPX96, liquidity, amountOut, false)
            : this.getNextSqrtPriceFromAmount0RoundingUp(sqrtPX96, liquidity, amountOut, false);
    }
    static getNextSqrtPriceFromAmount0RoundingUp(sqrtPX96, liquidity, amount, add) {
        if (amount === 0n)
            return sqrtPX96;
        const numerator1 = liquidity << 96n;
        if (add) {
            const product = multiplyIn256(amount, sqrtPX96);
            if (product / amount === sqrtPX96) {
                const denominator = addIn256(numerator1, product);
                if (denominator >= numerator1) {
                    return FullMath.mulDivRoundingUp(numerator1, sqrtPX96, denominator);
                }
            }
            return FullMath.mulDivRoundingUp(numerator1, 1n, numerator1 / sqrtPX96 + amount);
        }
        else {
            const product = multiplyIn256(amount, sqrtPX96);
            invariant(product / amount === sqrtPX96);
            invariant(numerator1 > product);
            const denominator = numerator1 - product;
            return FullMath.mulDivRoundingUp(numerator1, sqrtPX96, denominator);
        }
    }
    static getNextSqrtPriceFromAmount1RoundingDown(sqrtPX96, liquidity, amount, add) {
        if (add) {
            const quotient = amount <= MaxUint160
                ? (amount << 96n) / liquidity
                : (amount * Q96) / liquidity;
            return sqrtPX96 + quotient;
        }
        else {
            const quotient = FullMath.mulDivRoundingUp(amount, Q96, liquidity);
            invariant(sqrtPX96 > quotient);
            return sqrtPX96 - quotient;
        }
    }
}
