// sushi/_cjs/evm/abi/tridentRouterAbi/tridentRouterAbi_exactInputWithNativeToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tridentRouterAbi_exactInputWithNativeToken = void 0;
exports.tridentRouterAbi_exactInputWithNativeToken = [
    {
        inputs: [
            {
                components: [
                    { internalType: 'address', name: 'tokenIn', type: 'address' },
                    { internalType: 'uint256', name: 'amountIn', type: 'uint256' },
                    {
                        internalType: 'uint256',
                        name: 'amountOutMinimum',
                        type: 'uint256',
                    },
                    {
                        components: [
                            { internalType: 'address', name: 'pool', type: 'address' },
                            { internalType: 'bytes', name: 'data', type: 'bytes' },
                        ],
                        internalType: 'struct ITridentRouter.Path[]',
                        name: 'path',
                        type: 'tuple[]',
                    },
                ],
                internalType: 'struct ITridentRouter.ExactInputParams',
                name: 'params',
                type: 'tuple',
            },
        ],
        name: 'exactInputWithNativeToken',
        outputs: [{ internalType: 'uint256', name: 'amountOut', type: 'uint256' }],
        stateMutability: 'payable',
        type: 'function',
    },
];
