// @coinbase/onchainkit/dist/fund/utils/subscribeToWindowMessage.js
import { DEFAULT_ONRAMP_URL } from "../constants.js";
const MessageCodes = {
  AppParams: "app_params",
  PaymentLinkSuccess: "payment_link_success",
  PaymentLinkClosed: "payment_link_closed",
  GuestCheckoutRedirectSuccess: "guest_checkout_redirect_success",
  Success: "success",
  Event: "event"
};
function subscribeToWindowMessage({
  onMessage,
  allowedOrigin = DEFAULT_ONRAMP_URL,
  onValidateOrigin = () => Promise.resolve(true)
}) {
  const handleMessage = (event) => {
    if (!isAllowedOrigin({
      event,
      allowedOrigin
    })) {
      return;
    }
    const {
      eventName,
      data
    } = JSON.parse(event.data);
    if (eventName === "event") {
      (async () => {
        if (await onValidateOrigin(event.origin)) {
          onMessage(data);
        }
      })();
    }
  };
  window.addEventListener("message", handleMessage);
  return () => {
    window.removeEventListener("message", handleMessage);
  };
}
function isAllowedOrigin({
  event,
  allowedOrigin
}) {
  const isOriginAllowed = !allowedOrigin || event.origin === allowedOrigin;
  return isOriginAllowed;
}
export {
  MessageCodes,
  subscribeToWindowMessage
};
