// opensea-js/lib/api/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventOrderType = exports.AssetEventType = exports.TraitDisplayType = exports.CollectionOrderByOption = void 0;
var CollectionOrderByOption;
(function (CollectionOrderByOption) {
    CollectionOrderByOption["CREATED_DATE"] = "created_date";
    CollectionOrderByOption["ONE_DAY_CHANGE"] = "one_day_change";
    CollectionOrderByOption["SEVEN_DAY_VOLUME"] = "seven_day_volume";
    CollectionOrderByOption["SEVEN_DAY_CHANGE"] = "seven_day_change";
    CollectionOrderByOption["NUM_OWNERS"] = "num_owners";
    CollectionOrderByOption["MARKET_CAP"] = "market_cap";
})(CollectionOrderByOption || (exports.CollectionOrderByOption = CollectionOrderByOption = {}));
/**
 * Trait display type returned by OpenSea API.
 * @category API Models
 */
var TraitDisplayType;
(function (TraitDisplayType) {
    TraitDisplayType["NUMBER"] = "number";
    TraitDisplayType["BOOST_PERCENTAGE"] = "boost_percentage";
    TraitDisplayType["BOOST_NUMBER"] = "boost_number";
    TraitDisplayType["AUTHOR"] = "author";
    TraitDisplayType["DATE"] = "date";
    /** "None" is used for string traits */
    TraitDisplayType["NONE"] = "None";
})(TraitDisplayType || (exports.TraitDisplayType = TraitDisplayType = {}));
/**
 * Asset event type returned by OpenSea API.
 * @category API Models
 */
var AssetEventType;
(function (AssetEventType) {
    AssetEventType["ORDER"] = "order";
    AssetEventType["SALE"] = "sale";
    AssetEventType["TRANSFER"] = "transfer";
    AssetEventType["CANCEL"] = "cancel";
    AssetEventType["REDEMPTION"] = "redemption";
})(AssetEventType || (exports.AssetEventType = AssetEventType = {}));
/**
 * Order type for events.
 * @category API Models
 */
var EventOrderType;
(function (EventOrderType) {
    EventOrderType["LISTING"] = "listing";
    EventOrderType["ITEM_OFFER"] = "item_offer";
    EventOrderType["COLLECTION_OFFER"] = "collection_offer";
    EventOrderType["TRAIT_OFFER"] = "trait_offer";
})(EventOrderType || (exports.EventOrderType = EventOrderType = {}));
