// sushi/_cjs/evm/currency/native.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializedEvmNativeSchema = exports.EvmNative = void 0;
const z = __importStar(require("zod"));
const native_js_1 = require("../../generic/currency/native.js");
const chains_js_1 = require("../chain/chains.js");
const wrapped_native_js_1 = require("../config/tokens/wrapped-native.js");
const token_js_1 = require("./token.js");
class EvmNative extends native_js_1.Native {
    wrap() {
        return new token_js_1.EvmToken({
            ...wrapped_native_js_1.WNATIVE[this.chainId],
            metadata: structuredClone(this.metadata),
        });
    }
    toJSON() {
        return {
            chainId: this.chainId,
            symbol: this.symbol,
            name: this.name,
            decimals: this.decimals,
            type: this.type,
            metadata: this.metadata,
        };
    }
    static fromJSON(data) {
        return new this(data);
    }
    static fromChainId(chainId) {
        const chain = (0, chains_js_1.getEvmChainById)(chainId);
        return new EvmNative({
            chainId,
            symbol: chain.viemChain.nativeCurrency.symbol,
            name: chain.viemChain.nativeCurrency.name,
            decimals: chain.viemChain.nativeCurrency.decimals,
        });
    }
}
exports.EvmNative = EvmNative;
const serializedEvmNativeSchema = ({ metadata, } = {}) => z.object({
    chainId: z.number().int().refine(chains_js_1.isEvmChainId),
    symbol: z.string(),
    name: z.string(),
    decimals: z.number().int().nonnegative(),
    type: z.literal('native'),
    metadata: (metadata ||
        z.record(z.unknown()).optional().default({})),
});
exports.serializedEvmNativeSchema = serializedEvmNativeSchema;
