// sushi/_cjs/evm/currency/token.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializedEvmTokenSchema = exports.EvmToken = void 0;
exports.isEvmAddress = isEvmAddress;
const tiny_invariant_1 = __importDefault(require("tiny-invariant"));
const viem_1 = require("viem");
const z = __importStar(require("zod"));
const token_js_1 = require("../../generic/currency/token.js");
const chains_js_1 = require("../chain/chains.js");
function isEvmAddress(address) {
    return (0, viem_1.isAddress)(address, { strict: false });
}
class EvmToken extends token_js_1.Token {
    constructor({ origin, address, ...rest }) {
        super({ address: address.toLowerCase(), ...rest });
        this.origin = origin;
    }
    sortsBefore(other) {
        (0, tiny_invariant_1.default)(this.chainId === other.chainId, 'Tokens must be on the same chain');
        (0, tiny_invariant_1.default)(this.address.toLowerCase() !== other.address.toLowerCase(), 'Tokens must have different addresses');
        return this.address.toLowerCase() < other.address.toLowerCase();
    }
    wrap() {
        return this;
    }
    toJSON() {
        return {
            chainId: this.chainId,
            address: this.address,
            symbol: this.symbol,
            name: this.name,
            decimals: this.decimals,
            type: this.type,
            metadata: this.metadata,
        };
    }
    static fromJSON(data) {
        return new this(data);
    }
}
exports.EvmToken = EvmToken;
const serializedEvmTokenSchema = ({ metadata, } = {}) => z.object({
    chainId: z.number().int().refine(chains_js_1.isEvmChainId),
    address: z.string().refine(isEvmAddress),
    symbol: z.string(),
    name: z.string(),
    decimals: z.number().int().nonnegative(),
    type: z.literal('token'),
    metadata: (metadata ||
        z.record(z.unknown()).optional().default({})),
});
exports.serializedEvmTokenSchema = serializedEvmTokenSchema;
