// sushi/_esm/generic/chain/chains.js
import { evmChains } from '../../evm/chain/index.js';
import { kvmChains } from '../../kvm/chain/chains.js';
import { mvmChains } from '../../mvm/chain/chains.js';
import { tvmChains } from '../../tvm/chain/chains.js';
export const chains = [
    ...evmChains,
    ...mvmChains,
    ...tvmChains,
    ...kvmChains,
];
export const ChainId = /* @__PURE__ */ new Proxy({}, {
    get: (_, key) => {
        const _key = key.toLowerCase().replace(/_/g, '-');
        const chain = chains.find((chain) => chain.key === _key);
        if (!chain)
            return undefined;
        return chain.chainId;
    },
    ownKeys: () => chains.map((c) => c.key),
    getOwnPropertyDescriptor: () => {
        return { enumerable: true, configurable: true };
    },
});
export const chainIds = /* @__PURE__ */ chains.map((chain) => chain.chainId);
export function isChainId(chainId) {
    return chains.some((chain) => chain.chainId === chainId);
}
export function getChainById(chainId) {
    return chains.find((chain) => chain.chainId === chainId);
}
export function isMainnetChainId(chainId) {
    return chains.some((chain) => chain.chainId === chainId && chain.netType === 'mainnet');
}
export function isTestnetChainId(chainId) {
    return chains.some((chain) => chain.chainId === chainId && chain.netType === 'testnet');
}
export function isChainKey(key) {
    return chains.some((chain) => chain.key === key);
}
export function getChainByKey(key) {
    return chains.find((chain) => chain.key === key);
}
