// sushi/_cjs/generic/chain/chains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chainIds = exports.ChainId = exports.chains = void 0;
exports.isChainId = isChainId;
exports.getChainById = getChainById;
exports.isMainnetChainId = isMainnetChainId;
exports.isTestnetChainId = isTestnetChainId;
exports.isChainKey = isChainKey;
exports.getChainByKey = getChainByKey;
const index_js_1 = require("../../evm/chain/index.js");
const chains_js_1 = require("../../kvm/chain/chains.js");
const chains_js_2 = require("../../mvm/chain/chains.js");
const chains_js_3 = require("../../tvm/chain/chains.js");
exports.chains = [
    ...index_js_1.evmChains,
    ...chains_js_2.mvmChains,
    ...chains_js_3.tvmChains,
    ...chains_js_1.kvmChains,
];
exports.ChainId = new Proxy({}, {
    get: (_, key) => {
        const _key = key.toLowerCase().replace(/_/g, '-');
        const chain = exports.chains.find((chain) => chain.key === _key);
        if (!chain)
            return undefined;
        return chain.chainId;
    },
    ownKeys: () => exports.chains.map((c) => c.key),
    getOwnPropertyDescriptor: () => {
        return { enumerable: true, configurable: true };
    },
});
exports.chainIds = exports.chains.map((chain) => chain.chainId);
function isChainId(chainId) {
    return exports.chains.some((chain) => chain.chainId === chainId);
}
function getChainById(chainId) {
    return exports.chains.find((chain) => chain.chainId === chainId);
}
function isMainnetChainId(chainId) {
    return exports.chains.some((chain) => chain.chainId === chainId && chain.netType === 'mainnet');
}
function isTestnetChainId(chainId) {
    return exports.chains.some((chain) => chain.chainId === chainId && chain.netType === 'testnet');
}
function isChainKey(key) {
    return exports.chains.some((chain) => chain.key === key);
}
function getChainByKey(key) {
    return exports.chains.find((chain) => chain.key === key);
}
