// sushi/_esm/evm/pool/sushiswap-v3/entities/NonfungiblePositionManager.js
import invariant from 'tiny-invariant';
import { encodeFunctionData, maxUint128, toHex, zeroAddress, } from 'viem';
import { Amount } from '../../../../generic/currency/amount.js';
import { nonfungiblePositionManagerAbi_burn, nonfungiblePositionManagerAbi_collect, nonfungiblePositionManagerAbi_createAndInitializePoolIfNecessary, nonfungiblePositionManagerAbi_decreaseLiquidity, nonfungiblePositionManagerAbi_increaseLiquidity, nonfungiblePositionManagerAbi_mint, nonfungiblePositionManagerAbi_permit, nonfungiblePositionManagerAbi_safeTransferFrom, } from '../../../abi/nonfungiblePositionManagerAbi/index.js';
import { validateAndParseAddress } from '../utils/index.js';
import { Multicall } from './Multicall.js';
import { Payments } from './Payments.js';
import { Position } from './Position.js';
import { SelfPermit } from './SelfPermit.js';
// type guard
function isMint(options) {
    return Object.keys(options).some((k) => k === 'recipient');
}
export class NonfungiblePositionManager {
    /**
     * Cannot be constructed.
     */
    constructor() { }
    static encodeCreate(pool) {
        return encodeFunctionData({
            abi: nonfungiblePositionManagerAbi_createAndInitializePoolIfNecessary,
            functionName: 'createAndInitializePoolIfNecessary',
            args: [
                pool.token0.address,
                pool.token1.address,
                pool.fee,
                pool.sqrtRatioX96,
            ],
        });
    }
    static createCallParameters(pool) {
        return {
            calldata: this.encodeCreate(pool),
            value: toHex(0),
        };
    }
    static addCallParameters(position, options) {
        invariant(position.liquidity > 0n, 'ZERO_LIQUIDITY');
        const calldatas = [];
        // get amounts
        const { amount0: amount0Desired, amount1: amount1Desired } = position.mintAmounts;
        // adjust for slippage
        const minimumAmounts = position.mintAmountsWithSlippage(options.slippageTolerance);
        const amount0Min = minimumAmounts.amount0;
        const amount1Min = minimumAmounts.amount1;
        const deadline = BigInt(options.deadline);
        // create pool if needed
        if (isMint(options) && options.createPool) {
            calldatas.push(this.encodeCreate(position.pool));
        }
        // permits if necessary
        if (options.token0Permit) {
            calldatas.push(SelfPermit.encodePermit(position.pool.token0, options.token0Permit));
        }
        if (options.token1Permit) {
            calldatas.push(SelfPermit.encodePermit(position.pool.token1, options.token1Permit));
        }
        // mint
        if (isMint(options)) {
            const recipient = validateAndParseAddress(options.recipient);
            calldatas.push(encodeFunctionData({
                abi: nonfungiblePositionManagerAbi_mint,
                functionName: 'mint',
                args: [
                    {
                        token0: position.pool.token0.address,
                        token1: position.pool.token1.address,
                        fee: position.pool.fee,
                        tickLower: position.tickLower,
                        tickUpper: position.tickUpper,
                        amount0Desired: amount0Desired,
                        amount1Desired: amount1Desired,
                        amount0Min,
                        amount1Min,
                        recipient,
                        deadline,
                    },
                ],
            }));
        }
        else {
            // increase
            calldatas.push(encodeFunctionData({
                abi: nonfungiblePositionManagerAbi_increaseLiquidity,
                functionName: 'increaseLiquidity',
                args: [
                    {
                        tokenId: BigInt(options.tokenId),
                        amount0Desired: amount0Desired,
                        amount1Desired: amount1Desired,
                        amount0Min,
                        amount1Min,
                        deadline,
                    },
                ],
            }));
        }
        let value = toHex(0);
        if (options.useNative) {
            const wrapped = options.useNative.wrap();
            invariant(position.pool.token0.isSame(wrapped) ||
                position.pool.token1.isSame(wrapped), 'NO_WETH');
            const wrappedValue = position.pool.token0.isSame(wrapped)
                ? amount0Desired
                : amount1Desired;
            // we only need to refund if we're actually sending ETH
            if (wrappedValue > 0n) {
                calldatas.push(Payments.encodeRefundETH());
            }
            value = toHex(wrappedValue);
        }
        return {
            calldata: Multicall.encodeMulticall(calldatas),
            value,
        };
    }
    static encodeCollect(collectOptions) {
        const calldatas = [];
        const nativeBalancesByRecipient = new Map();
        const sweepBalancesByRecipient = new Map();
        for (const options of collectOptions) {
            const tokenId = BigInt(options.tokenId);
            const involvesETH = options.expectedCurrencyOwed0.currency.type === 'native' ||
                options.expectedCurrencyOwed1.currency.type === 'native';
            const recipient = validateAndParseAddress(options.recipient);
            calldatas.push(encodeFunctionData({
                abi: nonfungiblePositionManagerAbi_collect,
                functionName: 'collect',
                args: [
                    {
                        tokenId,
                        recipient: involvesETH ? zeroAddress : recipient,
                        amount0Max: maxUint128,
                        amount1Max: maxUint128,
                    },
                ],
            }));
            if (involvesETH) {
                const ethAmount = options.expectedCurrencyOwed0.currency.type === 'native'
                    ? options.expectedCurrencyOwed0.amount
                    : options.expectedCurrencyOwed1.amount;
                const token = options.expectedCurrencyOwed0.currency.type === 'native'
                    ? options.expectedCurrencyOwed1.currency
                    : options.expectedCurrencyOwed0.currency;
                const tokenAmount = options.expectedCurrencyOwed0.currency.type === 'native'
                    ? options.expectedCurrencyOwed1.amount
                    : options.expectedCurrencyOwed0.amount;
                const currentNativeBalance = nativeBalancesByRecipient.get(recipient) ?? 0n;
                nativeBalancesByRecipient.set(recipient, currentNativeBalance + ethAmount);
                const balanceKey = `${token.address}-${recipient}`;
                const accumulatedBalance = sweepBalancesByRecipient.get(balanceKey);
                if (accumulatedBalance) {
                    accumulatedBalance.amount += tokenAmount;
                }
                else {
                    sweepBalancesByRecipient.set(balanceKey, {
                        token,
                        amount: tokenAmount,
                        recipient,
                    });
                }
            }
        }
        for (const [recipient, nativeAmount,] of nativeBalancesByRecipient.entries()) {
            calldatas.push(Payments.encodeUnwrapWETH9(nativeAmount, recipient));
        }
        for (const { token, amount, recipient, } of sweepBalancesByRecipient.values()) {
            calldatas.push(Payments.encodeSweepToken(token, amount, recipient));
        }
        return calldatas;
    }
    static collectCallParameters(options) {
        const calldatas = NonfungiblePositionManager.encodeCollect(Array.isArray(options) ? options : [options]);
        return {
            calldata: Multicall.encodeMulticall(calldatas),
            value: toHex(0),
        };
    }
    /**
     * Produces the calldata for completely or partially exiting a position
     * @param position The position to exit
     * @param options Additional information necessary for generating the calldata
     * @returns The call parameters
     */
    static removeCallParameters(position, options) {
        const calldatas = [];
        const deadline = BigInt(options.deadline);
        const tokenId = BigInt(options.tokenId);
        // construct a partial position with a percentage of liquidity
        const partialPosition = new Position({
            pool: position.pool,
            liquidity: options.liquidityPercentage.mul(position.liquidity).quotient,
            tickLower: position.tickLower,
            tickUpper: position.tickUpper,
        });
        invariant(partialPosition.liquidity > 0n, 'ZERO_LIQUIDITY');
        // slippage-adjusted underlying amounts
        const { amount0: amount0Min, amount1: amount1Min } = partialPosition.burnAmountsWithSlippage(options.slippageTolerance);
        if (options.permit) {
            calldatas.push(encodeFunctionData({
                abi: nonfungiblePositionManagerAbi_permit,
                functionName: 'permit',
                args: [
                    validateAndParseAddress(options.permit.spender),
                    tokenId,
                    deadline,
                    options.permit.v,
                    options.permit.r,
                    options.permit.s,
                ],
            }));
        }
        // remove liquidity
        calldatas.push(encodeFunctionData({
            abi: nonfungiblePositionManagerAbi_decreaseLiquidity,
            functionName: 'decreaseLiquidity',
            args: [
                {
                    tokenId,
                    liquidity: partialPosition.liquidity,
                    amount0Min,
                    amount1Min,
                    deadline,
                },
            ],
        }));
        const { expectedCurrencyOwed0, expectedCurrencyOwed1, ...rest } = options.collectOptions;
        calldatas.push(...NonfungiblePositionManager.encodeCollect([
            {
                tokenId: toHex(BigInt(options.tokenId)),
                // add the underlying value to the expected currency already owed
                expectedCurrencyOwed0: expectedCurrencyOwed0.add(new Amount(expectedCurrencyOwed0.currency, amount0Min)),
                expectedCurrencyOwed1: expectedCurrencyOwed1.add(new Amount(expectedCurrencyOwed1.currency, amount1Min)),
                ...rest,
            },
        ]));
        if (options.liquidityPercentage.eq(1n)) {
            if (options.burnToken) {
                calldatas.push(encodeFunctionData({
                    abi: nonfungiblePositionManagerAbi_burn,
                    functionName: 'burn',
                    args: [tokenId],
                }));
            }
        }
        else {
            invariant(options.burnToken !== true, 'CANNOT_BURN');
        }
        return {
            calldata: Multicall.encodeMulticall(calldatas),
            value: toHex(0),
        };
    }
    static safeTransferFromParameters(options) {
        const recipient = validateAndParseAddress(options.recipient);
        const sender = validateAndParseAddress(options.sender);
        let calldata;
        if (options.data) {
            calldata = encodeFunctionData({
                abi: nonfungiblePositionManagerAbi_safeTransferFrom,
                functionName: 'safeTransferFrom',
                args: [sender, recipient, BigInt(options.tokenId), options.data],
            });
        }
        else {
            calldata = encodeFunctionData({
                abi: nonfungiblePositionManagerAbi_safeTransferFrom,
                functionName: 'safeTransferFrom',
                args: [sender, recipient, BigInt(options.tokenId)],
            });
        }
        return {
            calldata: calldata,
            value: toHex(0),
        };
    }
}
