// sushi/_esm/evm/pool/sushiswap-v3/entities/Payments.js
import { encodeFunctionData } from 'viem';
import { peripheryPaymentsWithFeeAbi_refundETH, peripheryPaymentsWithFeeAbi_sweepToken, peripheryPaymentsWithFeeAbi_sweepTokenWithFee, peripheryPaymentsWithFeeAbi_unwrapWETH9, peripheryPaymentsWithFeeAbi_unwrapWETH9WithFee, } from '../../../abi/peripheryPaymentsWithFeeAbi/index.js';
import { validateAndParseAddress } from '../utils/index.js';
export class Payments {
    /**
     * Cannot be constructed.
     */
    constructor() { }
    static encodeFeeBips(fee) {
        return fee.mul(10_000).quotient;
    }
    static encodeUnwrapWETH9(amountMinimum, recipient, feeOptions) {
        recipient = validateAndParseAddress(recipient);
        if (feeOptions) {
            const feeBips = this.encodeFeeBips(feeOptions.fee);
            const feeRecipient = validateAndParseAddress(feeOptions.recipient);
            return encodeFunctionData({
                abi: peripheryPaymentsWithFeeAbi_unwrapWETH9WithFee,
                functionName: 'unwrapWETH9WithFee',
                args: [amountMinimum, recipient, feeBips, feeRecipient],
            });
        }
        else {
            return encodeFunctionData({
                abi: peripheryPaymentsWithFeeAbi_unwrapWETH9,
                functionName: 'unwrapWETH9',
                args: [amountMinimum, recipient],
            });
        }
    }
    static encodeSweepToken(token, amountMinimum, recipient, feeOptions) {
        recipient = validateAndParseAddress(recipient);
        if (feeOptions) {
            const feeBips = this.encodeFeeBips(feeOptions.fee);
            const feeRecipient = validateAndParseAddress(feeOptions.recipient);
            return encodeFunctionData({
                abi: peripheryPaymentsWithFeeAbi_sweepTokenWithFee,
                functionName: 'sweepTokenWithFee',
                args: [token.address, amountMinimum, recipient, feeBips, feeRecipient],
            });
        }
        else {
            return encodeFunctionData({
                abi: peripheryPaymentsWithFeeAbi_sweepToken,
                functionName: 'sweepToken',
                args: [token.address, amountMinimum, recipient],
            });
        }
    }
    static encodeRefundETH() {
        return encodeFunctionData({
            abi: peripheryPaymentsWithFeeAbi_refundETH,
            functionName: 'refundETH',
        });
    }
}
