// sushi/_esm/evm/pool/sushiswap-v3/entities/SelfPermit.js
import { encodeFunctionData } from 'viem';
import { selfPermitAbi_selfPermit, selfPermitAbi_selfPermitAllowed, } from '../../../abi/selfPermitAbi/index.js';
function isAllowedPermit(permitOptions) {
    return 'nonce' in permitOptions;
}
export class SelfPermit {
    /**
     * Cannot be constructed.
     */
    constructor() { }
    static encodePermit(token, options) {
        if (isAllowedPermit(options)) {
            return encodeFunctionData({
                abi: selfPermitAbi_selfPermitAllowed,
                functionName: 'selfPermitAllowed',
                args: [
                    token.address,
                    BigInt(options.nonce),
                    BigInt(options.expiry),
                    options.v,
                    options.r,
                    options.s,
                ],
            });
        }
        return encodeFunctionData({
            abi: selfPermitAbi_selfPermit,
            functionName: 'selfPermit',
            args: [
                token.address,
                BigInt(options.amount),
                BigInt(options.deadline),
                options.v,
                options.r,
                options.s,
            ],
        });
    }
}
