// sushi/_cjs/evm/chain/chains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EvmChainId = exports.evmChainIds = exports.evmChains = void 0;
exports.isEvmChainId = isEvmChainId;
exports.getEvmChainById = getEvmChainById;
exports.isEvmMainnetChainId = isEvmMainnetChainId;
exports.isEvmTestnetChainId = isEvmTestnetChainId;
exports.isEvmChainKey = isEvmChainKey;
exports.getEvmChainByKey = getEvmChainByKey;
const viem_1 = require("viem");
const chains_1 = require("viem/chains");
const define_chain_js_1 = require("./define-chain.js");
const ethereum = (0, define_chain_js_1.defineEvmChain)(chains_1.mainnet, {
    key: 'ethereum',
    shortName: 'eth',
});
const polygon = (0, define_chain_js_1.defineEvmChain)(chains_1.polygon, {
    key: 'polygon',
    shortName: 'pol',
});
const fantom = (0, define_chain_js_1.defineEvmChain)(chains_1.fantom, {
    key: 'fantom',
    shortName: 'ftm',
});
const gnosis = (0, define_chain_js_1.defineEvmChain)(chains_1.gnosis, {
    key: 'gnosis',
    shortName: 'gno',
});
const bsc = (0, define_chain_js_1.defineEvmChain)(chains_1.bsc, {
    key: 'bsc',
    shortName: 'bnb',
});
const bobaBnb = (0, define_chain_js_1.defineEvmChain)((0, viem_1.defineChain)({
    id: 56288,
    name: 'Boba BNB',
    nativeCurrency: { name: 'Boba', symbol: 'BOBA', decimals: 18 },
    rpcUrls: {
        default: {
            http: ['https://bnb.boba.network'],
        },
        public: {
            http: ['https://bnb.boba.network'],
        },
    },
    blockExplorers: {
        default: {
            name: 'Boba BNB Mainnet Explorer',
            url: 'https://bnb.bobascan.com/',
        },
    },
}), {
    key: 'boba-bnb',
    shortName: 'boba-bnb',
    netType: 'mainnet',
    contracts: {
        multicall3: {
            address: '0x67da5f2ffaddff067ab9d5f025f8810634d84287',
            blockCreated: 18871,
        },
    },
    parentChainId: bsc.chainId,
});
const arbitrum = (0, define_chain_js_1.defineEvmChain)(chains_1.arbitrum, {
    key: 'arbitrum',
    shortName: 'arb1',
    parentChainId: ethereum.chainId,
});
const arbitrumNova = (0, define_chain_js_1.defineEvmChain)(chains_1.arbitrumNova, {
    key: 'arbitrum-nova',
    shortName: 'arb-nova',
    parentChainId: ethereum.chainId,
});
const avalanche = (0, define_chain_js_1.defineEvmChain)(chains_1.avalanche, {
    key: 'avalanche',
    shortName: 'avax',
});
const harmony = (0, define_chain_js_1.defineEvmChain)(chains_1.harmonyOne, {
    key: 'harmony',
    shortName: 'hmy-s0',
});
const celo = (0, define_chain_js_1.defineEvmChain)(chains_1.celo, {
    key: 'celo',
    shortName: 'celo',
});
const optimism = (0, define_chain_js_1.defineEvmChain)(chains_1.optimism, {
    key: 'optimism',
    shortName: 'oeth',
    parentChainId: ethereum.chainId,
});
const kava = (0, define_chain_js_1.defineEvmChain)(chains_1.kava, {
    key: 'kava',
    shortName: 'kava',
});
const metis = (0, define_chain_js_1.defineEvmChain)(chains_1.metis, {
    key: 'metis',
    shortName: 'metis-andromeda',
    parentChainId: ethereum.chainId,
});
const boba = (0, define_chain_js_1.defineEvmChain)(chains_1.boba, {
    key: 'boba',
    shortName: 'boba',
    parentChainId: ethereum.chainId,
});
const bttc = (0, define_chain_js_1.defineEvmChain)(chains_1.bitTorrent, {
    key: 'bttc',
    shortName: 'btt',
});
const polygonZkevm = (0, define_chain_js_1.defineEvmChain)(chains_1.polygonZkEvm, {
    key: 'polygon-zkevm',
    shortName: 'pol-zkevm',
    parentChainId: ethereum.chainId,
});
const thundercore = (0, define_chain_js_1.defineEvmChain)(chains_1.thunderCore, {
    key: 'thundercore',
    shortName: 'tt',
});
const filecoin = (0, define_chain_js_1.defineEvmChain)(chains_1.filecoin, {
    key: 'filecoin',
    shortName: 'fil',
});
const haqq = (0, define_chain_js_1.defineEvmChain)(chains_1.haqqMainnet, {
    key: 'haqq',
    contracts: {
        multicall3: {
            address: '0xfe2d04a5018ac1b366f599a13bf4e0c760b2ae6b',
            blockCreated: 6589598,
        },
    },
    shortName: 'haqq',
});
const core = (0, define_chain_js_1.defineEvmChain)(chains_1.coreDao, {
    key: 'core',
    shortName: 'core',
});
const zksync = (0, define_chain_js_1.defineEvmChain)(chains_1.zksync, {
    key: 'zksync-era',
    shortName: 'zksync-era',
    parentChainId: ethereum.chainId,
});
const linea = (0, define_chain_js_1.defineEvmChain)(chains_1.linea, {
    key: 'linea',
    shortName: 'linea',
    parentChainId: ethereum.chainId,
});
const base = (0, define_chain_js_1.defineEvmChain)(chains_1.base, {
    key: 'base',
    shortName: 'base',
    parentChainId: ethereum.chainId,
});
const scroll = (0, define_chain_js_1.defineEvmChain)(chains_1.scroll, {
    key: 'scroll',
    shortName: 'scr',
    parentChainId: ethereum.chainId,
});
const zetachain = (0, define_chain_js_1.defineEvmChain)(chains_1.zetachain, {
    key: 'zetachain',
    shortName: 'zetachain',
});
const cronos = (0, define_chain_js_1.defineEvmChain)(chains_1.cronos, {
    key: 'cronos',
    shortName: 'cro',
});
const blast = (0, define_chain_js_1.defineEvmChain)(chains_1.blast, {
    key: 'blast',
    shortName: 'blast',
    parentChainId: ethereum.chainId,
});
const skaleEuropa = (0, define_chain_js_1.defineEvmChain)(chains_1.skaleEuropa, {
    key: 'skale-europa',
    shortName: 'europa',
});
const rootstock = (0, define_chain_js_1.defineEvmChain)(chains_1.rootstock, {
    key: 'rootstock',
    shortName: 'rsk',
});
const mantle = (0, define_chain_js_1.defineEvmChain)(chains_1.mantle, {
    key: 'mantle',
    shortName: 'mantle',
    parentChainId: ethereum.chainId,
});
const manta = (0, define_chain_js_1.defineEvmChain)(chains_1.manta, {
    key: 'manta',
    shortName: 'manta',
    parentChainId: ethereum.chainId,
});
const mode = (0, define_chain_js_1.defineEvmChain)(chains_1.mode, {
    key: 'mode',
    shortName: 'mode',
    parentChainId: ethereum.chainId,
});
const taiko = (0, define_chain_js_1.defineEvmChain)(chains_1.taiko, {
    key: 'taiko',
    shortName: 'tko',
    parentChainId: ethereum.chainId,
});
const zklink = (0, define_chain_js_1.defineEvmChain)(chains_1.zkLinkNova, {
    key: 'zklink',
    contracts: {
        multicall3: {
            address: '0x825267e0fa5cae92f98540828a54198dcb3eaeb5',
            blockCreated: 146055,
        },
    },
    shortName: 'zklink',
});
const ape = (0, define_chain_js_1.defineEvmChain)(chains_1.apeChain, {
    key: 'ape',
    shortName: 'ape',
    parentChainId: arbitrum.chainId,
});
const sonic = (0, define_chain_js_1.defineEvmChain)(chains_1.sonic, {
    key: 'sonic',
    shortName: 'sonic',
});
const hemi = (0, define_chain_js_1.defineEvmChain)(chains_1.hemi, {
    key: 'hemi',
    contracts: {
        multicall3: {
            address: '0xca11bde05977b3631167028862be2a173976ca11',
            blockCreated: 484490,
        },
    },
    shortName: 'hemi',
    parentChainId: ethereum.chainId,
});
const katana = (0, define_chain_js_1.defineEvmChain)((0, viem_1.defineChain)({
    id: 747474,
    name: 'Katana',
    nativeCurrency: {
        decimals: 18,
        name: 'Ether',
        symbol: 'ETH',
    },
    rpcUrls: {
        default: {
            http: ['https://rpc.katanarpc.com'],
        },
    },
    blockExplorers: {
        default: {
            name: 'Katana Explorer',
            url: 'https://katanascan.com',
            apiUrl: 'https://api.katanascan.com/api',
        },
    },
    testnet: false,
    contracts: {
        multicall3: {
            address: '0xca11bde05977b3631167028862be2a173976ca11',
            blockCreated: 0,
        },
    },
}), {
    key: 'katana',
    shortName: 'katana',
    parentChainId: ethereum.chainId,
});
const hyperevm = (0, define_chain_js_1.defineEvmChain)((0, viem_1.defineChain)({
    id: 999,
    name: 'HyperEVM',
    nativeCurrency: {
        decimals: 18,
        name: 'Hyperliquid',
        symbol: 'HYPE',
    },
    rpcUrls: {
        default: {
            http: ['https://rpc.hyperliquid.xyz/evm'],
        },
    },
    blockExplorers: {
        default: {
            name: 'HyperEVM Explorer',
            url: 'https://hyperevmscan.io',
            apiUrl: 'https://api.hyperevmscan.io/api',
        },
    },
    testnet: false,
    contracts: {
        multicall3: {
            address: '0xca11bde05977b3631167028862be2a173976ca11',
            blockCreated: 13051,
        },
    },
}), {
    key: 'hyperevm',
    shortName: 'hyperevm',
});
const berachain = (0, define_chain_js_1.defineEvmChain)(chains_1.berachain, {
    key: 'berachain',
    shortName: 'bera',
});
const plasma = (0, define_chain_js_1.defineEvmChain)((0, viem_1.defineChain)({
    id: 9745,
    name: 'Plasma',
    nativeCurrency: {
        name: 'Plasma',
        symbol: 'XPL',
        decimals: 18,
    },
    rpcUrls: {
        default: {
            http: ['https://rpc.plasma.to'],
        },
    },
    blockExplorers: {
        default: {
            name: 'PlasmaScan',
            url: 'https://plasmascan.to',
        },
    },
    testnet: false,
    contracts: {
        multicall3: {
            address: '0xca11bde05977b3631167028862be2a173976ca11',
            blockCreated: 0,
        },
    },
}), {
    key: 'plasma',
    shortName: 'plasma',
});
const sepolia = (0, define_chain_js_1.defineEvmChain)(chains_1.sepolia, {
    key: 'sepolia',
    shortName: 'sep',
});
const arbitrumSepolia = (0, define_chain_js_1.defineEvmChain)(chains_1.arbitrumSepolia, {
    key: 'arbitrum-sepolia',
    shortName: 'arb-sep',
    parentChainId: sepolia.chainId,
});
const tatara = (0, define_chain_js_1.defineEvmChain)((0, viem_1.defineChain)({
    id: 129399,
    name: 'Tatara',
    nativeCurrency: {
        decimals: 18,
        name: 'Ether',
        symbol: 'ETH',
    },
    rpcUrls: {
        default: {
            http: ['https://rpc.tatara.katanarpc.com'],
        },
    },
    blockExplorers: {
        default: {
            name: 'Tatara Explorer',
            url: 'https://explorer.tatara.katana.network',
            apiUrl: 'https://explorer.tatara.katana.network/api',
        },
    },
    testnet: true,
    contracts: {
        multicall3: {
            address: '0xca11bde05977b3631167028862be2a173976ca11',
            blockCreated: 0,
        },
    },
}), {
    key: 'tatara',
    shortName: 'tatara',
    parentChainId: sepolia.chainId,
});
exports.evmChains = [
    ethereum,
    bobaBnb,
    polygon,
    fantom,
    gnosis,
    bsc,
    arbitrum,
    arbitrumNova,
    avalanche,
    harmony,
    celo,
    optimism,
    kava,
    metis,
    boba,
    bttc,
    polygonZkevm,
    thundercore,
    filecoin,
    haqq,
    core,
    zksync,
    linea,
    base,
    scroll,
    zetachain,
    cronos,
    blast,
    skaleEuropa,
    rootstock,
    mantle,
    manta,
    mode,
    taiko,
    zklink,
    ape,
    sonic,
    hemi,
    katana,
    hyperevm,
    berachain,
    plasma,
    sepolia,
    arbitrumSepolia,
    tatara,
];
exports.evmChainIds = exports.evmChains.map((chain) => chain.chainId);
const chainIdByKey = Object.freeze(exports.evmChains.reduce((acc, { key, chainId }) => {
    acc[key.toUpperCase().replace(/-/g, '_')] = chainId;
    return acc;
}, {}));
exports.EvmChainId = chainIdByKey;
const chainById = Object.freeze(Object.fromEntries(exports.evmChains.map((c) => [c.chainId, c])));
function isEvmChainId(chainId) {
    return Object.hasOwn(chainById, chainId);
}
function getEvmChainById(chainId) {
    return chainById[chainId];
}
function isEvmMainnetChainId(chainId) {
    return chainById[chainId]?.netType === 'mainnet';
}
function isEvmTestnetChainId(chainId) {
    return chainById[chainId]?.netType === 'testnet';
}
function isEvmChainKey(key) {
    return exports.evmChains.some((chain) => chain.key === key);
}
function getEvmChainByKey(key) {
    return exports.evmChains.find((chain) => chain.key === key);
}
