// sushi/_cjs/evm/chain/define-chain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defineEvmChain = defineEvmChain;
function defineEvmChain(viemChain, chain) {
    const chainId = (chain.chainId || viemChain.id);
    const blockExplorers = (chain.blockExplorers ||
        viemChain.blockExplorers);
    const name = (chain.name || viemChain.name);
    const nativeCurrency = (chain.nativeCurrency ||
        viemChain.nativeCurrency);
    const rpcUrls = (chain.rpcUrls || viemChain.rpcUrls);
    const contracts = (chain.contracts || viemChain.contracts);
    const netType = (viemChain.testnet ? 'testnet' : 'mainnet');
    if (!blockExplorers) {
        throw new Error('defineEvmChain: blockExplorers is required');
    }
    if (!contracts?.multicall3) {
        throw new Error('defineEvmChain: contracts->multicall3 is required');
    }
    return {
        type: 'evm',
        chainId,
        key: chain.key,
        name,
        shortName: chain.shortName,
        netType,
        blockExplorers,
        parentChainId: chain.parentChainId,
        viemChain: {
            ...viemChain,
            id: chainId,
            name: name,
            nativeCurrency: nativeCurrency,
            rpcUrls: rpcUrls,
            blockExplorers: blockExplorers,
            contracts: contracts,
        },
        getTransactionUrl: (input) => `${blockExplorers.default.url}/tx/${input}`,
        getAccountUrl: (input) => `${blockExplorers.default.url}/address/${input}`,
        getTokenUrl: (input) => `${blockExplorers.default.url}/token/${input}`,
    };
}
